---
title: "Interferon Gamma Signaling in Sarcoidosis RNAseq Data"
author: "Xiting Yan"
date: "11/28/2020"
output:
  html_notebook:
    code_folding: hide
    fig_caption: yes
    highlight: tango
    number_sections: no
    theme: united
    toc: yes
    toc_depth: 6
    toc_float: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(eval=TRUE,echo = TRUE,cache=TRUE,warning=FALSE,message = FALSE,results='hold',cache.lazy = FALSE)
knitr::opts_knit$set(eval.after = 'fig.cap',dev=c('png','postscript'))


library(captioner)
library(circlize)
library(clusterProfiler)
library(corrplot)
library(ComplexHeatmap)
library(dplyr)
library(EnsDb.Hsapiens.v86)
library(gdata)
library(ggplot2)
library(ggpubr)
library(ggrepel)
library(gplots)
library(grid)
library(gridExtra)
library(gridGraphics)
library(kableExtra)
library(knitr)
library(Matrix)
library(org.Hs.eg.db)
library(parallel)
library(reshape2)
library(scales)
library(Seurat)
library(SingleR)
library(tidyr)
library(tidyverse)
library(viridisLite)
library(xlsx)
library(randomcoloR)

knit_hooks$set(webgl = hook_webgl)

table_nums_1 <- captioner::captioner(prefix="Table",levels=1)
figure_nums_1<- captioner::captioner(prefix="Figure",levels=1)


table_nums_2 <- captioner::captioner(prefix="Table",levels=2)
figure_nums_2<- captioner::captioner(prefix="Figure",levels=2)

table_nums_3 <- captioner::captioner(prefix="Table",levels=3)
figure_nums_3 <- captioner::captioner(prefix="Figure",levels=3)

home.dir<-"/home/yanxiting/driver_Grace"
source(paste(home.dir,"/Rprogram/my_functions.R",sep=""))
output.dir<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38")
```


# Background

In this note, we explore the expression profiles of genes related to interferon gamma signaling to understand the inflammation signals in the blood of sarcoidosis patients.

We downloaded the Interferon Gamma Signaling network from Metacore and demonstrated the expression profiles of these genes in the Sarcoidosis patients.

# IFNG signaling expression
We load in the list of IFN-gamma signaling genes from metacore and demonstrated the expression profiles of these genes in the Sarcoidosis PBMC data.
```{r fig.height=16,fig.width=10}
# load in the clinical data
clinic.filepath<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/data/clin_data_visit1_trunc_XY.csv")
clinic.matrix=read.csv(clinic.filepath,check.names=F,stringsAsFactors=F)
rownames(clinic.matrix)<-clinic.matrix[,1]

# load in the IFN-gamma signaling gene
data.filepath<-file.path(output.dir,"data","Interferon_gamma_signaling_metacore.xls")
gene.list<-read.xls(data.filepath,check.names=F,skip=2,stringsAsFactors=F)
gene.list<-gene.list[gene.list$'Gene Symbol'!="",]
gene.list<-unique(gene.list$`Gene Symbol`) 
length(gene.list)

# load in the gene expression data
# fpkm.filepath<-"/Users/yanxiting/MyVolumes/GRACE/scratch/GRADS/SARC_results/Results_summary_BAL_hg38/FPKM_matrix_hg38_Corrected_allGenes_BAL209_withAnnot.txt"
data.dir<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38")
fpkm.filepath<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/data","TPM_baseline_276_clean_GRADSID.txt")

# load in the fpkm matrix
fpkm.matrix<-read.table(fpkm.filepath,sep="\t",header=T,check.names=F,as.is=TRUE,comment.char = "")
fpkm.matrix.anno<-fpkm.matrix[,1:5]
fpkm.matrix<-fpkm.matrix[,6:ncol(fpkm.matrix)]

# generate the heatmap for the genes in gene.list only
data.matrix<-fpkm.matrix[fpkm.matrix.anno[,1]%in%gene.list,]
data.matrix.anno<-fpkm.matrix.anno[fpkm.matrix.anno[,1]%in%gene.list,]
rownames(data.matrix)<-data.matrix.anno[,1]

unityNormalize <- function(x){
     return((x-min(x))/(max(x)-min(x)))
}

my.data.norm<-t(apply(as.matrix(data.matrix), 1, unityNormalize))  
my.clinic.matrix<-clinic.matrix[colnames(my.data.norm),]
my.clinic.matrix[my.clinic.matrix[,"cd4"]<0,"cd4"]<-NA
# sort the cells for cell types, disease and subjects
temp.meta.data<-data.frame(cellBarcode=1:ncol(my.data.norm),phenogroup=my.clinic.matrix[,"PHENGRP"],cd4=my.clinic.matrix[,"cd4"],subject.names=my.clinic.matrix[,1],stringsAsFactors=F)

column_order <- temp.meta.data %>% 
    arrange(phenogroup, cd4) %>%
    pull(cellBarcode)

phenogroup_order<-temp.meta.data %>% 
    arrange(phenogroup, cd4) %>%
    pull(phenogroup)

cd4_order <- temp.meta.data %>% 
    arrange(phenogroup, cd4) %>%
    pull(cd4)
    
subject_order <- temp.meta.data %>% 
    arrange(phenogroup, cd4) %>%
    pull(subject.names)

my.data<-my.data.norm[,column_order]

######
# define the color
my.colors<-randomColor(count = length(unique(my.clinic.matrix[,"PHENGRP"])),hue="random",luminosity="dark")
#my.colors<-distinctColorPalette(k=length(unique(my.clinic.matrix[,"PHENGRP"])),altCol=FALSE,runTsne=TRUE)
df.color<-list()

# colors for cell types
temp.vect<-my.colors
names(temp.vect)<-unique(my.clinic.matrix$PHENGRP)
df.color[[1]]<-temp.vect

# color for disease
#df.color[[2]]<-c("control"="lightgreen","asthma"="salmon")
df.color[[2]]<-colorRamp2(c(0, 1000, 2000), c("blue", "white", "red"))

  
# color for subject
temp.vect<-randomColor(length(unique(my.clinic.matrix[,1])))
names(temp.vect)<-unique(as.character(subject_order))
df.color[[3]]<-temp.vect

names(df.color)<-c("phenogroup","cd4","subject")

# create the heatmap annotation
#col.annotation<-HeatmapAnnotation(celltype=celltype_order,disease=disease_order,subject=subject_order,col=df.color,show_legend=FALSE,annotation_height=unit(7,"cm"))
col.annotation<-HeatmapAnnotation(phenogroup=phenogroup_order,cd4=cd4_order,show_legend=TRUE,annotation_height=unit(7,"cm"))


colors.viridis.unity <- colorRamp2(c(0, 1/10, 2/10, 3/10, 4/10, 5/10, 6/10, 7/10, 8/10, 1), c(viridis(n=9),viridis(n=9)[9]), space = "RGB")
colors.plasma.unity <- colorRamp2(c(0, 1/10, 2/10, 3/10, 4/10, 5/10, 6/10, 7/10, 8/10, 1), c(plasma(n=9),plasma(n=9)[9]), space = "RGB")
colors.inferno.unity <- colorRamp2(c(0, 1/10, 2/10, 3/10, 4/10, 5/10, 6/10, 7/10, 8/10, 1), c(inferno(n=9), inferno(n=9)[9]), space="RGB")


g<- grid.grabExpr(draw(Heatmap(as.matrix(my.data.norm), 
    col= colors.inferno.unity,
    #column_split= factor(celltype_order, levels=cluster.chosen),
    #row_split=factor(row_celltype_order,levels=cluster.chosen),
    #cluster_rows=TRUE,
    cluster_columns=FALSE,
    #column_order=my.column.order,
    #row_order=hclust(dist(my.data.norm))$order,
    cluster_column_slices=FALSE,
    top_annotation=col.annotation,
    show_column_names=FALSE,
    #column_title_side="top",
    #column_title_rot = 0,
    column_dend_reorder=TRUE,
    #row_names_gp = gpar(fontsize = 6),
    show_row_names = TRUE,
    #right_annotation=row.annotation,
    use_raster=FALSE,
    heatmap_legend_param=list(title = "Expression"),
    show_heatmap_legend = TRUE
    )))

output.filepath<-file.path(output.dir,"baseline","figures","heatmap_IFNGsignaling_CD4&Phenogroup.pdf")
ggsave(output.filepath,device="pdf",height=16,width=10,dpi=300,g)

Heatmap(as.matrix(my.data.norm), 
    col= colors.inferno.unity,
    #column_split= factor(celltype_order, levels=cluster.chosen),
    #row_split=factor(row_celltype_order,levels=cluster.chosen),
    #cluster_rows=TRUE,
    cluster_columns=FALSE,
    #column_order=my.column.order,
    #row_order=hclust(dist(my.data.norm))$order,
    cluster_column_slices=FALSE,
    top_annotation=col.annotation,
    show_column_names=FALSE,
    #column_title_side="top",
    #column_title_rot = 0,
    column_dend_reorder=TRUE,
    #row_names_gp = gpar(fontsize = 6),
    show_row_names = TRUE,
    #right_annotation=row.annotation,
    use_raster=FALSE,
    heatmap_legend_param=list(title = "Expression"),
    show_heatmap_legend = TRUE
    )
```

Here's a heatmap for all IFN-gamma signaling genes with annotation but clustered.

```{r fig.height=15, fig.width=10}
Heatmap(as.matrix(my.data.norm), 
    col= colors.inferno.unity,
    #column_split= factor(celltype_order, levels=cluster.chosen),
    #row_split=factor(row_celltype_order,levels=cluster.chosen),
    #cluster_rows=TRUE,
    cluster_columns=TRUE,
    #column_order=my.column.order,
    #row_order=hclust(dist(my.data.norm))$order,
    cluster_column_slices=FALSE,
    top_annotation=col.annotation,
    show_column_names=FALSE,
    #column_title_side="top",
    #column_title_rot = 0,
    column_dend_reorder=TRUE,
    #row_names_gp = gpar(fontsize = 6),
    show_row_names = TRUE,
    #right_annotation=row.annotation,
    use_raster=FALSE,
    heatmap_legend_param=list(title = "Expression"),
    show_heatmap_legend = TRUE
    )
g<-grid.grabExpr(draw(Heatmap(as.matrix(my.data.norm), 
    col= colors.inferno.unity,
    #column_split= factor(celltype_order, levels=cluster.chosen),
    #row_split=factor(row_celltype_order,levels=cluster.chosen),
    #cluster_rows=TRUE,
    cluster_columns=TRUE,
    #column_order=my.column.order,
    #row_order=hclust(dist(my.data.norm))$order,
    cluster_column_slices=FALSE,
    top_annotation=col.annotation,
    show_column_names=FALSE,
    #column_title_side="top",
    #column_title_rot = 0,
    column_dend_reorder=TRUE,
    #row_names_gp = gpar(fontsize = 6),
    show_row_names = TRUE,
    #right_annotation=row.annotation,
    use_raster=FALSE,
    heatmap_legend_param=list(title = "Expression"),
    show_heatmap_legend = TRUE
    )))
output.filepath<-file.path(output.dir,"baseline","figures","heatmap_IFNGsignaling_CD4&Phenogroup_clustered.pdf")
ggsave(output.filepath,device="pdf",height=15,width=10,dpi=300,g)

```

Here's the heatmap for all IFN-gamma signaling genes without any annotation for the columns.
```{r fig.height=16,fig.width=10}
Heatmap(as.matrix(my.data.norm), 
    col= colors.inferno.unity,
    #column_split= factor(celltype_order, levels=cluster.chosen),
    #row_split=factor(row_celltype_order,levels=cluster.chosen),
    #cluster_rows=TRUE,
    cluster_columns=TRUE,
    #column_order=my.column.order,
    #row_order=hclust(dist(my.data.norm))$order,
    cluster_column_slices=FALSE,
    #top_annotation=col.annotation,
    show_column_names=FALSE,
    #column_title_side="top",
    #column_title_rot = 0,
    column_dend_reorder=TRUE,
    #row_names_gp = gpar(fontsize = 6),
    show_row_names = TRUE,
    #right_annotation=row.annotation,
    use_raster=FALSE,
    heatmap_legend_param=list(title = "Expression"),
    show_heatmap_legend = TRUE
    )
```
