---
title: "SARC PBMC Supervised Analysis"
author: "Xiting Yan"
date: "11/29/2020"
output: 
  html_notebook:
    theme: united
    highlight: tango
    toc: true
    toc_depth: 6
    toc_float: true
    number_sections: true
    code_folding: hide
    fig_caption: true
    #bibliography: bibliography.bib
    #csl: biomed-central.csl
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,cache=TRUE,warning=FALSE,message = FALSE,results='markup',cache.lazy = FALSE)
knitr::opts_knit$set(eval.after = 'fig.cap',dev=c('png','postscript'))

library(kableExtra)
library(gdata)
library(knitr)
library(captioner)
library(nlme)
library(rgl)
library(gplots)

knit_hooks$set(webgl = hook_webgl)


table_nums_1 <- captioner::captioner(prefix="Table",levels=1)
figure_nums_1<- captioner::captioner(prefix="Figure",levels=1)


table_nums_2 <- captioner::captioner(prefix="Table",levels=2)
figure_nums_2<- captioner::captioner(prefix="Figure",levels=2)

table_nums_3 <- captioner::captioner(prefix="Table",levels=3)
figure_nums_3 <- captioner::captioner(prefix="Figure",levels=3)
# This rmd file is compiled on xiting's laptop. If needed to be compiled on a different machine, the file path will need to be changed.

home.dir<-"/home/yanxiting/driver_Grace"
```

# Method
To identify gene signatures associated with clinical traits in a supervised way, we conducted the supervised analysis to identify gene signatures associated with each given traits, including the phenotypes, demographics, PFTs and other clinical traits of sarcoidosis.

# Results

## original data

### Data Loading
First, we load in the fpkm and the clinical data.
```{r}

#########
# This part comes from 209_local_data_load_AM_XYversion.R
#this code loads the 209 BAL samples into the same variable names as clean_data_load.R
source(file.path(home.dir,"/Rprogram/my_functions.R"))
fpkm.filepath<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/data","TPM_baseline_276_clean_GRADSID.txt")
clinic.filepath<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/data/clin_data_visit1_trunc_XY.csv")
ct.filepath<-file.path(home.dir,"scratch/GRADS/SARC_results/Data/CT_data_20170712/Sarc_ct_reads_corrected.xls")

########################################
# load in the fpkm matrix
fpkm.matrix.clean<-read.table(fpkm.filepath,sep="\t",header=T,check.names=F,as.is=TRUE,comment.char = "")
rownames(fpkm.matrix.clean)<-paste(fpkm.matrix.clean[,1],"_",fpkm.matrix.clean[,2],sep="") # some genes have >1 isoforms due to two different annotations on two different chromosomes.
anno.matrix<-fpkm.matrix.clean[,1:5]
fpkm.matrix.clean<-fpkm.matrix.clean[,6:ncol(fpkm.matrix.clean)]
fpkm.matrix.clean<-as.matrix(fpkm.matrix.clean)

########################################
# load in the clinical data extracted from the database.

# clinical traits provided by Laura
clinic.matrix=read.csv(clinic.filepath,check.names=F,stringsAsFactors=F)
rownames(clinic.matrix)<-clinic.matrix[,1]

# load in the CT scan reads and merge them into the clinical data
ct.matrix<-read.xls(ct.filepath,stringsAsFactors=F)
rownames(ct.matrix)<-ct.matrix[,1]
ct.matrix<-ct.matrix[,c("Med_Lymphadenopathy","Hilar_Lymphadenopathy","Micronodule","Bronchial_Wall_Thickening","Traction_Bronchiectasis","Bronchiectasis_severity","Ground_Glass","Honeycombing","Reticular_Abnormality","Pulmonary_Art","Tree_in_bud")]
ct.matrix<-ct.matrix[rownames(clinic.matrix),]
rownames(ct.matrix)<-rownames(clinic.matrix)
clinic.matrix<-cbind(clinic.matrix,ct.matrix)

# only do this for the traits included in the BAL analysis
clinic.matrix.bal<-read.table(file.path(work.dir,"scratch/GRADS/SARC_results/Data/SARC_combine_clinical_data20170814.txt"),sep="\t",header=T,check.names=F,stringsAsFactors=F,quote="")
my.clinic.matrix<-clinic.matrix[,colnames(clinic.matrix)%in%colnames(clinic.matrix.bal)]
my.clinic.matrix<-cbind(my.clinic.matrix,clinic.matrix[,c("p_total","p_mono","p_neut","p_eos","p_baso","p_lymph","abs_lymph")])

# add the FEV1/FVC ratio
my.clinic.matrix<-cbind(my.clinic.matrix,my.clinic.matrix[,"BDFEV1"]/my.clinic.matrix[,"BDFVC"])
colnames(my.clinic.matrix)[ncol(my.clinic.matrix)]<-"FEV1FVCratio"

# output the merged clinical data matrix into a file.
output.filepath<-file.path(home.dir,"scratch/GRADS/SARC_results/Data_PBMC/SARC_combine_clinical_data20210112.xlsx")
write.xlsx(my.clinic.matrix,file=output.filepath,row.names=T,col.names=T,append=F)
cat("The combined clinical data has been saved as ",output.filepath,"\n",sep="")

# match the clinical data with the columns in the fpkm matrix
my.clinic.matrix<-my.clinic.matrix[colnames(fpkm.matrix.clean),]

# add the obstructive vs. restrictive variable FEV1FVCratio=0 is obstructive and =1 is restrictive
#temp.vect<-rep(NA,nrow(clinic.matrix))
#temp.vect[!is.na(clinic.matrix[,"FEV1FVCratio"]) & clinic.matrix[,"FEV1FVCratio"]<0.7]<- 0
#temp.vect[(!is.na(clinic.matrix[,"FEV1FVCratio"]) & clinic.matrix[,"FEV1FVCratio"]>=0.7) & (clinic.matrix[,"FVCPRED"]<80 & clinic.matrix[,"FVCPRED"]>=0)]<- 1
#clinic.matrix<-cbind(clinic.matrix,temp.vect)
#colnames(clinic.matrix)[ncol(clinic.matrix)]<-"FEV1FVCratioBinary"

# add the obstructive vs. normal variable obstractiveBinary=0 is normal and =1 is obstructive
#temp.vect<-rep(NA,nrow(clinic.matrix))
#temp.vect[(!is.na(clinic.matrix[,"FEV1FVCratio"]) & clinic.matrix[,"FEV1FVCratio"]<0.7)]<- 1
#temp.vect[(!is.na(clinic.matrix[,"FEV1FVCratio"]) & clinic.matrix[,"FEV1FVCratio"]>0.7) & (clinic.matrix[,"FVCPRED"]>80 & clinic.matrix[,"FVCPRED"]>=0)]<- 0
#clinic.matrix<-cbind(clinic.matrix,temp.vect)
#colnames(clinic.matrix)[ncol(clinic.matrix)]<-"obstructiveBinary"



# add the restrictive vs. normal variable restrictiveBinary=0 is normal and =1 is restrictive
#temp.vect<-rep(NA,nrow(clinic.matrix))
#temp.vect[(!is.na(clinic.matrix[,"FEV1FVCratio"]) & clinic.matrix[,"FEV1FVCratio"]>0.7) & (clinic.matrix[,"FVCPRED"]<80 & clinic.matrix[,"FVCPRED"]>=0)]<- 1
#temp.vect[(!is.na(clinic.matrix[,"FEV1FVCratio"]) & clinic.matrix[,"FEV1FVCratio"]>0.7) & (clinic.matrix[,"FVCPRED"]>80 & clinic.matrix[,"FVCPRED"]>=0)]<- 0
#clinic.matrix<-cbind(clinic.matrix,temp.vect)
#colnames(clinic.matrix)[ncol(clinic.matrix)]<-"restrictiveBinary"
cat("The dimension of the clinical data is\n",paste(dim(my.clinic.matrix),collapse=" "),"\n",sep="")
```

Second, we conduct the supervised analysis to identify genes associated with the given traits extracted above.

### Continuous traits
* Function to do the test for continuous tratis.
```{r}
#########
# This part comes from hg38_Supervised_analysis_AM_XYversion.R
################################
# 4.4. Identify genes associated with age and PFTs
#source("/home/fas/kaminski/am2952/scratch/SARC215/0_CleanDataLoad.R")
#source(file.path(work.dir,"Rprogram/209_local_data_load_AM.R")

my.con.cor<-function(fpkm.matrix.clean,feature.name="AGE",method.name="spearman",tissue.name="BAL",clinic.matrix,output.dir){

# fpkm.matrix.clean needs to have column names as GRADS IDs
# clinic.matrix needs to have row names as GRADS IDs 
temp.names<-intersect(colnames(fpkm.matrix.clean),rownames(clinic.matrix))
data.matrix<-fpkm.matrix.clean[,temp.names]
clinic.matrix<-clinic.matrix[temp.names,]
bal.fpkm<-data.matrix

# generate the clinical matrix
my.clinic.matrix<-clinic.matrix[colnames(data.matrix),]	

clinic.vect<-as.numeric(my.clinic.matrix[,colnames(my.clinic.matrix)==feature.name])
clinic.vect[clinic.vect<0]<-NA

# generate the correlation results
age.pvalue<-apply(bal.fpkm,1,my.cor.test,y=clinic.vect,method.name=method.name)
age.pvalue<-matrix(unlist(age.pvalue),ncol=2,byrow=T)
colnames(age.pvalue)<-c("pvalue","corr_coef")
age.fdr<-p.adjust(age.pvalue[,"pvalue"],method="fdr")
age.pvalue<-cbind(age.pvalue,age.fdr)
colnames(age.pvalue)[ncol(age.pvalue)]<-"fdr"
rownames(age.pvalue)<-rownames(bal.fpkm)

fpkm.info<-apply(bal.fpkm>0,1,sum)
fpkm.info<-cbind(fpkm.info,apply(bal.fpkm>0.1,1,sum))
fpkm.info<-cbind(fpkm.info,apply(bal.fpkm>1,1,sum))
fpkm.info<-cbind(fpkm.info,apply(bal.fpkm,1,median))
fpkm.info<-cbind(fpkm.info,apply(bal.fpkm,1,mean))
colnames(fpkm.info)<-c("fpkm>0_num","fpkm>0.1_num","fpkm>1_num","fpkm_median","fpkm_mean")


if(file.exists(output.dir)==F){
cat("WARNING: output.dir doesn't exists!\n")
return(1)
}

cmd.out<-cbind(age.pvalue,fpkm.info)
output.filepath<-file.path(output.dir,paste("correlation_",method.name,"_",feature.name,"_allgenes_",tissue.name,".txt",sep=""))
cat("gene_names\t",file=output.filepath,append=F)
write.table(cmd.out,file=output.filepath,append=T,sep="\t",row.names=T,col.names=T,quote=F)

output.filepath<-file.path(output.dir,paste("correlation_",method.name,"_",feature.name,"_fdr0.05_",tissue.name,".txt",sep=""))
cat("gene_names\t",file=output.filepath,append=F)
#write.table(age.pvalue[!is.na(age.pvalue[,3]) & age.pvalue[,3]<0.05,],file=output.filepath,append=T,sep="\t",row.names=T,col.names=T,quote=F)
write.table(cmd.out[!is.na(age.pvalue[,3]) & age.pvalue[,3]<0.05,],file=output.filepath,append=T,sep="\t",row.names=T,col.names=T,quote=F)

output.filepath<-file.path(output.dir,paste("correlation_",method.name,"_",feature.name,"_fdr0.05_",tissue.name,"_both_metacore.txt",sep=""))
cat("gene_names\tintensity\n",file=output.filepath,append=F)
sig.genes<-rownames(data.matrix)[!is.na(age.pvalue[,3]) & age.pvalue[,3]<0.05]
sig.genes<-unname(sapply(sig.genes,my.element.extract,splitchar="_",index=1))
sig.genes<-unique(sig.genes)
cmd.out<-cbind(sig.genes,rep(1,length(sig.genes)))
write.table(cmd.out,file=output.filepath,append=T,sep="\t",row.names=F,col.names=F,quote=F)


output.filepath<-file.path(output.dir,paste("correlation_",method.name,"_",feature.name,"_fdr0.05_",tissue.name,"_positive_metacore.txt",sep=""))
cat("gene_names\tintensity\n",file=output.filepath,append=F)
sig.genes<-rownames(data.matrix)[!is.na(age.pvalue[,3]) & age.pvalue[,3]<0.05 & age.pvalue[,2]>0]
sig.genes<-unname(sapply(sig.genes,my.element.extract,splitchar="_",index=1))
sig.genes<-unique(sig.genes)
cmd.out<-cbind(sig.genes,rep(1,length(sig.genes)))
write.table(cmd.out,file=output.filepath,append=T,sep="\t",row.names=F,col.names=F,quote=F)


output.filepath<-file.path(output.dir,paste("correlation_",method.name,"_",feature.name,"_fdr0.05_",tissue.name,"_negative_metacore.txt",sep=""))
cat("gene_names\tintensity\n",file=output.filepath,append=F)
sig.genes<-rownames(data.matrix)[!is.na(age.pvalue[,3]) & age.pvalue[,3]<0.05 & age.pvalue[,2]<0]
sig.genes<-unname(sapply(sig.genes,my.element.extract,splitchar="_",index=1))
sig.genes<-unique(sig.genes)
cmd.out<-cbind(sig.genes,rep(1,length(sig.genes)))
write.table(cmd.out,file=output.filepath,append=T,sep="\t",row.names=F,col.names=F,quote=F)

return(0)
}
```

* Apply the function above to the continuous traits
```{r}
# for age
#output.dir<-"~/driver_Naftali/GRADS/SARC_results/Results_summary/PFTs_correlated_genes"
output.dir<-"/home/yanxiting/Documents/Research/GRADS_SARC_PBMC/Demographics_correlated_genes"
if(file.exists(output.dir)==F){
  dir.create(output.dir)
}
my.con.cor(fpkm.matrix.clean=fpkm.matrix.clean,feature.name="AGE",method.name="spearman",tissue.name="PBMC",clinic.matrix=my.clinic.matrix,output.dir)
my.con.cor(fpkm.matrix.clean=fpkm.matrix.clean,feature.name="GENDER",method.name="spearman",tissue.name="PBMC",clinic.matrix=my.clinic.matrix,output.dir)
my.con.cor(fpkm.matrix.clean=fpkm.matrix.clean,feature.name="RACE",method.name="spearman",tissue.name="PBMC",clinic.matrix=my.clinic.matrix[my.clinic.matrix[,"RACE"]%in%c(1,2),],output.dir)


# for FVC, FEV1 and DLCO
output.dir<-"/home/yanxiting/Documents/Research/GRADS_SARC_PBMC/PFTs_correlated_genes"
if(file.exists(output.dir)==F){
  dir.create(output.dir)
}

my.con.cor(fpkm.matrix.clean,feature.name="BDFVC",method.name="spearman",tissue.name="PBMC",clinic.matrix=my.clinic.matrix,output.dir)
my.con.cor(fpkm.matrix.clean,feature.name="POSTFVC",method.name="spearman",tissue.name="PBMC",clinic.matrix=my.clinic.matrix,output.dir)
my.con.cor(fpkm.matrix.clean,feature.name="FVCPRED",method.name="spearman",tissue.name="PBMC",clinic.matrix=my.clinic.matrix,output.dir)
my.con.cor(fpkm.matrix.clean,feature.name="BDFEV1",method.name="spearman",tissue.name="PBMC",clinic.matrix=my.clinic.matrix,output.dir)
my.con.cor(fpkm.matrix.clean,feature.name="POSTFEV1",method.name="spearman",tissue.name="PBMC",clinic.matrix=my.clinic.matrix,output.dir)
my.con.cor(fpkm.matrix.clean,feature.name="FEV1PRED",method.name="spearman",tissue.name="PBMC",clinic.matrix=my.clinic.matrix,output.dir)
my.con.cor(fpkm.matrix.clean,feature.name="BDDLCO",method.name="spearman",tissue.name="PBMC",clinic.matrix=my.clinic.matrix,output.dir)
my.con.cor(fpkm.matrix.clean,feature.name="PREDDLCO",method.name="spearman",tissue.name="PBMC",clinic.matrix=my.clinic.matrix,output.dir)
my.con.cor(fpkm.matrix.clean,feature.name="FEV1FVCratio",method.name="spearman",tissue.name="PBMC",clinic.matrix=my.clinic.matrix,output.dir)


# for scadding treated as a continuous variable, potentially more reasonable than DEGs
#output.dir<-"~/driver_Naftali/GRADS/SARC_results/Results_summary/Disease_correlated_genes"
output.dir<-"/home/yanxiting/Documents/Research/GRADS_SARC_PBMC/Disease_correlated_genes"
if(file.exists(output.dir)==F){
  dir.create(output.dir)
}
my.con.cor(fpkm.matrix.clean,feature.name="SCADDING",method.name="spearman",tissue.name="PBMC",clinic.matrix=my.clinic.matrix,output.dir) 


output.dir<-"/home/yanxiting/Documents/Research/GRADS_SARC_PBMC/CellDifferentials_correlated_genes"
if(file.exists(output.dir)==F){
  dir.create(output.dir)
}

my.con.cor(fpkm.matrix.clean=fpkm.matrix.clean,feature.name="p_total",method.name="spearman",tissue.name="PBMC",clinic.matrix=my.clinic.matrix,output.dir=output.dir)
my.con.cor(fpkm.matrix.clean=fpkm.matrix.clean,feature.name="p_mono",method.name="spearman",tissue.name="PBMC",clinic.matrix=my.clinic.matrix,output.dir=output.dir)
my.con.cor(fpkm.matrix.clean=fpkm.matrix.clean,feature.name="p_neut",method.name="spearman",tissue.name="PBMC",clinic.matrix=my.clinic.matrix,output.dir=output.dir)
my.con.cor(fpkm.matrix.clean=fpkm.matrix.clean,feature.name="p_eos",method.name="spearman",tissue.name="PBMC",clinic.matrix=my.clinic.matrix,output.dir=output.dir)
my.con.cor(fpkm.matrix.clean=fpkm.matrix.clean,feature.name="p_baso",method.name="spearman",tissue.name="PBMC",clinic.matrix=my.clinic.matrix,output.dir=output.dir)
my.con.cor(fpkm.matrix.clean=fpkm.matrix.clean,feature.name="p_lymph",method.name="spearman",tissue.name="PBMC",clinic.matrix=my.clinic.matrix,output.dir=output.dir)
my.con.cor(fpkm.matrix.clean=fpkm.matrix.clean,feature.name="p_lymph",method.name="spearman",tissue.name="PBMC",clinic.matrix=my.clinic.matrix,output.dir=output.dir)


```

### Categorical traits
* Function to do the test for categorical traits.
```{r}
my.cat.cor<-function(fpkm.matrix.clean,feature.name="GENDER",method.name="wilcox",tissue.name="BAL",clinic.matrix,outputdir){

# extract the fpkm matrix for the given tissue
temp.names<-intersect(colnames(fpkm.matrix.clean),rownames(clinic.matrix))
data.matrix<-fpkm.matrix.clean[,temp.names]
clinic.matrix<-clinic.matrix[temp.names,]
bal.fpkm<-data.matrix

# generate the clinical matrix
my.clinic.matrix<-clinic.matrix[colnames(data.matrix),]	

clinic.vect<-as.numeric(as.factor(my.clinic.matrix[,colnames(my.clinic.matrix)==feature.name]))
if(length(table(clinic.vect))!=2){
cat("ERROR: the phenotype does not have two levels!\n")
return(1)
}

cat(feature.name,": baseline=",sort(unique(my.clinic.matrix[,colnames(my.clinic.matrix)==feature.name]),decreasing = F)[1],"\n",sep="")

# generate the correlation results
x1<-bal.fpkm[,!is.na(clinic.vect) & clinic.vect==as.numeric(names(table(clinic.vect))[1])]
x2<-bal.fpkm[,!is.na(clinic.vect) & clinic.vect==as.numeric(names(table(clinic.vect))[2])]

#age.pvalue<-apply(bal.fpkm,1,my.cor.test,y=clinic.vect,method.name=method.name)
age.pvalue.direct<-apply(cbind(x1,x2),1,my.wilcox,n=ncol(x1))
age.pvalue<-abs(age.pvalue.direct)

#age.pvalue<-matrix(unlist(age.pvalue),ncol=2,byrow=T)
age.pvalue<-cbind(age.pvalue,rep(NA,length(age.pvalue)))
colnames(age.pvalue)<-c("pvalue","corr_coef")
age.fdr<-p.adjust(age.pvalue[,"pvalue"],method="fdr")
age.pvalue<-cbind(age.pvalue,age.fdr)
colnames(age.pvalue)[ncol(age.pvalue)]<-"fdr"
rownames(age.pvalue)<-rownames(bal.fpkm)
temp.fold<-apply(x1,1,mean,na.rm=T)/apply(x2,1,mean,na.rm=T)
age.pvalue<-cbind(age.pvalue,temp.fold)
colnames(age.pvalue)[ncol(age.pvalue)]<-"fold_change"

age.pvalue<-cbind(age.pvalue,age.pvalue.direct)
colnames(age.pvalue)[ncol(age.pvalue)]<-"pval_dir"


fpkm.info<-apply(bal.fpkm>0,1,sum)
fpkm.info<-cbind(fpkm.info,apply(bal.fpkm>0.1,1,sum))
fpkm.info<-cbind(fpkm.info,apply(bal.fpkm>1,1,sum))
fpkm.info<-cbind(fpkm.info,apply(bal.fpkm,1,median))
fpkm.info<-cbind(fpkm.info,apply(bal.fpkm,1,mean))
colnames(fpkm.info)<-c("fpkm>0_num","fpkm>0.1_num","fpkm>1_num","fpkm_median","fpkm_mean")


if(file.exists(outputdir)==F){
cat("WARNING: output.dir doesn't exists!\n")
return(1)
}

cmd.out<-cbind(age.pvalue,fpkm.info)
output.filepath<-file.path(outputdir,paste("correlation_",method.name,"_",feature.name,"_allgenes_",tissue.name,".txt",sep=""))
cat("gene_names\t",file=output.filepath,append=F)
write.table(cmd.out,file=output.filepath,append=T,sep="\t",row.names=T,col.names=T,quote=F)


if(sum(!is.na(age.pvalue[,"fdr"]) & age.pvalue[,"fdr"]<0.05)==0){
  # When there's nothing that pass the FDR threshold, we use the threshold of p value<0.05 & fold change >2
  sig.genes<-rownames(data.matrix)[!is.na(age.pvalue[,"pvalue"]) & age.pvalue[,"pvalue"]<0.05 & (!is.na(age.pvalue[,"fold_change"])) & (age.pvalue[,"fold_change"]>2 | age.pvalue[,"fold_change"]< 1/2) ]

  cmd.out<-cbind(age.pvalue,fpkm.info)
  output.filepath<-file.path(outputdir,paste("correlation_",method.name,"_",feature.name,"_pvalue0.05_FC2.0_",tissue.name,".txt",sep=""))
  cat("gene_names\t",file=output.filepath,append=F)
  write.table(cmd.out[sig.genes,],file=output.filepath,append=T,sep="\t",row.names=T,col.names=T,quote=F)

  output.filepath<-file.path(outputdir,paste("correlation_",method.name,"_",feature.name,"_pvalue0.05_FC2.0_",tissue.name,"_both_metacore.txt",sep=""))
  cat("gene_names\tintensity\n",file=output.filepath,append=F)
  sig.genes<-rownames(data.matrix)[!is.na(age.pvalue[,"pvalue"]) & age.pvalue[,"pvalue"]<0.05 & (!is.na(age.pvalue[,"fold_change"])) & (age.pvalue[,"fold_change"]>2 | age.pvalue[,"fold_change"]< 1/2) ]
  sig.genes<-unname(sapply(sig.genes,my.element.extract,splitchar="_",index=1))
  sig.genes<-unique(sig.genes)
  cmd.out<-cbind(sig.genes,rep(1,length(sig.genes)))
  write.table(cmd.out,file=output.filepath,append=T,sep="\t",row.names=F,col.names=F,quote=F)
  
  output.filepath<-file.path(outputdir,paste("correlation_",method.name,"_",feature.name,"_pvalue0.05_FC2.0_",tissue.name,"_positive_metacore.txt",sep=""))
  cat("gene_names\tintensity\n",file=output.filepath,append=F)
  sig.genes<-rownames(data.matrix)[!is.na(age.pvalue[,"pvalue"]) & age.pvalue[,"pvalue"]<0.05 & (!is.na(age.pvalue[,"fold_change"])) & (age.pvalue[,"fold_change"]>2 | age.pvalue[,"fold_change"]< 1/2) ]
  sig.genes<-sig.genes[age.pvalue[sig.genes,"pval_dir"]>0]
  sig.genes<-unname(sapply(sig.genes,my.element.extract,splitchar="_",index=1))
  sig.genes<-unique(sig.genes)
  cmd.out<-cbind(sig.genes,rep(1,length(sig.genes)))
  write.table(cmd.out,file=output.filepath,append=T,sep="\t",row.names=F,col.names=F,quote=F)
  
  
  output.filepath<-file.path(outputdir,paste("correlation_",method.name,"_",feature.name,"_pvalue0.05_FC2.0_",tissue.name,"_negative_metacore.txt",sep=""))
  cat("gene_names\tintensity\n",file=output.filepath,append=F)
  sig.genes<-rownames(data.matrix)[!is.na(age.pvalue[,"pvalue"]) & age.pvalue[,"pvalue"]<0.05 & (!is.na(age.pvalue[,"fold_change"])) & (age.pvalue[,"fold_change"]>2 | age.pvalue[,"fold_change"]< 1/2) ]
  sig.genes<-sig.genes[age.pvalue[sig.genes,"pval_dir"]<0]
  sig.genes<-unname(sapply(sig.genes,my.element.extract,splitchar="_",index=1))
  sig.genes<-unique(sig.genes)
  cmd.out<-cbind(sig.genes,rep(1,length(sig.genes)))
  write.table(cmd.out,file=output.filepath,append=T,sep="\t",row.names=F,col.names=F,quote=F)  
  
}else{
  # if there are anything significant with FDR<0.05
  cmd.out<-cbind(age.pvalue,fpkm.info)
  output.filepath<-file.path(outputdir,paste("correlation_",method.name,"_",feature.name,"_fdr0.05_",tissue.name,".txt",sep=""))
  cat("gene_names\t",file=output.filepath,append=F)
  write.table(cmd.out[!is.na(age.pvalue[,"fdr"]) & age.pvalue[,"fdr"]<0.05,],file=output.filepath,append=T,sep="\t",row.names=T,col.names=T,quote=F)

  output.filepath<-file.path(outputdir,paste("correlation_",method.name,"_",feature.name,"_fdr0.05_",tissue.name,"_both_metacore.txt",sep=""))
  cat("gene_names\tintensity\n",file=output.filepath,append=F)
  sig.genes<-rownames(data.matrix)[!is.na(age.pvalue[,"fdr"]) & age.pvalue[,"fdr"]<0.05]
  sig.genes<-unname(sapply(sig.genes,my.element.extract,splitchar="_",index=1))
  sig.genes<-unique(sig.genes)
  cmd.out<-cbind(sig.genes,rep(1,length(sig.genes)))
  write.table(cmd.out,file=output.filepath,append=T,sep="\t",row.names=F,col.names=F,quote=F)
  
  
  output.filepath<-file.path(outputdir,paste("correlation_",method.name,"_",feature.name,"_fdr0.05_",tissue.name,"_positive_metacore.txt",sep=""))
  cat("gene_names\tintensity\n",file=output.filepath,append=F)
  sig.genes<-rownames(data.matrix)[!is.na(age.pvalue[,"fdr"]) & age.pvalue[,"fdr"]<0.05 & age.pvalue[,"pval_dir"]>0]
  sig.genes<-unname(sapply(sig.genes,my.element.extract,splitchar="_",index=1))
  sig.genes<-unique(sig.genes)
  cmd.out<-cbind(sig.genes,rep(1,length(sig.genes)))
  write.table(cmd.out,file=output.filepath,append=T,sep="\t",row.names=F,col.names=F,quote=F)
  
  
  output.filepath<-file.path(outputdir,paste("correlation_",method.name,"_",feature.name,"_fdr0.05_",tissue.name,"_negative_metacore.txt",sep=""))
  cat("gene_names\tintensity\n",file=output.filepath,append=F)
  sig.genes<-rownames(data.matrix)[!is.na(age.pvalue[,"fdr"]) & age.pvalue[,"fdr"]<0.05 & age.pvalue[,"pval_dir"]<0]
  sig.genes<-unname(sapply(sig.genes,my.element.extract,splitchar="_",index=1))
  sig.genes<-unique(sig.genes)
  cmd.out<-cbind(sig.genes,rep(1,length(sig.genes)))
  write.table(cmd.out,file=output.filepath,append=T,sep="\t",row.names=F,col.names=F,quote=F)
}
return(0)
}
```

* Apply the function to binary traits.
```{r}
#output.dir<-"/home/fas/kaminski/am2952/scratch/SARC209clean/Results_summary/Demographics_correlated_genes"

##Female n=112 Male n=97 Female=0 Male=1 Female is the first group so the positive and negative correlated are for F
##White n=153 Black n=49 White=1 Black=2 White is the first group so the positive and negative correlated are for W
output.dir<-"/home/yanxiting/Documents/Research/GRADS_SARC_PBMC/Demographics_correlated_genes"
if(file.exists(output.dir)==F){
  dir.create(output.dir)
}

my.cat.cor(fpkm.matrix.clean,feature.name="GENDER",tissue.name="PBMC",clinic.matrix=my.clinic.matrix,outputdir=output.dir)
my.cat.cor(fpkm.matrix.clean,feature.name="RACE",tissue.name="PBMC",clinic.matrix=my.clinic.matrix[my.clinic.matrix[,"RACE"]%in%c(0,1),],outputdir=output.dir)

# Stage II-III treatment effect
output.dir<-"/home/yanxiting/Documents/Research/GRADS_SARC_PBMC/Treatment_responsive_genes"
if(file.exists(output.dir)==F){
  dir.create(output.dir)
}

# stage II-III treatment
output.subdir<-file.path(output.dir,"Stage II-III")
if(file.exists(output.subdir)==F){
  dir.create(output.subdir)
}

#my.cat.cor(fpkm.matrix.clean,feature.name="PHENGRP",tissue.name="PBMC",clinic.matrix=my.clinic.matrix[my.clinic.matrix[,"PHENGRP"]%in%c(3,4),],outputdir=output.subdir)
my.cat.cor(fpkm.matrix.clean,feature.name="PHENGRP",tissue.name="PBMC",clinic.matrix=my.clinic.matrix[my.clinic.matrix[,"PHENGRP"]%in%c("Stage II-III, untreated","Stage II-III, treated"),],outputdir=output.subdir)

# stage IV treatment
output.subdir<-file.path(output.dir,"Stage IV")
if(file.exists(output.subdir)==F){
  dir.create(output.subdir)
}

#my.cat.cor(fpkm.matrix.clean,feature.name="PHENGRP",tissue.name="PBMC",clinic.matrix=my.clinic.matrix[my.clinic.matrix[,"PHENGRP"]%in%c(5,6),],outputdir=output.subdir)
my.cat.cor(fpkm.matrix.clean,feature.name="PHENGRP",tissue.name="PBMC",clinic.matrix=my.clinic.matrix[my.clinic.matrix[,"PHENGRP"]%in%c("Stage IV, untreated","Stage IV, treated"),],outputdir=output.subdir)

# pairwise comparison between different phenotypic groups
output.dir<-"/home/yanxiting/Documents/Research/GRADS_SARC_PBMC/Phenotype_DEGs"
if(file.exists(output.dir)==F){
  dir.create(output.dir)
}

other.pheno<-1:8
other.pheno<-setdiff(other.pheno,2)
other.pheno.names<-c(
"Multi-organ",
"Non-acute, Stage I, untreated",
"Stage II-III, treated",
"Stage II-III, untreated",
"Stage IV, treated",
"Stage IV, untreated",
"Acute Sarcoidosis, untreated",
"Remitting, untreated")
other.pheno.names<-other.pheno.names[-2]

for(i in 1:length(other.pheno)){
output.subdir<-file.path(output.dir,other.pheno.names[i])
if(file.exists(output.subdir)==F){
  dir.create(output.subdir)
}

my.cat.cor(fpkm.matrix.clean,feature.name="PHENGRP",tissue.name="PBMC",clinic.matrix=my.clinic.matrix[my.clinic.matrix[,"PHENGRP"]%in%c(other.pheno.names[i],"Non-acute, Stage I, untreated"),],outputdir=output.subdir)
}



# scadding stage pairwise comparison
output.dir<-"/home/yanxiting/Documents/Research/GRADS_SARC_PBMC/SCADDING_CAT_DEGs"

if(file.exists(output.dir)==F){
dir.create(output.dir)
}

other.pheno<-c("I","II","III","IV")
other.pheno.names<-paste("SCADDING_",other.pheno,sep="")

for(i in 1:length(other.pheno)){
output.subdir<-file.path(output.dir,other.pheno.names[i])
if(file.exists(output.subdir)==F){
  dir.create(output.subdir)
}

my.cat.cor(fpkm.matrix.clean,feature.name="SCADDING",tissue.name="PBMC",clinic.matrix=my.clinic.matrix[my.clinic.matrix[,"SCADDING"]%in%c(other.pheno[i],"0"),],outputdir=output.subdir)
}

#output.dir<-file.path(work.dir,"scratch/Antun/SUPERvisedA/Results_summary2/PFTs_correlated_genes")
#my.cat.cor(fpkm.matrix.clean,feature.name="FEV1FVCratioBinary",tissue.name="BAL",samplelist=sample.list.all,clinicmatrix=clinic.matrix,outputdir=output.dir)
#my.cat.cor(fpkm.matrix.clean,feature.name="obstructiveBinary",tissue.name="BAL",samplelist=sample.list.all,clinicmatrix=clinic.matrix,outputdir=output.dir)
#my.cat.cor(fpkm.matrix.clean,feature.name="restrictiveBinary",tissue.name="BAL",samplelist=sample.list.all,clinicmatrix=clinic.matrix,outputdir=output.dir)
```

### CT scan reads
* Identify genes associated with CT scan reads.
```{r}
################################
# 4.12. Identify genes associated with the CT scan variables

# take an average between the two slides for all the cell differentials
feature.names<-c(
"Med_Lymphadenopathy",
"Hilar_Lymphadenopathy",
"Micronodule",
"Bronchial_Wall_Thickening",
"Traction_Bronchiectasis",
"Bronchiectasis_severity",
"Ground_Glass",
"Honeycombing",
"Reticular_Abnormality",
"Pulmonary_Art",
"Tree_in_bud"
)

ct.matrix<-my.clinic.matrix[,feature.names]

################
# change the coding of the CT variables

# Med_Lymphadenopathy: 0=no, 1=left, 2=right, 3=bilateral
# v1: yes vs no
# v2: bilateral vs one side
# v3: numeric 0=no, 1=one side, 2=bilateral
temp.vect<-ct.matrix[,"Med_Lymphadenopathy"]  # yes versus no and billateral versus one side
temp1<-rep(NA,nrow(ct.matrix))
temp2<-temp1
temp3<-temp1

temp1[temp.vect==0]<-0
temp1[temp.vect>0]<-1

temp2[temp.vect==1 | temp.vect==2]<-0
temp2[temp.vect==3]<-1

temp3[temp.vect==0]<-0
temp3[temp.vect==1 | temp.vect==2]<-1
temp3[temp.vect==3]<-2

ct.matrix<-cbind(ct.matrix,temp1,temp2,temp3)
colnames(ct.matrix)[(ncol(ct.matrix)-2):ncol(ct.matrix)]<-c("Med_Lymphadenopathy_yes_no","Med_Lymphadenopathy_oneside_bi","Med_Lymphadenopathy_numeric")


# Hilar_Lymphadenopathy: 0=no, 1=left, 2=right, 3=bilateral
# v1: yes vs no
# v2: bilateral vs one side
# v3: numeric 0=no, 1=one side, 2=bilateral
temp.vect<-ct.matrix[,"Hilar_Lymphadenopathy"]  # yes versus no and billateral versus one side
temp1<-rep(NA,nrow(ct.matrix))
temp2<-temp1
temp3<-temp1

temp1[temp.vect==0]<-0
temp1[temp.vect>0]<-1

temp2[temp.vect==1 | temp.vect==2]<-0
temp2[temp.vect==3]<-1

temp3[temp.vect==0]<-0
temp3[temp.vect==1 | temp.vect==2]<-1
temp3[temp.vect==3]<-2

ct.matrix<-cbind(ct.matrix,temp1,temp2,temp3)
colnames(ct.matrix)[(ncol(ct.matrix)-2):ncol(ct.matrix)]<-c("Hilar_Lymphadenopathy_yes_no","Hilar_Lymphadenopathy_oneside_bi","Hilar_Lymphadenopathy_numeric")

# for original CT reads
output.dir<-"/home/yanxiting/Documents/Research/GRADS_SARC_PBMC/CTVariables_correlated_genes"

if(file.exists(output.dir)==F){
dir.create(output.dir)
}

my.con.cor(fpkm.matrix.clean=fpkm.matrix.clean,feature.name="Med_Lymphadenopathy",method.name="spearman",tissue.name="PBMC",clinic.matri=ct.matrix,output.dir=output.dir)
my.con.cor(fpkm.matrix.clean=fpkm.matrix.clean,feature.name="Hilar_Lymphadenopathy",method.name="spearman",tissue.name="PBMC",clinic.matri=ct.matrix,output.dir=output.dir)
my.con.cor(fpkm.matrix.clean=fpkm.matrix.clean,feature.name="Bronchial_Wall_Thickening",method.name="spearman",tissue.name="PBMC",clinic.matri=ct.matrix,output.dir=output.dir)
my.con.cor(fpkm.matrix.clean=fpkm.matrix.clean,feature.name="Traction_Bronchiectasis",method.name="spearman",tissue.name="PBMC",clinic.matri=ct.matrix,output.dir=output.dir)
my.con.cor(fpkm.matrix.clean=fpkm.matrix.clean,feature.name="Bronchiectasis_severity",method.name="spearman",tissue.name="PBMC",clinic.matri=ct.matrix,output.dir=output.dir)
my.con.cor(fpkm.matrix.clean=fpkm.matrix.clean,feature.name="Ground_Glass",method.name="spearman",tissue.name="PBMC",clinic.matri=ct.matrix,output.dir=output.dir)
my.con.cor(fpkm.matrix.clean=fpkm.matrix.clean,feature.name="Honeycombing",method.name="spearman",tissue.name="PBMC",clinic.matri=ct.matrix,output.dir=output.dir)
my.con.cor(fpkm.matrix.clean=fpkm.matrix.clean,feature.name="Reticular_Abnormality",method.name="spearman",tissue.name="PBMC",clinic.matri=ct.matrix,output.dir=output.dir)
my.con.cor(fpkm.matrix.clean=fpkm.matrix.clean,feature.name="Med_Lymphadenopathy_numeric",method.name="spearman",tissue.name="PBMC",clinic.matri=ct.matrix,output.dir=output.dir)
my.con.cor(fpkm.matrix.clean=fpkm.matrix.clean,feature.name="Hilar_Lymphadenopathy_numeric",method.name="spearman",tissue.name="PBMC",clinic.matri=ct.matrix,output.dir=output.dir)

# for the binary CT variables

output.dir<-"/home/yanxiting/Documents/Research/GRADS_SARC_PBMC/CTVariables_correlated_genes"
if(file.exists(output.dir)==F){
dir.create(output.dir)
}

cat("Hilar_Lymphadenopathy_yes_no ([pheno value] [# of obs]):\n")
write.table(table(ct.matrix$Hilar_Lymphadenopathy_yes_no),row.names=F,col.names = F,quote=F)
cat("\n",sep="")
my.cat.cor(fpkm.matrix.clean=fpkm.matrix.clean,feature.name="Hilar_Lymphadenopathy_yes_no",tissue.name="PBMC",clinic.matrix=ct.matrix,outputdir=output.dir)

cat("Hilar_Lymphadenopathy_oneside_bi ([pheno value] [# of obs]):\n")
write.table(table(ct.matrix$Hilar_Lymphadenopathy_oneside_bi),row.names=F,col.names = F,quote=F)
cat("\n",sep="")
my.cat.cor(fpkm.matrix.clean=fpkm.matrix.clean,feature.name="Hilar_Lymphadenopathy_oneside_bi",tissue.name="PBMC",clinic.matrix=ct.matrix,outputdir=output.dir)

cat("Med_Lymphadenopathy_yes_no ([pheno value] [# of obs]):\n")
write.table(table(ct.matrix$Med_Lymphadenopathy_yes_no),row.names=F,col.names = F,quote=F)
cat("\n",sep="")
my.cat.cor(fpkm.matrix.clean=fpkm.matrix.clean,feature.name="Med_Lymphadenopathy_yes_no",tissue.name="PBMC",clinic.matrix=ct.matrix,outputdir=output.dir)

cat("Med_Lymphadenopathy_oneside_bi ([pheno value] [# of obs]):\n")
write.table(table(ct.matrix$Med_Lymphadenopathy_oneside_bi),row.names=F,col.names = F,quote=F)
cat("\n",sep="")
my.cat.cor(fpkm.matrix.clean=fpkm.matrix.clean,feature.name="Med_Lymphadenopathy_oneside_bi",tissue.name="PBMC",clinic.matrix=ct.matrix,outputdir=output.dir)

cat("Pulmonary_Art ([pheno value] [# of obs]):\n")
write.table(table(ct.matrix$Pulmonary_Art),row.names=F,col.names = F,quote=F)
cat("\n",sep="")
my.cat.cor(fpkm.matrix.clean=fpkm.matrix.clean,feature.name="Pulmonary_Art",tissue.name="PBMC",clinic.matrix=ct.matrix,outputdir=output.dir)

cat("Tree_in_bud ([pheno value] [# of obs]):\n")
write.table(table(ct.matrix$Tree_in_bud),row.names=F,col.names = F,quote=F)
cat("\n",sep="")
my.cat.cor(fpkm.matrix.clean=fpkm.matrix.clean,feature.name="Tree_in_bud",tissue.name="PBMC",clinic.matrix=ct.matrix,outputdir=output.dir)

cat("Micronodule ([pheno value] [# of obs]):\n")
write.table(table(ct.matrix$Micronodule),row.names=F,col.names = F,quote=F)
cat("\n",sep="")
my.cat.cor(fpkm.matrix.clean=fpkm.matrix.clean,feature.name="Micronodule",tissue.name="PBMC",clinic.matrix=ct.matrix,outputdir=output.dir)
```

### Recoded Micronodule and Lymphadenopathy
We recode the micronodule and lymphadenopathy to get a redefined and likely more accurate phenotype definition. 

#### DEGs
We consider this phenotype as a categorical variable and identify differentially expressed genes between these four groups using the none group as baseline group.

* First, define the function to do the test.
```{r}
my.cat.cor<-function(fpkm.matrix.clean,feature.name="GENDER",method.name="wilcox",tissue.name="BAL",clinic.matrix,outputdir){

# extract the fpkm matrix for the given tissue
temp.names<-intersect(colnames(fpkm.matrix.clean),rownames(clinic.matrix))
data.matrix<-fpkm.matrix.clean[,temp.names]
clinic.matrix<-clinic.matrix[temp.names,]
bal.fpkm<-data.matrix

# generate the clinical matrix
my.clinic.matrix<-clinic.matrix[colnames(data.matrix),]	

clinic.vect<-as.numeric(as.factor(my.clinic.matrix[,colnames(my.clinic.matrix)==feature.name]))
if(length(table(clinic.vect))!=2){
cat("ERROR: the phenotype does not have two levels!\n")
return(1)
}

cat(feature.name,": baseline=",sort(unique(my.clinic.matrix[,colnames(my.clinic.matrix)==feature.name]),decreasing = F)[1],"\n",sep="")

# generate the correlation results
x1<-bal.fpkm[,!is.na(clinic.vect) & clinic.vect==as.numeric(names(table(clinic.vect))[1])]
x2<-bal.fpkm[,!is.na(clinic.vect) & clinic.vect==as.numeric(names(table(clinic.vect))[2])]

#age.pvalue<-apply(bal.fpkm,1,my.cor.test,y=clinic.vect,method.name=method.name)
age.pvalue.direct<-apply(cbind(x1,x2),1,my.wilcox,n=ncol(x1))
age.pvalue<-abs(age.pvalue.direct)

#age.pvalue<-matrix(unlist(age.pvalue),ncol=2,byrow=T)
age.pvalue<-cbind(age.pvalue,rep(NA,length(age.pvalue)))
colnames(age.pvalue)<-c("pvalue","corr_coef")
age.fdr<-p.adjust(age.pvalue[,"pvalue"],method="fdr")
age.pvalue<-cbind(age.pvalue,age.fdr)
colnames(age.pvalue)[ncol(age.pvalue)]<-"fdr"
rownames(age.pvalue)<-rownames(bal.fpkm)
temp.fold<-apply(x1,1,mean,na.rm=T)/apply(x2,1,mean,na.rm=T)
age.pvalue<-cbind(age.pvalue,temp.fold)
colnames(age.pvalue)[ncol(age.pvalue)]<-"fold_change"

age.pvalue<-cbind(age.pvalue,age.pvalue.direct)
colnames(age.pvalue)[ncol(age.pvalue)]<-"pval_dir"


fpkm.info<-apply(bal.fpkm>0,1,sum)
fpkm.info<-cbind(fpkm.info,apply(bal.fpkm>0.1,1,sum))
fpkm.info<-cbind(fpkm.info,apply(bal.fpkm>1,1,sum))
fpkm.info<-cbind(fpkm.info,apply(bal.fpkm,1,median))
fpkm.info<-cbind(fpkm.info,apply(bal.fpkm,1,mean))
colnames(fpkm.info)<-c("fpkm>0_num","fpkm>0.1_num","fpkm>1_num","fpkm_median","fpkm_mean")


if(file.exists(outputdir)==F){
cat("WARNING: output.dir doesn't exists!\n")
return(1)
}

cmd.out<-cbind(age.pvalue,fpkm.info)
output.filepath<-file.path(outputdir,paste("correlation_",method.name,"_",feature.name,"_allgenes_",tissue.name,".txt",sep=""))
cat("gene_names\t",file=output.filepath,append=F)
write.table(cmd.out,file=output.filepath,append=T,sep="\t",row.names=T,col.names=T,quote=F)


if(sum(!is.na(age.pvalue[,"fdr"]) & age.pvalue[,"fdr"]<0.05)==0){
  # When there's nothing that pass the FDR threshold, we use the threshold of p value<0.05 & fold change >2
  sig.genes<-rownames(data.matrix)[!is.na(age.pvalue[,"pvalue"]) & age.pvalue[,"pvalue"]<0.05 & (!is.na(age.pvalue[,"fold_change"])) & (age.pvalue[,"fold_change"]>2 | age.pvalue[,"fold_change"]< 1/2) ]

  cmd.out<-cbind(age.pvalue,fpkm.info)
  output.filepath<-file.path(outputdir,paste("correlation_",method.name,"_",feature.name,"_pvalue0.05_FC2.0_",tissue.name,".txt",sep=""))
  cat("gene_names\t",file=output.filepath,append=F)
  write.table(cmd.out[sig.genes,],file=output.filepath,append=T,sep="\t",row.names=T,col.names=T,quote=F)

  output.filepath<-file.path(outputdir,paste("correlation_",method.name,"_",feature.name,"_pvalue0.05_FC2.0_",tissue.name,"_both_metacore.txt",sep=""))
  cat("gene_names\tintensity\n",file=output.filepath,append=F)
  sig.genes<-rownames(data.matrix)[!is.na(age.pvalue[,"pvalue"]) & age.pvalue[,"pvalue"]<0.05 & (!is.na(age.pvalue[,"fold_change"])) & (age.pvalue[,"fold_change"]>2 | age.pvalue[,"fold_change"]< 1/2) ]
  sig.genes<-unname(sapply(sig.genes,my.element.extract,splitchar="_",index=1))
  sig.genes<-unique(sig.genes)
  cmd.out<-cbind(sig.genes,rep(1,length(sig.genes)))
  write.table(cmd.out,file=output.filepath,append=T,sep="\t",row.names=F,col.names=F,quote=F)
  
  output.filepath<-file.path(outputdir,paste("correlation_",method.name,"_",feature.name,"_pvalue0.05_FC2.0_",tissue.name,"_positive_metacore.txt",sep=""))
  cat("gene_names\tintensity\n",file=output.filepath,append=F)
  sig.genes<-rownames(data.matrix)[!is.na(age.pvalue[,"pvalue"]) & age.pvalue[,"pvalue"]<0.05 & (!is.na(age.pvalue[,"fold_change"])) & (age.pvalue[,"fold_change"]>2 | age.pvalue[,"fold_change"]< 1/2) ]
  sig.genes<-sig.genes[age.pvalue[sig.genes,"pval_dir"]>0]
  sig.genes<-unname(sapply(sig.genes,my.element.extract,splitchar="_",index=1))
  sig.genes<-unique(sig.genes)
  cmd.out<-cbind(sig.genes,rep(1,length(sig.genes)))
  write.table(cmd.out,file=output.filepath,append=T,sep="\t",row.names=F,col.names=F,quote=F)
  
  
  output.filepath<-file.path(outputdir,paste("correlation_",method.name,"_",feature.name,"_pvalue0.05_FC2.0_",tissue.name,"_negative_metacore.txt",sep=""))
  cat("gene_names\tintensity\n",file=output.filepath,append=F)
  sig.genes<-rownames(data.matrix)[!is.na(age.pvalue[,"pvalue"]) & age.pvalue[,"pvalue"]<0.05 & (!is.na(age.pvalue[,"fold_change"])) & (age.pvalue[,"fold_change"]>2 | age.pvalue[,"fold_change"]< 1/2) ]
  sig.genes<-sig.genes[age.pvalue[sig.genes,"pval_dir"]<0]
  sig.genes<-unname(sapply(sig.genes,my.element.extract,splitchar="_",index=1))
  sig.genes<-unique(sig.genes)
  cmd.out<-cbind(sig.genes,rep(1,length(sig.genes)))
  write.table(cmd.out,file=output.filepath,append=T,sep="\t",row.names=F,col.names=F,quote=F)  
  
}else{
  # if there are anything significant with FDR<0.05
  cmd.out<-cbind(age.pvalue,fpkm.info)
  output.filepath<-file.path(outputdir,paste("correlation_",method.name,"_",feature.name,"_fdr0.05_",tissue.name,".txt",sep=""))
  cat("gene_names\t",file=output.filepath,append=F)
  write.table(cmd.out[!is.na(age.pvalue[,"fdr"]) & age.pvalue[,"fdr"]<0.05,],file=output.filepath,append=T,sep="\t",row.names=T,col.names=T,quote=F)

  output.filepath<-file.path(outputdir,paste("correlation_",method.name,"_",feature.name,"_fdr0.05_",tissue.name,"_both_metacore.txt",sep=""))
  cat("gene_names\tintensity\n",file=output.filepath,append=F)
  sig.genes<-rownames(data.matrix)[!is.na(age.pvalue[,"fdr"]) & age.pvalue[,"fdr"]<0.05]
  sig.genes<-unname(sapply(sig.genes,my.element.extract,splitchar="_",index=1))
  sig.genes<-unique(sig.genes)
  cmd.out<-cbind(sig.genes,rep(1,length(sig.genes)))
  write.table(cmd.out,file=output.filepath,append=T,sep="\t",row.names=F,col.names=F,quote=F)
  
  
  output.filepath<-file.path(outputdir,paste("correlation_",method.name,"_",feature.name,"_fdr0.05_",tissue.name,"_positive_metacore.txt",sep=""))
  cat("gene_names\tintensity\n",file=output.filepath,append=F)
  sig.genes<-rownames(data.matrix)[!is.na(age.pvalue[,"fdr"]) & age.pvalue[,"fdr"]<0.05 & age.pvalue[,"pval_dir"]>0]
  sig.genes<-unname(sapply(sig.genes,my.element.extract,splitchar="_",index=1))
  sig.genes<-unique(sig.genes)
  cmd.out<-cbind(sig.genes,rep(1,length(sig.genes)))
  write.table(cmd.out,file=output.filepath,append=T,sep="\t",row.names=F,col.names=F,quote=F)
  
  
  output.filepath<-file.path(outputdir,paste("correlation_",method.name,"_",feature.name,"_fdr0.05_",tissue.name,"_negative_metacore.txt",sep=""))
  cat("gene_names\tintensity\n",file=output.filepath,append=F)
  sig.genes<-rownames(data.matrix)[!is.na(age.pvalue[,"fdr"]) & age.pvalue[,"fdr"]<0.05 & age.pvalue[,"pval_dir"]<0]
  sig.genes<-unname(sapply(sig.genes,my.element.extract,splitchar="_",index=1))
  sig.genes<-unique(sig.genes)
  cmd.out<-cbind(sig.genes,rep(1,length(sig.genes)))
  write.table(cmd.out,file=output.filepath,append=T,sep="\t",row.names=F,col.names=F,quote=F)
}
return(0)
}
```

* Second, load in the data.
```{r}

# load in the clinical data
clinic.filepath<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/data/clin_data_visit1_trunc_XY.csv")
clinic.matrix=read.csv(clinic.filepath,check.names=F,stringsAsFactors=F)
rownames(clinic.matrix)<-clinic.matrix[,1]

# load in the CT scan data.
ct.filepath<-file.path(home.dir,"scratch/GRADS/SARC_results/Data/CT_data_20170712/Sarc_ct_reads_corrected.xls")
ct.matrix<-read.xls(ct.filepath,as.is=TRUE)
rownames(ct.matrix)<-ct.matrix[,1]
ct.matrix<-ct.matrix[clinic.matrix$GRADSID,]
ct.matrix[,1]<-clinic.matrix$GRADSID
rownames(ct.matrix)<-ct.matrix[,1]
my.ct.matrix<-ct.matrix[,c("Micronodule","Micronodule_conglomerate","Med_Lymphadenopathy","Hilar_Lymphadenopathy")]
clinic.matrix<-cbind(clinic.matrix,ct.matrix[,c("Micronodule","Micronodule_conglomerate","Med_Lymphadenopathy","Hilar_Lymphadenopathy")],stringsAsFactors=F)

# generate the contingency table
micronodule.vect<-as.logical(my.ct.matrix[,"Micronodule"]) | as.logical(my.ct.matrix[,"Micronodule_conglomerate"])
temp1<-my.ct.matrix$Med_Lymphadenopathy
temp2<-my.ct.matrix$Hilar_Lymphadenopathy
temp1[!is.na(temp1) & temp1!=0]<-1
temp2[!is.na(temp2) & temp2!=0]<-1
lympha.vect<-as.logical(temp1) | as.logical(temp2)

cat("Here's a contingency table showing the overlap between presence of micronodule and lymphadenopathy:\n")
table(micronodule.vect,lympha.vect)
cat("\nHere's the breakdown of patients with no micronodule and no lymphadenopathy across the different SCADDING stage:\n")
table(clinic.matrix$SCADDING[(!micronodule.vect) & (!lympha.vect)])
cat("\n")

# define a new phenotype "nodule_lymph_pheno" based on these four groups
my.pheno<-rep("",nrow(clinic.matrix))
my.pheno[(!micronodule.vect) & (!lympha.vect)]<-"none"
my.pheno[micronodule.vect & (!lympha.vect)]<-"micronodule"
my.pheno[(!micronodule.vect) & lympha.vect]<-"lymph"
my.pheno[micronodule.vect & lympha.vect]<-"both"
my.pheno[my.pheno==""]<-NA
clinic.matrix<-cbind(clinic.matrix,my.pheno,stringsAsFactors=F)
colnames(clinic.matrix)[ncol(clinic.matrix)]<-"nodule_lymph_pheno"


# load in the gene expression data
# fpkm.filepath<-"/Users/yanxiting/MyVolumes/GRACE/scratch/GRADS/SARC_results/Results_summary_BAL_hg38/FPKM_matrix_hg38_Corrected_allGenes_BAL209_withAnnot.txt"
data.dir<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38")
fpkm.filepath<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/data","TPM_baseline_276_clean_GRADSID.txt")

# load in the fpkm matrix
fpkm.matrix<-read.table(fpkm.filepath,sep="\t",header=T,check.names=F,as.is=TRUE,comment.char = "")
fpkm.matrix.anno<-fpkm.matrix[,1:5]
fpkm.matrix<-fpkm.matrix[,6:ncol(fpkm.matrix)]
rownames(fpkm.matrix)<-paste(fpkm.matrix.anno[,"Gene_Id"],"_",fpkm.matrix.anno[,"Chr"],sep="")

```


* Apply the function to binary traits.
```{r}
#output.dir<-"/home/fas/kaminski/am2952/scratch/SARC209clean/Results_summary/Demographics_correlated_genes"

##Female n=112 Male n=97 Female=0 Male=1 Female is the first group so the positive and negative correlated are for F
##White n=153 Black n=49 White=1 Black=2 White is the first group so the positive and negative correlated are for W
source(file.path(home.dir,"Rprogram/my_functions.R"))
output.dir<-"/home/yanxiting/Documents/Research/GRADS_SARC_PBMC/Supervised_Analysis/NoduleLymphaPheno_DEGs"
if(file.exists(output.dir)==F){
  dir.create(output.dir)
}

# compare lymphadenopathy only to micronodule only
output.subdir<-file.path(output.dir,"lymph_micronodule_compare")
if(file.exists(output.subdir)==F){
  dir.create(output.subdir)
}

my.clinic.matrix<-clinic.matrix[colnames(fpkm.matrix),]
fpkm.matrix.clean<-cbind(fpkm.matrix[,(!is.na(as.character(my.clinic.matrix[,"nodule_lymph_pheno"]))) & as.character(my.clinic.matrix[,"nodule_lymph_pheno"])=="lymph"],fpkm.matrix[,(!is.na(as.character(my.clinic.matrix[,"nodule_lymph_pheno"]))) & as.character(my.clinic.matrix[,"nodule_lymph_pheno"])=="micronodule"])
my.clinic.matrix<-my.clinic.matrix[colnames(fpkm.matrix.clean),]
temp<-my.cat.cor(fpkm.matrix.clean,feature.name="nodule_lymph_pheno",tissue.name="PBMC",clinic.matrix=my.clinic.matrix,outputdir=output.dir)
output.filepath<-file.path(output.subdir,"readme.txt")
cat(temp,file=output.filepath,append=F)

# compare lymphadenopathy only to both
output.subdir<-file.path(output.dir,"both_lymph_compare")
if(file.exists(output.subdir)==F){
  dir.create(output.subdir)
}

my.clinic.matrix<-clinic.matrix[colnames(fpkm.matrix),]
fpkm.matrix.clean<-cbind(fpkm.matrix[,(!is.na(as.character(my.clinic.matrix[,"nodule_lymph_pheno"]))) & as.character(my.clinic.matrix[,"nodule_lymph_pheno"])=="lymph"],fpkm.matrix[,(!is.na(as.character(my.clinic.matrix[,"nodule_lymph_pheno"]))) & as.character(my.clinic.matrix[,"nodule_lymph_pheno"])=="both"])
my.clinic.matrix<-my.clinic.matrix[colnames(fpkm.matrix.clean),]
temp<-my.cat.cor(fpkm.matrix.clean,feature.name="nodule_lymph_pheno",tissue.name="PBMC",clinic.matrix=my.clinic.matrix,outputdir=output.dir)
output.filepath<-file.path(output.subdir,"readme.txt")
cat(temp,file=output.filepath,append=F)

# compare micronodule only to both
output.subdir<-file.path(output.dir,"both_micronodule_compare")
if(file.exists(output.subdir)==F){
  dir.create(output.subdir)
}

my.clinic.matrix<-clinic.matrix[colnames(fpkm.matrix),]
fpkm.matrix.clean<-cbind(fpkm.matrix[,(!is.na(as.character(my.clinic.matrix[,"nodule_lymph_pheno"]))) & as.character(my.clinic.matrix[,"nodule_lymph_pheno"])=="micronodule"],fpkm.matrix[,(!is.na(as.character(my.clinic.matrix[,"nodule_lymph_pheno"]))) & as.character(my.clinic.matrix[,"nodule_lymph_pheno"])=="both"])
my.clinic.matrix<-my.clinic.matrix[colnames(fpkm.matrix.clean),]
temp<-my.cat.cor(fpkm.matrix.clean,feature.name="nodule_lymph_pheno",tissue.name="PBMC",clinic.matrix=my.clinic.matrix,outputdir=output.dir)
output.filepath<-file.path(output.subdir,"readme.txt")
cat(temp,file=output.filepath,append=F)


```


### Overlap between different traits
* We show the number of genes significantly (FDR<0.05) associated with the different clinical traits.
```{r eval=TRUE, echo=FALSE, cache=TRUE,warning=FALSE,message=FALSE,results='asis'}
data.dir<-"/home/yanxiting/Documents/Research/GRADS_SARC_PBMC"

#data.dir<-"/home/antun/driver_Grace_xy48/scratch/Antun/SUPERvisedA/Results_summary2"
final.list<-list()
# load in the gene lists
var.names<-c("AGE","GENDER","RACE",
             "FVCPRED","BDFVC","FEV1PRED","BDFEV1","PREDDLCO","BDDLCO","FEV1FVCratio",
             "Bronchial_Wall_Thickening","Bronchiectasis_severity","Ground_Glass","Honeycombing","Reticular_Abnormality","Traction_Bronchiectasis","Med_Lymphadenopathy_yes_no","Hilar_Lymphadenopathy_yes_no","SCADDING","p_total","p_mono","p_eos","p_lymph","p_baso","p_neut","abs_lymph")
var.annot<-c("Age","Gender","Race",
             "FVC% PRED","FVC","FEV1% PRED","FEV1","DLCO% PRED","DLCO","FEV1/FVC ratio",
             "Bronchial Wall Thickening","Bronchiectasis Severity","Ground Glass","Honeycombing","Reticular Abnormality","Traction Bronchiestasis","Mediastinal Lymphadenopathy","Hilar Lymphadenopathy","SCADDING","Total Cell Counts","Monocyte %","Eosinophil %","Lymphocyte %","Basophil %","Neutrophil %","Absolute Lymphocyte Count")
var.cat<-c(rep("Demographics",3),
           rep("PFTs",7),
           rep("CTVariables",8),
           "Disease",
           rep("CellDifferentials",7))
for(i in 1:length(var.names)){
file.names<-list.files(file.path(data.dir,paste(var.cat[i],"_correlated_genes",sep="")))
data.filepath<-file.path(data.dir,paste(var.cat[i],"_correlated_genes",sep=""),file.names[grep(paste(var.names[i],"_allgenes_PBMC.txt",sep=""),file.names)])
if(length(data.filepath)>1){
  # For race and gender, there are two results: spearman and wilcox. We use the results by the Wilcoxon Rank Sum test.
  cat("There are ",length(data.filepath)," files:\n",sep="")
  cat(data.filepath,collapse="\n")
  data.filepath<-data.filepath[grep("wilcox",data.filepath)]
  cat("\nWe chose ",data.filepath,"\n",sep="")
}
#data.filepath<-file.path(data.dir,paste(var.cat[i],"_correlated_genes",sep=""),paste("correlation_spearman_",var.names[i],"_allgenes_BAL.txt",sep=""))
temp<-read.table(data.filepath,sep = "\t",header=T,check.names=F,row.names=1,comment.char="")
#  temp<-temp[!is.na(temp[,"fdr"]) & temp[,"fdr"]<0.05,]
final.list[[i]]<-temp
}
names(final.list)<-var.annot

# get all the gene names
name_genes<-character()
for(i in 1:length(final.list)){
temp<-final.list[[i]]

name_genes<-union(name_genes,rownames(temp)[!is.na(temp[,"fdr"]) & temp[,"fdr"]<0.05])
}

name_genes<-unname(sapply(name_genes,my.element.remove,splitchar="_",index=-1))
cat("\n There are in total ",length(name_genes)," significant genes by the supervised analysis on all the variables.\n",sep="")
```

```{r eval=TRUE,fig.width=10,fig.height=10,cache=TRUE,warning=FALSE,message=FALSE,results='asis'}
#http://www.sthda.com/english/wiki/ggplot2-quick-correlation-matrix-heatmap-r-software-and-data-visualization
# calculate the number of overlapping genes
output.dir<-"/home/yanxiting/Documents/Research/GRADS_SARC_PBMC/overlap"

if(file.exists(output.dir)==F){
  dir.create(output.dir)
}

overlap.n.matrix<-matrix(NA,nrow=length(final.list),ncol=length(final.list))
overlap.p.matrix<-matrix(NA,nrow=length(final.list),ncol=length(final.list))
overlap.n.matrix.pos<-overlap.n.matrix
overlap.n.matrix.neg<-overlap.n.matrix
overlap.n.matrix.directsummary<-overlap.n.matrix

for(i in 1:length(final.list)){
  for(j in 1:length(final.list)){
    if(i==j){
        overlap.n.matrix[i,j]<-sum((!is.na(final.list[[i]][,"fdr"])) & final.list[[i]][,"fdr"]<0.05)  

        overlap.n.matrix.pos[i,j]<-overlap.n.matrix[i,j]
        overlap.n.matrix.neg[i,j]<-0
      
        list.overlap<-rownames(final.list[[i]])[!is.na(final.list[[i]][,"fdr"]) & final.list[[i]][,"fdr"]<0.05]

      if("pval_dir"%in%colnames(final.list[[i]])){
        coef.1<-final.list[[i]][list.overlap,"pval_dir"]
      }else{
        coef.1<-final.list[[i]][list.overlap,"corr_coef"]
      }

      overlap.n.matrix.directsummary[i,j]<-paste(sum(coef.1>0)," +\n",sum(coef.1<0)," -",sep="")
      next
    }
    temp1<-final.list[[i]]
    temp2<-final.list[[j]]
    
    temp.names<-intersect(rownames(temp1),rownames(temp2))

    temp1<-temp1[temp.names,]
    temp2<-temp2[temp.names,]
    
    list1<-rownames(final.list[[i]])[!is.na(final.list[[i]][,"fdr"]) & final.list[[i]][,"fdr"]<0.05]
    list2<-rownames(final.list[[j]])[!is.na(final.list[[j]][,"fdr"]) & final.list[[j]][,"fdr"]<0.05]

    list.overlap<-intersect(list1,list2)
    
    if("pval_dir"%in%colnames(final.list[[i]])){
      coef.1<-final.list[[i]][list.overlap,"pval_dir"]
    }else{
      coef.1<-final.list[[i]][list.overlap,"corr_coef"]
    }

    if("pval_dir"%in%colnames(final.list[[j]])){
      coef.2<-final.list[[j]][list.overlap,"pval_dir"]
    }else{
      coef.2<-final.list[[j]][list.overlap,"corr_coef"]
    }
    
    list.overlap.pos<-list.overlap[coef.1 * coef.2 >0]
    list.overlap.neg<-setdiff(intersect(list1,list2),list.overlap.pos)
    
    if(length(list.overlap.pos)>length(list.overlap.neg)){
      list.1<-list.overlap[coef.1>0 & coef.2>0]
      list.2<-list.overlap[coef.1<0 & coef.2<0]
      overlap.n.matrix.directsummary[i,j]<-paste(length(list.1),"(++)\n",length(list.2),"(--)")
    }
    if(length(list.overlap.pos)==length(list.overlap.neg)){
      list.1<-character()
      list.2<-character()
      overlap.n.matrix.directsummary[i,j]<-paste("0(++)\n0(--)")
    }
    if(length(list.overlap.pos)<length(list.overlap.neg)){
      list.1<-list.overlap[coef.1>0 & coef.2<0]
      list.2<-list.overlap[coef.1<0 & coef.2>0]
      overlap.n.matrix.directsummary[i,j]<-paste(length(list.1),"(+-)\n",length(list.2),"(-+)")
    }
    
    pop.size<-nrow(temp1)
    success.n<-length(list1)
    sample.n<-length(list2)
    x<-length(intersect(list1,list2))
    
    overlap.n.matrix[i,j]<-length(intersect(list1,list2))
    overlap.p.matrix[i,j]<-1-phyper(x-1,success.n,pop.size-success.n,sample.n,lower.tail=TRUE)
    
    overlap.n.matrix.pos[i,j]<-length(list.overlap.pos)
    overlap.n.matrix.neg[i,j]<-length(list.overlap.neg)
    if(i>j){
    # output the list of overlap genes
    if(length(list.overlap)>0){
      output.filepath<-file.path(output.dir,paste(names(final.list)[i],"_VS_",names(final.list)[j],"_all.txt",sep=""))
      cmd.out<-final.list[[i]][list.overlap,]
      colnames(cmd.out)<-paste(colnames(cmd.out)," (",names(final.list)[i],")",sep="")
      cmd.out.1<-final.list[[j]][list.overlap,]
      colnames(cmd.out.1)<-paste(colnames(cmd.out.1)," (",names(final.list)[j],")",sep="")
      cmd.out<-cbind(cmd.out,cmd.out.1)
      cat("tracking_ID\t",file=output.filepath,append=F)
      write.table(cmd.out,file=output.filepath,append=T,sep="\t",row.names=T,col.names=T,quote=F)
    }
    # output the list of overlap genes with the same directions
    if(length(list.overlap.pos)>0){
      output.filepath<-file.path(output.dir,paste(names(final.list)[i],"_VS_",names(final.list)[j],"_samedirection.txt",sep=""))
      cmd.out<-final.list[[i]][list.overlap.pos,]
      colnames(cmd.out)<-paste(colnames(cmd.out)," (",names(final.list)[i],")",sep="")
      cmd.out.1<-final.list[[j]][list.overlap.pos,]
      colnames(cmd.out.1)<-paste(colnames(cmd.out.1)," (",names(final.list)[j],")",sep="")
      cmd.out<-cbind(cmd.out,cmd.out.1)
      cat("tracking_ID\t",file=output.filepath,append=F)
      write.table(cmd.out,file=output.filepath,append=T,sep="\t",row.names=T,col.names=T,quote=F)
    }
    # output the list of overlapped genes with opposite directions
     if(length(list.overlap.neg)>0){
      output.filepath<-file.path(output.dir,paste(names(final.list)[i],"_VS_",names(final.list)[j],"_oppositedirection.txt",sep=""))
      cmd.out<-final.list[[i]][list.overlap.neg,]
      colnames(cmd.out)<-paste(colnames(cmd.out)," (",names(final.list)[i],")",sep="")
      cmd.out.1<-final.list[[j]][list.overlap.neg,]
      colnames(cmd.out.1)<-paste(colnames(cmd.out.1)," (",names(final.list)[j],")",sep="")
      cmd.out<-cbind(cmd.out,cmd.out.1)
      cat("tracking_ID\t",file=output.filepath,append=F)
      write.table(cmd.out,file=output.filepath,append=T,sep="\t",row.names=T,col.names=T,quote=F)
     }
      
     if(length(list.1)>0){
       
       if(length(list.overlap.pos)>length(list.overlap.neg)){
          output.filepath<-file.path(output.dir,paste(names(final.list)[i],"_VS_",names(final.list)[j],"_pospos.txt",sep=""))
       }
       
       if(length(list.overlap.pos)<length(list.overlap.neg)){
          output.filepath<-file.path(output.dir,paste(names(final.list)[i],"_VS_",names(final.list)[j],"_posneg.txt",sep=""))
       }
        cmd.out<-final.list[[i]][list.1,]
        colnames(cmd.out)<-paste(colnames(cmd.out)," (",names(final.list)[i],")",sep="")
        cmd.out.1<-final.list[[j]][list.1,]
        colnames(cmd.out.1)<-paste(colnames(cmd.out.1)," (",names(final.list)[j],")",sep="")
        cmd.out<-cbind(cmd.out,cmd.out.1)
        cat("tracking_ID\t",file=output.filepath,append=F)
        write.table(cmd.out,file=output.filepath,append=T,sep="\t",row.names=T,col.names=T,quote=F)
     }
      
    if(length(list.2)>0){
       if(length(list.overlap.pos)>length(list.overlap.neg)){
          output.filepath<-file.path(output.dir,paste(names(final.list)[i],"_VS_",names(final.list)[j],"_negneg.txt",sep=""))
       }
       
       if(length(list.overlap.pos)<length(list.overlap.neg)){
          output.filepath<-file.path(output.dir,paste(names(final.list)[i],"_VS_",names(final.list)[j],"_negpos.txt",sep=""))
       }
          cmd.out<-final.list[[i]][list.2,]
          colnames(cmd.out)<-paste(colnames(cmd.out)," (",names(final.list)[i],")",sep="")
          cmd.out.1<-final.list[[j]][list.2,]
          colnames(cmd.out.1)<-paste(colnames(cmd.out.1)," (",names(final.list)[j],")",sep="")
          cmd.out<-cbind(cmd.out,cmd.out.1)
          cat("tracking_ID\t",file=output.filepath,append=F)
          write.table(cmd.out,file=output.filepath,append=T,sep="\t",row.names=T,col.names=T,quote=F)
    }
    }
  }
}


rownames(overlap.n.matrix)<-names(final.list)
colnames(overlap.n.matrix)<-names(final.list)

rownames(overlap.p.matrix)<-names(final.list)
colnames(overlap.p.matrix)<-names(final.list)

```


```{r eval=TRUE,fig.width=12,fig.height=12, echo=FALSE,cache=FALSE,results='asis'}
# Get lower triangle of the correlation matrix
get_lower_tri<-function(cormat){
  cormat[upper.tri(cormat)] <- NA
  return(cormat)
}
# Get upper triangle of the correlation matrix
get_upper_tri <- function(cormat){
  cormat[lower.tri(cormat)]<- NA
  return(cormat)
}

library(reshape2)
library(ggplot2)

# draw the overlap matrix using ggplot2
#c1<-signif(overlap.n.matrix, 2)
c1<-overlap.n.matrix
c2<-overlap.p.matrix
#c2<-signif(overlap.p.matrix, 2)
diag(c2)<-rep(0,length(diag(c2)))
upper_tri <- get_upper_tri(c1)
upper_tri_p<-get_upper_tri(c2)
melted_cormat_n <- melt(upper_tri, na.rm = TRUE)
melted_cormat_p<-melt(upper_tri_p, na.rm=TRUE)

melted_cormat<-melted_cormat_n
melted_cormat[,3]<-melted_cormat_p[,3]
melted_cormat<-cbind(melted_cormat,melted_cormat_n[,3])
colnames(melted_cormat)[4]<-"gnum"
melted_cormat[melted_cormat[,"value"]>=0.05,"value"]<-NA

g<-ggplot(data = melted_cormat, aes(Var2, Var1, fill = value))+
  geom_tile(color = "white")+
  scale_fill_gradient(low = "yellow", high = "white", na.value="lightgray", space = "Lab", name="# of overlapping genes") +
  #scale_fill_gradientn(colours = c("red", "white"),values=scales::rescale(c(0, 0.05, 1)), na.value="white", space = "Lab", name="# of overlapping genes") +
  theme_minimal()+ 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, size = 12, hjust = 1))+
  coord_fixed()
#melted_cormat[,3]<-melted_cormat_n[,3]
g<-g+
  geom_text(aes(Var2, Var1, label = gnum), color = "black", size = 4)
print(g)

output.subdir<-"/home/yanxiting/Documents/Research/GRADS_SARC_PBMC"
ggsave(filename = "Figure2a_SupervisedAnalysis_OverlapMatrix_withFEV1FVCratio.eps",path=output.subdir,width=12,height=12,device="eps")
```


Here we generate another figure that shows the breakdown of the genes based on the directionality of the correlation coefficients. Each entry will show the number of overlapping genes and the breakdown of these genes into genes with + and + or - and - correlation with the rows and columns if those two features are positively correlated. The breakdown will be the number of genes with + and - or - and + correlation with the row and column if those two features are negatively correlated.  
```{r fig.width=20,fig.height=20,eval=TRUE}
get_lower_tri<-function(cormat){
  cormat[upper.tri(cormat)] <- NA
  return(cormat)
}
# Get upper triangle of the correlation matrix
get_upper_tri <- function(cormat){
  cormat[lower.tri(cormat)]<- NA
  return(cormat)
}


# draw the overlap matrix using ggplot2
#c1<-signif(overlap.n.matrix, 2)
temp<-overlap.n.matrix.directsummary
temp[!is.na(overlap.p.matrix) & overlap.p.matrix>=0.05]<-""
temp[!is.na(overlap.p.matrix) & overlap.p.matrix<0.05]<-paste("\n",temp[!is.na(overlap.p.matrix) & overlap.p.matrix<0.05],sep="")
temp[is.na(overlap.p.matrix)]<-paste("\n",temp[is.na(overlap.p.matrix)],sep="")

#c1<-matrix(paste(overlap.n.matrix,"\n",overlap.n.matrix.directsummary,sep=""),nrow=nrow(overlap.n.matrix),byrow=F)
c1<-matrix(paste(overlap.n.matrix,temp,sep=""),nrow=nrow(overlap.n.matrix),byrow=F)
rownames(c1)<-rownames(overlap.n.matrix)
colnames(c1)<-rownames(overlap.n.matrix)
c2<-overlap.p.matrix
#c2<-signif(overlap.p.matrix, 2)
diag(c2)<-rep(0,length(diag(c2)))
upper_tri <- get_upper_tri(c1)
upper_tri_p<-get_upper_tri(c2)
melted_cormat_n <- melt(upper_tri, na.rm = TRUE)
melted_cormat_p<-melt(upper_tri_p, na.rm=TRUE)

melted_cormat<-melted_cormat_n
melted_cormat[,3]<-melted_cormat_p[,3]
melted_cormat<-cbind(melted_cormat,melted_cormat_n[,3])
colnames(melted_cormat)[4]<-"gnum"
melted_cormat[melted_cormat[,"value"]>=0.05,"value"]<-NA

g<-ggplot(data = melted_cormat, aes(Var2, Var1, fill = value))+
  geom_tile(color = "white")+
  scale_fill_gradient(low = "yellow", high = "white", na.value="lightgray", space = "Lab", name="# of overlapping genes") +
  #scale_fill_gradientn(colours = c("red", "white"),values=scales::rescale(c(0, 0.05, 1)), na.value="white", space = "Lab", name="# of overlapping genes") +
  theme_minimal()+ 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, size = 18, hjust = 1),axis.text.y = element_text( size = 18))+
  coord_fixed()
#melted_cormat[,3]<-melted_cormat_n[,3]
g<-g+
  geom_text(aes(Var2, Var1, label = gnum), color = "black", size = 4.5,lineheight = .8)
print(g)

#output.subdir<-file.path(output.dir,"scratch/GRADS/SARC_results/Results_summary_BAL_hg38/writings")
output.subdir<-"/home/yanxiting/Documents/Research/GRADS_SARC_PBMC"
ggsave(filename = "Figure2a_SupervisedAnalysis_OverlapMatrix_breakdown.eps",path=output.subdir,width=22,height=22,device="eps")
```







## DESeq2

### PCA

#### Unadjusted DESeq2 Normalized Data

First, we generate PCA plot using the unadjusted DESeq2 normalized data.
```{r}
library(xlsx)
home.dir<-"/home/yanxiting/driver_Grace"
my.distinct.colors<-c('#e6194b', '#3cb44b', '#ffe119', '#4363d8', '#f58231', '#911eb4', '#46f0f0', '#f032e6', '#bcf60c', '#fabebe', '#008080', '#e6beff', '#9a6324', '#fffac8', '#800000', '#aaffc3', '#808000', '#ffd8b1', '#000075', '#808080', '#ffffff', '#000000')

data.dir<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38")
fpkm.filepath<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/data","DESeq2_normalized_276_clean.txt")
masterkey.filepath<-file.path(home.dir,"scratch/GRADS/SARC_results/Data/GRADS Master Progress Key 6-13-16.xlsx")
clinic.filepath<-file.path(home.dir,"scratch/GRADS/SARC_results/Data_PBMC/clin_data_visit1_XY.csv")
output.dir<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/data_adjusted2")

# load in the fpkm matrix
fpkm.matrix<-read.table(fpkm.filepath,sep="\t",header=T,check.names=F,as.is=TRUE,comment.char="")
anno.matrix<-as.matrix(fpkm.matrix)[,1:6]
rownames(anno.matrix)<-anno.matrix[,"GeneID"]
rownames(fpkm.matrix)<-anno.matrix[,"GeneID"]
data.matrix<-fpkm.matrix[,7:ncol(fpkm.matrix)]

# load in the clinical data
clinic.matrix<-read.csv(clinic.filepath,check.names=F,header=T,as.is=TRUE)
rownames(clinic.matrix)<-as.matrix(clinic.matrix)[,1] # GRADS ID
clinic.matrix<-clinic.matrix[colnames(data.matrix),]

temp.data<-read.xlsx(masterkey.filepath,sheetIndex = 3, check.names=F, startRow=3)
rownames(temp.data)<-temp.data$`GRADS ID`
clinic.matrix<-cbind(clinic.matrix,as.matrix(temp.data)[rownames(clinic.matrix),"Phenotype/Genotype"])
colnames(clinic.matrix)[ncol(clinic.matrix)]<-"pheno_group"
gc(reset=TRUE)

data.matrix.orig<-data.matrix
clinic.matrix.orig<-clinic.matrix
```

We generate the 3-dimensional PCA plots.

```{r}
# We ran these codes on the workstation since it's easier to save and demonstrate the generated plots in Rstudio markdown files. 
###############################################################################################
# 1. identify outliers using pca plots from all 460 reactions
library(gdata)
library(rgl)
log2.transform=FALSE # do you want to log2 transform the data before PCA or not.
source(file.path(home.dir,"Rprogram","my_functions.R"))

#######################
# focus on some of the phenotypic groups
data.matrix<-data.matrix.orig[,!clinic.matrix.orig$pheno_group%in%c("Multi-organ","Cardiac defining therapy","Acute Sarcoidosis, untreated")]
clinic.matrix<-clinic.matrix.orig[!clinic.matrix.orig$pheno_group%in%c("Multi-organ","Cardiac defining therapy","Acute Sarcoidosis, untreated"),]

# Log2 transform
if(log2.transform){
data.matrix<-data.matrix+3e-208
data.matrix<-log(data.matrix,base=2)
}
# calculate the number of expressed genes for each sample
expgene.num<-apply(data.matrix>0,2,sum)
##################
# PCA analysis
library(gplots)
# calculate the SVD
temp.sd<-apply(data.matrix,1,sd)
data.matrix<-data.matrix[temp.sd>0,]  #23861 genes
data.matrix.norm<-my.normalize(data.matrix)
my.svd<-svd(data.matrix.norm)
cumsum(my.svd$d^2/sum(my.svd$d^2))[1:3]
my.projected<-t(data.matrix.norm)%*%my.svd$u 
my.colors<-factor(as.matrix(clinic.matrix)[,"pheno_group"],levels=c("Remitting, untreated","Non-acute, Stage I, untreated","Stage II-III, untreated","Stage II-III, treated","Stage IV, untreated","Stage IV, treated"))
my.colors<-my.distinct.colors[as.numeric(my.colors)]
open3d()
plot3d(my.projected[,1],my.projected[,2],my.projected[,3],xlab="PC 1",ylab="PC 2",zlab="PC 3",col=my.colors,type="s", size=1,box = F,axes = T)
par3d(windowRect = c(0, 0, 400, 400))
legend3d("topright", legend = c("Remitting, untreated","Non-acute, Stage I, untreated","Stage II-III, untreated","Stage II-III, treated","Stage IV, untreated","Stage IV, treated"), pch = 16, col = my.distinct.colors[1:length(unique(my.colors))], cex=0.65, inset=c(0))
rglwidget()
#filename<-writeWebGL(dir=file.path(pca.dir,"pca_3d"),reuse=TRUE)
# label the samples based on subjects and disease

#identify3d(my.projected[,1],my.projected[,2],my.projected[,3],labels=colnames(data.matrix.norm),n=ncol(data.matrix.norm),plot=TRUE)
# label samples using subject ID

```

No treated samples.

```{r}
# We ran these codes on the workstation since it's easier to save and demonstrate the generated plots in Rstudio markdown files. 
###############################################################################################
# 1. identify outliers using pca plots from all 460 reactions
library(gdata)
library(rgl)
log2.transform=FALSE # do you want to log2 transform the data before PCA or not.
source(file.path(home.dir,"Rprogram","my_functions.R"))

#######################
# focus on some of the phenotypic groups
data.matrix<-data.matrix.orig[,!clinic.matrix.orig$pheno_group%in%c("Multi-organ","Cardiac defining therapy","Acute Sarcoidosis, untreated","Stage IV, treated","Stage II-III, treated")]
clinic.matrix<-clinic.matrix.orig[!clinic.matrix.orig$pheno_group%in%c("Multi-organ","Cardiac defining therapy","Acute Sarcoidosis, untreated","Stage IV, treated","Stage II-III, treated"),]

# Log2 transform
if(log2.transform){
data.matrix<-data.matrix+3e-208
data.matrix<-log(data.matrix,base=2)
}
# calculate the number of expressed genes for each sample
expgene.num<-apply(data.matrix>0,2,sum)
##################
# PCA analysis
library(gplots)
# calculate the SVD
temp.sd<-apply(data.matrix,1,sd)
data.matrix<-data.matrix[temp.sd>0,]  #23861 genes
data.matrix.norm<-my.normalize(data.matrix)
my.svd<-svd(data.matrix.norm)
cumsum(my.svd$d^2/sum(my.svd$d^2))[1:3]
my.projected<-t(data.matrix.norm)%*%my.svd$u 
my.colors<-factor(as.matrix(clinic.matrix)[,"pheno_group"],levels=c("Remitting, untreated","Non-acute, Stage I, untreated","Stage II-III, untreated","Stage IV, untreated"))
my.colors<-my.distinct.colors[as.numeric(my.colors)]
open3d()
plot3d(my.projected[,1],my.projected[,2],my.projected[,3],xlab="PC 1",ylab="PC 2",zlab="PC 3",col=my.colors,type="s", size=1,box = F,axes = T)
par3d(windowRect = c(0, 0, 400, 400))
legend3d("topright", legend = c("Remitting, untreated","Non-acute, Stage I, untreated","Stage II-III, untreated","Stage IV, untreated"), pch = 16, col = my.distinct.colors[1:length(unique(my.colors))], cex=0.65, inset=c(0))
rglwidget()
#filename<-writeWebGL(dir=file.path(pca.dir,"pca_3d"),reuse=TRUE)
# label the samples based on subjects and disease

#identify3d(my.projected[,1],my.projected[,2],my.projected[,3],labels=colnames(data.matrix.norm),n=ncol(data.matrix.norm),plot=TRUE)
# label samples using subject ID

```

Stage 1 untreated vs others.

```{r}
# We ran these codes on the workstation since it's easier to save and demonstrate the generated plots in Rstudio markdown files. 
###############################################################################################
# 1. identify outliers using pca plots from all 460 reactions
library(gdata)
library(rgl)
log2.transform=FALSE # do you want to log2 transform the data before PCA or not.
source(file.path(home.dir,"Rprogram","my_functions.R"))

#######################
# focus on some of the phenotypic groups
data.matrix<-data.matrix.orig[,!clinic.matrix.orig$pheno_group%in%c("Multi-organ","Cardiac defining therapy","Acute Sarcoidosis, untreated","Stage IV, treated","Stage II-III, treated")]
clinic.matrix<-clinic.matrix.orig[!clinic.matrix.orig$pheno_group%in%c("Multi-organ","Cardiac defining therapy","Acute Sarcoidosis, untreated","Stage IV, treated","Stage II-III, treated"),]

temp.vect<-as.matrix(clinic.matrix)[,"pheno_group"]
temp.vect[temp.vect!="Non-acute, Stage I, untreated"]<-"Other"
clinic.matrix$pheno_group<-temp.vect
# Log2 transform
if(log2.transform){
data.matrix<-data.matrix+3e-208
data.matrix<-log(data.matrix,base=2)
}
# calculate the number of expressed genes for each sample
expgene.num<-apply(data.matrix>0,2,sum)
##################
# PCA analysis
library(gplots)
# calculate the SVD
temp.sd<-apply(data.matrix,1,sd)
data.matrix<-data.matrix[temp.sd>0,]  #23861 genes
data.matrix.norm<-my.normalize(data.matrix)
my.svd<-svd(data.matrix.norm)
cumsum(my.svd$d^2/sum(my.svd$d^2))[1:3]
my.projected<-t(data.matrix.norm)%*%my.svd$u 
my.colors<-factor(as.matrix(clinic.matrix)[,"pheno_group"],levels=c("Other","Non-acute, Stage I, untreated"))
my.colors<-my.distinct.colors[as.numeric(my.colors)]
open3d()
plot3d(my.projected[,1],my.projected[,2],my.projected[,3],xlab="PC 1",ylab="PC 2",zlab="PC 3",col=my.colors,type="s", size=1,box = F,axes = T)
par3d(windowRect = c(0, 0, 400, 400))
legend3d("topright", legend = c("Other","Non-acute, Stage I, untreated"), pch = 16, col = my.distinct.colors[1:length(unique(my.colors))], cex=0.65, inset=c(0))
rglwidget()
#filename<-writeWebGL(dir=file.path(pca.dir,"pca_3d"),reuse=TRUE)
# label the samples based on subjects and disease

#identify3d(my.projected[,1],my.projected[,2],my.projected[,3],labels=colnames(data.matrix.norm),n=ncol(data.matrix.norm),plot=TRUE)
# label samples using subject ID

```

#### Aadjusted DESeq2 Normalized Data

We also generate PCA plot using the DESeq2 normalized data adjusted for cell differentials.
```{r}
home.dir<-"/home/yanxiting/driver_Grace"
my.distinct.colors<-c('#e6194b', '#3cb44b', '#ffe119', '#4363d8', '#f58231', '#911eb4', '#46f0f0', '#f032e6', '#bcf60c', '#fabebe', '#008080', '#e6beff', '#9a6324', '#fffac8', '#800000', '#aaffc3', '#808000', '#ffd8b1', '#000075', '#808080', '#ffffff', '#000000')
data.dir<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38")
fpkm.filepath<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/data_adjusted2/DESeq2_normalized_276_clean_celldiffadjusted_withannot.txt")
#fpkm.filepath<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/data","DESeq2_normalized_276_clean.txt")
masterkey.filepath<-file.path(home.dir,"scratch/GRADS/SARC_results/Data/GRADS Master Progress Key 6-13-16.xlsx")
clinic.filepath<-file.path(home.dir,"scratch/GRADS/SARC_results/Data_PBMC/clin_data_visit1_XY.csv")
output.dir<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/data_adjusted2")

# load in the fpkm matrix
fpkm.matrix<-read.table(fpkm.filepath,sep="\t",header=T,check.names=F,as.is=TRUE,comment.char="")
anno.matrix<-as.matrix(fpkm.matrix)[,1:6]
rownames(anno.matrix)<-anno.matrix[,"GeneID"]
rownames(fpkm.matrix)<-anno.matrix[,"GeneID"]
data.matrix<-fpkm.matrix[,7:ncol(fpkm.matrix)]

# load in the clinical data
clinic.matrix<-read.csv(clinic.filepath,check.names=F,header=T,as.is=TRUE)
rownames(clinic.matrix)<-as.matrix(clinic.matrix)[,1] # GRADS ID
clinic.matrix<-clinic.matrix[colnames(data.matrix),]

temp.data<-read.xlsx(masterkey.filepath,sheetIndex = 3, check.names=F, startRow=3)
rownames(temp.data)<-temp.data$`GRADS ID`
clinic.matrix<-cbind(clinic.matrix,as.matrix(temp.data)[rownames(clinic.matrix),"Phenotype/Genotype"])
colnames(clinic.matrix)[ncol(clinic.matrix)]<-"pheno_group"
gc(reset=TRUE)

data.matrix.orig<-data.matrix
clinic.matrix.orig<-clinic.matrix
```

We generate the 3-dimensional PCA plots.

```{r}
# We ran these codes on the workstation since it's easier to save and demonstrate the generated plots in Rstudio markdown files. 
###############################################################################################
# 1. identify outliers using pca plots from all 460 reactions
library(gdata)
library(rgl)
log2.transform=FALSE # do you want to log2 transform the data before PCA or not.
source(file.path(home.dir,"Rprogram","my_functions.R"))

#######################
# focus on some of the phenotypic groups
data.matrix<-data.matrix.orig[,!clinic.matrix.orig$pheno_group%in%c("Multi-organ","Cardiac defining therapy","Acute Sarcoidosis, untreated")]
clinic.matrix<-clinic.matrix.orig[!clinic.matrix.orig$pheno_group%in%c("Multi-organ","Cardiac defining therapy","Acute Sarcoidosis, untreated"),]

# Log2 transform
if(log2.transform){
data.matrix<-data.matrix+3e-208
data.matrix<-log(data.matrix,base=2)
}
# calculate the number of expressed genes for each sample
expgene.num<-apply(data.matrix>0,2,sum)
##################
# PCA analysis
library(gplots)
# calculate the SVD
temp.sd<-apply(data.matrix,1,sd)
data.matrix<-data.matrix[temp.sd>0,]  #23861 genes
data.matrix.norm<-my.normalize(data.matrix)
my.svd<-svd(data.matrix.norm)
cumsum(my.svd$d^2/sum(my.svd$d^2))[1:3]
my.projected<-t(data.matrix.norm)%*%my.svd$u 
my.colors<-factor(as.matrix(clinic.matrix)[,"pheno_group"],levels=c("Remitting, untreated","Non-acute, Stage I, untreated","Stage II-III, untreated","Stage II-III, treated","Stage IV, untreated","Stage IV, treated"))
my.colors<-my.distinct.colors[as.numeric(my.colors)]
open3d()
plot3d(my.projected[,1],my.projected[,2],my.projected[,3],xlab="PC 1",ylab="PC 2",zlab="PC 3",col=my.colors,type="s", size=1,box = F,axes = T)
par3d(windowRect = c(0, 0, 400, 400))
legend3d("topright", legend = c("Remitting, untreated","Non-acute, Stage I, untreated","Stage II-III, untreated","Stage II-III, treated","Stage IV, untreated","Stage IV, treated"), pch = 16, col = my.distinct.colors[1:length(unique(my.colors))], cex=0.65, inset=c(0))
rglwidget()
#filename<-writeWebGL(dir=file.path(pca.dir,"pca_3d"),reuse=TRUE)
# label the samples based on subjects and disease

#identify3d(my.projected[,1],my.projected[,2],my.projected[,3],labels=colnames(data.matrix.norm),n=ncol(data.matrix.norm),plot=TRUE)
# label samples using subject ID

```

PCA plot with the untreated subjects only.

```{r}
# We ran these codes on the workstation since it's easier to save and demonstrate the generated plots in Rstudio markdown files. 
###############################################################################################
# 1. identify outliers using pca plots from all 460 reactions
library(gdata)
library(rgl)
log2.transform=FALSE # do you want to log2 transform the data before PCA or not.
source(file.path(home.dir,"Rprogram","my_functions.R"))

#######################
# focus on some of the phenotypic groups
data.matrix<-data.matrix.orig[,!clinic.matrix.orig$pheno_group%in%c("Multi-organ","Cardiac defining therapy","Acute Sarcoidosis, untreated","Stage II-III, treated","Stage IV, treated")]
clinic.matrix<-clinic.matrix.orig[!clinic.matrix.orig$pheno_group%in%c("Multi-organ","Cardiac defining therapy","Acute Sarcoidosis, untreated","Stage II-III, treated","Stage IV, treated"),]

# Log2 transform
if(log2.transform){
data.matrix<-data.matrix+3e-208
data.matrix<-log(data.matrix,base=2)
}
# calculate the number of expressed genes for each sample
expgene.num<-apply(data.matrix>0,2,sum)
##################
# PCA analysis
library(gplots)
# calculate the SVD
temp.sd<-apply(data.matrix,1,sd)
data.matrix<-data.matrix[temp.sd>0,]  #23861 genes
data.matrix.norm<-my.normalize(data.matrix)
my.svd<-svd(data.matrix.norm)
cumsum(my.svd$d^2/sum(my.svd$d^2))[1:3]
my.projected<-t(data.matrix.norm)%*%my.svd$u 
my.colors<-factor(as.matrix(clinic.matrix)[,"pheno_group"],levels=c("Remitting, untreated","Non-acute, Stage I, untreated","Stage II-III, untreated","Stage IV, untreated"))
my.colors<-my.distinct.colors[as.numeric(my.colors)]
open3d()
plot3d(my.projected[,1],my.projected[,2],my.projected[,3],xlab="PC 1",ylab="PC 2",zlab="PC 3",col=my.colors,type="s", size=1,box = F,axes = T)
par3d(windowRect = c(0, 0, 400, 400))
legend3d("topright", legend = c("Remitting, untreated","Non-acute, Stage I, untreated","Stage II-III, untreated","Stage IV, untreated"), pch = 16, col = my.distinct.colors[1:length(unique(my.colors))], cex=0.65, inset=c(0))
rglwidget()
#filename<-writeWebGL(dir=file.path(pca.dir,"pca_3d"),reuse=TRUE)
# label the samples based on subjects and disease

#identify3d(my.projected[,1],my.projected[,2],my.projected[,3],labels=colnames(data.matrix.norm),n=ncol(data.matrix.norm),plot=TRUE)
# label samples using subject ID

```

PCA plot only with non-acute Stage I untreated vs all others.

```{r}
# We ran these codes on the workstation since it's easier to save and demonstrate the generated plots in Rstudio markdown files. 
###############################################################################################
# 1. identify outliers using pca plots from all 460 reactions
library(gdata)
library(rgl)
log2.transform=FALSE # do you want to log2 transform the data before PCA or not.
source(file.path(home.dir,"Rprogram","my_functions.R"))

#######################
# focus on some of the phenotypic groups
data.matrix<-data.matrix.orig[,!clinic.matrix.orig$pheno_group%in%c("Multi-organ","Cardiac defining therapy","Acute Sarcoidosis, untreated","Stage II-III, treated","Stage IV, treated")]
clinic.matrix<-clinic.matrix.orig[!clinic.matrix.orig$pheno_group%in%c("Multi-organ","Cardiac defining therapy","Acute Sarcoidosis, untreated","Stage II-III, treated","Stage IV, treated"),]

temp.vect<-as.matrix(clinic.matrix)[,"pheno_group"]
temp.vect[temp.vect!="Non-acute, Stage I, untreated"]<-"Other"
clinic.matrix$pheno_group<-temp.vect

# Log2 transform
if(log2.transform){
data.matrix<-data.matrix+3e-208
data.matrix<-log(data.matrix,base=2)
}
# calculate the number of expressed genes for each sample
expgene.num<-apply(data.matrix>0,2,sum)
##################
# PCA analysis
library(gplots)
# calculate the SVD
temp.sd<-apply(data.matrix,1,sd)
data.matrix<-data.matrix[temp.sd>0,]  #23861 genes
data.matrix.norm<-my.normalize(data.matrix)
my.svd<-svd(data.matrix.norm)
cumsum(my.svd$d^2/sum(my.svd$d^2))[1:3]
my.projected<-t(data.matrix.norm)%*%my.svd$u 
my.colors<-factor(as.matrix(clinic.matrix)[,"pheno_group"],levels=c("Other","Non-acute, Stage I, untreated"))
my.colors<-my.distinct.colors[as.numeric(my.colors)]
open3d()
plot3d(my.projected[,1],my.projected[,2],my.projected[,3],xlab="PC 1",ylab="PC 2",zlab="PC 3",col=my.colors,type="s", size=1,box = F,axes = T)
par3d(windowRect = c(0, 0, 400, 400))
legend3d("topright", legend = c("Other","Non-acute, Stage I, untreated"), pch = 16, col = my.distinct.colors[1:length(unique(my.colors))], cex=0.65, inset=c(0))
rglwidget()
#filename<-writeWebGL(dir=file.path(pca.dir,"pca_3d"),reuse=TRUE)
# label the samples based on subjects and disease

#identify3d(my.projected[,1],my.projected[,2],my.projected[,3],labels=colnames(data.matrix.norm),n=ncol(data.matrix.norm),plot=TRUE)
# label samples using subject ID

```

### DEGs (GRADS PBMC)

We identify the differentially expressed genes between given groups of subjects using DESeq2. Do not include remitting group for this analysis.

```{r}
library(DESeq2)
library(apeglm)
source(file.path(home.dir,"Rprogram/my_functions.R"))
#count.filepath<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/data","DESeq2_normalized_276_clean.RDS")
count.filepath<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/data","rawcounts_baseline_276_clean.txt")
clinic.filepath<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/data/clinic_matrix_merged.RDS")
output.dir<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/DEGs/DESeq2")

# load in the raw counts matrix
count.matrix<-read.table(count.filepath,sep="\t",header=T,check.names=F,stringsAsFactors = F)
annot.matrix<-count.matrix[,1:6]
count.matrix<-count.matrix[,7:ncol(count.matrix)]
rownames(count.matrix)<-count.matrix$GeneID

# loading the clinic data and match it with the raw count matrix
clinic.matrix<-readRDS(clinic.filepath,refhook = NULL)
clinic.matrix<-clinic.matrix[colnames(count.matrix),]

# replace space in the PHENGRP values with -
clinic.matrix$PHENGRP<-sapply(clinic.matrix$PHENGRP,my.char.replace,orig=" ",replace="_")
clinic.matrix$PHENGRP<-sapply(clinic.matrix$PHENGRP,my.char.replace,orig=",",replace="")
clinic.matrix$PHENGRP<-sapply(clinic.matrix$PHENGRP,my.char.replace,orig="-",replace="_")
```

#### Stage II-III, untreated vs Stage IV, untreated:
```{r}
# comparing:
# Stage II-III untreated vs stage IV untreated
# Stage I untreated vs stage IV untreated
# Stage I, II-III untreated vs Stage IV untreated (BLOOM gene set had pro-inflammatory genes for validation)
#  Acute sarcoid untreated vs stage IV untreated
# patients on steroid vs not on steroid
my.coldata<-clinic.matrix
my.coldata<-my.coldata[my.coldata$PHENGRP%in%c("Stage_II_III_untreated","Stage_IV_untreated"),]
my.coldata$PHENGRP<-factor(my.coldata$PHENGRP,levels=c("Stage_II_III_untreated","Stage_IV_untreated"))
my.coldata$PHENGRP <- relevel(my.coldata$PHENGRP, ref = "Stage_II_III_untreated")
table(my.coldata$PHENGRP)
dds <- DESeqDataSetFromMatrix(countData = count.matrix[,rownames(my.coldata)],
                              colData = my.coldata,
                              design = ~ PHENGRP)
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]
my.annot.matrix<-annot.matrix[keep,]
dds<-DESeq(dds)
res<-results(dds)
resLFC <- lfcShrink(dds, coef="PHENGRP_Stage_IV_untreated_vs_Stage_II_III_untreated", type="apeglm")

res05<-results(dds,alpha=0.05)
summary(res05)

# output the results
output.filepath<-file.path(output.dir,paste("DEGs_all_",resultsNames(dds)[2],".csv",sep=""))
cmd.out<-cbind(my.annot.matrix,as.data.frame(res))
cmd.out<-cmd.out[order(res$pvalue,decreasing=F),]
resOrdered <- res[order(res$pvalue),]
write.csv(cmd.out, file=output.filepath,row.names = FALSE)

output.filepath<-file.path(output.dir,paste("DEGs_fdr0.05_",resultsNames(dds)[2],".csv",sep=""))
cmd.out<-cmd.out[cmd.out$padj<0.05,]
resSig <- subset(resOrdered, padj < 0.05)
write.csv(cmd.out, file=output.filepath,row.names=FALSE)

# describe how many genes pass the FDR threshold
cat("There are ",sum(res$padj<0.05)," genes with FDR<0.05.\n",sep="")
```

#### Stage I, untreated vs Stage IV, untreated:
```{r}
# comparing:
# Stage II-III untreated vs stage IV untreated
# Stage I untreated vs stage IV untreated
# Stage I, II-III untreated vs Stage IV untreated (BLOOM gene set had pro-inflammatory genes for validation)
#  Acute sarcoid untreated vs stage IV untreated
# patients on steroid vs not on steroid
my.coldata<-clinic.matrix
my.coldata<-my.coldata[my.coldata$PHENGRP%in%c("Non_acute_Stage_I_untreated","Stage_IV_untreated"),]
my.coldata$PHENGRP<-factor(my.coldata$PHENGRP,levels=c("Non_acute_Stage_I_untreated","Stage_IV_untreated"))
my.coldata$PHENGRP <- relevel(my.coldata$PHENGRP, ref = "Non_acute_Stage_I_untreated")
table(my.coldata$PHENGRP)
dds <- DESeqDataSetFromMatrix(countData = count.matrix[,rownames(my.coldata)],
                              colData = my.coldata,
                              design = ~ PHENGRP)
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]
my.annot.matrix<-annot.matrix[keep,]
dds<-DESeq(dds)
res<-results(dds)
resLFC <- lfcShrink(dds, coef=resultsNames(dds)[2], type="apeglm")

res05<-results(dds,alpha=0.05)
summary(res05)

# output the results
output.filepath<-file.path(output.dir,paste("DEGs_all_",resultsNames(dds)[2],".csv",sep=""))
cmd.out<-cbind(my.annot.matrix,as.data.frame(res))
cmd.out<-cmd.out[order(res$pvalue,decreasing=F),]
resOrdered <- res[order(res$pvalue),]
write.csv(cmd.out, file=output.filepath,row.names = FALSE)

output.filepath<-file.path(output.dir,paste("DEGs_fdr0.05_",resultsNames(dds)[2],".csv",sep=""))
cmd.out<-cmd.out[!is.na(cmd.out$padj) & cmd.out$padj<0.05,]
resSig <- subset(resOrdered, padj < 0.05)
write.csv(cmd.out, file=output.filepath,row.names=FALSE)

# describe how many genes pass the FDR threshold
cat("There are ",sum(res$padj<0.05,na.rm=T)," genes with FDR<0.05.\n",sep="")
```


#### Stage I, untreated vs Stage II-III, untreated:
```{r}
# comparing:
# Stage II-III untreated vs stage IV untreated
# Stage I untreated vs stage IV untreated
# Stage I, II-III untreated vs Stage IV untreated (BLOOM gene set had pro-inflammatory genes for validation)
#  Acute sarcoid untreated vs stage IV untreated
# patients on steroid vs not on steroid
my.coldata<-clinic.matrix
my.coldata<-my.coldata[my.coldata$PHENGRP%in%c("Non_acute_Stage_I_untreated","Stage_II_III_untreated"),]
my.coldata$PHENGRP<-factor(my.coldata$PHENGRP,levels=c("Non_acute_Stage_I_untreated","Stage_II_III_untreated"))
my.coldata$PHENGRP <- relevel(my.coldata$PHENGRP, ref = "Non_acute_Stage_I_untreated")
table(my.coldata$PHENGRP)
dds <- DESeqDataSetFromMatrix(countData = count.matrix[,rownames(my.coldata)],
                              colData = my.coldata,
                              design = ~ PHENGRP)
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]
my.annot.matrix<-annot.matrix[keep,]
dds<-DESeq(dds)
res<-results(dds)
resLFC <- lfcShrink(dds, coef=resultsNames(dds)[2], type="apeglm")

res05<-results(dds,alpha=0.05)
summary(res05)

# output the results
output.filepath<-file.path(output.dir,paste("DEGs_all_",resultsNames(dds)[2],".csv",sep=""))
cmd.out<-cbind(my.annot.matrix,as.data.frame(res))
cmd.out<-cmd.out[order(res$pvalue,decreasing=F),]
resOrdered <- res[order(res$pvalue),]
write.csv(cmd.out, file=output.filepath,row.names = FALSE)

output.filepath<-file.path(output.dir,paste("DEGs_fdr0.05_",resultsNames(dds)[2],".csv",sep=""))
cmd.out<-cmd.out[!is.na(cmd.out$padj) & cmd.out$padj<0.05,]
resSig <- subset(resOrdered, padj < 0.05)
write.csv(cmd.out, file=output.filepath,row.names=FALSE)

# describe how many genes pass the FDR threshold
table(my.coldata$PHENGRP)
cat("There are ",sum(res$padj<0.05,na.rm=T)," genes with FDR<0.05.\n",sep="")
```


#### Stage II-III, untreated vs Stage IV, untreated:
```{r}
# comparing:
# Stage II-III untreated vs stage IV untreated
# Stage I untreated vs stage IV untreated
# Stage I, II-III untreated vs Stage IV untreated (BLOOM gene set had pro-inflammatory genes for validation)
#  Acute sarcoid untreated vs stage IV untreated
# patients on steroid vs not on steroid
my.coldata<-clinic.matrix
my.coldata<-my.coldata[my.coldata$PHENGRP%in%c("Stage_II_III_untreated","Stage_IV_untreated"),]
my.coldata$PHENGRP<-factor(my.coldata$PHENGRP,levels=c("Stage_II_III_untreated","Stage_IV_untreated"))
my.coldata$PHENGRP <- relevel(my.coldata$PHENGRP, ref = "Stage_II_III_untreated")
dds <- DESeqDataSetFromMatrix(countData = count.matrix[,rownames(my.coldata)],
                              colData = my.coldata,
                              design = ~ PHENGRP)
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]
my.annot.matrix<-annot.matrix[keep,]
dds<-DESeq(dds)
res<-results(dds)
resLFC <- lfcShrink(dds, coef=resultsNames(dds)[2], type="apeglm")

res05<-results(dds,alpha=0.05)
summary(res05)

# output the results
output.filepath<-file.path(output.dir,paste("DEGs_all_",resultsNames(dds)[2],".csv",sep=""))
cmd.out<-cbind(my.annot.matrix,as.data.frame(res))
cmd.out<-cmd.out[order(res$pvalue,decreasing=F),]
resOrdered <- res[order(res$pvalue),]
write.csv(cmd.out, file=output.filepath,row.names = FALSE)

output.filepath<-file.path(output.dir,paste("DEGs_fdr0.05_",resultsNames(dds)[2],".csv",sep=""))
cmd.out<-cmd.out[!is.na(cmd.out$padj) & cmd.out$padj<0.05,]
resSig <- subset(resOrdered, padj < 0.05)
write.csv(cmd.out, file=output.filepath,row.names=FALSE)

# describe how many genes pass the FDR threshold
cat("There are ",sum(res$padj<0.05,na.rm=T)," genes with FDR<0.05.\n",sep="")
```

#### Stage I untreated vs Stage II-IV, untreated:
```{r}
# comparing:
# Stage II-III untreated vs stage IV untreated
# Stage I untreated vs stage IV untreated
# Stage I, II-III untreated vs Stage IV untreated (BLOOM gene set had pro-inflammatory genes for validation)
#  Acute sarcoid untreated vs stage IV untreated
# patients on steroid vs not on steroid
my.coldata<-clinic.matrix
my.coldata<-my.coldata[my.coldata$PHENGRP%in%c("Non_acute_Stage_I_untreated","Stage_II_III_untreated","Stage_IV_untreated"),]
my.coldata$PHENGRP[my.coldata$PHENGRP%in%c("Stage_II_III_untreated","Stage_IV_untreated")]<-"Stage_II_IV_untreated"
my.coldata$PHENGRP<-factor(my.coldata$PHENGRP,levels=c("Non_acute_Stage_I_untreated","Stage_II_IV_untreated"))
my.coldata$PHENGRP <- relevel(my.coldata$PHENGRP, ref = "Non_acute_Stage_I_untreated")
dds <- DESeqDataSetFromMatrix(countData = count.matrix[,rownames(my.coldata)],
                              colData = my.coldata,
                              design = ~ PHENGRP)
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]
my.annot.matrix<-annot.matrix[keep,]
dds<-DESeq(dds)
res<-results(dds)
resLFC <- lfcShrink(dds, coef=resultsNames(dds)[2], type="apeglm")

res05<-results(dds,alpha=0.05)
summary(res05)

# output the results
output.filepath<-file.path(output.dir,paste("DEGs_all_",resultsNames(dds)[2],".csv",sep=""))
cmd.out<-cbind(my.annot.matrix,as.data.frame(res))
cmd.out<-cmd.out[order(res$pvalue,decreasing=F),]
resOrdered <- res[order(res$pvalue),]
write.csv(cmd.out, file=output.filepath,row.names = FALSE)

output.filepath<-file.path(output.dir,paste("DEGs_fdr0.05_",resultsNames(dds)[2],".csv",sep=""))
cmd.out<-cmd.out[!is.na(cmd.out$padj) & cmd.out$padj<0.05,]
resSig <- subset(resOrdered, padj < 0.05)
write.csv(cmd.out, file=output.filepath,row.names=FALSE)

# describe how many genes pass the FDR threshold
table(my.coldata$PHENGRP)
cat("There are ",sum(res$padj<0.05,na.rm = T)," genes with FDR<0.05.\n",sep="")
```



Acute sarcoid, untreated vs Stage IV, untreated:
```{r}
# comparing:
# Stage II-III untreated vs stage IV untreated
# Stage I untreated vs stage IV untreated
# Stage I, II-III untreated vs Stage IV untreated (BLOOM gene set had pro-inflammatory genes for validation)
#  Acute sarcoid untreated vs stage IV untreated
# patients on steroid vs not on steroid
my.coldata<-clinic.matrix
my.coldata<-my.coldata[my.coldata$PHENGRP%in%c("Acute-Sarcoidosis-untreated","Stage-IV-untreated"),]
my.coldata$PHENGRP<-factor(my.coldata$PHENGRP,levels=c("Acute-Sarcoidosis-untreated","Stage-IV-untreated"))
my.coldata$PHENGRP <- relevel(my.coldata$PHENGRP, ref = "Acute-Sarcoidosis-untreated")
dds <- DESeqDataSetFromMatrix(countData = count.matrix[,rownames(my.coldata)],
                              colData = my.coldata,
                              design = ~ PHENGRP)
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]
my.annot.matrix<-annot.matrix[keep,]
dds<-DESeq(dds)
res<-results(dds)
resLFC <- lfcShrink(dds, coef=resultsNames(dds)[2], type="apeglm")

res05<-results(dds,alpha=0.05)
summary(res05)

# output the results
output.filepath<-file.path(output.dir,paste("DEGs_all_",resultsNames(dds)[2],".csv",sep=""))
cmd.out<-cbind(my.annot.matrix,as.data.frame(res))
cmd.out<-cmd.out[order(res$pvalue,decreasing=F),]
resOrdered <- res[order(res$pvalue),]
write.csv(cmd.out, file=output.filepath,row.names = FALSE)

output.filepath<-file.path(output.dir,paste("DEGs_fdr0.05_",resultsNames(dds)[2],".csv",sep=""))
cmd.out<-cmd.out[!is.na(cmd.out$padj) & cmd.out$padj<0.05,]
resSig <- subset(resOrdered, padj < 0.05)
write.csv(cmd.out, file=output.filepath,row.names=FALSE)

# describe how many genes pass the FDR threshold
cat("There are ",sum(res$padj<0.05,na.rm = T)," genes with FDR<0.05.\n",sep="")
```

Steroid vs non-steroid usage patients:
```{r}
# comparing:
# Stage II-III untreated vs stage IV untreated
# Stage I untreated vs stage IV untreated
# Stage I, II-III untreated vs Stage IV untreated (BLOOM gene set had pro-inflammatory genes for validation)
#  Acute sarcoid untreated vs stage IV untreated
# patients on steroid vs not on steroid
my.coldata<-clinic.matrix[!is.na(clinic.matrix$steroid_atv1),]
my.coldata$steroid_atv1<-factor(my.coldata$steroid_atv1,levels=c(0,1))
my.coldata$steroid_atv1 <- relevel(my.coldata$steroid_atv1, ref = "0")
dds <- DESeqDataSetFromMatrix(countData = count.matrix[,rownames(my.coldata)],
                              colData = my.coldata,
                              design = ~ steroid_atv1)
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]
my.annot.matrix<-annot.matrix[keep,]
dds<-DESeq(dds)
res<-results(dds)
resLFC <- lfcShrink(dds, coef=resultsNames(dds)[2], type="apeglm")

res05<-results(dds,alpha=0.05)
summary(res05)

# output the results
output.filepath<-file.path(output.dir,paste("DEGs_all_",resultsNames(dds)[2],".csv",sep=""))
cmd.out<-cbind(my.annot.matrix,as.data.frame(res))
cmd.out<-cmd.out[order(res$pvalue,decreasing=F),]
resOrdered <- res[order(res$pvalue),]
write.csv(cmd.out, file=output.filepath,row.names = FALSE)

output.filepath<-file.path(output.dir,paste("DEGs_fdr0.05_",resultsNames(dds)[2],".csv",sep=""))
cmd.out<-cmd.out[!is.na(cmd.out$padj) & cmd.out$padj<0.05,]
resSig <- subset(resOrdered, padj < 0.05)
write.csv(cmd.out, file=output.filepath,row.names=FALSE)

# describe how many genes pass the FDR threshold
cat("There are ",sum(res$padj<0.05,na.rm = T)," genes with FDR<0.05.\n",sep="")
```

Stage I vs Stage II-IV, untreated:
```{r}
# comparing:
# Non-acute, Stage I, untreated vs c("Stage II-III, untreated","Stage IV, untreated")
# Stage I untreated vs stage IV untreated
# Stage I, II-III untreated vs Stage IV untreated (BLOOM gene set had pro-inflammatory genes for validation)
#  Acute sarcoid untreated vs stage IV untreated
# patients on steroid vs not on steroid
my.coldata<-clinic.matrix
my.coldata<-my.coldata[my.coldata$PHENGRP%in%c("Non-acute-Stage-I-untreated","Stage-II-III-untreated","Stage-IV-untreated"),]
my.coldata$PHENGRP[my.coldata$PHENGRP%in%c("Stage-II-III-untreated","Stage-IV-untreated")]<-"Stage-II-IV-untreated"

my.coldata$PHENGRP<-factor(my.coldata$PHENGRP,levels=c("Non-acute-Stage-I-untreated","Stage-II-IV-untreated"))
my.coldata$PHENGRP <- relevel(my.coldata$PHENGRP, ref = "Non-acute-Stage-I-untreated")
dds <- DESeqDataSetFromMatrix(countData = count.matrix[,rownames(my.coldata)],
                              colData = my.coldata,
                              design = ~ PHENGRP)
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]
my.annot.matrix<-annot.matrix[keep,]
dds<-DESeq(dds)
res<-results(dds)
resLFC <- lfcShrink(dds, coef=resultsNames(dds)[2], type="apeglm")

res05<-results(dds,alpha=0.05)
summary(res05)

# output the results
output.filepath<-file.path(output.dir,paste("DEGs_all_",resultsNames(dds)[2],".csv",sep=""))
cmd.out<-cbind(my.annot.matrix,as.data.frame(res))
cmd.out<-cmd.out[order(res$pvalue,decreasing=F),]
resOrdered <- res[order(res$pvalue),]
write.csv(cmd.out, file=output.filepath,row.names = FALSE)

output.filepath<-file.path(output.dir,paste("DEGs_fdr0.05_",resultsNames(dds)[2],".csv",sep=""))
cmd.out<-cmd.out[!is.na(cmd.out$padj) & cmd.out$padj<0.05,]
resSig <- subset(resOrdered, padj < 0.05)
write.csv(cmd.out, file=output.filepath,row.names=FALSE)

# describe how many genes pass the FDR threshold
cat("There are ",sum(res$padj<0.05,na.rm = T)," genes with FDR<0.05.\n",sep="")
```

#### Batch effect check

Since we found big difference between stage I untreated and stage III untreated, we want to double check to see if this was caused by the batch effect. 

To assess the batch effect in this dataset, we performed principal component analysis (PCA) on the whole dataset and calculate the distance between every two samples in the space formed by the top 3 principal components (PCs). The correlation bewteen this distance and the number of days between the processing dates of the corresponding two samples was visualized and examined using Spearman correlation test. 


```{r}
source(file.path(home.dir,"Rprogram/my_functions.R"))
library(xlsx)
home.dir<-"/home/yanxiting/driver_Grace"
date.filepath<-file.path(home.dir,"scratch/GRADS/SARC_results/Data/Sample_SpreadSheet_20160826_updated.xlsx")

# extract the processing date of all samples
temp.data<-read.xlsx(date.filepath,sheetIndex = 1,check.names=F)

temp1<-temp.data$Run_date
names(temp1)<-as.matrix(temp.data)[,"Sample ID1"]
temp1<-temp1[!is.na(temp1)]
temp1<-temp1[(!is.na(names(temp1))) & (names(temp1)!="NA")]
temp.date<-temp1

temp1<-temp.data$Run_date
names(temp1)<-as.matrix(temp.data)[,"Sample ID2"]
temp1<-temp1[!is.na(temp1)]
temp1<-temp1[(!is.na(names(temp1))) & (names(temp1)!="NA")]
temp.date<-c(temp.date,temp1)

temp1<-temp.data$Run_date
names(temp1)<-as.matrix(temp.data)[,"Sample ID3"]
temp1<-temp1[!is.na(temp1)]
temp1<-temp1[(!is.na(names(temp1))) & (names(temp1)!="NA")]
temp.date<-c(temp.date,temp1)

# retrieve the sample ID with run names for the stage I and stage II-III samples
count.filepath.1<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/data","rawcounts_baseline_276_clean.txt")
count.filepath.2<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/data","rawcounts_corrected_all.txt")
matched.filepath<-file.path(home.dir,"scratch/GRADS/SARC_results/Data/GRADS Master Progress Key 6-13-16.xlsx")

count.matrix<-read.table(count.filepath.1,sep="\t",header=T,check.names=F,stringsAsFactors = F)
annot.matrix<-count.matrix[,1:6]
count.matrix<-count.matrix[,7:ncol(count.matrix)]
rownames(count.matrix)<-count.matrix$GeneID
count.matrix.1<-count.matrix
count.matrix.gradsid<-colnames(count.matrix.1)

# convert the colnames of count.matrix.1 into kit ID without run names
matched.data<-read.xlsx(matched.filepath,sheetIndex = 3,check.names=F)
matched.data<-matched.data[,1:5]
temp.names<-unname(as.matrix(matched.data)[2,])
matched.data<-matched.data[3:nrow(matched.data),]
colnames(matched.data)<-temp.names

my.matched.data<-matched.data[matched.data[,1]%in%colnames(count.matrix.1),]
rownames(my.matched.data)<-my.matched.data[,1]
my.matched.data<-my.matched.data[colnames(count.matrix.1),]
colnames(count.matrix.1)<-my.matched.data[,2]

# extract the colnames of count.matrix.2 without run names
count.matrix<-read.table(count.filepath.2,sep="\t",header=T,check.names=F,stringsAsFactors = F)
annot.matrix<-count.matrix[,1:6]
count.matrix<-count.matrix[,7:ncol(count.matrix)]
rownames(count.matrix)<-count.matrix$GeneID
count.matrix.2<-count.matrix

temp.vect<-unname(sapply(colnames(count.matrix.2),my.element.extract,splitchar="_",index=1))
temp<-colnames(count.matrix.2)[temp.vect%in%colnames(count.matrix.1)]
temp.vect<-temp.vect[temp.vect%in%colnames(count.matrix.1)]

temp.list<-split(temp,temp.vect)

my.temp.list<-temp.list[unlist(lapply(temp.list,length))>1]
my.matched.list<-list()
for(i in 1:length(my.temp.list)){
  my.matched.list[[i]]<-numeric()
  for(j in 1:length(my.temp.list[[i]])){
    my.matched.list[[i]]<-c(my.matched.list[[i]],sum(count.matrix.1[,names(my.temp.list)[i]]==count.matrix.2[,my.temp.list[[i]][j]]))
  }
}

for(i in 1:length(my.temp.list)){
  temp.list[names(temp.list)==names(my.temp.list)[i]][[1]]<-temp.list[names(temp.list)==names(my.temp.list)[i]][[1]][my.matched.list[[i]]==26485]
}

sample.matrix<-cbind(unname(unlist(temp.list)),names(temp.list))
names(count.matrix.gradsid)<-colnames(count.matrix.1)
sample.matrix<-cbind(sample.matrix,count.matrix.gradsid[sample.matrix[,2]])
colnames(sample.matrix)<-c("kit_run","kit","grads_id")
output.filepath<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/data/baseline_276_clean_samplematrix.txt")
write.table(sample.matrix,file=output.filepath,append=F,sep="\t",row.names=F,col.names=T,quote=F)
# extract the kit IDs for stage I and stage III
count.matrix<-count.matrix.1
colnames(count.matrix)<-count.matrix.gradsid
clinic.filepath<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/data/clinic_matrix_merged.RDS")
output.dir<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/DEGs/DESeq2")

# loading the clinic data and match it with the raw count matrix
clinic.matrix<-readRDS(clinic.filepath,refhook = NULL)
clinic.matrix<-clinic.matrix[colnames(count.matrix),]

# replace space in the PHENGRP values with -
clinic.matrix$PHENGRP<-sapply(clinic.matrix$PHENGRP,my.char.replace,orig=" ",replace="_")
clinic.matrix$PHENGRP<-sapply(clinic.matrix$PHENGRP,my.char.replace,orig=",",replace="")
clinic.matrix$PHENGRP<-sapply(clinic.matrix$PHENGRP,my.char.replace,orig="-",replace="_")

my.coldata<-clinic.matrix
my.coldata<-my.coldata[my.coldata$PHENGRP%in%c("Non_acute_Stage_I_untreated","Stage_II_III_untreated"),]
my.coldata$PHENGRP<-factor(my.coldata$PHENGRP,levels=c("Non_acute_Stage_I_untreated","Stage_II_III_untreated"))
my.coldata$PHENGRP <- relevel(my.coldata$PHENGRP, ref = "Non_acute_Stage_I_untreated")
table(my.coldata$PHENGRP)

temp.list<-split(rownames(my.coldata),my.coldata$PHENGRP)
rownames(sample.matrix)<-sample.matrix[,3]
temp.list[[1]]<-sample.matrix[temp.list[[1]],1]
temp.list[[2]]<-sample.matrix[temp.list[[2]],1]

date.list<-list()
for(i in 1:length(temp.list)){
  temp<-temp.date[temp.list[[i]]]
  rownames(sample.matrix)<-sample.matrix[,1]
  names(temp)<-sample.matrix[names(temp),3]
  date.list[[i]]<-temp
}
names(date.list)<-names(temp.list)

# compare the day difference between every two samples within each group to those bewteen group to see if they are significnatly different.

temp1<-as.Date(date.list[[1]],format="%Y-%m-%d")
temp2<-as.Date(date.list[[2]],format="%Y-%m-%d")

within.group.dist<-numeric()
for(i in 2:length(temp1)){
  for(j in 1:(i-1)){
    within.group.dist<-c(within.group.dist,abs(as.numeric(temp1[i]-temp1[j])))
  }
}
within.group.dist.1<-within.group.dist


within.group.dist<-numeric()
for(i in 2:length(temp2)){
  for(j in 1:(i-1)){
    within.group.dist<-c(within.group.dist,abs(as.numeric(temp2[i]-temp2[j])))
  }
}
within.group.dist.2<-within.group.dist


between.group.dist<-numeric()
for(i in 1:length(temp1)){
  for(j in 1:length(temp2)){
    between.group.dist<-c(between.group.dist,abs(as.numeric(temp1[i]-temp2[j])))
  }
}


wilcox.test(within.group.dist.1,between.group.dist)

wilcox.test(within.group.dist.2,between.group.dist)
```

### WGCNA (GRADS PBMC DEGs)

We take the PBMC DEGs as pre-defined gene set and assess its enrichment in the UCSF cohort data using GSEA.

#### data loading

First, we load in the gene expression data from UCSF cohort, which was normalized using DESeq2. We'll use the log2(DESeq2+1) as the expression data.

```{r}
# load in the normalized count matrix
library(xlsx)
home.dir<-"/home/yanxiting/driver_Grace"
source(file.path(home.dir,"Rprogram/my_functions.R"))

deseq2.filepath<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/Laura_Koth/data/DESeq2log2_celldiffadjusted_nooutliers.RDS")
my.data<-readRDS(deseq2.filepath,refhook = NULL)

anno.data<-my.data[,1:6]
rownames(my.data)<-as.matrix(my.data)[,1]
fpkm.matrix<-my.data[,7:ncol(my.data)]

#log2 transform, if yes, do the analysis on log2(fpkm.matrix+1) to avoid log2(0) transformation
log2.transform<-F
if(log2.transform){
fpkm.matrix<-fpkm.matrix+1
fpkm.matrix<-log(fpkm.matrix,base=2)
}
fpkm<-fpkm.matrix

# filter out genes with no variation and centralize the data 
temp.sd<-apply(fpkm,1,sd)
fpkm<-fpkm[temp.sd>0,]
anno.data<-anno.data[anno.data[,1]%in%rownames(fpkm),]
removed_genes <- rownames(fpkm)[temp.sd<=0] 

```

Second, we load in the phenotype data.

```{r}
home.dir<-"/home/yanxiting/driver_Grace"
data.filepath<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/Laura_Koth/data/cxr_merge_ucsf_seq.xlsx")
clinic.data<-read.xlsx(data.filepath,sheetIndex = 1,check.names=F,stringsAsFactors=FALSE)
```

We match the order of columns in fpkm and rows of clinic.data.

```{r}
temp.names<-intersect(colnames(fpkm),clinic.data$sample_names)
fpkm.data<-fpkm[,temp.names]
rownames(clinic.data)<-clinic.data$sample_names
clinic.data<-clinic.data[temp.names,]

# extract the data for Stage I and Stage II-IV
my.fpkm.data<-fpkm.data[,clinic.data$CXR_stage%in%c(1,2,3,4)]
my.clinic.data<-clinic.data[clinic.data$CXR_stage%in%c(1,2,3,4),]

```

We output the matched expression matrix and clinical data into text files.

```{r}
# extract the patients data
my.fpkm.data<-my.fpkm.data[,!is.na(my.clinic.data$CXR_stage)]
my.clinic.data<-my.clinic.data[!is.na(my.clinic.data$CXR_stage),]

output.dir<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/Laura_Koth/GSEA_PBMCDEGs_log2DESeq2_adjusted")
if(file.exists(output.dir)==F){
  dir.create(output.dir)
}

fpkm.output.filepath<-file.path(output.dir,"DESeq2_log2_sarconly_matched_nooutliers.RDS")
clinic.output.filepath<-file.path(output.dir,"clinic_data_sarconly_matched_nooutliers.RDS")
fpkm.annot.output.filepath<-file.path(output.dir,"DESeq2_log2_annot_sarconly_matched_nooutliers.RDS")


saveRDS(fpkm.data,file=fpkm.output.filepath,refhook = NULL)
saveRDS(clinic.data,file=clinic.output.filepath,refhook = NULL)
saveRDS(anno.data,file=fpkm.annot.output.filepath,refhook = NULL)
```

Last, load in the DEGs from GRADS PBMC data by comparing stage I untreated to stage II-IV untreated.

```{r}
data.filepath<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/DEGs/DESeq2/DEGs_fdr0.05_PHENGRP_Stage_II_IV_untreated_vs_Non_acute_Stage_I_untreated.csv")
gene.list<-read.csv(data.filepath)
```

#### GSEA input file generation
We first generate file containing both the clinical matrix and CT scan features so that both group 1 and group 2 in the GSEA analysis can be defined based on combinations of clinical traits and CT scan data.

We first generate the gene set file.

```{r}
output.folder<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/Laura_Koth/GSEA_PBMCDEGs_log2DESeq2_adjusted/no_outliers_withID")
if(file.exists(output.folder)==F){
  dir.create(output.folder)
}

output.filepath<-file.path(output.folder,"GRADS_PBMC_DEGs_I_II-IV_FDR0.05.gmt")

cmd.out<-paste("PBMC_DEGs","\t","DEGs of stage I vs II-IV in GRADS PBMC","\t",sep="")
cmd.out<-paste(cmd.out,paste(as.matrix(gene.list)[,1],collapse="\t"),sep="")
cat(cmd.out,file=output.filepath,append=F)

```


##### Stage I vs II-IV untreated
We generate input files for the differential expressed genes between nonacute stage I untreated vs II-IV. 

```{r}
# note this function is slightly different for DESeq2 Norm data and the TPM matrix
#my_gsea_filecreate_binary<-function(fpkm.filepath,clinic.filepath,var.names,group1.values,group2.values,group1.name="group1",group2.name="group2",output.dir,suffix.name){
my_gsea_filecreate_binary<-function(fpkm.matrix,clinic.matrix,fpkm.matrix.anno,var.names,group1.values,group2.values,group1.name="group1",group2.name="group2",output.dir,suffix.name){
  #####################################################################################################################################################
  # Arguments:
  # 
  # gexp.filepath is the file path of the TPM matrix with first 5 columns being annotation for genes 
  # (TPM_baseline_276_clean_celldiffadjusted_withannot.txt).
  # 
  # clinic.filpath is the file path of the clinic data (clinic_matrix_merged.txt). The first row needs to be the column names.
  # 
  # var.names is a vector containing the names of columns in clinic.filepath that you want to generate the phenotype file for.
  #
  # group1.values is a data.frame of string describing values of var.names that are considered as gorup 1. 
  # For example, group1.values=data.frame(PHENGRP="nonacute Stage I untreated"). Note that the columns in group1.values need to match those in var.names if there are
  # more than 1 variable to define the groups.
  #
  # group2.values has the same format as group1.values but is the settings for group2.
  #
  # group1.name is the name you want to call your group1 in the cls file.
  #
  # group2.name is the name you want to call your group2 in the cls file.
  # 
  # output.dir is the folder name where you want to save the created files.
  #
  # suffix.name is the suffix name in the name of the output files. The gct file will be named as gct_suffix.name.gct and cls file will be named as
  # cls_suffix.name.gct. For example, if suffix.name="test", the gct file will be named "gct_test.gct" and the cls file will be named as 
  # "cls_test.cls".
  # 
  # Value:
  #
  # This function will return 0 if successfully ran. Otherwise, it will return 1. If unsuccessful, information regarding what went wrong will be spit 
  # out as warning messages.
  #
  # When ran successfully, this function will create two files under the specified output.dir named gct_var.name.gct and cls_var.name.cls, where 
  # var.name will be the speficied value 
  # in the argument. These two files can be directly loaded into GSEA together with another gene set file that needs to be generated outside of this 
  # function.
  #####################################################################################################################################################
  
  #  generate the gene expression matrix input file for GSEA
  #fpkm.filepath<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/data_adjusted2","TPM_baseline_276_clean_celldiffadjusted_withannot.txt")
  #clinic.filepath<-"/home/yanxiting/driver_Grace/scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/data/clinic_matrix_merged.txt"
  #output.dir<-"/home/yanxiting/driver_Grace/scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/GSEA_knowngenes"
  # suffix.name<-"analysis1"
  
  
  # load in the fpkm matrix
  #fpkm.matrix<-read.table(fpkm.filepath,sep="\t",header=T,check.names=F,as.is=TRUE,comment.char = "")
  #fpkm.matrix.anno<-fpkm.matrix[,1:6]
  #fpkm.matrix<-fpkm.matrix[,7:ncol(fpkm.matrix)]
  
  # load in the clinical data and subset the samples to those listed in the fpkm.matrix
  #clinic.matrix=read.table(clinic.filepath,sep="\t",header=T,check.names=F,stringsAsFactors=F)
  #rownames(clinic.matrix)<-as.matrix(clinic.matrix)[,1] # GRADS ID
  clinic.matrix<-clinic.matrix[colnames(fpkm.matrix),]

  # based on var.name, group1.values, and group2.values, identify the samples to include in the files.
  if(file.exists(output.dir)==F){
    dir.create(output.dir)
  }
  
  gct.filepath<-file.path(output.dir,paste("gexp_",suffix.name,".gct",sep=""))
  cls.filepath<-file.path(output.dir,paste("cls_",suffix.name,".cls",sep=""))
  cls.filepath.2<-file.path(output.dir,paste("cls_",suffix.name,"_withIDs.cls",sep=""))
  # substract samples with var.name equal to the values in group1.values and group2. values
  temp1<-fpkm.matrix[,apply(as.data.frame(clinic.matrix[,var.names]),1,paste,collapse="_")%in%apply(group1.values,1,paste,collapse="_")]
  #cat("\n")
  #cat(apply(as.data.frame(clinic.matrix[,var.names]),1,paste,collapse="_"))
  #cat("\n")
  temp2<-fpkm.matrix[,apply(as.data.frame(clinic.matrix[,var.names]),1,paste,collapse="_")%in%apply(group2.values,1,paste,collapse="_")]
  cat("There are ",ncol(temp1)," and ", ncol(temp2)," samples for group1 and group2, respectively.\n")
  data.matrix<-cbind(temp1,temp2)
  data.matrix.id<-colnames(data.matrix)
  data.matrix<-cbind(fpkm.matrix.anno[,1],rep("na",nrow(data.matrix)),data.matrix)
  colnames(data.matrix)[1:2]<-c("NAME","Description")
  pheno.vect<-c(rep(0,ncol(temp1)),rep(1,ncol(temp2)))
  data.matrix.pheno<-clinic.matrix[data.matrix.id,]  
  # output the gene expression data
  cmd.out<-"#1.2\n"
  cmd.out<-paste(cmd.out,nrow(data.matrix),"\t",ncol(data.matrix)-2,"\n",sep="")
  cat(cmd.out,file=gct.filepath,append=F)
  write.table(data.matrix,sep="\t",file=gct.filepath,append=T,row.names=F,col.names=T,quote=F)
  
  #write.table(data.matrix,sep="\t",file=gct.filepath.2,append=F,row.names=T,col.names=T,quote=F)
  
  # output the phenotype file
  cmd.out<-paste(ncol(data.matrix)-2," ",2," ",1,"\n",sep="")
  cmd.out<-paste(cmd.out,"# group1 group2\n",sep="")
  cat(cmd.out,file=cls.filepath,append=F,sep="")
  cat(pheno.vect,file=cls.filepath,append=T,sep=" ")
  
  # output the whole clinic data
  write.table(data.matrix.pheno,file=cls.filepath.2,append=F,sep="\t",row.names = T,col.names=T,quote=F)
  
  results<-list(group1=temp1,group2=temp2)
  
  return(results)
}


output.folder<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/Laura_Koth/GSEA_PBMCDEGs_log2DESeq2_adjusted/no_outliers_withID")

# analysis 1b
varnames<-c("CXR_stage")
temp<-my_gsea_filecreate_binary(fpkm.matrix=fpkm.data,
                          clinic.matrix = clinic.data,
                          fpkm.matrix.anno = anno.data,
                          var.names = varnames,
                          group1.values = data.frame(CXR_stage=c("1")),
                          group2.values = data.frame(PHENGRP=c("2","3","4")),
                          output.dir=output.folder,
                          suffix.name="analysis1b"
                          )
```

Here's a table showing the breakdown of the patients based on CXR_stage.
```{r}
table(my.clinic.data$CXR_stage)
```




### PBMC DEGs in BAL WGCNA results

We look for the PBMC DEGs in the BAL WGCNA results to see if they cluster in certain module.

```{r}

# load in the PBMC DEGs list
deg.filepath<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/DEGs/DESeq2/DEGs_fdr0.05_PHENGRP_Stage_II_IV_untreated_vs_Non_acute_Stage_I_untreated.csv")
deg.matrix<-read.csv(deg.filepath,header=T)

# load in the list of member genes for each gene module from BAL WGCNA results.
module.file<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_BAL_hg38/WGCNA_BAL/CV0_compare/module_assignment_outliers_7_withnumbers.txt")
module.matrix<-read.table(module.file,sep="\t",header=T)

cat("There are ",sum(deg.matrix$GeneID%in%module.matrix$gene_names)," genes overlapped.\n")

temp.matrix<-module.matrix[module.matrix$gene_names%in%deg.matrix$GeneID,]

```

### DEGs (GRADS BAL)

We identify the differentially expressed genes between given groups of subjects using DESeq2 using the BAL samples to see if the PBMC DEGs can be validated in BAL. Do not include remitting group for this analysis.

#### data loading

```{r}
library(DESeq2)
library(apeglm)
source(file.path(home.dir,"Rprogram/my_functions.R"))
#count.filepath<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/data","DESeq2_normalized_276_clean.RDS")
count.filepath<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_BAL_hg38","rawcounts_matrix_corrected_BAL209_withAnnot.txt")
clinic.filepath<-file.path(home.dir,"scratch/GRADS/SARC_results/Data/SARC_combine_clinical_data20170814.txt")
output.dir<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/DEGs/DESeq2_GRADS_BAL")

if(file.exists(output.dir)==F){
  dir.create(output.dir)
}

# load in the raw counts matrix
count.matrix<-read.table(count.filepath,sep="\t",header=T,check.names=F,stringsAsFactors = F)
annot.matrix<-count.matrix[,1:6]
count.matrix<-count.matrix[,7:ncol(count.matrix)]
rownames(count.matrix)<-count.matrix$GeneID
colnames(count.matrix)<-unname(sapply(colnames(count.matrix),my.element.extract,splitchar="_",index=1))
# loading the clinic data and match it with the raw count matrix
#clinic.matrix<-readRDS(clinic.filepath,refhook = NULL)

# coding for PHENGRP:
#1=Multi-organ
#2=Non-acute, Stage I, untreated
#3=Stage II-III, treated
#4=Stage II-III, untreated
#5=Stage IV, treated
#6=Stage IV, untreated
#7=Acute Sarcoidosis, untreated
#8=Remitting, untreated
#9=Cardiac defining therapy

clinic.matrix<-read.table(clinic.filepath,sep="\t",header=T,check.names=F,as.is=TRUE)

# extract the sample IDs for all the rows in clinic.matrix
mkey<-read.xls(file.path(home.dir,"scratch/GRADS/SARC_results/Data/GRADS Master Progress Key 6-13-16.xlsx"),sheet=3,check.names=F,stringsAsFactors = F)
mkey<-mkey[3:nrow(mkey),1:5]
colnames(mkey)<-c("GRADS_ID","BloodKitsEnrollment","BloodKitsMonth6","BALKits","Phenotype")
#grep only the sarcoidosis grads ids (3rd letter S)
temp<-grep("^.{2}S", mkey[,1])
mkey<-mkey[temp,]
#grep only the 209 sarcoidosis BAL kit ids (3rd letter S)
temp<-colnames(count.matrix)
mkey<-mkey[mkey[,4]%in%temp,]
rownames(mkey)<-mkey[,1]
clinic.matrix<-cbind(clinic.matrix,mkey[clinic.matrix[,1],4])
colnames(clinic.matrix)[ncol(clinic.matrix)]<-"kit_id"
rownames(clinic.matrix)<-clinic.matrix$kit_id

clinic.matrix<-clinic.matrix[colnames(count.matrix),]

# replace space in the PHENGRP values with -
#clinic.matrix$PHENGRP<-sapply(clinic.matrix$PHENGRP,my.char.replace,orig=" ",replace="_")
#clinic.matrix$PHENGRP<-sapply(clinic.matrix$PHENGRP,my.char.replace,orig=",",replace="")
#clinic.matrix$PHENGRP<-sapply(clinic.matrix$PHENGRP,my.char.replace,orig="-",replace="_")
```



#### Stage II-III, untreated vs Stage IV, untreated:
```{r}
# comparing:
# Stage II-III untreated vs stage IV untreated
# Stage I untreated vs stage IV untreated
# Stage I, II-III untreated vs Stage IV untreated (BLOOM gene set had pro-inflammatory genes for validation)
#  Acute sarcoid untreated vs stage IV untreated
# patients on steroid vs not on steroid

#1=Multi-organ
#2=Non-acute, Stage I, untreated
#3=Stage II-III, treated
#4=Stage II-III, untreated
#5=Stage IV, treated
#6=Stage IV, untreated
#7=Acute Sarcoidosis, untreated
#8=Remitting, untreated
#9=Cardiac defining therapy

my.coldata<-clinic.matrix
my.coldata<-my.coldata[my.coldata$PHENGRP%in%c(4,6),]
temp.vect<-my.coldata$PHENGRP
temp.vect[temp.vect==4]<-"Stage_II_III_untreated"
temp.vect[temp.vect==6]<-"Stage_IV_untreated"
my.coldata$PHENGRP<-temp.vect
table(my.coldata$PHENGRP)
my.coldata$PHENGRP<-factor(my.coldata$PHENGRP,levels=c("Stage_II_III_untreated","Stage_IV_untreated"))
my.coldata$PHENGRP <- relevel(my.coldata$PHENGRP, ref = "Stage_II_III_untreated")

dds <- DESeqDataSetFromMatrix(countData = count.matrix[,rownames(my.coldata)],
                              colData = my.coldata,
                              design = ~ PHENGRP)
# remove genes with too low count
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]
my.annot.matrix<-annot.matrix[keep,]
dds<-DESeq(dds)
res<-results(dds,name="PHENGRP_Stage_IV_untreated_vs_Stage_II_III_untreated",alpha=0.05)
resLFC <- lfcShrink(dds, coef="PHENGRP_Stage_IV_untreated_vs_Stage_II_III_untreated", type="apeglm")

res05<-results(dds,name="PHENGRP_Stage_IV_untreated_vs_Stage_II_III_untreated",alpha=0.05)
summary(res05)
plotMA(res05)

# output the results
output.filepath<-file.path(output.dir,paste("DEGs_all_",resultsNames(dds)[2],".csv",sep=""))
cmd.out<-cbind(my.annot.matrix,as.data.frame(res))
cmd.out<-cmd.out[order(res$pvalue,decreasing=F),]
resOrdered <- res[order(res$pvalue),]
write.csv(cmd.out, file=output.filepath,row.names = FALSE)

output.filepath<-file.path(output.dir,paste("DEGs_fdr0.05_",resultsNames(dds)[2],".csv",sep=""))
cmd.out<-cmd.out[cmd.out$padj<0.05,]
resSig <- subset(resOrdered, padj < 0.05)
write.csv(cmd.out, file=output.filepath,row.names=FALSE)

# describe how many genes pass the FDR threshold
cat("There are ",sum(res$padj<0.05,na.rm=T)," genes with FDR<0.05.\n",sep="")
```

#### Stage I, untreated vs Stage IV, untreated:
```{r}
# comparing:
# Stage II-III untreated vs stage IV untreated
# Stage I untreated vs stage IV untreated
# Stage I, II-III untreated vs Stage IV untreated (BLOOM gene set had pro-inflammatory genes for validation)
#  Acute sarcoid untreated vs stage IV untreated
# patients on steroid vs not on steroid

#1=Multi-organ
#2=Non-acute, Stage I, untreated
#3=Stage II-III, treated
#4=Stage II-III, untreated
#5=Stage IV, treated
#6=Stage IV, untreated
#7=Acute Sarcoidosis, untreated
#8=Remitting, untreated
#9=Cardiac defining therapy

my.coldata<-clinic.matrix
my.coldata<-my.coldata[my.coldata$PHENGRP%in%c(2,6),]
temp.vect<-my.coldata$PHENGRP
temp.vect[temp.vect==2]<-"Stage_I_untreated"
temp.vect[temp.vect==6]<-"Stage_IV_untreated"
my.coldata$PHENGRP<-temp.vect
table(my.coldata$PHENGRP)
my.coldata$PHENGRP<-factor(my.coldata$PHENGRP,levels=c("Stage_I_untreated","Stage_IV_untreated"))
my.coldata$PHENGRP <- relevel(my.coldata$PHENGRP, ref = "Stage_I_untreated")

dds <- DESeqDataSetFromMatrix(countData = count.matrix[,rownames(my.coldata)],
                              colData = my.coldata,
                              design = ~ PHENGRP)
# remove genes with too low count
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]
my.annot.matrix<-annot.matrix[keep,]
dds<-DESeq(dds)
res<-results(dds,name="PHENGRP_Stage_IV_untreated_vs_Stage_I_untreated",alpha=0.05)
resLFC <- lfcShrink(dds, coef="PHENGRP_Stage_IV_untreated_vs_Stage_I_untreated", type="apeglm")

res05<-results(dds,name="PHENGRP_Stage_IV_untreated_vs_Stage_I_untreated",alpha=0.05)
summary(res05)
plotMA(res05)

# output the results
output.filepath<-file.path(output.dir,paste("DEGs_all_",resultsNames(dds)[2],".csv",sep=""))
cmd.out<-cbind(my.annot.matrix,as.data.frame(res))
cmd.out<-cmd.out[order(res$pvalue,decreasing=F),]
resOrdered <- res[order(res$pvalue),]
write.csv(cmd.out, file=output.filepath,row.names = FALSE)

output.filepath<-file.path(output.dir,paste("DEGs_fdr0.05_",resultsNames(dds)[2],".csv",sep=""))
cmd.out<-cmd.out[cmd.out$padj<0.05,]
resSig <- subset(resOrdered, padj < 0.05)
write.csv(cmd.out, file=output.filepath,row.names=FALSE)

# describe how many genes pass the FDR threshold
cat("There are ",sum(res$padj<0.05,na.rm=T)," genes with FDR<0.05.\n",sep="")
```


#### Stage I, untreated vs Stage II-III, untreated:
```{r}
# comparing:
# Stage II-III untreated vs stage IV untreated
# Stage I untreated vs stage IV untreated
# Stage I, II-III untreated vs Stage IV untreated (BLOOM gene set had pro-inflammatory genes for validation)
#  Acute sarcoid untreated vs stage IV untreated
# patients on steroid vs not on steroid

#1=Multi-organ
#2=Non-acute, Stage I, untreated
#3=Stage II-III, treated
#4=Stage II-III, untreated
#5=Stage IV, treated
#6=Stage IV, untreated
#7=Acute Sarcoidosis, untreated
#8=Remitting, untreated
#9=Cardiac defining therapy

my.coldata<-clinic.matrix
my.coldata<-my.coldata[my.coldata$PHENGRP%in%c(2,4),]
temp.vect<-my.coldata$PHENGRP
temp.vect[temp.vect==2]<-"Stage_I_untreated"
temp.vect[temp.vect==4]<-"Stage_II_III_untreated"
my.coldata$PHENGRP<-temp.vect
table(my.coldata$PHENGRP)
my.coldata$PHENGRP<-factor(my.coldata$PHENGRP,levels=c("Stage_I_untreated","Stage_II_III_untreated"))
my.coldata$PHENGRP <- relevel(my.coldata$PHENGRP, ref = "Stage_I_untreated")

dds <- DESeqDataSetFromMatrix(countData = count.matrix[,rownames(my.coldata)],
                              colData = my.coldata,
                              design = ~ PHENGRP)
# remove genes with too low count
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]
my.annot.matrix<-annot.matrix[keep,]
dds<-DESeq(dds)
res<-results(dds,name="PHENGRP_Stage_II_III_untreated_vs_Stage_I_untreated",alpha=0.05)
resLFC <- lfcShrink(dds, coef="PHENGRP_Stage_II_III_untreated_vs_Stage_I_untreated", type="apeglm")

res05<-results(dds,name="PHENGRP_Stage_II_III_untreated_vs_Stage_I_untreated",alpha=0.05)
summary(res05)
plotMA(res05)

# output the results
output.filepath<-file.path(output.dir,paste("DEGs_all_",resultsNames(dds)[2],".csv",sep=""))
cmd.out<-cbind(my.annot.matrix,as.data.frame(res))
cmd.out<-cmd.out[order(res$pvalue,decreasing=F),]
resOrdered <- res[order(res$pvalue),]
write.csv(cmd.out, file=output.filepath,row.names = FALSE)

output.filepath<-file.path(output.dir,paste("DEGs_fdr0.05_",resultsNames(dds)[2],".csv",sep=""))
cmd.out<-cmd.out[cmd.out$padj<0.05,]
resSig <- subset(resOrdered, padj < 0.05)
write.csv(cmd.out, file=output.filepath,row.names=FALSE)

# describe how many genes pass the FDR threshold
cat("There are ",sum(res$padj<0.05,na.rm=T)," genes with FDR<0.05.\n",sep="")
```

#### Stage I untreated vs Stage II-IV, untreated:
```{r}
# comparing:
# Stage II-III untreated vs stage IV untreated
# Stage I untreated vs stage IV untreated
# Stage I, II-III untreated vs Stage IV untreated (BLOOM gene set had pro-inflammatory genes for validation)
#  Acute sarcoid untreated vs stage IV untreated
# patients on steroid vs not on steroid

#1=Multi-organ
#2=Non-acute, Stage I, untreated
#3=Stage II-III, treated
#4=Stage II-III, untreated
#5=Stage IV, treated
#6=Stage IV, untreated
#7=Acute Sarcoidosis, untreated
#8=Remitting, untreated
#9=Cardiac defining therapy

my.coldata<-clinic.matrix
my.coldata<-my.coldata[my.coldata$PHENGRP%in%c(2,4,6),]
temp.vect<-my.coldata$PHENGRP
temp.vect[temp.vect==2]<-"Non_acute_Stage_I_untreated"
temp.vect[temp.vect==4 | temp.vect==6]<-"Stage_II_IV_untreated"
my.coldata$PHENGRP<-temp.vect
table(my.coldata$PHENGRP)
my.coldata$PHENGRP<-factor(my.coldata$PHENGRP,levels=c("Non_acute_Stage_I_untreated","Stage_II_IV_untreated"))
my.coldata$PHENGRP <- relevel(my.coldata$PHENGRP, ref = "Non_acute_Stage_I_untreated")

dds <- DESeqDataSetFromMatrix(countData = count.matrix[,rownames(my.coldata)],
                              colData = my.coldata,
                              design = ~ PHENGRP)
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]
my.annot.matrix<-annot.matrix[keep,]
dds<-DESeq(dds)
resultsNames(dds)
res<-results(dds,name="PHENGRP_Stage_II_IV_untreated_vs_Non_acute_Stage_I_untreated")
resLFC <- lfcShrink(dds, coef=resultsNames(dds)[2], type="apeglm")

res05<-results(dds,alpha=0.05)
summary(res05)

# output the results
output.filepath<-file.path(output.dir,paste("DEGs_all_",resultsNames(dds)[2],".csv",sep=""))
cmd.out<-cbind(my.annot.matrix,as.data.frame(res))
cmd.out<-cmd.out[order(res$pvalue,decreasing=F),]
resOrdered <- res[order(res$pvalue),]
write.csv(cmd.out, file=output.filepath,row.names = FALSE)

output.filepath<-file.path(output.dir,paste("DEGs_fdr0.05_",resultsNames(dds)[2],".csv",sep=""))
cmd.out<-cmd.out[!is.na(cmd.out$padj) & cmd.out$padj<0.05,]
resSig <- subset(resOrdered, padj < 0.05)
write.csv(cmd.out, file=output.filepath,row.names=FALSE)

# describe how many genes pass the FDR threshold
cat("There are ",sum(res$padj<0.05,na.rm = T)," genes with FDR<0.05.\n",sep="")
```


#### Other comparisons
Acute sarcoid, untreated vs Stage IV, untreated:
```{r}
# comparing:
# Stage II-III untreated vs stage IV untreated
# Stage I untreated vs stage IV untreated
# Stage I, II-III untreated vs Stage IV untreated (BLOOM gene set had pro-inflammatory genes for validation)
#  Acute sarcoid untreated vs stage IV untreated
# patients on steroid vs not on steroid
my.coldata<-clinic.matrix
my.coldata<-my.coldata[my.coldata$PHENGRP%in%c("Acute-Sarcoidosis-untreated","Stage-IV-untreated"),]
my.coldata$PHENGRP<-factor(my.coldata$PHENGRP,levels=c("Acute-Sarcoidosis-untreated","Stage-IV-untreated"))
my.coldata$PHENGRP <- relevel(my.coldata$PHENGRP, ref = "Acute-Sarcoidosis-untreated")
dds <- DESeqDataSetFromMatrix(countData = count.matrix[,rownames(my.coldata)],
                              colData = my.coldata,
                              design = ~ PHENGRP)
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]
my.annot.matrix<-annot.matrix[keep,]
dds<-DESeq(dds)
res<-results(dds)
resLFC <- lfcShrink(dds, coef=resultsNames(dds)[2], type="apeglm")

res05<-results(dds,alpha=0.05)
summary(res05)

# output the results
output.filepath<-file.path(output.dir,paste("DEGs_all_",resultsNames(dds)[2],".csv",sep=""))
cmd.out<-cbind(my.annot.matrix,as.data.frame(res))
cmd.out<-cmd.out[order(res$pvalue,decreasing=F),]
resOrdered <- res[order(res$pvalue),]
write.csv(cmd.out, file=output.filepath,row.names = FALSE)

output.filepath<-file.path(output.dir,paste("DEGs_fdr0.05_",resultsNames(dds)[2],".csv",sep=""))
cmd.out<-cmd.out[!is.na(cmd.out$padj) & cmd.out$padj<0.05,]
resSig <- subset(resOrdered, padj < 0.05)
write.csv(cmd.out, file=output.filepath,row.names=FALSE)

# describe how many genes pass the FDR threshold
cat("There are ",sum(res$padj<0.05,na.rm = T)," genes with FDR<0.05.\n",sep="")
```

Steroid vs non-steroid usage patients:
```{r}
# comparing:
# Stage II-III untreated vs stage IV untreated
# Stage I untreated vs stage IV untreated
# Stage I, II-III untreated vs Stage IV untreated (BLOOM gene set had pro-inflammatory genes for validation)
#  Acute sarcoid untreated vs stage IV untreated
# patients on steroid vs not on steroid
my.coldata<-clinic.matrix[!is.na(clinic.matrix$steroid_atv1),]
my.coldata$steroid_atv1<-factor(my.coldata$steroid_atv1,levels=c(0,1))
my.coldata$steroid_atv1 <- relevel(my.coldata$steroid_atv1, ref = "0")
dds <- DESeqDataSetFromMatrix(countData = count.matrix[,rownames(my.coldata)],
                              colData = my.coldata,
                              design = ~ steroid_atv1)
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]
my.annot.matrix<-annot.matrix[keep,]
dds<-DESeq(dds)
res<-results(dds)
resLFC <- lfcShrink(dds, coef=resultsNames(dds)[2], type="apeglm")

res05<-results(dds,alpha=0.05)
summary(res05)

# output the results
output.filepath<-file.path(output.dir,paste("DEGs_all_",resultsNames(dds)[2],".csv",sep=""))
cmd.out<-cbind(my.annot.matrix,as.data.frame(res))
cmd.out<-cmd.out[order(res$pvalue,decreasing=F),]
resOrdered <- res[order(res$pvalue),]
write.csv(cmd.out, file=output.filepath,row.names = FALSE)

output.filepath<-file.path(output.dir,paste("DEGs_fdr0.05_",resultsNames(dds)[2],".csv",sep=""))
cmd.out<-cmd.out[!is.na(cmd.out$padj) & cmd.out$padj<0.05,]
resSig <- subset(resOrdered, padj < 0.05)
write.csv(cmd.out, file=output.filepath,row.names=FALSE)

# describe how many genes pass the FDR threshold
cat("There are ",sum(res$padj<0.05,na.rm = T)," genes with FDR<0.05.\n",sep="")
```

Stage I vs Stage II-IV, untreated:
```{r}
# comparing:
# Non-acute, Stage I, untreated vs c("Stage II-III, untreated","Stage IV, untreated")
# Stage I untreated vs stage IV untreated
# Stage I, II-III untreated vs Stage IV untreated (BLOOM gene set had pro-inflammatory genes for validation)
#  Acute sarcoid untreated vs stage IV untreated
# patients on steroid vs not on steroid
my.coldata<-clinic.matrix
my.coldata<-my.coldata[my.coldata$PHENGRP%in%c("Non-acute-Stage-I-untreated","Stage-II-III-untreated","Stage-IV-untreated"),]
my.coldata$PHENGRP[my.coldata$PHENGRP%in%c("Stage-II-III-untreated","Stage-IV-untreated")]<-"Stage-II-IV-untreated"

my.coldata$PHENGRP<-factor(my.coldata$PHENGRP,levels=c("Non-acute-Stage-I-untreated","Stage-II-IV-untreated"))
my.coldata$PHENGRP <- relevel(my.coldata$PHENGRP, ref = "Non-acute-Stage-I-untreated")
dds <- DESeqDataSetFromMatrix(countData = count.matrix[,rownames(my.coldata)],
                              colData = my.coldata,
                              design = ~ PHENGRP)
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]
my.annot.matrix<-annot.matrix[keep,]
dds<-DESeq(dds)
res<-results(dds)
resLFC <- lfcShrink(dds, coef=resultsNames(dds)[2], type="apeglm")

res05<-results(dds,alpha=0.05)
summary(res05)

# output the results
output.filepath<-file.path(output.dir,paste("DEGs_all_",resultsNames(dds)[2],".csv",sep=""))
cmd.out<-cbind(my.annot.matrix,as.data.frame(res))
cmd.out<-cmd.out[order(res$pvalue,decreasing=F),]
resOrdered <- res[order(res$pvalue),]
write.csv(cmd.out, file=output.filepath,row.names = FALSE)

output.filepath<-file.path(output.dir,paste("DEGs_fdr0.05_",resultsNames(dds)[2],".csv",sep=""))
cmd.out<-cmd.out[!is.na(cmd.out$padj) & cmd.out$padj<0.05,]
resSig <- subset(resOrdered, padj < 0.05)
write.csv(cmd.out, file=output.filepath,row.names=FALSE)

# describe how many genes pass the FDR threshold
cat("There are ",sum(res$padj<0.05,na.rm = T)," genes with FDR<0.05.\n",sep="")
```


### PCA plot (old)
We generate PCA plot of all the samples labelled using the phenotypic group.

```{r}
library(DESeq2)
library(apeglm)
source(file.path(home.dir,"Rprogram/my_functions.R"))
count.filepath<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/data","rawcounts_baseline_276_clean.txt")
clinic.filepath<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/data/clinic_matrix_merged.RDS")
output.dir<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/DEGs/DESeq2")

# load in the raw counts matrix
count.matrix<-read.table(count.filepath,sep="\t",header=T,check.names=F,stringsAsFactors = F)
annot.matrix<-count.matrix[,1:6]
count.matrix<-count.matrix[,7:ncol(count.matrix)]
rownames(count.matrix)<-count.matrix$GeneID

# loading the clinic data and match it with the raw count matrix
clinic.matrix<-readRDS(clinic.filepath,refhook = NULL)
clinic.matrix<-clinic.matrix[colnames(count.matrix),]

# replace space in the PHENGRP values with -
clinic.matrix$PHENGRP<-sapply(clinic.matrix$PHENGRP,my.char.replace,orig=" ",replace="-")
clinic.matrix$PHENGRP<-sapply(clinic.matrix$PHENGRP,my.char.replace,orig=",",replace="-")
clinic.matrix$PHENGRP<-sapply(clinic.matrix$PHENGRP,my.char.replace,orig="--",replace="-")
```

## Monocyte Proportion

We compare the monocyte cell differential difference between stage I and II-IV to see if the monocyte cell differential is the reason for the enrichment of genes expressed in monocytes in the DEGs.

```{r fig.width=8,fig.height=8}
temp1<-clinic.matrix$p_mono[clinic.matrix$PHENGRP=="Non_acute_Stage_I_untreated"]
temp2<-clinic.matrix$p_mono[clinic.matrix$PHENGRP%in%c("Stage_II_III_untreated")]
wilcox.test(temp1,temp2,alternative = "less")



library(ggplot2)
#boxplot(clinic.matrix$p_mono~clinic.matrix$PHENGRP,besides=T)
temp.data<-data.frame(PHENGRP=clinic.matrix$PHENGRP,monocyte_perc=clinic.matrix$p_mono)
p<-ggplot(temp.data,aes(x=PHENGRP,y=monocyte_perc))+geom_boxplot()+geom_jitter(shape=16,position=position_jitter(0.2))+theme(axis.text.x=element_text(angle=45,vjust=,hjust=1))
p

temp.data<-temp.data[temp.data$PHENGRP%in%c("Non_acute_Stage_I_untreated","Stage_II_III_untreated","Stage_IV_untreated"),]
p<-ggplot(temp.data,aes(x=PHENGRP,y=monocyte_perc))+geom_boxplot()+geom_jitter(shape=16,position=position_jitter(0.2))+theme(axis.text.x=element_text(angle=45,vjust=,hjust=1))
p

```




## edgeR
We identify the differentially expressed genes between give groups of subjects using edgeR.

