---
title: "GSEA analysis of SARC associated genes"
author: "Xiting Yan"
date: "10/02/2019"
output:
  html_notebook:
    code_folding: hide
    fig_caption: yes
    highlight: tango
    number_sections: no
    theme: united
    toc: yes
    toc_depth: 6
    toc_float: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(eval=TRUE,echo = TRUE,cache=TRUE,warning=FALSE,message = FALSE,results='hold',cache.lazy = FALSE)
knitr::opts_knit$set(eval.after = 'fig.cap',dev=c('png','postscript'))


library(captioner)
library(circlize)
library(clusterProfiler)
library(corrplot)
library(ComplexHeatmap)
library(dplyr)
library(EnsDb.Hsapiens.v86)
library(gdata)
library(ggplot2)
library(ggpubr)
library(ggrepel)
library(gplots)
library(grid)
library(gridExtra)
library(gridGraphics)
library(kableExtra)
library(knitr)
library(Matrix)
library(org.Hs.eg.db)
library(parallel)
library(reshape2)
library(scales)
library(Seurat)
library(SingleR)
library(tidyr)
library(tidyverse)
library(viridisLite)
library(xlsx)
library(randomcoloR)

library(kableExtra)
library(gdata)
library(knitr)
library(captioner)
library(nlme)
library(rgl)
library(gplots)
library(WGCNA)
library(xlsx)
library(randomcoloR)
library(ComplexHeatmap)

knit_hooks$set(webgl = hook_webgl)


table_nums_1 <- captioner::captioner(prefix="Table",levels=1)
figure_nums_1<- captioner::captioner(prefix="Figure",levels=1)


table_nums_2 <- captioner::captioner(prefix="Table",levels=2)
figure_nums_2<- captioner::captioner(prefix="Figure",levels=2)

table_nums_3 <- captioner::captioner(prefix="Table",levels=3)
figure_nums_3 <- captioner::captioner(prefix="Figure",levels=3)

home.dir<-"/home/yanxiting/driver_Grace"
#home.dir<-"/home/xy48"
source(paste(home.dir,"/Rprogram/my_functions.R",sep=""))
output.dir<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline")
```

In this note, we generate the input files for GSEA to conduct gene set enrichment analysis of the identified genes differentially expressed between given groups of patients.

# STAGE I vs others

We generate input files for the differential expressed genes between nonacute stage I untreated vs all other groups based on Laura's observations. 



```{r}

# output the merged clinical data matrix as a txt file
clinic.filepath<-"/home/yanxiting/driver_Grace/scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/data/clinic_matrix_merged.RDS"
clinic.matrix<-readRDS(clinic.filepath,refhook = NULL)
output.filepath<-"/home/yanxiting/driver_Grace/scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/data/clinic_matrix_merged.txt"
write.table(clinic.matrix,file=output.filepath,sep="\t",row.names=F,col.names=T,quote=F,append=F)



my_gsea_filecreate_binary<-function(fpkm.filepath,clinic.filepath,var.names,group1.values,group2.values,group1.name="group1",group2.name="group2",output.dir,suffix.name){
  #####################################################################################################################################################
  # Arguments:
  # 
  # gexp.filepath is the file path of the TPM matrix with first 5 columns being annotation for genes 
  # (TPM_baseline_276_clean_celldiffadjusted_withannot.txt).
  # 
  # clinic.filpath is the file path of the clinic data (clinic_matrix_merged.txt). The first row needs to be the column names.
  # 
  # var.names is a vector containing the names of columns in clinic.filepath that you want to generate the phenotype file for.
  #
  # group1.values is a data.frame of string describing values of var.names that are considered as gorup 1. 
  # For example, group1.values=data.frame(PHENGRP="nonacute Stage I untreated"). Note that the columns in group1.values need to match those in var.names if there are
  # more than 1 variable to define the groups.
  #
  # group2.values has the same format as group1.values but is the settings for group2.
  #
  # group1.name is the name you want to call your group1 in the cls file.
  #
  # group2.name is the name you want to call your group2 in the cls file.
  # 
  # output.dir is the folder name where you want to save the created files.
  #
  # suffix.name is the suffix name in the name of the output files. The gct file will be named as gct_suffix.name.gct and cls file will be named as
  # cls_suffix.name.gct. For example, if suffix.name="test", the gct file will be named "gct_test.gct" and the cls file will be named as 
  # "cls_test.cls".
  # 
  # Value:
  #
  # This function will return 0 if successfully ran. Otherwise, it will return 1. If unsuccessful, information regarding what went wrong will be spit 
  # out as warning messages.
  #
  # When ran successfully, this function will create two files under the specified output.dir named gct_var.name.gct and cls_var.name.cls, where 
  # var.name will be the speficied value 
  # in the argument. These two files can be directly loaded into GSEA together with another gene set file that needs to be generated outside of this 
  # function.
  #####################################################################################################################################################
  
  #  generate the gene expression matrix input file for GSEA
  #fpkm.filepath<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/data_adjusted2","TPM_baseline_276_clean_celldiffadjusted_withannot.txt")
  #clinic.filepath<-"/home/yanxiting/driver_Grace/scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/data/clinic_matrix_merged.txt"
  #output.dir<-"/home/yanxiting/driver_Grace/scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/GSEA_knowngenes"
  # suffix.name<-"analysis1"
  
  
  # load in the fpkm matrix
  fpkm.matrix<-read.table(fpkm.filepath,sep="\t",header=T,check.names=F,as.is=TRUE,comment.char = "")
  fpkm.matrix.anno<-fpkm.matrix[,1:5]
  fpkm.matrix<-fpkm.matrix[,6:ncol(fpkm.matrix)]
  
  # load in the clinical data and subset the samples to those listed in the fpkm.matrix
  clinic.matrix=read.table(clinic.filepath,sep="\t",header=T,check.names=F,stringsAsFactors=F)
  rownames(clinic.matrix)<-as.matrix(clinic.matrix)[,1] # GRADS ID
  clinic.matrix<-clinic.matrix[colnames(fpkm.matrix),]

  # based on var.name, group1.values, and group2.values, identify the samples to include in the files.
  if(file.exists(output.dir)==F){
    dir.create(output.dir)
  }
  
  gct.filepath<-file.path(output.dir,paste("gexp_",suffix.name,".gct",sep=""))
  cls.filepath<-file.path(output.dir,paste("cls_",suffix.name,".cls",sep=""))
  
  # substract samples with var.name equal to the values in group1.values and group2. values
  temp1<-fpkm.matrix[,apply(as.data.frame(clinic.matrix[,var.names]),1,paste,collapse="_")%in%apply(group1.values,1,paste,collapse="_")]
  temp2<-fpkm.matrix[,apply(as.data.frame(clinic.matrix[,var.names]),1,paste,collapse="_")%in%apply(group2.values,1,paste,collapse="_")]
  cat("There are ",ncol(temp1)," and ", ncol(temp2)," samples for group1 and group2, respectively.\n")
  data.matrix<-cbind(temp1,temp2)
  data.matrix<-cbind(fpkm.matrix.anno[,1],rep("na",nrow(data.matrix)),data.matrix)
  colnames(data.matrix)[1:2]<-c("NAME","Description")
  pheno.vect<-c(rep(0,ncol(temp1)),rep(1,ncol(temp2)))
  
  # output the gene expression data
  cmd.out<-"#1.2\n"
  cmd.out<-paste(cmd.out,nrow(data.matrix),"\t",ncol(data.matrix)-2,"\n",sep="")
  cat(cmd.out,file=gct.filepath,append=F)
  write.table(data.matrix,file=gct.filepath,append=T,row.names=F,col.names=T,quote=F)
  
  # output the phenotype file
  cmd.out<-paste(ncol(data.matrix)-2,"\t",2,"\t",1,"\n",sep="")
  cmd.out<-paste(cmd.out,"# group1 group2\n",sep="")
  cat(cmd.out,file=cls.filepath,append=F,sep="")
  cat(pheno.vect,file=cls.filepath,append=T,sep=" ")
  
  return(0)
}



fpkm.file=file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/data_adjusted2","TPM_baseline_276_clean_celldiffadjusted_withannot.txt")
clinic.file<-"/home/yanxiting/driver_Grace/scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/data/clinic_matrix_merged.txt"
output.folder<-"/home/yanxiting/driver_Grace/scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/GSEA_knowngenes"

# generate the files for analysis 1
varnames<-c("PHENGRP")
my_gsea_filecreate_binary(fpkm.filepath=fpkm.file,
                          clinic.filepath = clinic.file,
                          var.names = varnames,
                          group1.values = data.frame(PHENGRP=c("Non-acute, Stage I, untreated")),
                          group2.values = data.frame(PHENGRP=c("Stage II-III, untreated","Stage IV, untreated")),
                          output.dir=output.folder,
                          suffix.name="analysis1"
                          )

# generate the files for analysis 2
varnames<-c("PHENGRP")
my_gsea_filecreate_binary(fpkm.filepath=fpkm.file,
                          clinic.filepath = clinic.file,
                          var.names = varnames,
                          group1.values = data.frame(PHENGRP=c("Remitting, untreated")),
                          group2.values = data.frame(PHENGRP=c("Non-acute, Stage I, untreated","Stage II-III, untreated","Stage IV, untreated")),
                          output.dir=output.folder,
                          suffix.name="analysis2"
                          )

# generate the files for analysis 3
varnames<-c("PHENGRP")
my_gsea_filecreate_binary(fpkm.filepath=fpkm.file,
                          clinic.filepath = clinic.file,
                          var.names = varnames,
                          group1.values = data.frame(PHENGRP=c("Stage IV, untreated")),
                          group2.values = data.frame(PHENGRP=c("Non-acute, Stage I, untreated","Stage II-III, untreated")),
                          output.dir=output.folder,
                          suffix.name="analysis3"
                          )
# generate the files for analysis 4. If we directly specify group1.values using numbers, it won't work. I had to specify them to be a space+number as a character.
varnames<-c("nodule_lymph_pheno","steroid_atv1","dmard_atv1")
my_gsea_filecreate_binary(fpkm.filepath=fpkm.file,
                          clinic.filepath = clinic.file,
                          var.names = varnames,
                          group1.values = data.frame(nodule_lymph_pheno=c("lymph"),steroid_atv1=c(" 0"),dmard_atv1=c(" 0")),
                          group2.values = data.frame(nodule_lymph_pheno=c("micronodule","both"),steroid_atv1=c(" 0"," 0"),dmard_atv1=c(" 0"," 0")),
                          output.dir=output.folder,
                          suffix.name="analysis4"
                          )


```


