---
title: "GSEA analysis of SARC associated genes"
author: "Xiting Yan"
date: "10/02/2019"
output:
  html_notebook:
    code_folding: hide
    fig_caption: yes
    highlight: tango
    number_sections: no
    theme: united
    toc: yes
    toc_depth: 6
    toc_float: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(eval=TRUE,echo = TRUE,cache=TRUE,warning=FALSE,message = FALSE,results='hold',cache.lazy = FALSE)
knitr::opts_knit$set(eval.after = 'fig.cap',dev=c('png','postscript'))


library(captioner)
library(circlize)
library(clusterProfiler)
library(corrplot)
library(ComplexHeatmap)
library(dplyr)
library(EnsDb.Hsapiens.v86)
library(gdata)
library(ggplot2)
library(ggpubr)
library(ggrepel)
library(gplots)
library(grid)
library(gridExtra)
library(gridGraphics)
library(kableExtra)
library(knitr)
library(Matrix)
library(org.Hs.eg.db)
library(parallel)
library(reshape2)
library(scales)
library(Seurat)
library(SingleR)
library(tidyr)
library(tidyverse)
library(viridisLite)
library(xlsx)
library(randomcoloR)

library(kableExtra)
library(gdata)
library(knitr)
library(captioner)
library(nlme)
library(rgl)
library(gplots)
library(WGCNA)
library(xlsx)
library(randomcoloR)
library(ComplexHeatmap)

knit_hooks$set(webgl = hook_webgl)


table_nums_1 <- captioner::captioner(prefix="Table",levels=1)
figure_nums_1<- captioner::captioner(prefix="Figure",levels=1)


table_nums_2 <- captioner::captioner(prefix="Table",levels=2)
figure_nums_2<- captioner::captioner(prefix="Figure",levels=2)

table_nums_3 <- captioner::captioner(prefix="Table",levels=3)
figure_nums_3 <- captioner::captioner(prefix="Figure",levels=3)

home.dir<-"/home/yanxiting/driver_Grace"
#home.dir<-"/home/xy48"
source(paste(home.dir,"/Rprogram/my_functions.R",sep=""))
output.dir<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline")
```

In this note, we generate the input files for GSEA to conduct gene set enrichment analysis of genes differentially expressed between different phenotypic groups in UCSF data.

# Data loading

First, we load in the gene expression data, which was normalized using DESeq2. We'll use the log2(DESeq2+1) as the expression data.

```{r}
# load in the normalized count matrix
home.dir<-"/home/yanxiting/driver_Grace"
source(file.path(home.dir,"Rprogram/my_functions.R"))

deseq2.filepath<-file.path(home.dir,"scratch60/Laura_Koth","DESeq2_normalized.RDS")
my.data<-readRDS(deseq2.filepath,refhook = NULL)

anno.data<-my.data[,1:6]
rownames(my.data)<-as.matrix(my.data)[,1]
fpkm.matrix<-my.data[,7:ncol(my.data)]

#log2 transform, if yes, do the analysis on log2(fpkm.matrix+1) to avoid log2(0) transformation
log2.transform<-T
if(log2.transform){
fpkm.matrix<-fpkm.matrix+1
fpkm.matrix<-log(fpkm.matrix,base=2)
}
fpkm<-fpkm.matrix

# filter out genes with no variation and centralize the data 
temp.sd<-apply(fpkm,1,sd)
fpkm<-fpkm[temp.sd>0,]
anno.data<-anno.data[anno.data[,1]%in%rownames(fpkm),]
removed_genes <- rownames(fpkm)[temp.sd<=0] 

# remove outliers
fpkm<-fpkm[,colnames(fpkm)!="sarcoidosis_X235"]

```

Second, we load in the phenotype data.

```{r}
home.dir<-"/home/yanxiting/driver_Grace"
data.filepath<-file.path(home.dir,"scratch60/Laura_Koth/data/cxr_merge_ucsf_seq.xlsx")
clinic.data<-read.xlsx(data.filepath,sheetIndex = 1,check.names=F,stringsAsFactors=FALSE)
```

We match the order of columns in fpkm and rows of clinic.data.

```{r}
temp.names<-intersect(colnames(fpkm),clinic.data$sample_names)
fpkm.data<-fpkm[,temp.names]
rownames(clinic.data)<-clinic.data$sample_names
clinic.data<-clinic.data[temp.names,]
```

We output the matched expression matrix and clinical data into text files.

```{r}
#substract the patients data
fpkm.data<-fpkm.data[,!is.na(clinic.data$CXR_stage)]
clinic.data<-clinic.data[!is.na(clinic.data$CXR_stage),]
fpkm.output.filepath<-file.path(home.dir,"scratch60/Laura_Koth/GSEA_knowngenes_log2DESeq2/DESeq2_log2_sarconly_matched_all.RDS")
clinic.output.filepath<-file.path(home.dir,"scratch60/Laura_Koth/GSEA_knowngenes_log2DESeq2/clinic_data_sarconly_matched_all.RDS")
fpkm.annot.output.filepath<-file.path(home.dir,"scratch60/Laura_Koth/GSEA_knowngenes_log2DESeq2/DESeq2_log2_annot_sarconly_matched_all.RDS")


saveRDS(fpkm.data,file=fpkm.output.filepath,refhook = NULL)
saveRDS(clinic.data,file=clinic.output.filepath,refhook = NULL)
saveRDS(anno.data,file=fpkm.annot.output.filepath,refhook = NULL)
```

# GSEA input files generation (all samples)
We first generate file containing both the clinical matrix and CT scan features so that both group 1 and group 2 in the GSEA analysis can be defined based on combinations of clinical triats and CT scan data.


We generate input files for the differential expressed genes between nonacute stage I untreated vs all other groups based on Laura's observations. 

```{r}
# note this function is slightly different for DESeq2 Norm data and the TPM matrix
#my_gsea_filecreate_binary<-function(fpkm.filepath,clinic.filepath,var.names,group1.values,group2.values,group1.name="group1",group2.name="group2",output.dir,suffix.name){
my_gsea_filecreate_binary<-function(fpkm.matrix,clinic.matrix,fpkm.matrix.anno,var.names,group1.values,group2.values,group1.name="group1",group2.name="group2",output.dir,suffix.name){
  #####################################################################################################################################################
  # Arguments:
  # 
  # gexp.filepath is the file path of the TPM matrix with first 5 columns being annotation for genes 
  # (TPM_baseline_276_clean_celldiffadjusted_withannot.txt).
  # 
  # clinic.filpath is the file path of the clinic data (clinic_matrix_merged.txt). The first row needs to be the column names.
  # 
  # var.names is a vector containing the names of columns in clinic.filepath that you want to generate the phenotype file for.
  #
  # group1.values is a data.frame of string describing values of var.names that are considered as gorup 1. 
  # For example, group1.values=data.frame(PHENGRP="nonacute Stage I untreated"). Note that the columns in group1.values need to match those in var.names if there are
  # more than 1 variable to define the groups.
  #
  # group2.values has the same format as group1.values but is the settings for group2.
  #
  # group1.name is the name you want to call your group1 in the cls file.
  #
  # group2.name is the name you want to call your group2 in the cls file.
  # 
  # output.dir is the folder name where you want to save the created files.
  #
  # suffix.name is the suffix name in the name of the output files. The gct file will be named as gct_suffix.name.gct and cls file will be named as
  # cls_suffix.name.gct. For example, if suffix.name="test", the gct file will be named "gct_test.gct" and the cls file will be named as 
  # "cls_test.cls".
  # 
  # Value:
  #
  # This function will return 0 if successfully ran. Otherwise, it will return 1. If unsuccessful, information regarding what went wrong will be spit 
  # out as warning messages.
  #
  # When ran successfully, this function will create two files under the specified output.dir named gct_var.name.gct and cls_var.name.cls, where 
  # var.name will be the speficied value 
  # in the argument. These two files can be directly loaded into GSEA together with another gene set file that needs to be generated outside of this 
  # function.
  #####################################################################################################################################################
  
  #  generate the gene expression matrix input file for GSEA
  #fpkm.filepath<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/data_adjusted2","TPM_baseline_276_clean_celldiffadjusted_withannot.txt")
  #clinic.filepath<-"/home/yanxiting/driver_Grace/scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/data/clinic_matrix_merged.txt"
  #output.dir<-"/home/yanxiting/driver_Grace/scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/GSEA_knowngenes"
  # suffix.name<-"analysis1"
  
  
  # load in the fpkm matrix
  #fpkm.matrix<-read.table(fpkm.filepath,sep="\t",header=T,check.names=F,as.is=TRUE,comment.char = "")
  #fpkm.matrix.anno<-fpkm.matrix[,1:6]
  #fpkm.matrix<-fpkm.matrix[,7:ncol(fpkm.matrix)]
  
  # load in the clinical data and subset the samples to those listed in the fpkm.matrix
  #clinic.matrix=read.table(clinic.filepath,sep="\t",header=T,check.names=F,stringsAsFactors=F)
  #rownames(clinic.matrix)<-as.matrix(clinic.matrix)[,1] # GRADS ID
  clinic.matrix<-clinic.matrix[colnames(fpkm.matrix),]

  # based on var.name, group1.values, and group2.values, identify the samples to include in the files.
  if(file.exists(output.dir)==F){
    dir.create(output.dir)
  }
  
  gct.filepath<-file.path(output.dir,paste("gexp_",suffix.name,".gct",sep=""))
  cls.filepath<-file.path(output.dir,paste("cls_",suffix.name,".cls",sep=""))
  
  # substract samples with var.name equal to the values in group1.values and group2. values
  temp1<-fpkm.matrix[,apply(as.data.frame(clinic.matrix[,var.names]),1,paste,collapse="_")%in%apply(group1.values,1,paste,collapse="_")]
  #cat("\n")
  #cat(apply(as.data.frame(clinic.matrix[,var.names]),1,paste,collapse="_"))
  #cat("\n")
  temp2<-fpkm.matrix[,apply(as.data.frame(clinic.matrix[,var.names]),1,paste,collapse="_")%in%apply(group2.values,1,paste,collapse="_")]
  cat("There are ",ncol(temp1)," and ", ncol(temp2)," samples for group1 and group2, respectively.\n")
  data.matrix<-cbind(temp1,temp2)
  data.matrix<-cbind(fpkm.matrix.anno[,1],rep("na",nrow(data.matrix)),data.matrix)
  colnames(data.matrix)[1:2]<-c("NAME","Description")
  pheno.vect<-c(rep(0,ncol(temp1)),rep(1,ncol(temp2)))
  
  # output the gene expression data
  cmd.out<-"#1.2\n"
  cmd.out<-paste(cmd.out,nrow(data.matrix),"\t",ncol(data.matrix)-2,"\n",sep="")
  cat(cmd.out,file=gct.filepath,append=F)
  write.table(data.matrix,sep="\t",file=gct.filepath,append=T,row.names=F,col.names=T,quote=F)
  
  # output the phenotype file
  cmd.out<-paste(ncol(data.matrix)-2," ",2," ",1,"\n",sep="")
  cmd.out<-paste(cmd.out,"# group1 group2\n",sep="")
  cat(cmd.out,file=cls.filepath,append=F,sep="")
  cat(pheno.vect,file=cls.filepath,append=T,sep=" ")
  
  results<-list(group1=temp1,group2=temp2)
  
  return(results)
}

output.folder<-file.path(home.dir,"scratch60/Laura_Koth/GSEA_knowngenes_log2DESeq2/all_samples")


# analysis 3b
varnames<-c("CXR_stage")
temp<-my_gsea_filecreate_binary(fpkm.matrix=fpkm.data,
                          clinic.matrix = clinic.data,
                          fpkm.matrix.anno = anno.data,
                          var.names = varnames,
                          group1.values = data.frame(CXR_stage=c("4")),
                          group2.values = data.frame(CXR_stage=c("1","2","3")),
                          output.dir=output.folder,
                          suffix.name="analysis3b"
                          )

# analysis 1b
varnames<-c("CXR_stage")
temp<-my_gsea_filecreate_binary(fpkm.matrix=fpkm.data,
                          clinic.matrix = clinic.data,
                          fpkm.matrix.anno = anno.data,
                          var.names = varnames,
                          group1.values = data.frame(CXR_stage=c("1")),
                          group2.values = data.frame(PHENGRP=c("2","3","4")),
                          output.dir=output.folder,
                          suffix.name="analysis1b"
                          )

```

Here's a table showing the breakdown of the patients based on CXR_stage.
```{r}
table(clinic.data$CXR_stage)
```


# GSEA input files generation (no outliers)

We then generate files containing both the clinical matrix and CT scan features so that both group 1 and group 2 in the GSEA analysis can be defined based on combinations of clinical triats and CT scan data.

First, we load in the list of outliers.
```{r}
outliers.filepath<-file.path(home.dir,"scratch60/Laura_Koth/DESeq2_log2_outlierlist.RDS")
outlier.list<-readRDS(outliers.filepath,refhook = NULL)

#substract the patients data
clinic.data<-clinic.data[!colnames(fpkm.data)%in%outlier.list,]
fpkm.data<-fpkm.data[,!colnames(fpkm.data)%in%outlier.list]

fpkm.output.filepath<-file.path(home.dir,"scratch60/Laura_Koth/GSEA_knowngenes_log2DESeq2/DESeq2_log2_sarconly_matched_nooutlier.RDS")
clinic.output.filepath<-file.path(home.dir,"scratch60/Laura_Koth/GSEA_knowngenes_log2DESeq2/clinic_data_sarconly_matched_nooutlier.RDS")
fpkm.annot.output.filepath<-file.path(home.dir,"scratch60/Laura_Koth/GSEA_knowngenes_log2DESeq2/DESeq2_log2_annot_sarconly_matched_nooutlier.RDS")


saveRDS(fpkm.data,file=fpkm.output.filepath,refhook = NULL)
saveRDS(clinic.data,file=clinic.output.filepath,refhook = NULL)
saveRDS(anno.data,file=fpkm.annot.output.filepath,refhook = NULL)
```

We generate input files for the differential expressed genes between nonacute stage I untreated vs all other groups based on Laura's observations. 

```{r}
# note this function is slightly different for DESeq2 Norm data and the TPM matrix
#my_gsea_filecreate_binary<-function(fpkm.filepath,clinic.filepath,var.names,group1.values,group2.values,group1.name="group1",group2.name="group2",output.dir,suffix.name){
output.folder<-file.path(home.dir,"scratch60/Laura_Koth/GSEA_knowngenes_log2DESeq2/no_outliers")


# analysis 3b
varnames<-c("CXR_stage")
temp<-my_gsea_filecreate_binary(fpkm.matrix=fpkm.data,
                          clinic.matrix = clinic.data,
                          fpkm.matrix.anno = anno.data,
                          var.names = varnames,
                          group1.values = data.frame(CXR_stage=c("4")),
                          group2.values = data.frame(CXR_stage=c("1","2","3")),
                          output.dir=output.folder,
                          suffix.name="analysis3b"
                          )

# analysis 1b
varnames<-c("CXR_stage")
temp<-my_gsea_filecreate_binary(fpkm.matrix=fpkm.data,
                          clinic.matrix = clinic.data,
                          fpkm.matrix.anno = anno.data,
                          var.names = varnames,
                          group1.values = data.frame(CXR_stage=c("1")),
                          group2.values = data.frame(PHENGRP=c("2","3","4")),
                          output.dir=output.folder,
                          suffix.name="analysis1b"
                          )

```

Here's a table showing the breakdown of the patients based on CXR_stage.
```{r}
table(clinic.data$CXR_stage)
```
__________________________________________________________________________________________________
Codes below these lines are not run but serve as backup codes.
__________________________________________________________________________________________________


```{r leftover,eval=F}
# These are left over codes from the previous codes.
# generate the files for analysis 2
varnames<-c("CXR_stage")
temp<-my_gsea_filecreate_binary(fpkm.filepath=fpkm.file,
                          clinic.filepath = clinic.file,
                          var.names = varnames,
                          group1.values = data.frame(PHENGRP=c("Remitting, untreated")),
                          group2.values = data.frame(PHENGRP=c("Non-acute, Stage I, untreated","Stage II-III, untreated","Stage IV, untreated")),
                          output.dir=output.folder,
                          suffix.name="analysis2"
                          )

# generate the files for analysis 3
varnames<-c("PHENGRP")
temp<-my_gsea_filecreate_binary(fpkm.filepath=fpkm.file,
                          clinic.filepath = clinic.file,
                          var.names = varnames,
                          group1.values = data.frame(PHENGRP=c("Stage IV, untreated")),
                          group2.values = data.frame(PHENGRP=c("Non-acute, Stage I, untreated","Stage II-III, untreated")),
                          output.dir=output.folder,
                          suffix.name="analysis3"
                          )
# generate the files for analysis 4. If we directly specify group1.values using numbers, it won't work. I had to specify them to be a space+number as a character.
varnames<-c("nodule_lymph_pheno","steroid_atv1","dmard_atv1")
temp<-my_gsea_filecreate_binary(fpkm.filepath=fpkm.file,
                          clinic.filepath = clinic.file,
                          var.names = varnames,
                          group1.values = data.frame(nodule_lymph_pheno=c("lymph"),steroid_atv1=c(" 0"),dmard_atv1=c(" 0")),
                          group2.values = data.frame(nodule_lymph_pheno=c("micronodule","both"),steroid_atv1=c(" 0"," 0"),dmard_atv1=c(" 0"," 0")),
                          output.dir=output.folder,
                          suffix.name="analysis4"
                          )

# generate the files for analysis 5. 
varnames<-c("PHENGRP")
temp<-my_gsea_filecreate_binary(fpkm.filepath=fpkm.file,
                          clinic.filepath = clinic.file,
                          var.names = varnames,
                          group1.values = data.frame(PHENGRP=c("Remitting, untreated")),
                          group2.values = data.frame(PHENGRP=c("Stage II-III, untreated")),
                          output.dir=output.folder,
                          suffix.name="analysis5"
                          )


# generate the files for analysis 4a. If we directly specify group1.values using numbers, it won't work. I had to specify them to be a space+number as a character.
clinic.filepath<-"/home/yanxiting/driver_Grace/scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/data/clinic_ct_merge_matrix_withgex.RDS"
clinic.matrix<-readRDS(clinic.filepath,refhook = NULL)
clinic.file<-"/home/yanxiting/driver_Grace/scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/data/clinic_ct_merge_matrix_withgex.txt"

varnames<-c("nodule_lymph_pheno","steroid_atv1","dmard_atv1","Ground_Glass","Honeycombing","Reticular_Abnormality")
temp<-my_gsea_filecreate_binary(fpkm.filepath=fpkm.file,
                          clinic.filepath = clinic.file,
                          var.names = varnames,
                          group1.values = data.frame(nodule_lymph_pheno=c("lymph"),steroid_atv1=c(" 0"),dmard_atv1=c(" 0"),Ground_Glass=c(" 0"),Honeycombing=c(" 0"),Reticular_Abnormality=c(" 0")),
                          group2.values = data.frame(nodule_lymph_pheno=c("micronodule","both"),steroid_atv1=c(" 0"," 0"),dmard_atv1=c(" 0"," 0"),Ground_Glass=c(" 0"," 0"),Honeycombing=c(" 0"," 0"),Reticular_Abnormality=c(" 0"," 0")),
                          output.dir=output.folder,
                          suffix.name="analysis4a"
                          )

# generate the files for analysis 4b. If we directly specify group1.values using numbers, it won't work. I had to specify them to be a space+number as a character.
clinic.filepath<-"/home/yanxiting/driver_Grace/scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/data/clinic_ct_merge_matrix_withgex.RDS"
clinic.matrix<-readRDS(clinic.filepath,refhook = NULL)
clinic.file<-"/home/yanxiting/driver_Grace/scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/data/clinic_ct_merge_matrix_withgex.txt"

varnames<-c("nodule_lymph_pheno","steroid_atv1","dmard_atv1","Ground_Glass","Honeycombing","Reticular_Abnormality")
temp<-my_gsea_filecreate_binary(fpkm.filepath=fpkm.file,
                          clinic.filepath = clinic.file,
                          var.names = varnames,
                          group1.values = data.frame(nodule_lymph_pheno=c("lymph"),steroid_atv1=c(" 0"),dmard_atv1=c(" 0"),Ground_Glass=c(" 0"),Honeycombing=c(" 0"),Reticular_Abnormality=c(" 0")),
                          group2.values = data.frame(nodule_lymph_pheno=c("micronodule"),steroid_atv1=c(" 0"),dmard_atv1=c(" 0"),Ground_Glass=c(" 0"),Honeycombing=c(" 0"),Reticular_Abnormality=c(" 0")),
                          output.dir=output.folder,
                          suffix.name="analysis4b"
                          )
```

We also generate GSEA input files for patients from given lists including the following:

* Stage I, untreated and lymphadenopathy only overlapped patients 
* Stage I, untread only patients
* lymphadenopathy only patients

These three analysis is to see if the overlapped patients drove the significant results.

```{r}
grads.id.list<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/GSEA_knowngenes_DESeq2_log2adjusted/gradsid_overlap_nonoverlap.xlsx")
grads.id.matrix<-read.xls(grads.id.list,sheet=1)
overlap.id<-as.character(grads.id.matrix[,1])[as.character(grads.id.matrix[,1])!=""]
stage1only.id<-as.character(grads.id.matrix[,2])[as.character(grads.id.matrix[,2])!=""]
lymphonly.id<-as.character(grads.id.matrix[,3])[as.character(grads.id.matrix[,3])!=""]


#----------------------------------------------------------------------------------------------------------------
# Part 1: 12 stage I, untreated and lymph only patients
#----------------------------------------------------------------------------------------------------------------
# load in the data
fpkm.file=file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/data_adjusted2","DESeq2_normalized_276_clean_log2_celldiffadjusted_withannot.txt")
clinic.file<-"/home/yanxiting/driver_Grace/scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/data/clinic_matrix_merged.txt"
output.folder<-"/home/yanxiting/driver_Grace/scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/GSEA_knowngenes_DESeq2_log2adjusted"
suffix.name<-"analysis6"
var.names=c("PHENGRP")
group2.values = data.frame(PHENGRP=c("Stage II-III, untreated","Stage IV, untreated"))

# load in the fpkm matrix
fpkm.matrix<-read.table(fpkm.file,sep="\t",header=T,check.names=F,as.is=TRUE,comment.char = "")
fpkm.matrix.anno<-fpkm.matrix[,1:6]
fpkm.matrix<-fpkm.matrix[,7:ncol(fpkm.matrix)]
  
# load in the clinical data and subset the samples to those listed in the fpkm.matrix
clinic.matrix=read.table(clinic.file,sep="\t",header=T,check.names=F,stringsAsFactors=F)
rownames(clinic.matrix)<-as.matrix(clinic.matrix)[,1] # GRADS ID
clinic.matrix<-clinic.matrix[colnames(fpkm.matrix),]

# based on var.name, group1.values, and group2.values, identify the samples to include in the files.
if(file.exists(output.folder)==F){
  dir.create(output.folder)
}
  
gct.filepath<-file.path(output.folder,paste("gexp_",suffix.name,".gct",sep=""))
cls.filepath<-file.path(output.folder,paste("cls_",suffix.name,".cls",sep=""))
  
# substract samples with var.name equal to the values in group1.values and group2. values
#temp1<-fpkm.matrix[,apply(as.data.frame(clinic.matrix[,var.names]),1,paste,collapse="_")%in%apply(group1.values,1,paste,collapse="_")]
temp1<-fpkm.matrix[,overlap.id]
temp2<-fpkm.matrix[,apply(as.data.frame(clinic.matrix[,var.names]),1,paste,collapse="_")%in%apply(group2.values,1,paste,collapse="_")]

cat("There are ",ncol(temp1)," and ", ncol(temp2)," samples for group1 and group2, respectively.\n")
data.matrix<-cbind(temp1,temp2)
data.matrix<-cbind(fpkm.matrix.anno[,1],rep("na",nrow(data.matrix)),data.matrix)
colnames(data.matrix)[1:2]<-c("NAME","Description")
pheno.vect<-c(rep(0,ncol(temp1)),rep(1,ncol(temp2)))

# output the gene expression data
cmd.out<-"#1.2\n"
cmd.out<-paste(cmd.out,nrow(data.matrix),"\t",ncol(data.matrix)-2,"\n",sep="")
cat(cmd.out,file=gct.filepath,append=F)
write.table(data.matrix,sep="\t",file=gct.filepath,append=T,row.names=F,col.names=T,quote=F)

# output the phenotype file
cmd.out<-paste(ncol(data.matrix)-2," ",2," ",1,"\n",sep="")
cmd.out<-paste(cmd.out,"# group1 group2\n",sep="")
cat(cmd.out,file=cls.filepath,append=F,sep="")
cat(pheno.vect,file=cls.filepath,append=T,sep=" ")
#----------------------------------------------------------------------------------------------------------------

# part 2: stage I only
#----------------------------------------------------------------------------------------------------------------
# load in the data
fpkm.file=file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/data_adjusted2","DESeq2_normalized_276_clean_log2_celldiffadjusted_withannot.txt")
clinic.file<-"/home/yanxiting/driver_Grace/scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/data/clinic_matrix_merged.txt"
output.folder<-"/home/yanxiting/driver_Grace/scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/GSEA_knowngenes_DESeq2_log2adjusted"
suffix.name<-"analysis7"
var.names=c("PHENGRP")
group2.values = data.frame(PHENGRP=c("Stage II-III, untreated","Stage IV, untreated"))

# load in the fpkm matrix
fpkm.matrix<-read.table(fpkm.file,sep="\t",header=T,check.names=F,as.is=TRUE,comment.char = "")
fpkm.matrix.anno<-fpkm.matrix[,1:6]
fpkm.matrix<-fpkm.matrix[,7:ncol(fpkm.matrix)]
  
# load in the clinical data and subset the samples to those listed in the fpkm.matrix
clinic.matrix=read.table(clinic.file,sep="\t",header=T,check.names=F,stringsAsFactors=F)
rownames(clinic.matrix)<-as.matrix(clinic.matrix)[,1] # GRADS ID
clinic.matrix<-clinic.matrix[colnames(fpkm.matrix),]

# based on var.name, group1.values, and group2.values, identify the samples to include in the files.
if(file.exists(output.folder)==F){
  dir.create(output.folder)
}
  
gct.filepath<-file.path(output.folder,paste("gexp_",suffix.name,".gct",sep=""))
cls.filepath<-file.path(output.folder,paste("cls_",suffix.name,".cls",sep=""))
  
# substract samples with var.name equal to the values in group1.values and group2. values
#temp1<-fpkm.matrix[,apply(as.data.frame(clinic.matrix[,var.names]),1,paste,collapse="_")%in%apply(group1.values,1,paste,collapse="_")]
temp1<-fpkm.matrix[,stage1only.id]
temp2<-fpkm.matrix[,apply(as.data.frame(clinic.matrix[,var.names]),1,paste,collapse="_")%in%apply(group2.values,1,paste,collapse="_")]

cat("There are ",ncol(temp1)," and ", ncol(temp2)," samples for group1 and group2, respectively.\n")
data.matrix<-cbind(temp1,temp2)
data.matrix<-cbind(fpkm.matrix.anno[,1],rep("na",nrow(data.matrix)),data.matrix)
colnames(data.matrix)[1:2]<-c("NAME","Description")
pheno.vect<-c(rep(0,ncol(temp1)),rep(1,ncol(temp2)))

# output the gene expression data
cmd.out<-"#1.2\n"
cmd.out<-paste(cmd.out,nrow(data.matrix),"\t",ncol(data.matrix)-2,"\n",sep="")
cat(cmd.out,file=gct.filepath,append=F)
write.table(data.matrix,sep="\t",file=gct.filepath,append=T,row.names=F,col.names=T,quote=F)

# output the phenotype file
cmd.out<-paste(ncol(data.matrix)-2," ",2," ",1,"\n",sep="")
cmd.out<-paste(cmd.out,"# group1 group2\n",sep="")
cat(cmd.out,file=cls.filepath,append=F,sep="")
cat(pheno.vect,file=cls.filepath,append=T,sep=" ")
#----------------------------------------------------------------------------------------------------------------
# part 3: lymph only
#----------------------------------------------------------------------------------------------------------------
# load in the data
fpkm.file=file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/data_adjusted2","DESeq2_normalized_276_clean_log2_celldiffadjusted_withannot.txt")
clinic.file<-"/home/yanxiting/driver_Grace/scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/data/clinic_matrix_merged.txt"
output.folder<-"/home/yanxiting/driver_Grace/scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/GSEA_knowngenes_DESeq2_log2adjusted"
suffix.name<-"analysis8"
var.names=c("PHENGRP")
group2.values = data.frame(PHENGRP=c("Stage II-III, untreated","Stage IV, untreated"))

# load in the fpkm matrix
fpkm.matrix<-read.table(fpkm.file,sep="\t",header=T,check.names=F,as.is=TRUE,comment.char = "")
fpkm.matrix.anno<-fpkm.matrix[,1:6]
fpkm.matrix<-fpkm.matrix[,7:ncol(fpkm.matrix)]
  
# load in the clinical data and subset the samples to those listed in the fpkm.matrix
clinic.matrix=read.table(clinic.file,sep="\t",header=T,check.names=F,stringsAsFactors=F)
rownames(clinic.matrix)<-as.matrix(clinic.matrix)[,1] # GRADS ID
clinic.matrix<-clinic.matrix[colnames(fpkm.matrix),]

# based on var.name, group1.values, and group2.values, identify the samples to include in the files.
if(file.exists(output.folder)==F){
  dir.create(output.folder)
}
  
gct.filepath<-file.path(output.folder,paste("gexp_",suffix.name,".gct",sep=""))
cls.filepath<-file.path(output.folder,paste("cls_",suffix.name,".cls",sep=""))
  
# substract samples with var.name equal to the values in group1.values and group2. values
#temp1<-fpkm.matrix[,apply(as.data.frame(clinic.matrix[,var.names]),1,paste,collapse="_")%in%apply(group1.values,1,paste,collapse="_")]
temp1<-fpkm.matrix[,lymphonly.id]
temp2<-fpkm.matrix[,apply(as.data.frame(clinic.matrix[,var.names]),1,paste,collapse="_")%in%apply(group2.values,1,paste,collapse="_")]

cat("There are ",ncol(temp1)," and ", ncol(temp2)," samples for group1 and group2, respectively.\n")
data.matrix<-cbind(temp1,temp2)
data.matrix<-cbind(fpkm.matrix.anno[,1],rep("na",nrow(data.matrix)),data.matrix)
colnames(data.matrix)[1:2]<-c("NAME","Description")
pheno.vect<-c(rep(0,ncol(temp1)),rep(1,ncol(temp2)))

# output the gene expression data
cmd.out<-"#1.2\n"
cmd.out<-paste(cmd.out,nrow(data.matrix),"\t",ncol(data.matrix)-2,"\n",sep="")
cat(cmd.out,file=gct.filepath,append=F)
write.table(data.matrix,sep="\t",file=gct.filepath,append=T,row.names=F,col.names=T,quote=F)

# output the phenotype file
cmd.out<-paste(ncol(data.matrix)-2," ",2," ",1,"\n",sep="")
cmd.out<-paste(cmd.out,"# group1 group2\n",sep="")
cat(cmd.out,file=cls.filepath,append=F,sep="")
cat(pheno.vect,file=cls.filepath,append=T,sep=" ")
#----------------------------------------------------------------------------------------------------------------
```

We examine the CT features of these three groups of patients.
```{r}
# load in the list of GRADS IDs in all three groups
grads.id.list<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/GSEA_knowngenes_DESeq2_log2adjusted/gradsid_overlap_nonoverlap.xlsx")
grads.id.matrix<-read.xls(grads.id.list,sheet=1)
overlap.id<-as.character(grads.id.matrix[,1])[as.character(grads.id.matrix[,1])!=""]
stage1only.id<-as.character(grads.id.matrix[,2])[as.character(grads.id.matrix[,2])!=""]
lymphonly.id<-as.character(grads.id.matrix[,3])[as.character(grads.id.matrix[,3])!=""]

# load in the clinical data
clinic.filepath<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/data/clinic_matrix_merged.RDS")
clinic.matrix<-readRDS(clinic.filepath,refhook = NULL)
clinic.matrix<-clinic.matrix[,c("GENDER","RACE","ethn","AGE","ethor","wbc","cd4","cal","d25","d125","crp","p_lymph","p_mono","p_neut","p_eos","p_baso","FVCPRED","FEV1PRED","PREDDLCO","SCADDING","smoke","pk_yr","steroid_atv1","dmard_atv1","Micronodule","Med_Lymphadenopathy","Hilar_Lymphadenopathy","nodule_lymph_pheno")]
clinic.matrix<-clinic.matrix[c(overlap.id,stage1only.id,lymphonly.id),]

# generate the summary table

var.names<-c("GENDER","RACE","AGE","ethor","wbc","cd4","cal","d25","d125","crp","p_lymph","p_mono","p_neut","p_eos","p_baso","FVCPRED","FEV1PRED","PREDDLCO","SCADDING","smoke","pk_yr","steroid_atv1","dmard_atv1","Micronodule","Med_Lymphadenopathy","Hilar_Lymphadenopathy","nodule_lymph_pheno")

var.names.cat<-c("GENDER","RACE","ethor","smoke","steroid_atv1","dmard_atv1","SCADDING","Micronodule","Med_Lymphadenopathy","Hilar_Lymphadenopathy","nodule_lymph_pheno")
var.names.con<-c("AGE","wbc","cd4","cal","d25","d125","crp","p_lymph","p_mono","p_neut","p_eos","p_baso","FVCPRED","FEV1PRED","PREDDLCO","pk_yr")

var.type<-rep("con",length(var.names))
var.type[var.names%in%var.names.cat]<-"cat"
var.annot<-var.names

summary.table<-character()

for(i in 1:length(var.names)){
  if(var.type[i]=="cat"){
    temp.matrix<-as.matrix(table(clinic.matrix[overlap.id,var.names[i]]))
    rownames(temp.matrix)<-paste(var.annot[i]," - ",rownames(temp.matrix))
    
    temp.sum<-sum(temp.matrix)
    temp.matrix2<-temp.matrix/temp.sum
    summary.table<-rbind(summary.table,as.matrix(paste(temp.matrix,"(",round(temp.matrix2,digits=2),")",sep="")))
    rownames(summary.table)[(nrow(summary.table)-nrow(temp.matrix)+1):nrow(summary.table)]<-rownames(temp.matrix)
    }else{
    temp.mean<-mean(clinic.matrix[overlap.id,var.names[i]],na.rm=T)
    temp.sd<-sd(clinic.matrix[overlap.id,var.names[i]],na.rm=T)
    temp.mean<-round(temp.mean,digits=2)
    temp.sd<-round(temp.sd,digits=2)
    summary.table<-rbind(summary.table,paste(temp.mean,"±",temp.sd,sep=""))
    rownames(summary.table)[nrow(summary.table)]<-paste(var.annot[i],", mean±SD",sep="")
  }
  
}

colnames(summary.table)[1]<-"Overlap"
summary.table.total<-summary.table


summary.table<-character()

for(i in 1:length(var.names)){
  if(var.type[i]=="cat"){
    temp.matrix<-as.matrix(table(clinic.matrix[stage1only.id,var.names[i]]))
    rownames(temp.matrix)<-paste(var.annot[i]," - ",rownames(temp.matrix))
    
    temp.sum<-sum(temp.matrix)
    temp.matrix2<-temp.matrix/temp.sum
    summary.table<-rbind(summary.table,as.matrix(paste(temp.matrix,"(",round(temp.matrix2,digits=2),")",sep="")))
    rownames(summary.table)[(nrow(summary.table)-nrow(temp.matrix)+1):nrow(summary.table)]<-rownames(temp.matrix)
    }else{
    temp.mean<-mean(clinic.matrix[stage1only.id,var.names[i]],na.rm=T)
    temp.sd<-sd(clinic.matrix[stage1only.id,var.names[i]],na.rm=T)
    temp.mean<-round(temp.mean,digits=2)
    temp.sd<-round(temp.sd,digits=2)
    summary.table<-rbind(summary.table,paste(temp.mean,"±",temp.sd,sep=""))
    rownames(summary.table)[nrow(summary.table)]<-paste(var.annot[i],", mean±SD",sep="")
  }
  
}

colnames(summary.table)[1]<-"Stage I only"

temp.names<-union(rownames(summary.table.total),rownames(summary.table))
temp<-matrix("",nrow=length(temp.names),ncol=2)
rownames(temp)<-temp.names
temp[rownames(summary.table.total),1]<-summary.table.total[,1]
temp[rownames(summary.table),2]<-summary.table[,1]
summary.table.total<-temp


summary.table<-character()

for(i in 1:length(var.names)){
  if(var.type[i]=="cat"){
    temp.matrix<-as.matrix(table(clinic.matrix[lymphonly.id,var.names[i]]))
    rownames(temp.matrix)<-paste(var.annot[i]," - ",rownames(temp.matrix))
    
    temp.sum<-sum(temp.matrix)
    temp.matrix2<-temp.matrix/temp.sum
    summary.table<-rbind(summary.table,as.matrix(paste(temp.matrix,"(",round(temp.matrix2,digits=2),")",sep="")))
    rownames(summary.table)[(nrow(summary.table)-nrow(temp.matrix)+1):nrow(summary.table)]<-rownames(temp.matrix)
    }else{
    temp.mean<-mean(clinic.matrix[lymphonly.id,var.names[i]],na.rm=T)
    temp.sd<-sd(clinic.matrix[lymphonly.id,var.names[i]],na.rm=T)
    temp.mean<-round(temp.mean,digits=2)
    temp.sd<-round(temp.sd,digits=2)
    summary.table<-rbind(summary.table,paste(temp.mean,"±",temp.sd,sep=""))
    rownames(summary.table)[nrow(summary.table)]<-paste(var.annot[i],", mean±SD",sep="")
  }
  
}

colnames(summary.table)[1]<-"Lymph only"

temp.names<-union(rownames(summary.table.total),rownames(summary.table))
temp<-matrix("",nrow=length(temp.names),ncol=3)
rownames(temp)<-temp.names
temp[rownames(summary.table.total),1]<-summary.table.total[,1]
temp[rownames(summary.table.total),2]<-summary.table.total[,2]
temp[rownames(summary.table),3]<-summary.table[,1]

summary.table.total<-temp

colnames(summary.table.total)<-c("overlap","Stage 1 only","Lymph only")


# calculate the p values to assess the differences between the three groups of patients for each trait

pvalue.vect<-numeric()

for(i in 1:length(var.names)){
  
  if(var.type[i]=="cat"){
    temp.matrix<-as.matrix(table(clinic.matrix[c(overlap.id,stage1only.id,lymphonly.id),var.names[i]]))
    rownames(temp.matrix)<-paste(var.annot[i]," - ",rownames(temp.matrix))
    
    temp1<-clinic.matrix[c(overlap.id,stage1only.id,lymphonly.id),var.names[i]]
    temp2<-c(rep("overlap",length(overlap.id)),rep("stage1",length(stage1only.id)),rep("lymph",length(lymphonly.id)))
    temp.index<-1:length(temp1)
    temp.index<-temp.index[!(is.na(temp1) | is.na(temp2))]
    temp1<-temp1[temp.index]
    temp2<-temp2[temp.index]
    
    if(length(unique(temp1))==1){
      pvalue.vect<-c(pvalue.vect, rep(1,length(unique(temp1))))
      names(pvalue.vect)[length(pvalue.vect)]<-rownames(temp.matrix)
    }else{
      pvalue.vect<-c(pvalue.vect, rep(chisq.test(temp1,as.factor(temp2))$p.value,length(unique(temp1))))  
      names(pvalue.vect)[(length(pvalue.vect)-length(unique(temp1))+1):length(pvalue.vect)]<-rownames(temp.matrix)
    }
    
    
    }else{
    temp1<-clinic.matrix[c(overlap.id,stage1only.id,lymphonly.id),var.names[i]]
    temp2<-c(rep("overlap",length(overlap.id)),rep("stage1",length(stage1only.id)),rep("lymph",length(lymphonly.id)))
    pvalue.vect<-c(pvalue.vect, kruskal.test(temp1~as.factor(temp2))$p.value)
    names(pvalue.vect)[length(pvalue.vect)]<-paste(var.annot[i],", mean±SD",sep="")
  }
  
}

summary.table.total<-cbind(summary.table.total,pvalue.vect[rownames(summary.table.total)])
colnames(summary.table.total)[4]<-"p value"

output.filepath<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/GSEA_knowngenes_DESeq2_log2adjusted/gradsid_overlap_nonoverlap_clinicfeatures.xlsx")
write.xlsx(summary.table.total,file=output.filepath,row.names=T,col.names=T,append=F)

```

We also generate the feature table for the CT scan variables.

```{r}
# load in the list of GRADS IDs in all three groups
grads.id.list<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/GSEA_knowngenes_DESeq2_log2adjusted/gradsid_overlap_nonoverlap.xlsx")
grads.id.matrix<-read.xls(grads.id.list,sheet=1)
overlap.id<-as.character(grads.id.matrix[,1])[as.character(grads.id.matrix[,1])!=""]
stage1only.id<-as.character(grads.id.matrix[,2])[as.character(grads.id.matrix[,2])!=""]
lymphonly.id<-as.character(grads.id.matrix[,3])[as.character(grads.id.matrix[,3])!=""]

# load in the clinical data
ct.filepath<-file.path(home.dir,"scratch/GRADS/SARC_results/Data/CT_data_20170712/Sarc_ct_reads_corrected.xls")
ct.matrix<-read.xls(ct.filepath,sheet=1)
rownames(ct.matrix)<-as.matrix(ct.matrix)[,"GRADSID"]
ct.matrix<-ct.matrix[,c("Med_Lymphadenopathy","Hilar_Lymphadenopathy","Micronodule","Bronchial_Wall_Thickening","Traction_Bronchiectasis","Bronchiectasis_severity","Ground_Glass","Honeycombing","Reticular_Abnormality","Pulmonary_Art","Tree_in_bud")]

ct.matrix<-ct.matrix[c(overlap.id,stage1only.id,lymphonly.id),]
group.vect<-c(rep("overlap",length(overlap.id)),rep("Stage I",length(stage1only.id)),rep("Lymph",length(lymphonly.id)))

# recode Med_Lymphadenopathy and Hilar_Lymphadenopathy
temp<-ct.matrix[,"Med_Lymphadenopathy"]
temp1<-rep(0,length(temp))
temp1[temp%in%c(1,2)]<-1
temp1[temp==3]<-2   # 0=no, 1=one side, 2=billateral
ct.matrix[,"Med_Lymphadenopathy"]<-temp1


temp<-ct.matrix[,"Hilar_Lymphadenopathy"]
temp1<-rep(0,length(temp))
temp1[temp%in%c(1,2)]<-1
temp1[temp==3]<-2   # 0=no, 1=one side, 2=billateral
ct.matrix[,"Hilar_Lymphadenopathy"]<-temp1


# generate the summary table
var.names<-c("Med_Lymphadenopathy","Hilar_Lymphadenopathy","Micronodule","Bronchial_Wall_Thickening","Traction_Bronchiectasis","Bronchiectasis_severity","Ground_Glass","Honeycombing","Reticular_Abnormality","Pulmonary_Art","Tree_in_bud")

summary.table<-character()

for(i in 1:length(var.names)){
  
    temp.matrix<-as.matrix(table(ct.matrix[overlap.id,var.names[i]]))
    rownames(temp.matrix)<-paste(var.names[i]," - ",rownames(temp.matrix))
    
    temp.sum<-sum(temp.matrix)
    temp.matrix2<-temp.matrix/temp.sum
    summary.table<-rbind(summary.table,as.matrix(paste(temp.matrix,"(",round(temp.matrix2,digits=2),")",sep="")))
    rownames(summary.table)[(nrow(summary.table)-nrow(temp.matrix)+1):nrow(summary.table)]<-rownames(temp.matrix)
}

colnames(summary.table)[1]<-"Overlap"
summary.table.total<-summary.table

summary.table<-character()

for(i in 1:length(var.names)){
  
    temp.matrix<-as.matrix(table(ct.matrix[stage1only.id,var.names[i]]))
    rownames(temp.matrix)<-paste(var.names[i]," - ",rownames(temp.matrix))
    
    temp.sum<-sum(temp.matrix)
    temp.matrix2<-temp.matrix/temp.sum
    summary.table<-rbind(summary.table,as.matrix(paste(temp.matrix,"(",round(temp.matrix2,digits=2),")",sep="")))
    rownames(summary.table)[(nrow(summary.table)-nrow(temp.matrix)+1):nrow(summary.table)]<-rownames(temp.matrix)
}

colnames(summary.table)[1]<-"Stage I"


temp.names<-union(rownames(summary.table.total),rownames(summary.table))
temp<-matrix("",nrow=length(temp.names),ncol=2)
rownames(temp)<-temp.names
temp[rownames(summary.table.total),1]<-summary.table.total[,1]
temp[rownames(summary.table),2]<-summary.table[,1]
summary.table.total<-temp


summary.table<-character()

for(i in 1:length(var.names)){
  
    temp.matrix<-as.matrix(table(ct.matrix[lymphonly.id,var.names[i]]))
    rownames(temp.matrix)<-paste(var.names[i]," - ",rownames(temp.matrix))
    
    temp.sum<-sum(temp.matrix)
    temp.matrix2<-temp.matrix/temp.sum
    summary.table<-rbind(summary.table,as.matrix(paste(temp.matrix,"(",round(temp.matrix2,digits=2),")",sep="")))
    rownames(summary.table)[(nrow(summary.table)-nrow(temp.matrix)+1):nrow(summary.table)]<-rownames(temp.matrix)
}

colnames(summary.table)[1]<-"Lymph only"


temp.names<-union(rownames(summary.table.total),rownames(summary.table))
temp<-matrix("",nrow=length(temp.names),ncol=3)
rownames(temp)<-temp.names
temp[rownames(summary.table.total),1]<-summary.table.total[,1]
temp[rownames(summary.table.total),2]<-summary.table.total[,2]
temp[rownames(summary.table),3]<-summary.table[,1]

summary.table.total<-temp

colnames(summary.table.total)<-c("overlap","Stage 1 only","Lymph only")


# calculate the p values to assess the differences between the three groups of patients for each trait

pvalue.vect<-numeric()

for(i in 1:length(var.names)){
  
    temp.matrix<-as.matrix(table(ct.matrix[c(overlap.id,stage1only.id,lymphonly.id),var.names[i]]))
    rownames(temp.matrix)<-paste(var.names[i]," - ",rownames(temp.matrix))
    
    temp1<-ct.matrix[c(overlap.id,stage1only.id,lymphonly.id),var.names[i]]
    temp2<-c(rep("overlap",length(overlap.id)),rep("stage1",length(stage1only.id)),rep("lymph",length(lymphonly.id)))
    temp.index<-1:length(temp1)
    temp.index<-temp.index[!(is.na(temp1) | is.na(temp2))]
    temp1<-temp1[temp.index]
    temp2<-temp2[temp.index]
    
    if(length(unique(temp1))==1){
      pvalue.vect<-c(pvalue.vect, rep(1,length(unique(temp1))))
      names(pvalue.vect)[length(pvalue.vect)]<-rownames(temp.matrix)
    }else{
      pvalue.vect<-c(pvalue.vect, rep(chisq.test(temp1,as.factor(temp2))$p.value,length(unique(temp1))))  
      names(pvalue.vect)[(length(pvalue.vect)-length(unique(temp1))+1):length(pvalue.vect)]<-rownames(temp.matrix)
    }
}

summary.table.total<-cbind(summary.table.total,pvalue.vect[rownames(summary.table.total)])
colnames(summary.table.total)[4]<-"p value"

output.filepath<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/GSEA_knowngenes_DESeq2_log2adjusted/gradsid_overlap_nonoverlap_ctfeatures.xlsx")
write.xlsx(summary.table.total,file=output.filepath,row.names=T,col.names=T,append=F)
```


We also generated the GSEA input files for three three groups of patients using the function we wrote.
```{r}

# output the merged clinical data matrix as a txt file
clinic.filepath<-"/home/yanxiting/driver_Grace/scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/data/clinic_matrix_merged.RDS"
clinic.matrix<-readRDS(clinic.filepath,refhook = NULL)
output.filepath<-"/home/yanxiting/driver_Grace/scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/data/clinic_matrix_merged.txt"
write.table(clinic.matrix,file=output.filepath,sep="\t",row.names=F,col.names=T,quote=F,append=F)


# note this function is slightly different for DESeq2 Norm data and the TPM matrix
my_gsea_filecreate_binary<-function(fpkm.filepath,clinic.filepath,var.names,group1.values,group2.values,group1.name="group1",group2.name="group2",output.dir,suffix.name){
  #####################################################################################################################################################
  # Arguments:
  # 
  # gexp.filepath is the file path of the TPM matrix with first 5 columns being annotation for genes 
  # (TPM_baseline_276_clean_celldiffadjusted_withannot.txt).
  # 
  # clinic.filpath is the file path of the clinic data (clinic_matrix_merged.txt). The first row needs to be the column names.
  # 
  # var.names is a vector containing the names of columns in clinic.filepath that you want to generate the phenotype file for.
  #
  # group1.values is a data.frame of string describing values of var.names that are considered as gorup 1. 
  # For example, group1.values=data.frame(PHENGRP="nonacute Stage I untreated"). Note that the columns in group1.values need to match those in var.names if there are
  # more than 1 variable to define the groups.
  #
  # group2.values has the same format as group1.values but is the settings for group2.
  #
  # group1.name is the name you want to call your group1 in the cls file.
  #
  # group2.name is the name you want to call your group2 in the cls file.
  # 
  # output.dir is the folder name where you want to save the created files.
  #
  # suffix.name is the suffix name in the name of the output files. The gct file will be named as gct_suffix.name.gct and cls file will be named as
  # cls_suffix.name.gct. For example, if suffix.name="test", the gct file will be named "gct_test.gct" and the cls file will be named as 
  # "cls_test.cls".
  # 
  # Value:
  #
  # This function will return 0 if successfully ran. Otherwise, it will return 1. If unsuccessful, information regarding what went wrong will be spit 
  # out as warning messages.
  #
  # When ran successfully, this function will create two files under the specified output.dir named gct_var.name.gct and cls_var.name.cls, where 
  # var.name will be the speficied value 
  # in the argument. These two files can be directly loaded into GSEA together with another gene set file that needs to be generated outside of this 
  # function.
  #####################################################################################################################################################
  
  #  generate the gene expression matrix input file for GSEA
  #fpkm.filepath<-file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/data_adjusted2","TPM_baseline_276_clean_celldiffadjusted_withannot.txt")
  #clinic.filepath<-"/home/yanxiting/driver_Grace/scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/data/clinic_matrix_merged.txt"
  #output.dir<-"/home/yanxiting/driver_Grace/scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/GSEA_knowngenes"
  # suffix.name<-"analysis1"
  
  
  # load in the fpkm matrix
  fpkm.matrix<-read.table(fpkm.filepath,sep="\t",header=T,check.names=F,as.is=TRUE,comment.char = "")
  fpkm.matrix.anno<-fpkm.matrix[,1:6]
  fpkm.matrix<-fpkm.matrix[,7:ncol(fpkm.matrix)]
  
  # load in the clinical data and subset the samples to those listed in the fpkm.matrix
  clinic.matrix=read.table(clinic.filepath,sep="\t",header=T,check.names=F,stringsAsFactors=F)
  rownames(clinic.matrix)<-as.matrix(clinic.matrix)[,1] # GRADS ID
  clinic.matrix<-clinic.matrix[colnames(fpkm.matrix),]

  # based on var.name, group1.values, and group2.values, identify the samples to include in the files.
  if(file.exists(output.dir)==F){
    dir.create(output.dir)
  }
  
  gct.filepath<-file.path(output.dir,paste("gexp_",suffix.name,".gct",sep=""))
  cls.filepath<-file.path(output.dir,paste("cls_",suffix.name,".cls",sep=""))
  
  # substract samples with var.name equal to the values in group1.values and group2. values
  temp1<-fpkm.matrix[,apply(as.data.frame(clinic.matrix[,var.names]),1,paste,collapse="_")%in%apply(group1.values,1,paste,collapse="_")]
  temp2<-fpkm.matrix[,apply(as.data.frame(clinic.matrix[,var.names]),1,paste,collapse="_")%in%apply(group2.values,1,paste,collapse="_")]
  cat("There are ",ncol(temp1)," and ", ncol(temp2)," samples for group1 and group2, respectively.\n")
  data.matrix<-cbind(temp1,temp2)
  data.matrix<-cbind(fpkm.matrix.anno[,1],rep("na",nrow(data.matrix)),data.matrix)
  colnames(data.matrix)[1:2]<-c("NAME","Description")
  pheno.vect<-c(rep(0,ncol(temp1)),rep(1,ncol(temp2)))
  
  # output the gene expression data
  cmd.out<-"#1.2\n"
  cmd.out<-paste(cmd.out,nrow(data.matrix),"\t",ncol(data.matrix)-2,"\n",sep="")
  cat(cmd.out,file=gct.filepath,append=F)
  write.table(data.matrix,file=gct.filepath,append=T,row.names=F,col.names=T,quote=F)
  
  # output the phenotype file
  cmd.out<-paste(ncol(data.matrix)-2,"\t",2,"\t",1,"\n",sep="")
  cmd.out<-paste(cmd.out,"# group1 group2\n",sep="")
  cat(cmd.out,file=cls.filepath,append=F,sep="")
  cat(pheno.vect,file=cls.filepath,append=T,sep=" ")
  
  return(0)
}



fpkm.file=file.path(home.dir,"scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/data_adjusted2","DESeq2_normalized_276_clean_log2_celldiffadjusted_withannot.txt")
clinic.file<-"/home/yanxiting/driver_Grace/scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/data/clinic_matrix_merged.txt"
output.folder<-"/home/yanxiting/driver_Grace/scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/GSEA_knowngenes_DESeq2_log2adjusted"

# generate the files for analysis 9: overlapped patients
varnames<-c("PHENGRP","nodule_lymph_pheno")
my_gsea_filecreate_binary(fpkm.filepath=fpkm.file,
                          clinic.filepath = clinic.file,
                          var.names = varnames,
                          group1.values = data.frame(PHENGRP=c("Non-acute, Stage I, untreated"),nodule_lymph_pheno=c("lymph")),
                          group2.values = data.frame(PHENGRP=c("Stage II-III, untreated","Stage II-III, untreated","Stage II-III, untreated","Stage IV, untreated","Stage IV, untreated","Stage IV, untreated"),nodule_lymph_pheno=c("lymph","micronodule","both","lymph","micronodule","both")),
                          output.dir=output.folder,
                          suffix.name="analysis9"
                          )

# generate the files for analysis 10: Stage I only
varnames<-c("PHENGRP","nodule_lymph_pheno")
my_gsea_filecreate_binary(fpkm.filepath=fpkm.file,
                          clinic.filepath = clinic.file,
                          var.names = varnames,
                          group1.values = data.frame(PHENGRP=c("Non-acute, Stage I, untreated","Non-acute, Stage I, untreated"),nodule_lymph_pheno=c("micronodule","both")),
                          group2.values = data.frame(PHENGRP=c("Stage II-III, untreated","Stage II-III, untreated","Stage II-III, untreated","Stage IV, untreated","Stage IV, untreated","Stage IV, untreated"),nodule_lymph_pheno=c("lymph","micronodule","both","lymph","micronodule","both")),
                          output.dir=output.folder,
                          suffix.name="analysis10"
                          )

# generate the files for analysis 11: lymph only (no stage I)
varnames<-c("PHENGRP","nodule_lymph_pheno")
phen.value<-c("Non-acute, Stage I, untreated","Cardiac defining therapy","Stage IV, treated","Stage IV, untreated","Stage II-III, untreated","Remitting, untreated","Multi-organ", "Stage II-III, treated","Acute Sarcoidosis, untreated")
phen.value<-setdiff(phen.value,"Non-acute, Stage I, untreated")

my_gsea_filecreate_binary(fpkm.filepath=fpkm.file,
                          clinic.filepath = clinic.file,
                          var.names = varnames,
                          group1.values = data.frame(PHENGRP=phen.value,nodule_lymph_pheno=rep("lymph",length(phen.value))),
                          group2.values = data.frame(PHENGRP=c("Stage II-III, untreated","Stage II-III, untreated","Stage II-III, untreated","Stage IV, untreated","Stage IV, untreated","Stage IV, untreated"),nodule_lymph_pheno=c("lymph","micronodule","both","lymph","micronodule","both")),
                          output.dir=output.folder,
                          suffix.name="analysis11"
                          )


```

## Comparing to 10X Genomicx data

We downloaded a previously published 10X Genomicx data and compared the genes in Bloom data to see if they are enriched in cell type markers or which cell types are expressing them.


### Standard Seurat

First, we load in the raw count data and redo the preprocessing using standard Seurat pipeline.
```{r fig.width=8,fig.width=8}
# load in the scRNA-seq data of SARC PBMC samples with annotations that was previously published
library(Matrix)
count.filepath<-"/home/yanxiting/Documents/Research/GRADS_SARC_PBMC/Data/GSE132338_matrix_counts.mtx"
coldata.filepath<-"/home/yanxiting/Documents/Research/GRADS_SARC_PBMC/Data/GSE132338_matrix_colData.txt"
rowdata.filepath<-"/home/yanxiting/Documents/Research/GRADS_SARC_PBMC/Data/GSE132338_matrix_rowData.mtx"
data.filepath<-"/home/yanxiting/Documents/Research/GRADS_SARC_PBMC/Data/GSE132338_SingleCellExperiment.RDS"
sc.data<-readRDS(data.filepath,refhook=NULL)

count.data<-readMM(count.filepath)
mycol.data<-read.table(coldata.filepath,sep="\t",row.names=1,header=T)
myrow.data<-readLines(rowdata.filepath)

# create a seurat object with the count data and put the integrated data in
rownames(count.data)<-myrow.data[2:length(myrow.data)]
colnames(count.data)<-rownames(mycol.data)

# create a Seurat object using the count matrix
my.data<-CreateSeuratObject(counts=count.data,project="PBMC",min.cells=0,min.features = 0)
my.data[["percent.mt"]] <- PercentageFeatureSet(my.data, pattern = "^MT-")

# Visualize QC metrics as a violin plot
VlnPlot(my.data, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

plot1 <- FeatureScatter(my.data, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(my.data, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
grid.arrange(plot1 , plot2,ncol=2)

my.data <- NormalizeData(my.data, normalization.method = "LogNormalize", scale.factor = 10000)
my.data <- FindVariableFeatures(my.data, selection.method = "vst", nfeatures = 2000)

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(my.data), 10)

# plot variable features with and without labels
plot1 <- VariableFeaturePlot(my.data)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
grid.arrange(plot1,plot2,ncol=2)
```

```{r fig.height=8,fig.width=8,message=FALSE}
# scale the data
all.genes <- rownames(my.data)
my.data <- ScaleData(my.data, features = all.genes)

# conduct PCA
my.data <- RunPCA(my.data, features = VariableFeatures(object = my.data))
VizDimLoadings(my.data, dims = 1:2, reduction = "pca")
DimPlot(my.data, reduction = "pca")
```


```{r fig.width=10,fig.height=24}
DimHeatmap(my.data, dims = 1:15, cells = 500, balanced = TRUE)
```

```{r fig.width=8,fig.height=8,message=FALSE}
# decide the number of PC to use
my.data <- JackStraw(my.data, dims = 40,num.replicate = 100)
my.data <- ScoreJackStraw(my.data, dims = 1:40)
JackStrawPlot(my.data, dims = 1:40)
```

We decides to use the top 20 PCs.
```{r fig.width=8, fig.height=8}
num.pc<-20
my.data <- FindNeighbors(my.data, dims = 1:num.pc)
my.data <- FindClusters(my.data, resolution = 0.5)
my.data <- RunUMAP(my.data, dims = 1:num.pc)
DimPlot(my.data, reduction = "umap")
```

Draw the UMAP with annotation from GEO.

```{r fig.width=8,fig.height=8,message=FALSE}
# add the cell type annotation into the data
temp.data<-sc.data@meta.data
temp.data<-temp.data[rownames(my.data@meta.data),]
my.data@meta.data<-cbind(my.data@meta.data,temp.data)
Idents(my.data)<-my.data@meta.data$celltype

my.distinct.colors<-c('#e6194b' , '#3cb44b', '#ffe119', '#4363d8', '#f58231', '#911eb4', '#46f0f0', '#f032e6', '#bcf60c', '#fabebe', '#008080', '#e6beff', '#9a6324', '#808080', '#800000',  '#808000', '#ffd8b1', '#000075',  '#ffffff', '#000000','#fffac8','#aaffc3')
celltype.colors<-my.distinct.colors[1:length(unique(my.data@meta.data$celltype))]

DimPlot(my.data, reduction = "umap",cols=celltype.colors,shuffle=T)

Idents(my.data)<-my.data@meta.data$donor
DimPlot(my.data, reduction = "umap",shuffle=T)
```
Save the Seurat object.
```{r}
saveRDS(my.data, file ="/home/yanxiting/Documents/Research/GRADS_SARC_PBMC/scRNA_compare/GSE132338_seurat_nointegration.rds")
```


We check the expression of the analysis4b genes in all cell types.
```{r fig.width=10,fig.height=5,message=FALSE}
my.distinct.colors<-c('#e6194b' , '#3cb44b', '#ffe119', '#4363d8', '#f58231', '#911eb4', '#46f0f0', '#f032e6', '#bcf60c', '#fabebe', '#008080', '#e6beff', '#9a6324', '#808080', '#800000',  '#808000', '#ffd8b1', '#000075',  '#ffffff', '#000000','#fffac8','#aaffc3')
celltype.colors<-my.distinct.colors[1:length(unique(my.data@meta.data$celltype))]
bloom.filepath<-"/home/yanxiting/Documents/Research/GRADS_SARC_PBMC/Data/analysis4b_BLOOM.tsv"
bloom.data<-as.data.frame(read_tsv(bloom.filepath))
bloom.genes<-bloom.data[bloom.data$`CORE ENRICHMENT`=="Yes","SYMBOL"]

temp.names<-intersect(bloom.genes,rownames(sc.data@assays$integrated@counts))

#output.filepath<-"~/driver_Grace/scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/scRNA_compare/featureplot_bloom_4b.pdf"
#output.filepath<-"/home/yanxiting/Documents/Research/GRADS_SARC_PBMC/scRNA_compare/featureplot_bloom_4b.pdf"
#pdf(output.filepath,width = 8,height=8,onefile = T)
#g.list<-list()

for(i in 1:length(temp.names)){
#temp.data<-my.data@meta.data
#g.1<-ggplot(temp.data,aes(x=UMAP_1,y=UMAP_2,colour=celltype))+geom_point(size=0.1,shape=19)+scale_color_manual(values=celltype.colors)+guides(color = guide_legend(override.aes = list(size = 8)))

Idents(my.data)<-my.data@meta.data$celltype
g.1<-DimPlot(my.data, reduction = "umap",cols=celltype.colors,shuffle=T)
g.2<-FeaturePlot(my.data,features=temp.names[i],pt.size = 0.1,order=T)
g<-grid.arrange(g.1,g.2,ncol=2)

output.filepath<-file.path("/home/yanxiting/Documents/Research/GRADS_SARC_PBMC/scRNA_compare",paste("featureplot_bloom_4b_",temp.names[i],".pdf"))
ggsave(output.filepath,g,device="pdf",width=10,height=5)
}

```


### Integrated Analysis
Conduct integrated analysis to remove donor effect. We have not succeeded in the integration analysis due to the very small number of cells per subject.
```{r fig.width=8,fig.height=8,message=FALSE}
# split the dataset into a list of two seurat objects (stim and CTRL)
ifnb.list <- SplitObject(my.data, split.by = "donor")
temp.size<-matrix(unlist(lapply(ifnb.list,dim)),ncol=2,byrow=T)[,2]
#ifnb.list<-ifnb.list[temp.size>=30]

# normalize and identify variable features for each dataset independently
ifnb.list <- lapply(X = ifnb.list, FUN = function(x) {
    x <- NormalizeData(x)
    x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 2000)
})

# select features that are repeatedly variable across datasets for integration
features <- SelectIntegrationFeatures(object.list = ifnb.list)
ifnb.list <- lapply(X = ifnb.list, FUN = function(x) {
    x <- ScaleData(x, features = features, verbose = FALSE)
    x <- RunPCA(x, features = features, verbose = FALSE, approx=FALSE)
})

immune.anchors <- FindIntegrationAnchors(object.list = ifnb.list, anchor.features = features, reduction = "rpca")
immune.combined <- IntegrateData(anchorset = immune.anchors)


my.anchors <- FindIntegrationAnchors(object.list = ifnb.list, anchor.features = features)
my.combined <- IntegrateData(anchorset = my.anchors)

DefaultAssay(my.combined) <- "integrated"

# Run the standard workflow for visualization and clustering
my.combined <- ScaleData(my.combined, verbose = FALSE)
my.combined <- RunPCA(my.combined, npcs = 30, verbose = FALSE)
my.combined <- RunUMAP(my.combined, reduction = "pca", dims = 1:30)
my.combined <- FindNeighbors(my.combined, reduction = "pca", dims = 1:30)
my.combined <- FindClusters(my.combined, resolution = 0.5)
p1 <- DimPlot(my.combined, reduction = "umap", group.by = "donor")
p2 <- DimPlot(my.combined, reduction = "umap", label = TRUE, repel = TRUE,shuffle=T)
```



```{r}
# left over
my.data@assays$integrated<-sc.data@assays$integrated
my.data@meta.data<-cbind(my.data@meta.data,mycol.data)

Idents(my.data)<-my.data@meta.data$celltype
# identify the cell type marker genes
#Idents(sc.data)<-sc.data@meta.data$celltype
celltype.names<-unique(as.character(my.data@meta.data$celltype))
sc.markers.result<-list()

for(i in 1:length(celltype.names)){
  sc.markers.result[[i]]<-FindMarkers(my.data,ident.1=celltype.names[i],min.pct=0.05)
}

# save the marker lists
#output.filepath<-file.path("~/driver_Grace/scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/data/GSE132338_SingleCellExperiment_markerlist.RDS")
output.filepath<-"/home/yanxiting/Documents/Research/GRADS_SARC_PBMC/scRNA_compare/GSE132338_SingleCellExperiment_markerlist.RDS"
saveRDS(sc.markers.result,file=output.filepath,refhook = NULL)
```

Draw the UMAPs to show the expression of bloom genes in this 10X data.
```{r fig.width=8,fig.height=8}
#bloom.filepath<-"~/driver_Grace/scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/scRNA_compare/analysis4b_BLOOM.tsv"
my.distinct.colors<-c('#e6194b' , '#3cb44b', '#ffe119', '#4363d8', '#f58231', '#911eb4', '#46f0f0', '#f032e6', '#bcf60c', '#fabebe', '#008080', '#e6beff', '#9a6324', '#808080', '#800000',  '#808000', '#ffd8b1', '#000075',  '#ffffff', '#000000','#fffac8','#aaffc3')
celltype.colors<-my.distinct.colors[1:length(unique(my.data@meta.data$celltype))]
bloom.filepath<-"/home/yanxiting/Documents/Research/GRADS_SARC_PBMC/Data/analysis4b_BLOOM.tsv"
bloom.data<-as.data.frame(read_tsv(bloom.filepath))
bloom.genes<-bloom.data[bloom.data$`CORE ENRICHMENT`=="Yes","SYMBOL"]

temp.names<-intersect(bloom.genes,rownames(sc.data@assays$integrated@counts))

#output.filepath<-"~/driver_Grace/scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/scRNA_compare/featureplot_bloom_4b.pdf"
output.filepath<-"/home/yanxiting/Documents/Research/GRADS_SARC_PBMC/scRNA_compare/featureplot_bloom_4b.pdf"
pdf(output.filepath,width = 8,height=8,onefile = T)
#g.list<-list()

for(i in 1:length(temp.names)){
temp.data<-my.data@meta.data
g.1<-ggplot(temp.data,aes(x=UMAP_1,y=UMAP_2,colour=celltype))+geom_point(size=0.1,shape=19)+scale_color_manual(values=celltype.colors)+guides(color = guide_legend(override.aes = list(size = 8)))

temp.color<-sc.data@assays$integrated@counts[temp.names[i],]
temp.data<-cbind(sc.data@meta.data,temp.color)
colnames(temp.data)[ncol(temp.data)]<-"col"
g.2<-ggplot(temp.data,aes(x=UMAP_1,y=UMAP_2,colour=col))+geom_point(size=0.01,shape=19)+scale_color_gradient(low="grey",high="red")+ggtitle(temp.names[i])

#output.filepath<-file.path("~/driver_Grace/scratch/GRADS/SARC_results/Results_summary_PBMC_hg38/baseline/scRNA_compare",paste("featureplot_bloom_4b_",temp.names[i],".pdf"))
g<-grid.arrange(g.1,g.2,ncol=2)
g
output.filepath<-file.path("/home/yanxiting/Documents/Research/GRADS_SARC_PBMC/scRNA_compare",paste("featureplot_bloom_4b_",temp.names[i],".pdf"))
ggsave(output.filepath,g,device="pdf",width=8,height=8)

}
dev.off()


```

First, we evaluate the enrichment of BLOOM genes in the marker gene lists
