---
title: "BAL manuscript"
author: "Xiting Yan"
date: "11/29/2018"
output:
  html_notebook:
    code_folding: hide
    fig_caption: yes
    highlight: tango
    number_sections: yes
    theme: united
    toc: yes
    toc_depth: 6
    toc_float: yes
header-includes: \usepackage[dvips]{graphicx}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(eval=TRUE,echo = TRUE,cache.lazy = FALSE,cache=TRUE)
knitr::opts_knit$set(eval.after = 'fig.cap',dev=c('png','postscript'))

library(kableExtra)
library(gdata)
library(knitr)
library(captioner)
library(nlme)
library(gridExtra)
library(reshape2)
library(ggplot2)
library(gplots)
library(WGCNA)
library(networkD3)
library(htmlwidgets)
library(NbClust)
library(clValid)
library(ggplotify)
library(grid)

table_nums <- captioner::captioner(prefix="Table",levels=1)
figure_nums<- captioner::captioner(prefix="Figure",levels=1)

table_nums_3 <- captioner::captioner(prefix="Table",levels=3)
figure_nums_3 <- captioner::captioner(prefix="Figure",levels=3)

work.dir<-"/home/yanxiting/driver_Grace"
#work.dir<-"/Users/yanxiting/MyVolumes/GRACE"
source(paste(work.dir,"/Rprogram/my_functions.R",sep=""))

```

# Introduction


# Methods

## Data Preprocessing

### Data Quality Control
We assess the quality of the raw sequencing data using both the quality summary from the Ion Torrent Server and the FastQC program. Experiments with uneven loading and uneven distribution of reads were repeated. In the raw sequencing data, no significant adapter sequences were found by FastQC and therefore, we did not trim the raw reads for any adapter sequences. 

### Mapping and Gene Expression Level Estimation
The raw sequencing data was mapped to the human genome (UCSC hg38) using a two-stage mapping strategy suggested by the Ion Torrent company. In the first stage, the reads were mapped to the human genome using STAR. In the second stage, reads that couldn't be mapped in the first stage were again aligned to the human genome using Bowtie2 with local alignment optimization. The mapping results from these two stages were merged using samtools and further fed to cufflinks to estimate the gene expression levels. 

### Batch Effect Assessment
To assess the batch effect in this dataset, we performed principal component analysis (PCA) on the whole dataset and calculate the distance between every two samples in the space formed by the top 3 principal components (PCs). The correlation bewteen this distance and the number of days between the processing dates of the corresponding two samples was visualized and examined using Spearman correlation test. 

### Mislabeling Correction
It is common for study with this scale to have sample mislabeling. We tried to correct for the mislabeling in two ways. First, we visualized all the samples, including the BAL and the PBMC samples, using PCA plots. The BAL samples and the PBMC samples formed two separate big clusters. Some BAL samples, however, were mixed with the PBMC samples in the PCA plots indicating that these samples are actually mislabelled as BAL samples. Secondly, we performed a SNP calling analysis on all the samples and also visualize the data using the PCA on the identified SNPs. Based on the fact that the BAL and the PBMC samples from the same subjects are supposed to have the same mutations, we were able to identify 12 samples that were mislabelled between the two compartments and find their true labels. The other mislabelled BAL samples for which we were not able to find the corresponding subjects using the SNP calling analysis were removed from further analysis. 

### Outlier Removal
After the mislabling correction, we performed the PCA on the BAL samples only to identify outliers. The identified outliers were removed from further analysis. 


## Supervised Analysis
We conducted the supervised analysis to identify gene signatures associated with each clinical trait. For continuous clinical traits, spearman correlation test was used. For categorical clinical traits, chi-square test was used to assess the significance. 

## Unsupervised Analysis
To identify gene signatures associated with clinical traits in an unsupervised way, we applied the WGCNA analysis to the dataset, which clusters the genes into gene modules and evaluates the correlation between the eigen gene of each module with the clinical traits. The following analysis pipeline was used. First, we filter out genes with no variation. Second, based on the hierarchical clustering tree of the samples using the genes from the first step, we chose a few cutting heights on the hierarchical clustering tree to remove outliers. For each given cutting height, samples merging into the tree with a higher height will be considered as outliers and removed. Third, for each cutting height, we apply the WGCNA with default setting to the filtered samples to obtain the gene modules. Finally, we compare the clustering resuts of different cutting heights to choose the highest cutting heights that has a stable clustering results.   

## Validation by an Independent Cohort
We validated our findings using an independent cohort from Freiburg, Germany. This cohort has recruited patients with Sarcoidosis. BAL samples were collected from the patients and profiled for gene expression using the Affymetrix HuGene 2.0 Chip. 

# Results

## Data Summary

### Figure 1. Study workflow

```{r fig.cap=figure_nums(name="fig1a_study_workflow_2",caption="Diagram explaining the work flow of the study (horizontal version)."),echo=FALSE,cache=FALSE,results='hide'}
```
`r figure_nums(name="fig1a_study_workflow_2")`
```{r eval=TRUE,echo=TRUE,cache=FALSE}
knitr::include_graphics(file.path(work.dir,"scratch/GRADS/SARC_results/Results_summary_BAL_hg38/workflow_chart/GRADS_SARC_flowchart_horizontal_v2.jpg"))
```

### Table 1a. Patient Demographics

```{r table1a_xy}
clinic.data<-read.table(paste(work.dir,"/scratch/GRADS/SARC_results/Data/SARC_combine_clinical_data20170814.txt",sep=""),sep="\t",header=T,check.names=F,stringsAsFactors=F)
var.names<-c("SCADDING","GENDER","RACE","AGE","BDFEV1","FEV1PRED","BDFVC","FVCPRED","BDDLCO","PREDDLCO")
var.type<-c("cat","cat","cat","con","con","con","con","con","con","con")

# coding for the phenotype
#1 ---- Multi-organ
#2 ---- Non-acute, Stage I, untreated
#3 ---- Stage II-III, treated
#4 ---- Stage II-III, untreated
#5 ---- Stage IV, treated
#6 ---- Stage IV, untreated
#7 ---- Acute Sarcoidosis, untreated
#8 ---- Remitting, untreated
#9 ---- Cardiac defining therapy

phengrp.annot<-c("Multi-organ",
                 "Non-acute, Stage I, untreated",
                 "Stage II-III, treated",
                 "Stage II-III, untreated",
                 "Stage IV, treated",
                 "Stage IV, untreated",
                 "Acute Sarcoidosis, untreated",
                 "Remitting, untreated",
                 "Cardiac defining therapy")

cmd.out<-character()
for(i in 1:length(var.names)){
  if(var.type[i]=="cat"){
    temp<-table(clinic.data[,var.names[i]],clinic.data[,"PHENGRP"],useNA="ifany")
    temp.matrix<-as.matrix(temp)
    rownames(temp.matrix)<-paste(var.names[i]," ",row.names(temp),sep="")
    colnames(temp.matrix)<-phengrp.annot[as.numeric(colnames(temp.matrix))]
    temp2<-temp.matrix/matrix(rep(apply(temp.matrix,2,sum),nrow(temp.matrix)),byrow=T,nrow=nrow(temp.matrix))
    temp2<-temp2*100
    my.matrix<-matrix(paste(temp.matrix,"(",formatC(temp2,digits = 1,format="f"),"%)",sep=""),byrow=F,nrow=nrow(temp.matrix))
    rownames(my.matrix)<-rownames(temp.matrix)
    colnames(my.matrix)<-colnames(temp.matrix)
    
    cmd.out<-rbind(cmd.out,my.matrix)
  
  }else{
    temp.list<-split(clinic.data[,var.names[i]],clinic.data$PHENGRP)
    temp.mean<-unlist(lapply(temp.list,mean,na.rm=T))
    temp.sd<-unlist(lapply(temp.list,sd,na.rm=T))
    temp.out<-paste(formatC(temp.mean,digits = 2,format="f"),"±",formatC(temp.sd,digits = 2,format="f"),sep="")
    names(temp.out)<-phengrp.annot[as.numeric(names(temp.mean))]
    if(i>1){
      cmd.out<-rbind(cmd.out,temp.out[colnames(cmd.out)])
      rownames(cmd.out)[nrow(cmd.out)]<-paste(var.names[i],", mean ± SD",sep="")
    }
    
    if(var.names[i]=="AGE"){
      temp.min<-unlist(lapply(temp.list,min,na.rm=T))
      temp.max<-unlist(lapply(temp.list,max,na.rm=T))
      temp.out<-paste(formatC(temp.min,digits = 2,format="f"),"-",formatC(temp.max,digits = 2,format="f"),sep="")
      names(temp.out)<-phengrp.annot[as.numeric(names(temp.min))]

      if(i>1){
        cmd.out<-rbind(cmd.out,temp.out[colnames(cmd.out)])
        rownames(cmd.out)[nrow(cmd.out)]<-paste(var.names[i],", range",sep="")
      }
    }
  }
}


output.filepath<-file.path(work.dir,"scratch/GRADS/SARC_results/Results_summary_BAL_hg38/writings/table1_XY.csv")
write.csv(cmd.out,file=output.filepath, append = F,row.names=T)
```

```{r eval=FALSE,echo=FALSE,results='hide'}
# These codes summarize the table from the original data files
library(gdata)
bigClinical<-read.xls(paste(work.dir,"/scratch/Antun/BAL_clinical/full_clinical975_209.xlsx",sep=""),sheet=1,header=T,check.names=F,stringsAsFactors=F)
smallClinical<-read.table(paste(work.dir,"/scratch/GRADS/SARC_results/Data/SARC_combine_clinical_data20170814.txt",sep=""),sep="\t",header=T,check.names=F,stringsAsFactors=F)
# coding for the phenotype
#1 ---- Multi-organ
#2 ---- Non-acute, Stage I, untreated
#3 ---- Stage II-III, treated
#4 ---- Stage II-III, untreated
#5 ---- Stage IV, treated
#6 ---- Stage IV, untreated
#7 ---- Acute Sarcoidosis, untreated
#8 ---- Remitting, untreated
#9 ---- Cardiac defining therapy
table(bigClinical$smdsc.phengrp)

# Acute Sarcoidosis, untreated                   Multi-organ 
#                           14                            23 
#Non-acute, Stage I, untreated          Remitting, untreated 
#                           25                            40 
#        Stage II-III, treated       Stage II-III, untreated 
#                           34                            42 
#            Stage IV, treated           Stage IV, untreated 
#                           19                            12 
table(smallClinical$PHENGRP)
# 1  2  3  4  5  6  7  8 
#23 25 34 42 19 12 14 40 

#matches in both, no Cardiac defining therapy also


ccol<-c("PHENOTYPE GROUPS", "1.MULTI ORGAN", "2.NON ACUTE STAGE I UNTREATED","3.STAGE II-III TREATED","4.STAGE II-III UNTREATED","5.STAGE IV TREATED","6.STAGE IV UNTREATED","7.ACUTE UNTREATED","8.REMITTING UNTREATED")
totaln<-unname(table(smallClinical$PHENGRP))
totalnc<-c("TOTAL",totaln)

###SCADDING
scad<-c(0,1,2,3,4)
for(i in 1:8){
	
	temp<-cbind(smallClinical$PHENGRP,smallClinical$SCADDING)
	temp<-temp[temp[,1]==i,]
	#remove rows with NA
	temp<-temp[complete.cases(temp),]
	
	scadph<-c()
	
		for(x in 0:4){
			
			temp2<-temp[temp[,2]==x,2]
			scadph<-c(scadph,length(temp2))
			
			}

	scad<-cbind(scad,scadph)
	
	
	}

scadtab<-data.frame(scad[,2:ncol(scad)])

for(n in 1:length(totaln)){
	scadtab[,n]<-paste(scadtab[,n],"(",round((scadtab[,n]/totaln[n])*100,2),"%)",sep="")
	}

namsca<-c("SCADDING 0 n(%)","SCADDING 1 n(%)","SCADDING 2 n(%)","SCADDING 3 n(%)","SCADDING 4 n(%)")
scadtab<-data.frame(namsca,scadtab,stringsAsFactors=F)
colnames(scadtab)<-ccol

mytab<-rbind(totalnc,scadtab)
###GENDER
genderv<-c()
for(z in 1:8){
	
	temp<-cbind(smallClinical$PHENGRP,smallClinical$GENDER)
	temp<-temp[temp[,1]==z,]
	
	temp2<-temp[temp[,2]==0,2]
	#genderv<-c(genderv,length(temp2))
		
	genderv<-c(genderv,paste(length(temp2),"(",round((length(temp2)/totaln[z])*100,2),"%)",sep=""))	
		
		
	}

genderv<-c("GENDER n(%)FEMALE",genderv)

mytab<-rbind(mytab,genderv)

###AGE

agev<-c()
ager<-c()
for(z in 1:8){
	
	temp<-cbind(smallClinical$PHENGRP,smallClinical$AGE)
	temp<-temp[temp[,1]==z,]
	
	agev<-c(agev,paste(round(mean(temp[,2],na.rm=T),2),"±",round(sd(temp[,2],na.rm=T),2),sep=""))	
	ager<-c(ager,paste(round(min(temp[,2],na.rm=T),2),"-",round(max(temp[,2],na.rm=T),2),sep=""))	
		
	}

agev<-c("AGE, years, mean±SD",agev)
ager<-c("AGE, range",ager)

mytab<-rbind(mytab,agev,ager)


###RACE
#temp<-cbind(smallClinical$GRADSID,smallClinical$RACE)
#2- black
#1- white
#3- other (american indian, hawaiian)
#6- multi racial or no primary race

racev<-c()
for(i in 1:8){
	
	temp<-cbind(smallClinical$PHENGRP,smallClinical$RACE)
	temp<-temp[temp[,1]==i,]
	
	tempv<-c(length(temp[temp[,2]==1,2]), length(temp[temp[,2]==2,2]), length(temp[temp[,2]==3 | temp[,2]==6,2]))
	
	racev<-cbind(racev,tempv)
	
	}

racev<-cbind(c("Caucasian/White","African American/Black","Other"),racev)
colnames(racev)<-ccol

mytab<-rbind(mytab,racev)


###FEV1
piz<-smallClinical$BDFEV1
piz[piz<0]<-NA
smallClinical$BDFEV1<-piz

fevv<-c()
for(z in 1:8){
	
	temp<-cbind(smallClinical$PHENGRP,smallClinical$BDFEV1)
	temp<-temp[temp[,1]==z,]
	
	fevv<-c(fevv,paste(round(mean(temp[,2],na.rm=T),2),"±",round(sd(temp[,2],na.rm=T),2),sep=""))
	
	}

fevv<-c("FEV1(L) mean±SD",fevv)
mytab<-rbind(mytab,fevv)

###FEV1%
piz<-smallClinical$FEV1PRED
piz[piz<0]<-NA
smallClinical$FEV1PRED<-piz

fevv<-c()
for(z in 1:8){
	
	temp<-cbind(smallClinical$PHENGRP,smallClinical$FEV1PRED)
	temp<-temp[temp[,1]==z,]
	
	fevv<-c(fevv,paste(round(mean(temp[,2],na.rm=T),2),"±",round(sd(temp[,2],na.rm=T),2),sep=""))
	
	}

fevv<-c("FEV1% mean±SD",fevv)
mytab<-rbind(mytab,fevv)


###FVC
piz<-smallClinical$BDFVC
piz[piz<0]<-NA
smallClinical$BDFVC<-piz

fevv<-c()
for(z in 1:8){
	
	temp<-cbind(smallClinical$PHENGRP,smallClinical$BDFVC)
	temp<-temp[temp[,1]==z,]
	
	fevv<-c(fevv,paste(round(mean(temp[,2],na.rm=T),2),"±",round(sd(temp[,2],na.rm=T),2),sep=""))
	
	}

fevv<-c("FVC(L) mean±SD",fevv)
mytab<-rbind(mytab,fevv)


###FVC%
piz<-smallClinical$FVCPRED
piz[piz<0]<-NA
smallClinical$FVCPRED<-piz

fevv<-c()
for(z in 1:8){
	
	temp<-cbind(smallClinical$PHENGRP,smallClinical$FVCPRED)
	temp<-temp[temp[,1]==z,]
	
	fevv<-c(fevv,paste(round(mean(temp[,2],na.rm=T),2),"±",round(sd(temp[,2],na.rm=T),2),sep=""))
	
	}

fevv<-c("FVC% mean±SD",fevv)
mytab<-rbind(mytab,fevv)

###DLCO
piz<-smallClinical$BDDLCO
piz[piz<0]<-NA
smallClinical$BDDLCO<-piz

fevv<-c()
for(z in 1:8){
	
	temp<-cbind(smallClinical$PHENGRP,smallClinical$BDDLCO)
	temp<-temp[temp[,1]==z,]
	
	fevv<-c(fevv,paste(round(mean(temp[,2],na.rm=T),2),"±",round(sd(temp[,2],na.rm=T),2),sep=""))
	
	}

fevv<-c("DLCO(ml/min/mmHg) mean±SD",fevv)
mytab<-rbind(mytab,fevv)


###DLCO%
piz<-smallClinical$PREDDLCO
piz[piz<0]<-NA
smallClinical$PREDDLCO<-piz

fevv<-c()
for(z in 1:8){
	
	temp<-cbind(smallClinical$PHENGRP,smallClinical$PREDDLCO)
	temp<-temp[temp[,1]==z,]
	
	fevv<-c(fevv,paste(round(mean(temp[,2],na.rm=T),2),"±",round(sd(temp[,2],na.rm=T),2),sep=""))
	
	}

fevv<-c("DLCO% mean±SD",fevv)
mytab<-rbind(mytab,fevv)


write.table(mytab,file="/home/antun/driver_Grace_xy48/scratch/GRADS/SARC_results/Rmarkdown/Antun/DemographicsTable.txt",append=F,sep="\t",row.names=F,col.names=T,quote=F)



###check the big Clinical table if it is the same
# coding for the phenotype
#1 ---- Multi-organ
#2 ---- Non-acute, Stage I, untreated
#3 ---- Stage II-III, treated
#4 ---- Stage II-III, untreated
#5 ---- Stage IV, treated
#6 ---- Stage IV, untreated
#7 ---- Acute Sarcoidosis, untreated
#8 ---- Remitting, untreated
#9 ---- Cardiac defining therapy

bigClinical[bigClinical[,909]=="Multi-organ",909]<-1
bigClinical[bigClinical[,909]=="Non-acute, Stage I, untreated",909]<-2
bigClinical[bigClinical[,909]=="Stage II-III, treated",909]<-3
bigClinical[bigClinical[,909]=="Stage II-III, untreated",909]<-4
bigClinical[bigClinical[,909]=="Stage IV, treated",909]<-5
bigClinical[bigClinical[,909]=="Stage IV, untreated",909]<-6
bigClinical[bigClinical[,909]=="Acute Sarcoidosis, untreated",909]<-7
bigClinical[bigClinical[,909]=="Remitting, untreated",909]<-8
bigClinical[bigClinical[,909]=="Cardiac defining therapy",909]<-9

bigClinical[,909]<-as.integer(bigClinical[,909])


ccol<-c("PHENOTYPE GROUPS", "1.MULTI ORGAN", "2.NON ACUTE STAGE I UNTREATED","3.STAGE II-III TREATED","4.STAGE II-III UNTREATED","5.STAGE IV TREATED","6.STAGE IV UNTREATED","7.ACUTE UNTREATED","8.REMITTING UNTREATED")
totaln<-unname(table(bigClinical$smdsc.phengrp))
totalnc<-c("TOTAL",totaln)

###SCADDING
scad<-c(0,1,2,3,4)
for(i in 1:8){
	
	temp<-cbind(bigClinical$smdsc.phengrp,bigClinical$SCADDING)
	temp<-temp[temp[,1]==i,]
	#remove rows with NA
	temp<-temp[complete.cases(temp),]
	
	scadph<-c()
	
		for(x in 0:4){
			
			temp2<-temp[temp[,2]==x,2]
			scadph<-c(scadph,length(temp2))
			
			}

	scad<-cbind(scad,scadph)
	
	
	}

scadtab<-data.frame(scad[,2:ncol(scad)])

for(n in 1:length(totaln)){
	scadtab[,n]<-paste(scadtab[,n],"(",round((scadtab[,n]/totaln[n])*100,2),"%)",sep="")
	}

namsca<-c("SCADDING 0 n(%)","SCADDING 1 n(%)","SCADDING 2 n(%)","SCADDING 3 n(%)","SCADDING 4 n(%)")
scadtab<-data.frame(namsca,scadtab,stringsAsFactors=F)
colnames(scadtab)<-ccol

mytab<-rbind(totalnc,scadtab)

###GENDER
genderv<-c()
for(z in 1:8){
	
	temp<-cbind(bigClinical$smdsc.phengrp,bigClinical$GENDER)
	temp<-temp[temp[,1]==z,]
	
	temp2<-temp[temp[,2]==0,2]
	#genderv<-c(genderv,length(temp2))
		
	genderv<-c(genderv,paste(length(temp2),"(",round((length(temp2)/totaln[z])*100,2),"%)",sep=""))	
		
		
	}

genderv<-c("GENDER n(%)FEMALE",genderv)

mytab<-rbind(mytab,genderv)

###AGE

agev<-c()
ager<-c()
for(z in 1:8){
	
	temp<-cbind(bigClinical$smdsc.phengrp,bigClinical$AGE)
	temp<-temp[temp[,1]==z,]
	
	agev<-c(agev,paste(round(mean(temp[,2],na.rm=T),2),"±",round(sd(temp[,2],na.rm=T),2),sep=""))	
	ager<-c(ager,paste(round(min(temp[,2],na.rm=T),2),"-",round(max(temp[,2],na.rm=T),2),sep=""))	
		
	}

agev<-c("AGE, years, mean±SD",agev)
ager<-c("AGE, range",ager)

mytab<-rbind(mytab,agev,ager)


###RACE
#temp<-cbind(smallClinical$GRADSID,smallClinical$RACE)
#2- black
#1- white
#3- other (american indian, hawaiian)
#6- multi racial or no primary race

racev<-c()
for(i in 1:8){
	
	temp<-cbind(bigClinical$smdsc.phengrp,bigClinical$RACE)
	temp<-temp[temp[,1]==i,]
	
	tempv<-c(length(temp[temp[,2]==1,2]), length(temp[temp[,2]==2,2]), length(temp[temp[,2]==3 | temp[,2]==6,2]))
	
	racev<-cbind(racev,tempv)
	
	}

racev<-cbind(c("Caucasian/White","African American/Black","Other"),racev)
colnames(racev)<-ccol

mytab<-rbind(mytab,racev)


###FEV1

fevv<-c()
for(z in 1:8){
	
	temp<-cbind(bigClinical$smdsc.phengrp,bigClinical$BDFEV1)
	temp<-temp[temp[,1]==z,]
	
	fevv<-c(fevv,paste(round(mean(temp[,2],na.rm=T),2),"±",round(sd(temp[,2],na.rm=T),2),sep=""))
	
	}

fevv<-c("FEV1(L) mean±SD",fevv)
mytab<-rbind(mytab,fevv)

###FEV1%

fevv<-c()
for(z in 1:8){
	
	temp<-cbind(bigClinical$smdsc.phengrp,bigClinical$FEV1PRED)
	temp<-temp[temp[,1]==z,]
	
	fevv<-c(fevv,paste(round(mean(temp[,2],na.rm=T),2),"±",round(sd(temp[,2],na.rm=T),2),sep=""))
	
	}

fevv<-c("FEV1% mean±SD",fevv)
mytab<-rbind(mytab,fevv)


###FVC

fevv<-c()
for(z in 1:8){
	
	temp<-cbind(bigClinical$smdsc.phengrp,bigClinical$BDFVC)
	temp<-temp[temp[,1]==z,]
	
	fevv<-c(fevv,paste(round(mean(temp[,2],na.rm=T),2),"±",round(sd(temp[,2],na.rm=T),2),sep=""))
	
	}

fevv<-c("FVC(L) mean±SD",fevv)
mytab<-rbind(mytab,fevv)


###FVC%

fevv<-c()
for(z in 1:8){
	
	temp<-cbind(bigClinical$smdsc.phengrp,bigClinical$FVCPRED)
	temp<-temp[temp[,1]==z,]
	
	fevv<-c(fevv,paste(round(mean(temp[,2],na.rm=T),2),"±",round(sd(temp[,2],na.rm=T),2),sep=""))
	
	}

fevv<-c("FVC% mean±SD",fevv)
mytab<-rbind(mytab,fevv)

###DLCO

fevv<-c()
for(z in 1:8){
	
	temp<-cbind(bigClinical$smdsc.phengrp,bigClinical$BDDLCO)
	temp<-temp[temp[,1]==z,]
	
	fevv<-c(fevv,paste(round(mean(temp[,2],na.rm=T),2),"±",round(sd(temp[,2],na.rm=T),2),sep=""))
	
	}

fevv<-c("DLCO(ml/min/mmHg) mean±SD",fevv)
mytab<-rbind(mytab,fevv)


###DLCO%

fevv<-c()
for(z in 1:8){
	
	temp<-cbind(bigClinical$smdsc.phengrp,bigClinical$PREDDLCO)
	temp<-temp[temp[,1]==z,]
	
	fevv<-c(fevv,paste(round(mean(temp[,2],na.rm=T),2),"±",round(sd(temp[,2],na.rm=T),2),sep=""))
	
	}

fevv<-c("DLCO% mean±SD",fevv)
mytab<-rbind(mytab,fevv)


write.table(mytab,file=paste(work.dir,"/scratch/GRADS/SARC_results/Rmarkdown/Antun/DemographicsTableBIG.txt",sep=""),append=F,sep="\t",row.names=F,col.names=T,quote=F)
```

```{r eval=TRUE,cache=FALSE,fig.cap=table_nums(name="table1_demo",caption="Patient demographics table"),results='hide'}
```
`r table_nums(name="table1_demo")`
```{r eval=TRUE,cache=FALSE,echo=FALSE,warning=FALSE,message=FALSE,results='asis'}

# This is the data directory for Xiting's Laptop
data.dir<-paste(work.dir,"/scratch/GRADS/SARC_results/Rmarkdown/Antun/data",sep="")

# This is the data directory for Antun's workstation
#data.dir<-"/home/antun/driver_Grace_xy48/scratch/GRADS/SARC_results/Rmarkdown/Antun"

mytab<-read.table(file.path(data.dir,"DemographicsTableBIG.txt"),sep="\t",header=T,check.names=F,stringsAsFactors=F)
kable(mytab,row.names = FALSE)%>% kable_styling(bootstrap_options = c("striped", "hover","condensed","responsive"),latex_options="scale_down",full_width = F,position="left")
```

### Table 1b. Freiburgh patient characteristics 

```{r eval=TRUE}
data.filepath<-file.path(work.dir,"scratch/Antun/microarray_Jonas/Clinical_data_for_Yale20180924.xls")

clinic.data<-read.xls(data.filepath,check.names=F,stringsAsFactors=F)
clinic.data<-clinic.data[clinic.data[,"Disease"]=="SARC",]

# change the SCADDING into integer
temp.scadding<-unname(sapply(clinic.data[,"Scadding"],my.element.extract,splitchar="SARC",index=-1))
temp.vect<-rep(1,length(temp.scadding))
temp.vect[temp.scadding=="II"]<-2
temp.vect[temp.scadding=="III"]<-3
temp.vect[temp.scadding=="IV"]<-4
clinic.data[,"Scadding"]<-temp.vect

# add the FEV1/FVC ratio
clinic.data<-cbind(clinic.data,clinic.data[,"FEV1"]/clinic.data[,"FVC"])
colnames(clinic.data)[ncol(clinic.data)]<-"FEV1/FVC ratio"

var.names<-c("Scadding","Age","Sex","AM","Ly","N.G","E.G","FEV1","FEV1%","FVC","FVC%","FEV1/FVC ratio")
var.type<-rep("con",length(var.names))
var.annot<-c("SCADDING","AGE","GENDER","MACROPHAGE","LYMPHOCYTE","NEUTROPHIL","EOSINOPHIL","FEV1","FEV1% predicted","FVC","FVC% predicted","FEV1/FVC ratio")
var.type[var.names%in%c("Scadding","Sex")]<-"cat"

# get the patient characteristics table
summary.table<-character()

for(i in 1:length(var.names)){
  #cat("i=",i,"\n")
  if(var.type[i]=="cat"){
    temp.matrix<-as.matrix(table(clinic.data[,var.names[i]]))
    rownames(temp.matrix)<-paste(var.annot[i]," - ",rownames(temp.matrix))
    
    if(var.names[i]=="Scadding" & sum(clinic.data[,var.names[i]]==0)==0){
      temp.matrix<-rbind(0,temp.matrix)
      rownames(temp.matrix)[1]<-paste(var.annot[i]," - 0")
    }
    temp.sum<-sum(temp.matrix)
    temp.matrix2<-temp.matrix/temp.sum
    summary.table<-rbind(summary.table,as.matrix(paste(temp.matrix,"(",temp.matrix2,")",sep="")))
    rownames(summary.table)[(nrow(summary.table)-nrow(temp.matrix)+1):nrow(summary.table)]<-rownames(temp.matrix)
    
  }else{
    temp.mean<-mean(clinic.data[,var.names[i]],na.rm=T)
    temp.sd<-sd(clinic.data[,var.names[i]],na.rm=T)
    temp.mean<-round(temp.mean,digits=4)
    temp.sd<-round(temp.sd,digits=4)
    summary.table<-rbind(summary.table,paste(temp.mean,"±",temp.sd,sep=""))
    rownames(summary.table)[nrow(summary.table)]<-paste(var.annot[i],", mean±SD",sep="")
  }
  
}
colnames(summary.table)[1]<-"Freiburgh"

###############
# extract the same info for GRADS cohort
grads.clinic.data<-read.table(paste(work.dir,"/scratch/GRADS/SARC_results/Data/SARC_combine_clinical_data20170814.txt",sep=""),sep="\t",header=T,check.names=F,stringsAsFactors=F)
# take the average of BALD1 and BALD2 as the final cell differential.
grads.clinic.data<-cbind(grads.clinic.data,apply(grads.clinic.data[,c("ALVBALD1","ALVBALD2")],1,mean,na.rm=T))
colnames(grads.clinic.data)[ncol(grads.clinic.data)]<-"ALVBALD"
grads.clinic.data<-cbind(grads.clinic.data,apply(grads.clinic.data[,c("EOSBALD1","EOSBALD1")],1,mean,na.rm=T))
colnames(grads.clinic.data)[ncol(grads.clinic.data)]<-"EOSBALD"
grads.clinic.data<-cbind(grads.clinic.data,apply(grads.clinic.data[,c("LYMBALD1","LYMBALD2")],1,mean,na.rm=T))
colnames(grads.clinic.data)[ncol(grads.clinic.data)]<-"LYMBALD"
grads.clinic.data<-cbind(grads.clinic.data,apply(grads.clinic.data[,c("NEUBALD1","NEUBALD2")],1,mean,na.rm=T))
colnames(grads.clinic.data)[ncol(grads.clinic.data)]<-"NEUBALD"

temp<-rep("w",nrow(grads.clinic.data))
temp[grads.clinic.data[,"GENDER"]==1]<-"m"
grads.clinic.data[,"GENDER"]<-temp

grads.clinic.data<-cbind(grads.clinic.data,grads.clinic.data[,"BDFEV1"]/grads.clinic.data[,"BDFVC"])
colnames(grads.clinic.data)[ncol(grads.clinic.data)]<-"FEV1/FVC ratio"


var.names<-c("SCADDING","AGE","GENDER","ALVBALD","LYMBALD","NEUBALD","EOSBALD","BDFEV1","FEV1PRED","BDFVC","FVCPRED","FEV1/FVC ratio")
var.type<-rep("con",length(var.names))
var.annot<-c("SCADDING","AGE","GENDER","MACROPHAGE","LYMPHOCYTE","NEUTROPHIL","EOSINOPHIL","FEV1","FEV1% predicted","FVC","FVC% predicted","FEV1/FVC ratio")
var.type[var.names%in%c("SCADDING","GENDER")]<-"cat"

summary.table2<-character()

for(i in 1:length(var.names)){
  if(var.type[i]=="cat"){
    temp.matrix<-as.matrix(table(grads.clinic.data[,var.names[i]]))
    rownames(temp.matrix)<-paste(var.annot[i]," - ",rownames(temp.matrix))
    
    temp.sum<-sum(temp.matrix)
    temp.matrix2<-temp.matrix/temp.sum
    summary.table2<-rbind(summary.table2,as.matrix(paste(temp.matrix,"(",round(temp.matrix2,digits=4),")",sep="")))
    rownames(summary.table2)[(nrow(summary.table2)-nrow(temp.matrix)+1):nrow(summary.table2)]<-rownames(temp.matrix)
    }else{
    temp.mean<-mean(grads.clinic.data[,var.names[i]],na.rm=T)
    temp.sd<-sd(grads.clinic.data[,var.names[i]],na.rm=T)
    temp.mean<-round(temp.mean,digits=4)
    temp.sd<-round(temp.sd,digits=4)
    summary.table2<-rbind(summary.table2,paste(temp.mean,"±",temp.sd,sep=""))
    rownames(summary.table2)[nrow(summary.table2)]<-paste(var.annot[i],", mean±SD",sep="")
  }
  
}
colnames(summary.table2)[1]<-"GRADS"

# calculate the p value comparing the difference between the two cohorts.
var.names.f<-c("Scadding","Age","Sex","AM","Ly","N.G","E.G","FEV1","FEV1%","FVC","FVC%","FEV1/FVC ratio")
var.names.g<-c("SCADDING","AGE","GENDER","ALVBALD","LYMBALD","NEUBALD","EOSBALD","BDFEV1","FEV1PRED","BDFVC","FVCPRED","FEV1/FVC ratio")
var.type<-rep("con",length(var.names.f))
var.type[var.names.f%in%c("Scadding","Sex")]<-"cat"


clinic.data.f<-clinic.data
clinic.data.g<-grads.clinic.data

pvalue.vect<-numeric()
for(i in 1:length(var.names.f)){
  #calculate the difference between the two cohort using wilcoxon test
  temp.f<-clinic.data.f[,var.names.f[i]]
  temp.g<-clinic.data.g[,var.names.g[i]]
  
  temp.f<-temp.f[!is.na(temp.f)]
  temp.g<-temp.g[!is.na(temp.g)]

  if(var.type[i]=="con"){
    pvalue.vect<-c(pvalue.vect,wilcox.test(temp.f,temp.g)$p.value)
  }else{
    temp.levels<-sort(unique(c(temp.f[!is.na(temp.f)],temp.g[!is.na(temp.g)])))
  
  
    temp.obs<-cbind(as.matrix(table(temp.f))[match(as.character(temp.levels),rownames(as.matrix(table(temp.f)))),],as.matrix(table(temp.g))[match(as.character(temp.levels),rownames(as.matrix(table(temp.g)))),])
    temp.obs[is.na(temp.obs)]<-0
    rownames(temp.obs)<-as.character(temp.levels)
  
    pvalue.vect<-c(pvalue.vect,chisq.test(temp.obs)$p.value)
  }
}

names(pvalue.vect)<-var.annot
pvalue.vect


output.filepath<-file.path(work.dir,"scratch/GRADS/SARC_results/Results_summary_BAL_hg38/writings/table1b_clinicaltable_bothcohorts.txt")
cat("\t",file=output.filepath,append=F)
write.table(cbind(summary.table2,summary.table),file=output.filepath,append=T,sep="\t",row.names=T,col.names=T,quote=F)


```



## Supervised Analysis

```{r supervised_analysis,eval=FALSE}

#########
# This part comes from 209_local_data_load_AM_XYversion.R
#this code loads the 209 BAL samples into the same variable names as clean_data_load.R

source(file.path(work.dir,"/Rprogram/my_functions.R"))

fpkm.matrix.clean<-read.table(file.path(work.dir,"scratch/GRADS/SARC_results/Results_summary_BAL_hg38/FPKM_matrix_hg38_Corrected_allGenes_BAL209.txt"),sep="\t",header=T,check.names=F,stringsAsFactors=F)

row.names(fpkm.matrix.clean)<-fpkm.matrix.clean[,1]
fpkm.matrix.clean<-fpkm.matrix.clean[,-1]
fpkm.matrix.clean<-as.matrix(fpkm.matrix.clean)

anno.matrix<-read.table(file.path(work.dir,"scratch/GRADS/SARC_results/Results_summary_BAL_hg38/FPKM_matrix_hg38_Corrected_allGenes_BAL209_withAnnot.txt"),sep="\t",header=T,check.names=F,stringsAsFactors=F)
anno.matrix<-anno.matrix[,1:9]
row.names(anno.matrix)<-paste(anno.matrix[,1],anno.matrix[,6],sep="_")

#clinic.matrix<-read.table("/home/antun/Desktop/SUPERvisedA/combine_clinical_data20170814.txt",sep="\t",header=T,check.names=F,stringsAsFactors=F,quote="")
clinic.matrix<-read.table(file.path(work.dir,"scratch/GRADS/SARC_results/Data/SARC_combine_clinical_data20170814.txt"),sep="\t",header=T,check.names=F,stringsAsFactors=F,quote="")

row.names(clinic.matrix)<-clinic.matrix[,1]

#make sure the GRADSID and the BAL KitID match
library(gdata)
mkey<-read.xls(file.path(work.dir,"scratch/GRADS/SARC_results/Data/GRADS Master Progress Key 6-13-16.xlsx"),sheet=3,check.names=F,stringsAsFactors = F)
mkey<-mkey[3:nrow(mkey),1:5]
colnames(mkey)<-c("GRADS_ID","BloodKitsEnrollment","BloodKitsMonth6","BALKits","Phenotype")

#grep only the sarcoidosis grads ids (3rd letter S)
temp<-grep("^.{2}S", mkey[,1])
mkey<-mkey[temp,]
#grep only the 209 sarcoidosis BAL kit ids (3rd letter S)
temp<-unname(sapply(colnames(fpkm.matrix.clean),my.element.extract,splitchar="_",index=1))
mkey<-mkey[mkey[,4]%in%temp,]

#create a "sample.list.all" 
sample.list.all<-matrix(NA,nrow=nrow(clinic.matrix),ncol=16)
sample.list.all[,1]<-mkey[,1]
sample.list.all[,3]<-mkey[,4]

data.matrix<-fpkm.matrix.clean
colnames(data.matrix)<-sapply(colnames(data.matrix),my.element.extract,splitchar="_",index=1)
sample.matrix<-sample.list.all
rownames(sample.matrix)<-sample.matrix[,3]
my.clinic.matrix<-clinic.matrix

# add the FEV1/FVC ratio
clinic.matrix<-cbind(clinic.matrix,clinic.matrix[,"BDFEV1"]/clinic.matrix[,"BDFVC"])
colnames(clinic.matrix)[ncol(clinic.matrix)]<-"FEV1FVCratio"

# add the obstructive vs. restrictive variable FEV1FVCratio=0 is obstructive and =1 is restrictive
temp.vect<-rep(NA,nrow(clinic.matrix))
temp.vect[!is.na(clinic.matrix[,"FEV1FVCratio"]) & clinic.matrix[,"FEV1FVCratio"]<0.7]<- 0
temp.vect[(!is.na(clinic.matrix[,"FEV1FVCratio"]) & clinic.matrix[,"FEV1FVCratio"]>=0.7) & (clinic.matrix[,"FVCPRED"]<80 & clinic.matrix[,"FVCPRED"]>=0)]<- 1
clinic.matrix<-cbind(clinic.matrix,temp.vect)
colnames(clinic.matrix)[ncol(clinic.matrix)]<-"FEV1FVCratioBinary"

# add the obstructive vs. normal variable obstractiveBinary=0 is normal and =1 is obstructive
temp.vect<-rep(NA,nrow(clinic.matrix))
temp.vect[(!is.na(clinic.matrix[,"FEV1FVCratio"]) & clinic.matrix[,"FEV1FVCratio"]<0.7)]<- 1
temp.vect[(!is.na(clinic.matrix[,"FEV1FVCratio"]) & clinic.matrix[,"FEV1FVCratio"]>0.7) & (clinic.matrix[,"FVCPRED"]>80 & clinic.matrix[,"FVCPRED"]>=0)]<- 0
clinic.matrix<-cbind(clinic.matrix,temp.vect)
colnames(clinic.matrix)[ncol(clinic.matrix)]<-"obstructiveBinary"



# add the restrictive vs. normal variable restrictiveBinary=0 is normal and =1 is restrictive
temp.vect<-rep(NA,nrow(clinic.matrix))
temp.vect[(!is.na(clinic.matrix[,"FEV1FVCratio"]) & clinic.matrix[,"FEV1FVCratio"]>0.7) & (clinic.matrix[,"FVCPRED"]<80 & clinic.matrix[,"FVCPRED"]>=0)]<- 1
temp.vect[(!is.na(clinic.matrix[,"FEV1FVCratio"]) & clinic.matrix[,"FEV1FVCratio"]>0.7) & (clinic.matrix[,"FVCPRED"]>80 & clinic.matrix[,"FVCPRED"]>=0)]<- 0
clinic.matrix<-cbind(clinic.matrix,temp.vect)
colnames(clinic.matrix)[ncol(clinic.matrix)]<-"restrictiveBinary"








#########
# This part comes from hg38_Supervised_analysis_AM_XYversion.R
################################
# 4.4. Identify genes associated with age and PFTs
#source("/home/fas/kaminski/am2952/scratch/SARC215/0_CleanDataLoad.R")
#source(file.path(work.dir,"Rprogram/209_local_data_load_AM.R")

my.con.cor<-function(fpkm.matrix.clean,feature.name="AGE",method.name="spearman",tissue.name="BAL",sample.list.all,clinic.matrix,output.dir){

# extract the fpkm matrix for the given tissue
data.matrix<-fpkm.matrix.clean
sample.names<-colnames(data.matrix)
names(sample.names)<-unname(sapply(sample.names,my.element.extract,splitchar="_",index=1))
colnames(data.matrix)<-names(sample.names)

if(tissue.name=="BAL"){
bal.fpkm<-data.matrix[,substr(colnames(data.matrix),3,3)=="B"]
}else{
bal.fpkm<-data.matrix[,substr(colnames(data.matrix),3,3)=="S"]
}

# generate the clinical matrix
if(tissue.name=="BAL"){
temp.list<-sample.list.all
rownames(temp.list)<-sample.list.all[,3]
temp.list<-temp.list[colnames(bal.fpkm),]
my.clinic.matrix<-clinic.matrix[temp.list[,1],]
}else{
temp.list<-sample.list.all
rownames(temp.list)<-sample.list.all[,10]
temp.list<-temp.list[colnames(bal.fpkm),]
my.clinic.matrix<-clinic.matrix[temp.list[,1],]	
}

#bal.fpkm<-data.matrix[,substr(colnames(data.matrix),3,3)=="B"]
#pbmc.fpkm<-data.matrix[,substr(colnames(data.matrix),3,3)=="S"]

clinic.vect<-as.numeric(my.clinic.matrix[,colnames(my.clinic.matrix)==feature.name])
clinic.vect[clinic.vect<0]<-NA
# generate the correlation results
age.pvalue<-apply(bal.fpkm,1,my.cor.test,y=clinic.vect,method.name=method.name)
age.pvalue<-matrix(unlist(age.pvalue),ncol=2,byrow=T)
colnames(age.pvalue)<-c("pvalue","corr_coef")
age.fdr<-p.adjust(age.pvalue[,"pvalue"],method="fdr")
age.pvalue<-cbind(age.pvalue,age.fdr)
colnames(age.pvalue)[ncol(age.pvalue)]<-"fdr"
rownames(age.pvalue)<-rownames(bal.fpkm)

fpkm.info<-apply(bal.fpkm>0,1,sum)
fpkm.info<-cbind(fpkm.info,apply(bal.fpkm>0.1,1,sum))
fpkm.info<-cbind(fpkm.info,apply(bal.fpkm>1,1,sum))
fpkm.info<-cbind(fpkm.info,apply(bal.fpkm,1,median))
fpkm.info<-cbind(fpkm.info,apply(bal.fpkm,1,mean))
colnames(fpkm.info)<-c("fpkm>0_num","fpkm>0.1_num","fpkm>1_num","fpkm_median","fpkm_mean")


if(file.exists(output.dir)==F){
cat("WARNING: output.dir doesn't exists!\n")
return(1)
}

cmd.out<-cbind(age.pvalue,fpkm.info)
output.filepath<-file.path(output.dir,paste("correlation_",method.name,"_",feature.name,"_allgenes_",tissue.name,".txt",sep=""))
cat("gene_names\t",file=output.filepath,append=F)
write.table(cmd.out,file=output.filepath,append=T,sep="\t",row.names=T,col.names=T,quote=F)

output.filepath<-file.path(output.dir,paste("correlation_",method.name,"_",feature.name,"_fdr0.05_",tissue.name,".txt",sep=""))
cat("gene_names\t",file=output.filepath,append=F)
#write.table(age.pvalue[!is.na(age.pvalue[,3]) & age.pvalue[,3]<0.05,],file=output.filepath,append=T,sep="\t",row.names=T,col.names=T,quote=F)
write.table(cmd.out[!is.na(age.pvalue[,3]) & age.pvalue[,3]<0.05,],file=output.filepath,append=T,sep="\t",row.names=T,col.names=T,quote=F)

output.filepath<-file.path(output.dir,paste("correlation_",method.name,"_",feature.name,"_fdr0.05_",tissue.name,"_both_metacore.txt",sep=""))
cat("gene_names\tintensity\n",file=output.filepath,append=F)
sig.genes<-rownames(data.matrix)[!is.na(age.pvalue[,3]) & age.pvalue[,3]<0.05]
sig.genes<-unname(sapply(sig.genes,my.element.extract,splitchar="_",index=1))
sig.genes<-unique(sig.genes)
cmd.out<-cbind(sig.genes,rep(1,length(sig.genes)))
write.table(cmd.out,file=output.filepath,append=T,sep="\t",row.names=F,col.names=F,quote=F)


output.filepath<-file.path(output.dir,paste("correlation_",method.name,"_",feature.name,"_fdr0.05_",tissue.name,"_positive_metacore.txt",sep=""))
cat("gene_names\tintensity\n",file=output.filepath,append=F)
sig.genes<-rownames(data.matrix)[!is.na(age.pvalue[,3]) & age.pvalue[,3]<0.05 & age.pvalue[,2]>0]
sig.genes<-unname(sapply(sig.genes,my.element.extract,splitchar="_",index=1))
sig.genes<-unique(sig.genes)
cmd.out<-cbind(sig.genes,rep(1,length(sig.genes)))
write.table(cmd.out,file=output.filepath,append=T,sep="\t",row.names=F,col.names=F,quote=F)


output.filepath<-file.path(output.dir,paste("correlation_",method.name,"_",feature.name,"_fdr0.05_",tissue.name,"_negative_metacore.txt",sep=""))
cat("gene_names\tintensity\n",file=output.filepath,append=F)
sig.genes<-rownames(data.matrix)[!is.na(age.pvalue[,3]) & age.pvalue[,3]<0.05 & age.pvalue[,2]<0]
sig.genes<-unname(sapply(sig.genes,my.element.extract,splitchar="_",index=1))
sig.genes<-unique(sig.genes)
cmd.out<-cbind(sig.genes,rep(1,length(sig.genes)))
write.table(cmd.out,file=output.filepath,append=T,sep="\t",row.names=F,col.names=F,quote=F)

return(0)
}


# for age
#output.dir<-"~/driver_Naftali/GRADS/SARC_results/Results_summary/PFTs_correlated_genes"
output.dir<-file.path(work.dir,"scratch/Antun/SUPERvisedA/Results_summary2/Demographics_correlated_genes")
my.con.cor(fpkm.matrix.clean,feature.name="AGE",method.name="spearman",tissue.name="BAL",sample.list.all,clinic.matrix,output.dir)
#my.con.cor(fpkm.matrix.clean,feature.name="AGE",method.name="spearman",tissue.name="PBMC",sample.list.all,clinic.matrix,output.dir)
my.con.cor(fpkm.matrix.clean,feature.name="GENDER",method.name="spearman",tissue.name="BAL",sample.list.all,clinic.matrix,output.dir)
#my.con.cor(fpkm.matrix.clean,feature.name="GENDER",method.name="spearman",tissue.name="PBMC",sample.list.all,clinic.matrix,output.dir)
my.con.cor(fpkm.matrix.clean,feature.name="RACE",method.name="spearman",tissue.name="BAL",sample.list.all,clinic.matrix[clinic.matrix[,"RACE"]%in%c(1,2),],output.dir)
#my.con.cor(fpkm.matrix.clean,feature.name="RACE",method.name="spearman",tissue.name="PBMC",sample.list.all,clinic.matrix[clinic.matrix[,"RACE"]%in%c(1,2),],output.dir)


# for FVC, FEV1 and DLCO
output.dir<-file.path(work.dir,"scratch/Antun/SUPERvisedA/Results_summary2/PFTs_correlated_genes")
my.con.cor(fpkm.matrix.clean,feature.name="BDFVC",method.name="spearman",tissue.name="BAL",sample.list.all,clinic.matrix,output.dir)
#my.con.cor(fpkm.matrix.clean,feature.name="BDFVC",method.name="spearman",tissue.name="PBMC",sample.list.all,clinic.matrix,output.dir)

my.con.cor(fpkm.matrix.clean,feature.name="POSTFVC",method.name="spearman",tissue.name="BAL",sample.list.all,clinic.matrix,output.dir)
#my.con.cor(fpkm.matrix.clean,feature.name="POSTFVC",method.name="spearman",tissue.name="PBMC",sample.list.all,clinic.matrix,output.dir)

my.con.cor(fpkm.matrix.clean,feature.name="FVCPRED",method.name="spearman",tissue.name="BAL",sample.list.all,clinic.matrix,output.dir)
#my.con.cor(fpkm.matrix.clean,feature.name="FVCPRED",method.name="spearman",tissue.name="PBMC",sample.list.all,clinic.matrix,output.dir)

my.con.cor(fpkm.matrix.clean,feature.name="BDFEV1",method.name="spearman",tissue.name="BAL",sample.list.all,clinic.matrix,output.dir)
#my.con.cor(fpkm.matrix.clean,feature.name="BDFEV1",method.name="spearman",tissue.name="PBMC",sample.list.all,clinic.matrix,output.dir)

my.con.cor(fpkm.matrix.clean,feature.name="POSTFEV1",method.name="spearman",tissue.name="BAL",sample.list.all,clinic.matrix,output.dir)
#my.con.cor(fpkm.matrix.clean,feature.name="POSTFEV1",method.name="spearman",tissue.name="PBMC",sample.list.all,clinic.matrix,output.dir)

my.con.cor(fpkm.matrix.clean,feature.name="FEV1PRED",method.name="spearman",tissue.name="BAL",sample.list.all,clinic.matrix,output.dir)
#my.con.cor(fpkm.matrix.clean,feature.name="FEV1PRED",method.name="spearman",tissue.name="PBMC",sample.list.all,clinic.matrix,output.dir)

my.con.cor(fpkm.matrix.clean,feature.name="BDDLCO",method.name="spearman",tissue.name="BAL",sample.list.all,clinic.matrix,output.dir)
#my.con.cor(fpkm.matrix.clean,feature.name="BDDLCO",method.name="spearman",tissue.name="PBMC",sample.list.all,clinic.matrix,output.dir)

my.con.cor(fpkm.matrix.clean,feature.name="PREDDLCO",method.name="spearman",tissue.name="BAL",sample.list.all,clinic.matrix,output.dir)
#my.con.cor(fpkm.matrix.clean,feature.name="PREDDLCO",method.name="spearman",tissue.name="PBMC",sample.list.all,clinic.matrix,output.dir)

my.con.cor(fpkm.matrix.clean,feature.name="FEV1FVCratio",method.name="spearman",tissue.name="BAL",sample.list.all,clinic.matrix,output.dir)


# for scadding
#output.dir<-"~/driver_Naftali/GRADS/SARC_results/Results_summary/Disease_correlated_genes"
output.dir<-file.path(work.dir,"scratch/Antun/SUPERvisedA/Results_summary2/Disease_correlated_genes")
my.con.cor(fpkm.matrix.clean,feature.name="SCADDING",method.name="spearman",tissue.name="BAL",sample.list.all,clinic.matrix,output.dir)
#my.con.cor(fpkm.matrix.clean,feature.name="SCADDING",method.name="spearman",tissue.name="PBMC",sample.list.all,clinic.matrix,output.dir)


#my.con.cor(fpkm.matrix.clean,feature.name="SCADDING",method.name="spearman",tissue.name="BAL",sample.list.all,clinic.matrix,output.dir)
#my.con.cor(fpkm.matrix.clean,feature.name="SCADDING",method.name="spearman",tissue.name="PBMC",sample.list.all,clinic.matrix,output.dir)


#source("/home/antun/Desktop/SUPERvisedA/r_codes/209_local_data_load_AM.R")


#output.dir<-"/home/antun/Desktop/SUPERvisedA/Results_summary/Demographics_correlated_genes/Directional_p"


my.cat.cor<-function(fpkm.matrix.clean,feature.name="GENDER",method.name="wilcox",tissue.name="BAL",samplelist,clinicmatrix,outputdir){

# extract the fpkm matrix for the given tissue
data.matrix<-fpkm.matrix.clean
sample.names<-colnames(data.matrix)
names(sample.names)<-unname(sapply(sample.names,my.element.extract,splitchar="_",index=1))
colnames(data.matrix)<-names(sample.names)
if(tissue.name=="BAL"){
bal.fpkm<-data.matrix[,substr(colnames(data.matrix),3,3)=="B"]
}else{
bal.fpkm<-data.matrix[,substr(colnames(data.matrix),3,3)=="S"]
}
# generate the clinical matrix
rownames(samplelist)<-samplelist[,3]
if(tissue.name=="BAL"){
#temp.list<-samplelist
#rownames(temp.list)<-samplelist[,3]
#temp.list<-temp.list[colnames(bal.fpkm),]
temp.list<-samplelist[colnames(bal.fpkm),]
temp.match<-match(temp.list[,1],as.matrix(clinicmatrix)[,1])
#my.clinic.matrix<-clinic.matrix[temp.list[,1],]
my.clinic.matrix<-clinicmatrix[temp.match,]
}else{
#temp.list<-samplelist
#rownames(temp.list)<-samplelist[,10]
#temp.list<-temp.list[colnames(bal.fpkm),]
temp.list<-samplelist[colnames(bal.fpkm),]
temp.match<-match(temp.list[,1],as.matrix(clinicmatrix)[,1])
#my.clinic.matrix<-clinic.matrix[temp.list[,1],]
my.clinic.matrix<-clinicmatrix[temp.match,]
}
#bal.fpkm<-data.matrix[,substr(colnames(data.matrix),3,3)=="B"]
#pbmc.fpkm<-data.matrix[,substr(colnames(data.matrix),3,3)=="S"]

clinic.vect<-as.numeric(my.clinic.matrix[,colnames(my.clinic.matrix)==feature.name])
if(length(table(clinic.vect))!=2){
cat("ERROR: the phenotype does not have two levels!\n")
return(1)
}


# generate the correlation results
x1<-bal.fpkm[,!is.na(clinic.vect) & clinic.vect==as.numeric(names(table(clinic.vect))[1])]
x2<-bal.fpkm[,!is.na(clinic.vect) & clinic.vect==as.numeric(names(table(clinic.vect))[2])]

#age.pvalue<-apply(bal.fpkm,1,my.cor.test,y=clinic.vect,method.name=method.name)
age.pvalue.direct<-apply(cbind(x1,x2),1,my.wilcox,n=ncol(x1))
age.pvalue<-abs(age.pvalue.direct)

#age.pvalue<-matrix(unlist(age.pvalue),ncol=2,byrow=T)
age.pvalue<-cbind(age.pvalue,rep(NA,length(age.pvalue)))
colnames(age.pvalue)<-c("pvalue","corr_coef")
age.fdr<-p.adjust(age.pvalue[,"pvalue"],method="fdr")
age.pvalue<-cbind(age.pvalue,age.fdr)
colnames(age.pvalue)[ncol(age.pvalue)]<-"fdr"
rownames(age.pvalue)<-rownames(bal.fpkm)
temp.fold<-apply(x1,1,mean,na.rm=T)/apply(x2,1,mean,na.rm=T)
age.pvalue<-cbind(age.pvalue,temp.fold)
colnames(age.pvalue)[ncol(age.pvalue)]<-"fold_change"

age.pvalue<-cbind(age.pvalue,age.pvalue.direct)
colnames(age.pvalue)[ncol(age.pvalue)]<-"pval_dir"


fpkm.info<-apply(bal.fpkm>0,1,sum)
fpkm.info<-cbind(fpkm.info,apply(bal.fpkm>0.1,1,sum))
fpkm.info<-cbind(fpkm.info,apply(bal.fpkm>1,1,sum))
fpkm.info<-cbind(fpkm.info,apply(bal.fpkm,1,median))
fpkm.info<-cbind(fpkm.info,apply(bal.fpkm,1,mean))
colnames(fpkm.info)<-c("fpkm>0_num","fpkm>0.1_num","fpkm>1_num","fpkm_median","fpkm_mean")


if(file.exists(outputdir)==F){
cat("WARNING: output.dir doesn't exists!\n")
return(1)
}

cmd.out<-cbind(age.pvalue,fpkm.info)
output.filepath<-file.path(outputdir,paste("correlation_",method.name,"_",feature.name,"_allgenes_",tissue.name,".txt",sep=""))
cat("gene_names\t",file=output.filepath,append=F)
write.table(cmd.out,file=output.filepath,append=T,sep="\t",row.names=T,col.names=T,quote=F)


if(sum(!is.na(age.pvalue[,"fdr"]) & age.pvalue[,"fdr"]<0.05)==0){
  # When there's nothing that pass the FDR threshold, we use the threshold of p value<0.05 & fold change >2
  sig.genes<-rownames(data.matrix)[!is.na(age.pvalue[,"pvalue"]) & age.pvalue[,"pvalue"]<0.05 & (!is.na(age.pvalue[,"fold_change"])) & (age.pvalue[,"fold_change"]>2 | age.pvalue[,"fold_change"]< 1/2) ]

  cmd.out<-cbind(age.pvalue,fpkm.info)
  output.filepath<-file.path(outputdir,paste("correlation_",method.name,"_",feature.name,"_pvalue0.05_FC2.0_",tissue.name,".txt",sep=""))
  cat("gene_names\t",file=output.filepath,append=F)
  write.table(cmd.out[sig.genes,],file=output.filepath,append=T,sep="\t",row.names=T,col.names=T,quote=F)

  output.filepath<-file.path(outputdir,paste("correlation_",method.name,"_",feature.name,"_pvalue0.05_FC2.0_",tissue.name,"_both_metacore.txt",sep=""))
  cat("gene_names\tintensity\n",file=output.filepath,append=F)
  sig.genes<-rownames(data.matrix)[!is.na(age.pvalue[,"pvalue"]) & age.pvalue[,"pvalue"]<0.05 & (!is.na(age.pvalue[,"fold_change"])) & (age.pvalue[,"fold_change"]>2 | age.pvalue[,"fold_change"]< 1/2) ]
  sig.genes<-unname(sapply(sig.genes,my.element.extract,splitchar="_",index=1))
  sig.genes<-unique(sig.genes)
  cmd.out<-cbind(sig.genes,rep(1,length(sig.genes)))
  write.table(cmd.out,file=output.filepath,append=T,sep="\t",row.names=F,col.names=F,quote=F)
  
  output.filepath<-file.path(outputdir,paste("correlation_",method.name,"_",feature.name,"_pvalue0.05_FC2.0_",tissue.name,"_positive_metacore.txt",sep=""))
  cat("gene_names\tintensity\n",file=output.filepath,append=F)
  sig.genes<-rownames(data.matrix)[!is.na(age.pvalue[,"pvalue"]) & age.pvalue[,"pvalue"]<0.05 & (!is.na(age.pvalue[,"fold_change"])) & (age.pvalue[,"fold_change"]>2 | age.pvalue[,"fold_change"]< 1/2) ]
  sig.genes<-sig.genes[age.pvalue[sig.genes,"pval_dir"]>0]
  sig.genes<-unname(sapply(sig.genes,my.element.extract,splitchar="_",index=1))
  sig.genes<-unique(sig.genes)
  cmd.out<-cbind(sig.genes,rep(1,length(sig.genes)))
  write.table(cmd.out,file=output.filepath,append=T,sep="\t",row.names=F,col.names=F,quote=F)
  
  
  output.filepath<-file.path(outputdir,paste("correlation_",method.name,"_",feature.name,"_pvalue0.05_FC2.0_",tissue.name,"_negative_metacore.txt",sep=""))
  cat("gene_names\tintensity\n",file=output.filepath,append=F)
  sig.genes<-rownames(data.matrix)[!is.na(age.pvalue[,"pvalue"]) & age.pvalue[,"pvalue"]<0.05 & (!is.na(age.pvalue[,"fold_change"])) & (age.pvalue[,"fold_change"]>2 | age.pvalue[,"fold_change"]< 1/2) ]
  sig.genes<-sig.genes[age.pvalue[sig.genes,"pval_dir"]<0]
  sig.genes<-unname(sapply(sig.genes,my.element.extract,splitchar="_",index=1))
  sig.genes<-unique(sig.genes)
  cmd.out<-cbind(sig.genes,rep(1,length(sig.genes)))
  write.table(cmd.out,file=output.filepath,append=T,sep="\t",row.names=F,col.names=F,quote=F)  
  
}else{
  # if there are anything significant with FDR<0.05
  cmd.out<-cbind(age.pvalue,fpkm.info)
  output.filepath<-file.path(outputdir,paste("correlation_",method.name,"_",feature.name,"_fdr0.05_",tissue.name,".txt",sep=""))
  cat("gene_names\t",file=output.filepath,append=F)
  write.table(cmd.out[!is.na(age.pvalue[,"fdr"]) & age.pvalue[,"fdr"]<0.05,],file=output.filepath,append=T,sep="\t",row.names=T,col.names=T,quote=F)

  output.filepath<-file.path(outputdir,paste("correlation_",method.name,"_",feature.name,"_fdr0.05_",tissue.name,"_both_metacore.txt",sep=""))
  cat("gene_names\tintensity\n",file=output.filepath,append=F)
  sig.genes<-rownames(data.matrix)[!is.na(age.pvalue[,"fdr"]) & age.pvalue[,"fdr"]<0.05]
  sig.genes<-unname(sapply(sig.genes,my.element.extract,splitchar="_",index=1))
  sig.genes<-unique(sig.genes)
  cmd.out<-cbind(sig.genes,rep(1,length(sig.genes)))
  write.table(cmd.out,file=output.filepath,append=T,sep="\t",row.names=F,col.names=F,quote=F)
  
  
  output.filepath<-file.path(outputdir,paste("correlation_",method.name,"_",feature.name,"_fdr0.05_",tissue.name,"_positive_metacore.txt",sep=""))
  cat("gene_names\tintensity\n",file=output.filepath,append=F)
  sig.genes<-rownames(data.matrix)[!is.na(age.pvalue[,"fdr"]) & age.pvalue[,"fdr"]<0.05 & age.pvalue[,"pval_dir"]>0]
  sig.genes<-unname(sapply(sig.genes,my.element.extract,splitchar="_",index=1))
  sig.genes<-unique(sig.genes)
  cmd.out<-cbind(sig.genes,rep(1,length(sig.genes)))
  write.table(cmd.out,file=output.filepath,append=T,sep="\t",row.names=F,col.names=F,quote=F)
  
  
  output.filepath<-file.path(outputdir,paste("correlation_",method.name,"_",feature.name,"_fdr0.05_",tissue.name,"_negative_metacore.txt",sep=""))
  cat("gene_names\tintensity\n",file=output.filepath,append=F)
  sig.genes<-rownames(data.matrix)[!is.na(age.pvalue[,"fdr"]) & age.pvalue[,"fdr"]<0.05 & age.pvalue[,"pval_dir"]<0]
  sig.genes<-unname(sapply(sig.genes,my.element.extract,splitchar="_",index=1))
  sig.genes<-unique(sig.genes)
  cmd.out<-cbind(sig.genes,rep(1,length(sig.genes)))
  write.table(cmd.out,file=output.filepath,append=T,sep="\t",row.names=F,col.names=F,quote=F)
}
return(0)
}

#output.dir<-"/home/fas/kaminski/am2952/scratch/SARC209clean/Results_summary/Demographics_correlated_genes"

##Female n=112 Male n=97 Female=0 Male=1 Female is the first group so the positive and negative correlated are for F
##White n=153 Black n=49 White=1 Black=2 White is the first group so the positive and negative correlated are for W
output.dir<-file.path(work.dir,"scratch/Antun/SUPERvisedA/Results_summary2/Demographics_correlated_genes")

rownames(sample.list.all)<-sample.list.all[,3]
rownames(clinic.matrix)<-as.matrix(clinic.matrix)[,1]
my.cat.cor(fpkm.matrix.clean,feature.name="GENDER",tissue.name="BAL",samplelist=sample.list.all,clinicmatrix=clinic.matrix,outputdir=output.dir)
clinic.matrix.new<-clinic.matrix
clinic.matrix.new[!clinic.matrix.new[,"RACE"]%in%c(1,2),"RACE"]<-NA
rownames(clinic.matrix.new)<-as.matrix(clinic.matrix.new)[,1]
my.cat.cor(fpkm.matrix.clean,feature.name="RACE",tissue.name="BAL",samplelist=sample.list.all,clinicmatrix=clinic.matrix.new,outputdir=output.dir)


output.dir<-file.path(work.dir,"scratch/Antun/SUPERvisedA/Results_summary2/PFTs_correlated_genes")
my.cat.cor(fpkm.matrix.clean,feature.name="FEV1FVCratioBinary",tissue.name="BAL",samplelist=sample.list.all,clinicmatrix=clinic.matrix,outputdir=output.dir)
my.cat.cor(fpkm.matrix.clean,feature.name="obstructiveBinary",tissue.name="BAL",samplelist=sample.list.all,clinicmatrix=clinic.matrix,outputdir=output.dir)
my.cat.cor(fpkm.matrix.clean,feature.name="restrictiveBinary",tissue.name="BAL",samplelist=sample.list.all,clinicmatrix=clinic.matrix,outputdir=output.dir)


#module load Apps/R/3.3.2-generic
#source("https://bioconductor.org/biocLite.R")
#biocLite("qvalue")

################################
# 4.6. Identify genes responsive to treatment
# matched data. columns of data.matrix are matched with rows in sample.matrix and my.clinic.matrix
# data.matrix	  	---- fpkm.matrix
# sample.matrix   	---- matched.list
# my.clinic.matrix	---- clinic.matrix
#source("/home/fas/kaminski/xy48/Rprogram/GRADS/0_CleanDataLoad.R")

#source("/home/antun/Desktop/SUPERvisedA/r_codes/209_local_data_load_AM.R")
#source(file.path(work.dir,"Rprogram/GRADS/BAL_final_hg38_20180726/SupervisedAnalysis/209_local_data_load_AM_XYversion.R"))


my.grads.wilcox<-function(x,n,alternative="two.sided"){
temp<-wilcox.test(x[1:n],x[(n+1):length(x)],alternative=alternative)
if(median(x[1:n])<median(x[(n+1):length(x)])){
result<-c(mean(x[1:n]),mean(x[(n+1):length(x)]),-mean(x[(n+1):length(x)])/mean(x[1:n]),sd(x)/mean(x),-temp$p.value)
}else{
result<-c(mean(x[1:n]),mean(x[(n+1):length(x)]),mean(x[1:n])/mean(x[(n+1):length(x)]),sd(x)/mean(x),temp$p.value)
}
names(result)<-c("mean treated","mean untreated","Fold Change (treated vs. untreated)","CV","Wilcoxon P value")
return(result)
}

#output.dir<-"/home/antun/Desktop/SUPERvisedA/Results_summary/Treatment_responsive_genes"
output.dir<-file.path(work.dir,"scratch/Antun/SUPERvisedA/Results_summary2/Treatment_responsive_genes")


# for BAL samples
data.matrix<-fpkm.matrix.clean
colnames(data.matrix)<-sapply(colnames(data.matrix),my.element.extract,splitchar="_",index=1)
sample.matrix<-sample.list.all
rownames(sample.matrix)<-sample.matrix[,3]
my.clinic.matrix<-clinic.matrix

bal.fpkm<-data.matrix[,substr(colnames(data.matrix),3,3)=="B"]
bal.sample.matrix<-sample.matrix[colnames(bal.fpkm),]
bal.clinic.matrix<-my.clinic.matrix[as.matrix(bal.sample.matrix)[,1],]
rownames(bal.clinic.matrix)<-colnames(bal.fpkm)

#pbmc.fpkm<-data.matrix[,substr(colnames(data.matrix),3,3)=="S"]
#pbmc.sample.matrix<-sample.matrix[colnames(pbmc.fpkm),]
#pbmc.clinic.matrix<-my.clinic.matrix[as.matrix(pbmc.sample.matrix)[,1],]
#rownames(pbmc.clinic.matrix)<-colnames(pbmc.fpkm)

# coding for the phenotype
#1 ---- Multi-organ
#2 ---- Non-acute, Stage I, untreated
#3 ---- Stage II-III, treated
#4 ---- Stage II-III, untreated
#5 ---- Stage IV, treated
#6 ---- Stage IV, untreated
#7 ---- Acute Sarcoidosis, untreated
#8 ---- Remitting, untreated
#9 ---- Cardiac defining therapy

#############
# compare stage II-III vs V untreated for wonnie
temp.pheno<-bal.clinic.matrix[,"PHENGRP"]
temp1<-bal.fpkm[,temp.pheno==4]
temp2<-bal.fpkm[,temp.pheno==6]

temp1.sd<-apply(temp1,1,sd)
temp2.sd<-apply(temp2,1,sd)

temp.data<-cbind(temp1,temp2)
temp.data<-temp.data[temp1.sd>0 | temp2.sd>0,]
bal.result<-t(apply(temp.data,1,my.grads.wilcox,n=ncol(temp1),alternative="two.sided"))
fdr.vect<-p.adjust(abs(bal.result[,"Wilcoxon P value"]),method="fdr")
library(qvalue)
q.vect<-qvalue(abs(bal.result[,"Wilcoxon P value"]))$qvalues
bal.result<-cbind(bal.result,fdr.vect,q.vect)
colnames(bal.result)[(ncol(bal.result)-1):ncol(bal.result)]<-c("FDR","q values")

rownames(anno.matrix)<-paste(anno.matrix[,1],"_",anno.matrix[,"tss_id"],sep="")

fpkm.info<-apply(temp.data>0,1,sum)
fpkm.info<-cbind(fpkm.info,apply(temp.data>0.1,1,sum))
fpkm.info<-cbind(fpkm.info,apply(temp.data>1,1,sum))
fpkm.info<-cbind(fpkm.info,apply(temp.data,1,median))
fpkm.info<-cbind(fpkm.info,apply(temp.data,1,mean))
colnames(fpkm.info)<-c("fpkm>0_num","fpkm>0.1_num","fpkm>1_num","fpkm_median","fpkm_mean")


output.filepath<-file.path(output.dir,"bal_treatmenteffect_result_stageII-III_V.txt")
cmd.out<-cbind(anno.matrix[rownames(bal.result),],bal.result,fpkm.info)
write.table(cmd.out,file=output.filepath,append=F,sep="\t",row.names=F,col.names=T,quote=F)

cmd.out<-cmd.out[!is.nan(as.numeric(cmd.out[,"Fold Change (treated vs. untreated)"])) & abs(as.numeric(cmd.out[,"Fold Change (treated vs. untreated)"]))>2 & abs(as.numeric(cmd.out[,"Wilcoxon P value"]))<0.05,]
output.filepath<-file.path(output.dir,"bal_treatmenteffect_result_stageII-III_V_pvalue0.05_FC2.0.txt")
write.table(cmd.out,file=output.filepath,append=F,sep="\t",row.names=F,col.names=T,quote=F)

temp<-paste(cmd.out[,1],"_",cmd.out[,"tss_id"],sep="")
output.filepath<-file.path(output.dir,"bal_treatmenteffect_result_stageII-III_V_pvalue0.05_FC2.0_genelist.txt")
cat(temp,file=output.filepath,append=F,sep="\n")

cmd.out<-unique(cmd.out[,1])
cmd.out<-cbind(cmd.out,rep(1,length(cmd.out)))
output.filepath<-file.path(output.dir,"bal_treatmenteffect_result_stageII-III_V_pvalue0.05_FC2.0_genelist_metacore.txt")
write.table(cmd.out,file=output.filepath,append=F,sep="\t",row.names=F,col.names=F,quote=F)



#############
# for stage II-III

## for BAL
temp.pheno<-bal.clinic.matrix[,"PHENGRP"]
temp1<-bal.fpkm[,temp.pheno==3]
temp2<-bal.fpkm[,temp.pheno==4]

temp1.sd<-apply(temp1,1,sd)
temp2.sd<-apply(temp2,1,sd)

temp.data<-cbind(temp1,temp2)
temp.data<-temp.data[temp1.sd>0 | temp2.sd>0,]
bal.result<-t(apply(temp.data,1,my.grads.wilcox,n=ncol(temp1),alternative="two.sided"))
fdr.vect<-p.adjust(abs(bal.result[,"Wilcoxon P value"]),method="fdr")
library(qvalue)
q.vect<-qvalue(abs(bal.result[,"Wilcoxon P value"]))$qvalues
bal.result<-cbind(bal.result,fdr.vect,q.vect)
colnames(bal.result)[(ncol(bal.result)-1):ncol(bal.result)]<-c("FDR","q values")




rownames(anno.matrix)<-paste(anno.matrix[,1],"_",anno.matrix[,"tss_id"],sep="")

fpkm.info<-apply(temp.data>0,1,sum)
fpkm.info<-cbind(fpkm.info,apply(temp.data>0.1,1,sum))
fpkm.info<-cbind(fpkm.info,apply(temp.data>1,1,sum))
fpkm.info<-cbind(fpkm.info,apply(temp.data,1,median))
fpkm.info<-cbind(fpkm.info,apply(temp.data,1,mean))
colnames(fpkm.info)<-c("fpkm>0_num","fpkm>0.1_num","fpkm>1_num","fpkm_median","fpkm_mean")



output.filepath<-file.path(output.dir,"bal_treatmenteffect_result_stageII-III.txt")
cmd.out<-cbind(anno.matrix[rownames(bal.result),],bal.result,fpkm.info)
write.table(cmd.out,file=output.filepath,append=F,sep="\t",row.names=F,col.names=T,quote=F)

cmd.out<-cmd.out[!is.nan(as.numeric(cmd.out[,"Fold Change (treated vs. untreated)"])) & abs(as.numeric(cmd.out[,"Fold Change (treated vs. untreated)"]))>2 & abs(as.numeric(cmd.out[,"Wilcoxon P value"]))<0.05,]
output.filepath<-file.path(output.dir,"bal_treatmenteffect_result_stageII-III_pvalue0.05_FC2.0.txt")
write.table(cmd.out,file=output.filepath,append=F,sep="\t",row.names=F,col.names=T,quote=F)

temp<-paste(cmd.out[,1],"_",cmd.out[,"tss_id"],sep="")
output.filepath<-file.path(output.dir,"bal_treatmenteffect_result_stageII-III_pvalue0.05_FC2.0_genelist.txt")
cat(temp,file=output.filepath,append=F,sep="\n")

cmd.out<-unique(cmd.out[,1])
cmd.out<-cbind(cmd.out,rep(1,length(cmd.out)))
output.filepath<-file.path(output.dir,"bal_treatmenteffect_result_stageII-III_pvalue0.05_FC2.0_genelist_metacore.txt")
write.table(cmd.out,file=output.filepath,append=F,sep="\t",row.names=F,col.names=F,quote=F)

## for PBMC
#temp.pheno<-pbmc.clinic.matrix[,"PHENGRP"]
#temp1<-pbmc.fpkm[,temp.pheno==3]
#temp2<-pbmc.fpkm[,temp.pheno==4]

#temp1.sd<-apply(temp1,1,sd)
#temp2.sd<-apply(temp2,1,sd)

#temp.data<-cbind(temp1,temp2)
#temp.data<-temp.data[temp1.sd>0 | temp2.sd>0,]
#pbmc.result<-t(apply(temp.data,1,my.grads.wilcox,n=ncol(temp1),alternative="two.sided"))
#fdr.vect<-p.adjust(abs(pbmc.result[,"Wilcoxon P value"]),method="fdr")
#library(qvalue)
#q.vect<-qvalue(abs(pbmc.result[,"Wilcoxon P value"]))$qvalues
#pbmc.result<-cbind(pbmc.result,fdr.vect,q.vect)
#colnames(pbmc.result)[(ncol(pbmc.result)-1):ncol(pbmc.result)]<-c("FDR","q values")

#rownames(anno.matrix)<-paste(anno.matrix[,1],"_",anno.matrix[,"tss_id"],sep="")

#fpkm.info<-apply(temp.data>0,1,sum)
#fpkm.info<-cbind(fpkm.info,apply(temp.data>0.1,1,sum))
#fpkm.info<-cbind(fpkm.info,apply(temp.data>1,1,sum))
#fpkm.info<-cbind(fpkm.info,apply(temp.data,1,median))
#fpkm.info<-cbind(fpkm.info,apply(temp.data,1,mean))
#colnames(fpkm.info)<-c("fpkm>0_num","fpkm>0.1_num","fpkm>1_num","fpkm_median","fpkm_mean")

#output.filepath<-file.path(output.dir,"pbmc_treatmenteffect_result_stageII-III.txt")
#cmd.out<-cbind(anno.matrix[rownames(pbmc.result),],pbmc.result,fpkm.info)
#write.table(cmd.out,file=output.filepath,append=F,sep="\t",row.names=F,col.names=T,quote=F)

# select genes with pvalue<0.05 and fold change>2
#cmd.out<-cmd.out[!is.nan(as.numeric(cmd.out[,"Fold Change (treated vs. untreated)"])) & abs(as.numeric(cmd.out[,"Fold Change (treated vs. untreated)"]))>2 & abs(as.numeric(cmd.out[,"Wilcoxon P value"]))<0.05,]
#output.filepath<-file.path(output.dir,"pbmc_treatmenteffect_result_stageII-III_pvalue0.05_FC2.0.txt")
#write.table(cmd.out,file=output.filepath,append=F,sep="\t",row.names=F,col.names=T,quote=F)

#temp<-paste(cmd.out[,1],"_",cmd.out[,"tss_id"],sep="")
#output.filepath<-file.path(output.dir,"pbmc_treatmenteffect_result_stageII-III_pvalue0.05_FC2.0_genelist.txt")
#cat(temp,file=output.filepath,append=F,sep="\n")


#cmd.out<-unique(cmd.out[,1])
#cmd.out<-cbind(cmd.out,rep(1,length(cmd.out)))
#output.filepath<-file.path(output.dir,"pbmc_treatmenteffect_result_stageII-III_pvalue0.05_FC2.0_genelist_metacore.txt")
#write.table(cmd.out,file=output.filepath,append=F,sep="\t",row.names=F,col.names=F,quote=F)




#############
# for stage IV

## for BAL
temp.pheno<-bal.clinic.matrix[,"PHENGRP"]
temp1<-bal.fpkm[,temp.pheno==5]
temp2<-bal.fpkm[,temp.pheno==6]

temp1.sd<-apply(temp1,1,sd)
temp2.sd<-apply(temp2,1,sd)

temp.data<-cbind(temp1,temp2)
temp.data<-temp.data[temp1.sd>0 | temp2.sd>0,]
bal.result<-t(apply(temp.data,1,my.grads.wilcox,n=ncol(temp1),alternative="two.sided"))
fdr.vect<-p.adjust(abs(bal.result[,"Wilcoxon P value"]),method="fdr")
library(qvalue)
q.vect<-qvalue(abs(bal.result[,"Wilcoxon P value"]))$qvalues
bal.result<-cbind(bal.result,fdr.vect,q.vect)
colnames(bal.result)[(ncol(bal.result)-1):ncol(bal.result)]<-c("FDR","q values")

rownames(anno.matrix)<-paste(anno.matrix[,1],"_",anno.matrix[,"tss_id"],sep="")


fpkm.info<-apply(temp.data>0,1,sum)
fpkm.info<-cbind(fpkm.info,apply(temp.data>0.1,1,sum))
fpkm.info<-cbind(fpkm.info,apply(temp.data>1,1,sum))
fpkm.info<-cbind(fpkm.info,apply(temp.data,1,median))
fpkm.info<-cbind(fpkm.info,apply(temp.data,1,mean))
colnames(fpkm.info)<-c("fpkm>0_num","fpkm>0.1_num","fpkm>1_num","fpkm_median","fpkm_mean")


output.filepath<-file.path(output.dir,"bal_treatmenteffect_result_stageIV.txt")
cmd.out<-cbind(anno.matrix[rownames(bal.result),],bal.result,fpkm.info)
write.table(cmd.out,file=output.filepath,append=F,sep="\t",row.names=F,col.names=T,quote=F)


cmd.out<-cmd.out[!is.nan(as.numeric(cmd.out[,"Fold Change (treated vs. untreated)"])) & abs(as.numeric(cmd.out[,"Fold Change (treated vs. untreated)"]))>2 & abs(as.numeric(cmd.out[,"Wilcoxon P value"]))<0.05,]
output.filepath<-file.path(output.dir,"bal_treatmenteffect_result_stageIV_pvalue0.05_FC2.0.txt")
write.table(cmd.out,file=output.filepath,append=F,sep="\t",row.names=F,col.names=T,quote=F)

temp<-paste(cmd.out[,1],"_",cmd.out[,"tss_id"],sep="")
output.filepath<-file.path(output.dir,"bal_treatmenteffect_result_stageIV_pvalue0.05_FC2.0_genelist.txt")
cat(temp,file=output.filepath,append=F,sep="\n")


cmd.out<-unique(cmd.out[,1])
cmd.out<-cbind(cmd.out,rep(1,length(cmd.out)))
output.filepath<-file.path(output.dir,"bal_treatmenteffect_result_stageIV_pvalue0.05_FC2.0_genelist_metacore.txt")
write.table(cmd.out,file=output.filepath,append=F,sep="\t",row.names=F,col.names=F,quote=F)


## for PBMC
#temp.pheno<-pbmc.clinic.matrix[,"PHENGRP"]
#temp1<-pbmc.fpkm[,temp.pheno==5]
#temp2<-pbmc.fpkm[,temp.pheno==6]

#temp1.sd<-apply(temp1,1,sd)
#temp2.sd<-apply(temp2,1,sd)

#temp.data<-cbind(temp1,temp2)
#temp.data<-temp.data[temp1.sd>0 | temp2.sd>0,]
#pbmc.result<-t(apply(temp.data,1,my.grads.wilcox,n=ncol(temp1),alternative="two.sided"))
#fdr.vect<-p.adjust(abs(pbmc.result[,"Wilcoxon P value"]),method="fdr")
#library(qvalue)
#q.vect<-qvalue(abs(pbmc.result[,"Wilcoxon P value"]))$qvalues
#pbmc.result<-cbind(pbmc.result,fdr.vect,q.vect)
#colnames(pbmc.result)[(ncol(pbmc.result)-1):ncol(pbmc.result)]<-c("FDR","q values")

#rownames(anno.matrix)<-paste(anno.matrix[,1],"_",anno.matrix[,"tss_id"],sep="")


#fpkm.info<-apply(temp.data>0,1,sum)
#fpkm.info<-cbind(fpkm.info,apply(temp.data>0.1,1,sum))
#fpkm.info<-cbind(fpkm.info,apply(temp.data>1,1,sum))
#fpkm.info<-cbind(fpkm.info,apply(temp.data,1,median))
#fpkm.info<-cbind(fpkm.info,apply(temp.data,1,mean))
#colnames(fpkm.info)<-c("fpkm>0_num","fpkm>0.1_num","fpkm>1_num","fpkm_median","fpkm_mean")


#output.filepath<-file.path(output.dir,"pbmc_treatmenteffect_result_stageIV.txt")
#cmd.out<-cbind(anno.matrix[rownames(pbmc.result),],pbmc.result,fpkm.info)
#write.table(cmd.out,file=output.filepath,append=F,sep="\t",row.names=F,col.names=T,quote=F)


#cmd.out<-cmd.out[!is.nan(as.numeric(cmd.out[,"Fold Change (treated vs. untreated)"])) & abs(as.numeric(cmd.out[,"Fold Change (treated vs. untreated)"]))>2 & abs(as.numeric(cmd.out[,"Wilcoxon P value"]))<0.05,]
#output.filepath<-file.path(output.dir,"pbmc_treatmenteffect_result_stageIV_pvalue0.05_FC2.0.txt")
#write.table(cmd.out,file=output.filepath,append=F,sep="\t",row.names=F,col.names=T,quote=F)

#temp<-paste(cmd.out[,1],"_",cmd.out[,"tss_id"],sep="")
#output.filepath<-file.path(output.dir,"pbmc_treatmenteffect_result_stageIV_pvalue0.05_FC2.0_genelist.txt")
#cat(temp,file=output.filepath,append=F,sep="\n")


#cmd.out<-unique(cmd.out[,1])
#cmd.out<-cbind(cmd.out,rep(1,length(cmd.out)))
#output.filepath<-file.path(output.dir,"pbmc_treatmenteffect_result_stageIV_pvalue0.05_FC2.0_genelist_metacore.txt")
#write.table(cmd.out,file=output.filepath,append=F,sep="\t",row.names=F,col.names=F,quote=F)





################################
# 4.7. Identify genes differentially expressed between each phenotype group and non-accute stage I, untreated
# matched data. columns of data.matrix are matched with rows in sample.matrix and my.clinic.matrix
# data.matrix	  	---- fpkm.matrix
# sample.matrix   	---- matched.list
# my.clinic.matrix	---- clinic.matrix
#source("/home/fas/kaminski/xy48/Rprogram/GRADS/0_CleanDataLoad.R")

source("/home/antun/Desktop/SUPERvisedA/r_codes/209_local_data_load_AM.R")




my.grads.wilcox<-function(x,n,alternative="two.sided"){
temp<-wilcox.test(x[1:n],x[(n+1):length(x)],alternative=alternative)
if(median(x[1:n])<median(x[(n+1):length(x)])){
result<-c(mean(x[1:n]),mean(x[(n+1):length(x)]),-mean(x[(n+1):length(x)])/mean(x[1:n]),sd(x)/mean(x),-temp$p.value)
}else{
result<-c(mean(x[1:n]),mean(x[(n+1):length(x)]),mean(x[1:n])/mean(x[(n+1):length(x)]),sd(x)/mean(x),temp$p.value)
}
names(result)<-c("mean other","mean nonaccute stage I, untreated","Fold Change (other vs. nonaccute stage I, untreated)","CV","Wilcoxon P value")
return(result)
}

output.dir<-"/home/antun/Desktop/SUPERvisedA/Results_summary/Phenotype_DEGs"

# for BAL samples
bal.fpkm<-data.matrix[,substr(colnames(data.matrix),3,3)=="B"]
bal.sample.matrix<-sample.matrix[colnames(bal.fpkm),]
bal.clinic.matrix<-my.clinic.matrix[as.matrix(bal.sample.matrix)[,1],]
rownames(bal.clinic.matrix)<-colnames(bal.fpkm)

#pbmc.fpkm<-data.matrix[,substr(colnames(data.matrix),3,3)=="S"]
#pbmc.sample.matrix<-sample.matrix[colnames(pbmc.fpkm),]
#pbmc.clinic.matrix<-my.clinic.matrix[as.matrix(pbmc.sample.matrix)[,1],]
#rownames(pbmc.clinic.matrix)<-colnames(pbmc.fpkm)

# coding for the phenotype
#1 ---- Multi-organ
#2 ---- Non-acute, Stage I, untreated
#3 ---- Stage II-III, treated
#4 ---- Stage II-III, untreated
#5 ---- Stage IV, treated
#6 ---- Stage IV, untreated
#7 ---- Acute Sarcoidosis, untreated
#8 ---- Remitting, untreated
#9 ---- Cardiac defining therapy

## for BAL
other.pheno<-1:8
other.pheno<-setdiff(other.pheno,2)
other.pheno.names<-c(
"Multi-organ",
"Non-acute, Stage I, untreated",
"Stage II-III, treated",
"Stage II-III, untreated",
"Stage IV, treated",
"Stage IV, untreated",
"Acute Sarcoidosis, untreated",
"Remitting, untreated")
other.pheno.names<-other.pheno.names[-2]

for(j in 1:length(other.pheno)){
temp.pheno<-bal.clinic.matrix[,"PHENGRP"]
temp2<-bal.fpkm[,temp.pheno==2]
temp1<-bal.fpkm[,temp.pheno==other.pheno[j]]

temp1.sd<-apply(temp1,1,sd)
temp2.sd<-apply(temp2,1,sd)

temp.data<-cbind(temp1,temp2)
temp.data<-temp.data[temp1.sd>0 | temp2.sd>0,]
bal.result<-t(apply(temp.data,1,my.grads.wilcox,n=ncol(temp1),alternative="two.sided"))
fdr.vect<-p.adjust(abs(bal.result[,"Wilcoxon P value"]),method="fdr")
library(qvalue)
q.vect<-qvalue(abs(bal.result[,"Wilcoxon P value"]))$qvalues
bal.result<-cbind(bal.result,fdr.vect,q.vect)
colnames(bal.result)[(ncol(bal.result)-1):ncol(bal.result)]<-c("FDR","q values")

rownames(anno.matrix)<-paste(anno.matrix[,1],"_",anno.matrix[,"tss_id"],sep="")

fpkm.info<-apply(temp.data>0,1,sum)
fpkm.info<-cbind(fpkm.info,apply(temp.data>0.1,1,sum))
fpkm.info<-cbind(fpkm.info,apply(temp.data>1,1,sum))
fpkm.info<-cbind(fpkm.info,apply(temp.data,1,median))
fpkm.info<-cbind(fpkm.info,apply(temp.data,1,mean))
colnames(fpkm.info)<-c("fpkm>0_num","fpkm>0.1_num","fpkm>1_num","fpkm_median","fpkm_mean")

output.filepath<-file.path(output.dir,paste("bal_DEGs_",other.pheno.names[j],"_all.txt",sep=""))
cmd.out<-cbind(anno.matrix[rownames(bal.result),],bal.result,fpkm.info)
write.table(cmd.out,file=output.filepath,append=F,sep="\t",row.names=F,col.names=T,quote=F)

cmd.out<-cmd.out[!is.nan(as.numeric(cmd.out[,"Fold Change (other vs. nonaccute stage I, untreated)"])) & abs(as.numeric(cmd.out[,"Fold Change (other vs. nonaccute stage I, untreated)"]))>2 & abs(as.numeric(cmd.out[,"Wilcoxon P value"]))<0.05,]
output.filepath<-file.path(output.dir,paste("bal_DEGs_",other.pheno.names[j],"_pvalue0.05_FC2.0.txt",sep=""))
write.table(cmd.out,file=output.filepath,append=F,sep="\t",row.names=F,col.names=T,quote=F)

temp<-paste(cmd.out[,1],"_",cmd.out[,"tss_id"],sep="")
output.filepath<-file.path(output.dir,paste("bal_DEGs_",other.pheno.names[j],"_pvalue0.05_FC2.0_genelist.txt",sep=""))
cat(temp,file=output.filepath,append=F,sep="\n")

cmd.out<-unique(cmd.out[,1])
cmd.out<-cbind(cmd.out,rep(1,length(cmd.out)))
output.filepath<-file.path(output.dir,paste("bal_DEGs_",other.pheno.names[j],"_pvalue0.05_FC2.0_genelist_metacore.txt",sep=""))
write.table(cmd.out,file=output.filepath,append=F,sep="\t",row.names=F,col.names=F,quote=F)

## for PBMC
#temp.pheno<-pbmc.clinic.matrix[,"PHENGRP"]
#temp1<-pbmc.fpkm[,temp.pheno==2]
#temp2<-pbmc.fpkm[,temp.pheno==other.pheno[j]]

#temp1.sd<-apply(temp1,1,sd)
#temp2.sd<-apply(temp2,1,sd)

#temp.data<-cbind(temp1,temp2)
#temp.data<-temp.data[temp1.sd>0 | temp2.sd>0,]
#pbmc.result<-t(apply(temp.data,1,my.grads.wilcox,n=ncol(temp1),alternative="two.sided"))
#fdr.vect<-p.adjust(abs(pbmc.result[,"Wilcoxon P value"]),method="fdr")
#library(qvalue)
#q.vect<-qvalue(abs(pbmc.result[,"Wilcoxon P value"]))$qvalues
#pbmc.result<-cbind(pbmc.result,fdr.vect,q.vect)
#colnames(pbmc.result)[(ncol(pbmc.result)-1):ncol(pbmc.result)]<-c("FDR","q values")

#rownames(anno.matrix)<-paste(anno.matrix[,1],"_",anno.matrix[,"tss_id"],sep="")

#fpkm.info<-apply(temp.data>0,1,sum)
#fpkm.info<-cbind(fpkm.info,apply(temp.data>0.1,1,sum))
#fpkm.info<-cbind(fpkm.info,apply(temp.data>1,1,sum))
#fpkm.info<-cbind(fpkm.info,apply(temp.data,1,median))
#fpkm.info<-cbind(fpkm.info,apply(temp.data,1,mean))
#colnames(fpkm.info)<-c("fpkm>0_num","fpkm>0.1_num","fpkm>1_num","fpkm_median","fpkm_mean")

#output.filepath<-file.path(output.dir,paste("pbmc_DEGs_",other.pheno.names[j],"_all.txt",sep=""))
#cmd.out<-cbind(anno.matrix[rownames(pbmc.result),],pbmc.result,fpkm.info)
#write.table(cmd.out,file=output.filepath,append=F,sep="\t",row.names=F,col.names=T,quote=F)

# select genes with pvalue<0.05 and fold change>2
#cmd.out<-cmd.out[!is.nan(as.numeric(cmd.out[,"Fold Change (other vs. nonaccute stage I, untreated)"])) & abs(as.numeric(cmd.out[,"Fold Change (other vs. nonaccute stage I, untreated)"]))>2 & abs(as.numeric(cmd.out[,"Wilcoxon P value"]))<0.05,]
#output.filepath<-file.path(output.dir,paste("pbmc_DEGs_",other.pheno.names[j],"_pvalue0.05_FC2.0.txt",sep=""))
#write.table(cmd.out,file=output.filepath,append=F,sep="\t",row.names=F,col.names=T,quote=F)

#temp<-paste(cmd.out[,1],"_",cmd.out[,"tss_id"],sep="")
#output.filepath<-file.path(output.dir,paste("pbmc_DEGs_",other.pheno.names[j],"_pvalue0.05_FC2.0_genelist.txt",sep=""))
#cat(temp,file=output.filepath,append=F,sep="\n")


#cmd.out<-unique(cmd.out[,1])
#cmd.out<-cbind(cmd.out,rep(1,length(cmd.out)))
#output.filepath<-file.path(output.dir,paste("pbmc_DEGs_",other.pheno.names[j],"_pvalue0.05_FC2.0_genelist_metacore.txt",sep=""))
#write.table(cmd.out,file=output.filepath,append=F,sep="\t",row.names=F,col.names=F,quote=F)

}

# generate the heatmap to show the variation across different phenotype groups using the genes identified above.









################################
# 4.8. Identify genes differentially expressed between each SCADDING stage and SCADDING stage 0
# matched data. columns of data.matrix are matched with rows in sample.matrix and my.clinic.matrix
# data.matrix	  	---- fpkm.matrix
# sample.matrix   	---- matched.list
# my.clinic.matrix	---- clinic.matrix
#source("/home/fas/kaminski/xy48/Rprogram/GRADS/0_CleanDataLoad.R")

source("/home/antun/Desktop/SUPERvisedA/r_codes/209_local_data_load_AM.R")



my.grads.wilcox<-function(x,n,alternative="two.sided"){
temp<-wilcox.test(x[1:n],x[(n+1):length(x)],alternative=alternative)
if(median(x[1:n])<median(x[(n+1):length(x)])){
result<-c(mean(x[1:n]),mean(x[(n+1):length(x)]),-mean(x[(n+1):length(x)])/mean(x[1:n]),sd(x)/mean(x),-temp$p.value)
}else{
result<-c(mean(x[1:n]),mean(x[(n+1):length(x)]),mean(x[1:n])/mean(x[(n+1):length(x)]),sd(x)/mean(x),temp$p.value)
}
names(result)<-c("mean other","mean SCADDING 0","Fold Change (other vs. SCADDING_0)","CV","Wilcoxon P value")
return(result)
}

output.dir<-"/home/antun/Desktop/SUPERvisedA/Results_summary/SCADDING_CAT_DEGs"

if(file.exists(output.dir)==F){
dir.create(output.dir)
}

# for BAL samples
bal.fpkm<-data.matrix[,substr(colnames(data.matrix),3,3)=="B"]
bal.sample.matrix<-sample.matrix[colnames(bal.fpkm),]
bal.clinic.matrix<-my.clinic.matrix[as.matrix(bal.sample.matrix)[,1],]
rownames(bal.clinic.matrix)<-colnames(bal.fpkm)

#pbmc.fpkm<-data.matrix[,substr(colnames(data.matrix),3,3)=="S"]
#pbmc.sample.matrix<-sample.matrix[colnames(pbmc.fpkm),]
#pbmc.clinic.matrix<-my.clinic.matrix[as.matrix(pbmc.sample.matrix)[,1],]
#rownames(pbmc.clinic.matrix)<-colnames(pbmc.fpkm)

## for BAL
other.pheno<-1:4
other.pheno.names<-paste("SCADDING_",other.pheno,sep="")

for(j in 1:length(other.pheno)){
temp.pheno<-bal.clinic.matrix[,"SCADDING"]

#changing the one sample with NA SCADDING in clinical to make the loop work
temp.pheno[is.na(temp.pheno)] <- "NA"

temp2<-bal.fpkm[,temp.pheno==0]
temp1<-bal.fpkm[,temp.pheno==other.pheno[j]]

temp1.sd<-apply(temp1,1,sd)
temp2.sd<-apply(temp2,1,sd)

temp.data<-cbind(temp1,temp2)
temp.data<-temp.data[temp1.sd>0 | temp2.sd>0,]
bal.result<-t(apply(temp.data,1,my.grads.wilcox,n=ncol(temp1),alternative="two.sided"))
fdr.vect<-p.adjust(abs(bal.result[,"Wilcoxon P value"]),method="fdr")
library(qvalue)
q.vect<-qvalue(abs(bal.result[,"Wilcoxon P value"]))$qvalues
bal.result<-cbind(bal.result,fdr.vect,q.vect)
colnames(bal.result)[(ncol(bal.result)-1):ncol(bal.result)]<-c("FDR","q values")

rownames(anno.matrix)<-paste(anno.matrix[,1],"_",anno.matrix[,"tss_id"],sep="")

fpkm.info<-apply(temp.data>0,1,sum)
fpkm.info<-cbind(fpkm.info,apply(temp.data>0.1,1,sum))
fpkm.info<-cbind(fpkm.info,apply(temp.data>1,1,sum))
fpkm.info<-cbind(fpkm.info,apply(temp.data,1,median))
fpkm.info<-cbind(fpkm.info,apply(temp.data,1,mean))
colnames(fpkm.info)<-c("fpkm>0_num","fpkm>0.1_num","fpkm>1_num","fpkm_median","fpkm_mean")

output.filepath<-file.path(output.dir,paste("bal_DEGs_",other.pheno.names[j],"_all.txt",sep=""))
cmd.out<-cbind(anno.matrix[rownames(bal.result),],bal.result,fpkm.info)
write.table(cmd.out,file=output.filepath,append=F,sep="\t",row.names=F,col.names=T,quote=F)

cmd.out<-cmd.out[!is.nan(as.numeric(cmd.out[,"Fold Change (other vs. SCADDING_0)"])) & abs(as.numeric(cmd.out[,"Fold Change (other vs. SCADDING_0)"]))>2 & abs(as.numeric(cmd.out[,"Wilcoxon P value"]))<0.05,]
output.filepath<-file.path(output.dir,paste("bal_DEGs_",other.pheno.names[j],"_pvalue0.05_FC2.0.txt",sep=""))
write.table(cmd.out,file=output.filepath,append=F,sep="\t",row.names=F,col.names=T,quote=F)

temp<-paste(cmd.out[,1],"_",cmd.out[,"tss_id"],sep="")
output.filepath<-file.path(output.dir,paste("bal_DEGs_",other.pheno.names[j],"_pvalue0.05_FC2.0_genelist.txt",sep=""))
cat(temp,file=output.filepath,append=F,sep="\n")

cmd.out<-unique(cmd.out[,1])
cmd.out<-cbind(cmd.out,rep(1,length(cmd.out)))
output.filepath<-file.path(output.dir,paste("bal_DEGs_",other.pheno.names[j],"_pvalue0.05_FC2.0_genelist_metacore.txt",sep=""))
write.table(cmd.out,file=output.filepath,append=F,sep="\t",row.names=F,col.names=F,quote=F)

## for PBMC
#temp.pheno<-pbmc.clinic.matrix[,"SCADDING"]
#temp1<-pbmc.fpkm[,temp.pheno==0]
#temp2<-pbmc.fpkm[,temp.pheno==other.pheno[j]]

#temp1.sd<-apply(temp1,1,sd)
#temp2.sd<-apply(temp2,1,sd)

#temp.data<-cbind(temp1,temp2)
#temp.data<-temp.data[temp1.sd>0 | temp2.sd>0,]
#pbmc.result<-t(apply(temp.data,1,my.grads.wilcox,n=ncol(temp1),alternative="two.sided"))
#fdr.vect<-p.adjust(abs(pbmc.result[,"Wilcoxon P value"]),method="fdr")
#library(qvalue)
#q.vect<-qvalue(abs(pbmc.result[,"Wilcoxon P value"]))$qvalues
#pbmc.result<-cbind(pbmc.result,fdr.vect,q.vect)
#colnames(pbmc.result)[(ncol(pbmc.result)-1):ncol(pbmc.result)]<-c("FDR","q values")

#rownames(anno.matrix)<-paste(anno.matrix[,1],"_",anno.matrix[,"tss_id"],sep="")


#fpkm.info<-apply(temp.data>0,1,sum)
#fpkm.info<-cbind(fpkm.info,apply(temp.data>0.1,1,sum))
#fpkm.info<-cbind(fpkm.info,apply(temp.data>1,1,sum))
#fpkm.info<-cbind(fpkm.info,apply(temp.data,1,median))
#fpkm.info<-cbind(fpkm.info,apply(temp.data,1,mean))
#colnames(fpkm.info)<-c("fpkm>0_num","fpkm>0.1_num","fpkm>1_num","fpkm_median","fpkm_mean")

#output.filepath<-file.path(output.dir,paste("pbmc_DEGs_",other.pheno.names[j],"_all.txt",sep=""))
#cmd.out<-cbind(anno.matrix[rownames(pbmc.result),],pbmc.result,fpkm.info)
#write.table(cmd.out,file=output.filepath,append=F,sep="\t",row.names=F,col.names=T,quote=F)

# select genes with pvalue<0.05 and fold change>2
#cmd.out<-cmd.out[!is.nan(as.numeric(cmd.out[,"Fold Change (other vs. SCADDING_0)"])) & abs(as.numeric(cmd.out[,"Fold Change (other vs. SCADDING_0)"]))>2 & abs(as.numeric(cmd.out[,"Wilcoxon P value"]))<0.05,]
#output.filepath<-file.path(output.dir,paste("pbmc_DEGs_",other.pheno.names[j],"_pvalue0.05_FC2.0.txt",sep=""))
#write.table(cmd.out,file=output.filepath,append=F,sep="\t",row.names=F,col.names=T,quote=F)

#temp<-paste(cmd.out[,1],"_",cmd.out[,"tss_id"],sep="")
#output.filepath<-file.path(output.dir,paste("pbmc_DEGs_",other.pheno.names[j],"_pvalue0.05_FC2.0_genelist.txt",sep=""))
#cat(temp,file=output.filepath,append=F,sep="\n")


#cmd.out<-unique(cmd.out[,1])
#cmd.out<-cbind(cmd.out,rep(1,length(cmd.out)))
#output.filepath<-file.path(output.dir,paste("pbmc_DEGs_",other.pheno.names[j],"_pvalue0.05_FC2.0_genelist_metacore.txt",sep=""))
#write.table(cmd.out,file=output.filepath,append=F,sep="\t",row.names=F,col.names=F,quote=F)

}




#module load Apps/R/3.3.2-generic

################################
# 4.9. Categorizing PFTs and identify genes differentially expressed between the lowest category with every other category
# matched data. columns of data.matrix are matched with rows in sample.matrix and my.clinic.matrix
# data.matrix	  	---- fpkm.matrix
# sample.matrix   	---- matched.list
# my.clinic.matrix	---- clinic.matrix
#source("/home/fas/kaminski/xy48/Rprogram/GRADS/0_CleanDataLoad.R")

#source("/home/fas/kaminski/am2952/scratch/SARC215/0_CleanDataLoad.R")

source("/home/antun/Desktop/SUPERvisedA/r_codes/209_local_data_load_AM.R")



my.grads.wilcox<-function(x,n,alternative="two.sided",var.name=""){
temp<-wilcox.test(x[1:n],x[(n+1):length(x)],alternative=alternative)
if(median(x[1:n])<median(x[(n+1):length(x)])){
result<-c(mean(x[1:n]),mean(x[(n+1):length(x)]),-mean(x[(n+1):length(x)])/mean(x[1:n]),sd(x)/mean(x),-temp$p.value)
}else{
result<-c(mean(x[1:n]),mean(x[(n+1):length(x)]),mean(x[1:n])/mean(x[(n+1):length(x)]),sd(x)/mean(x),temp$p.value)
}
names(result)<-c("mean other",paste("mean ",var.name,sep=""),paste("Fold Change (other vs. ",var.name,")",sep=""),"CV","Wilcoxon P value")
return(result)
}


output.dir<-"/home/antun/Desktop/SUPERvisedA/Results_summary/PFTs_CAT_DEGs"

if(file.exists(output.dir)==F){
dir.create(output.dir)
}

pft.names<-c("FVCPRED","FEV1PRED","PREDDLCO")
pft.threshold<-matrix(c(35,50,80,30,50,80,35,50,80),nrow=3,byrow=T)

for(k in 1:length(pft.names)){

bal.fpkm<-data.matrix[,substr(colnames(data.matrix),3,3)=="B"]
bal.sample.matrix<-sample.matrix[colnames(bal.fpkm),]
bal.clinic.matrix<-my.clinic.matrix[as.matrix(bal.sample.matrix)[,1],]
rownames(bal.clinic.matrix)<-colnames(bal.fpkm)

#pbmc.fpkm<-data.matrix[,substr(colnames(data.matrix),3,3)=="S"]
#pbmc.sample.matrix<-sample.matrix[colnames(pbmc.fpkm),]
#pbmc.clinic.matrix<-my.clinic.matrix[as.matrix(pbmc.sample.matrix)[,1],]
#rownames(pbmc.clinic.matrix)<-colnames(pbmc.fpkm)

other.pheno<-rbind(pft.threshold[k,],c(pft.threshold[k,2:ncol(pft.threshold)],"+"))
other.pheno[2,1:(ncol(other.pheno)-1)]<-paste("-",other.pheno[2,1:(ncol(other.pheno)-1)],sep="")
other.pheno.names<-apply(other.pheno,2,paste,collapse="")

library(qvalue)

## for BAL
temp.pheno<-bal.clinic.matrix[,pft.names[k]]
temp.pheno[temp.pheno<0]<-NA
for(j in 1:length(other.pheno.names)){
temp2<-bal.fpkm[,(!is.na(temp.pheno)) & (temp.pheno>=pft.threshold[k,2] & temp.pheno<pft.threshold[k,3]) & (temp.pheno>=0)]

if(j==2){
next
}

if(j<length(other.pheno.names)){
temp1<-bal.fpkm[,(!is.na(temp.pheno)) & (temp.pheno>=pft.threshold[k,j] & temp.pheno<pft.threshold[k,j+1]) & (temp.pheno>=0)]
}else{
temp1<-bal.fpkm[,(!is.na(temp.pheno)) & (temp.pheno>=pft.threshold[k,j]) & (temp.pheno>=0)]	
}

if(length(temp1)<10*nrow(bal.fpkm)){
next
}

temp1.sd<-apply(temp1,1,sd)
temp2.sd<-apply(temp2,1,sd)

temp.data<-cbind(temp1,temp2)
temp.data<-temp.data[temp1.sd>0 | temp2.sd>0,]
bal.result<-t(apply(temp.data,1,my.grads.wilcox,n=ncol(temp1),alternative="two.sided",var.name="50-80"))
fdr.vect<-p.adjust(abs(bal.result[,"Wilcoxon P value"]),method="fdr")

q.vect<-qvalue(abs(bal.result[,"Wilcoxon P value"]))$qvalues
bal.result<-cbind(bal.result,fdr.vect,q.vect)
colnames(bal.result)[(ncol(bal.result)-1):ncol(bal.result)]<-c("FDR","q values")

rownames(anno.matrix)<-paste(anno.matrix[,1],"_",anno.matrix[,"tss_id"],sep="")


fpkm.info<-apply(temp.data>0,1,sum)
fpkm.info<-cbind(fpkm.info,apply(temp.data>0.1,1,sum))
fpkm.info<-cbind(fpkm.info,apply(temp.data>1,1,sum))
fpkm.info<-cbind(fpkm.info,apply(temp.data,1,median))
fpkm.info<-cbind(fpkm.info,apply(temp.data,1,mean))
colnames(fpkm.info)<-c("fpkm>0_num","fpkm>0.1_num","fpkm>1_num","fpkm_median","fpkm_mean")

output.filepath<-file.path(output.dir,paste(pft.names[k],"_bal_DEGs_",other.pheno.names[j],"_vs_50-80_all.txt",sep=""))
cmd.out<-cbind(anno.matrix[rownames(bal.result),],bal.result,fpkm.info)
write.table(cmd.out,file=output.filepath,append=F,sep="\t",row.names=F,col.names=T,quote=F)

#cmd.out<-cmd.out[!is.nan(as.numeric(cmd.out[,"Fold Change (other vs. SCADDING_0)"])) & abs(as.numeric(cmd.out[,"Fold Change (other vs. SCADDING_0)"]))>2 & abs(as.numeric(cmd.out[,"Wilcoxon P value"]))<0.05,]
cmd.out<-cmd.out[!is.nan(as.numeric(cmd.out[,grep("Fold Change",colnames(cmd.out))])) & abs(as.numeric(cmd.out[,grep("Fold Change",colnames(cmd.out))]))>2 & abs(as.numeric(cmd.out[,grep("Wilcoxon",colnames(cmd.out))]))<0.05,]
output.filepath<-file.path(output.dir,paste(pft.names[k],"_bal_DEGs_",other.pheno.names[j],"_vs_50-80_pvalue0.05_FC2.0.txt",sep=""))
write.table(cmd.out,file=output.filepath,append=F,sep="\t",row.names=F,col.names=T,quote=F)

temp<-paste(cmd.out[,1],"_",cmd.out[,"tss_id"],sep="")
output.filepath<-file.path(output.dir,paste(pft.names[k],"_bal_DEGs_",other.pheno.names[j],"_vs_50-80_pvalue0.05_FC2.0_genelist.txt",sep=""))
cat(temp,file=output.filepath,append=F,sep="\n")

cmd.out<-unique(cmd.out[,1])
cmd.out<-cbind(cmd.out,rep(1,length(cmd.out)))
output.filepath<-file.path(output.dir,paste(pft.names[k],"_bal_DEGs_",other.pheno.names[j],"_vs_50-80_pvalue0.05_FC2.0_genelist_metacore.txt",sep=""))
write.table(cmd.out,file=output.filepath,append=F,sep="\t",row.names=F,col.names=F,quote=F)
}


## for PBMC
#temp.pheno<-pbmc.clinic.matrix[,pft.names[k]]
#temp.pheno[temp.pheno<0]<-NA
#for(j in 1:length(other.pheno.names)){
#temp2<-pbmc.fpkm[,(!is.na(temp.pheno)) & (temp.pheno>=pft.threshold[k,1] & temp.pheno<pft.threshold[k,2]) & (temp.pheno>=0)]

#if(j==2){
#next
#}

#if(j<length(other.pheno.names)){
#temp1<-pbmc.fpkm[,(!is.na(temp.pheno)) & (temp.pheno>=pft.threshold[k,j] & temp.pheno<pft.threshold[k,j+1]) & (temp.pheno>=0)]
#}else{
#temp1<-pbmc.fpkm[,(!is.na(temp.pheno)) & (temp.pheno>=pft.threshold[k,j]) & (temp.pheno>=0)]	
#}

#if(length(temp1)<10*nrow(pbmc.fpkm)){
#next
#}

#temp1.sd<-apply(temp1,1,sd)
#temp2.sd<-apply(temp2,1,sd)

#temp.data<-cbind(temp1,temp2)
#temp.data<-temp.data[temp1.sd>0 | temp2.sd>0,]
#pbmc.result<-t(apply(temp.data,1,my.grads.wilcox,n=ncol(temp1),alternative="two.sided",var.name="50-80"))
#fdr.vect<-p.adjust(abs(pbmc.result[,"Wilcoxon P value"]),method="fdr")

#q.vect<-qvalue(abs(pbmc.result[,"Wilcoxon P value"]))$qvalues
#pbmc.result<-cbind(pbmc.result,fdr.vect,q.vect)
#colnames(pbmc.result)[(ncol(pbmc.result)-1):ncol(pbmc.result)]<-c("FDR","q values")

#rownames(anno.matrix)<-paste(anno.matrix[,1],"_",anno.matrix[,"tss_id"],sep="")

#fpkm.info<-apply(temp.data>0,1,sum)
#fpkm.info<-cbind(fpkm.info,apply(temp.data>0.1,1,sum))
#fpkm.info<-cbind(fpkm.info,apply(temp.data>1,1,sum))
#fpkm.info<-cbind(fpkm.info,apply(temp.data,1,median))
#fpkm.info<-cbind(fpkm.info,apply(temp.data,1,mean))
#colnames(fpkm.info)<-c("fpkm>0_num","fpkm>0.1_num","fpkm>1_num","fpkm_median","fpkm_mean")

#output.filepath<-file.path(output.dir,paste(pft.names[k],"_pbmc_DEGs_",other.pheno.names[j],"_vs_50-80_all.txt",sep=""))
#cmd.out<-cbind(anno.matrix[rownames(pbmc.result),],pbmc.result,fpkm.info)
#write.table(cmd.out,file=output.filepath,append=F,sep="\t",row.names=F,col.names=T,quote=F)

#cmd.out<-cmd.out[!is.nan(as.numeric(cmd.out[,"Fold Change (other vs. SCADDING_0)"])) & abs(as.numeric(cmd.out[,"Fold Change (other vs. SCADDING_0)"]))>2 & abs(as.numeric(cmd.out[,"Wilcoxon P value"]))<0.05,]
#cmd.out<-cmd.out[!is.nan(as.numeric(cmd.out[,grep("Fold Change",colnames(cmd.out))])) & abs(as.numeric(cmd.out[,grep("Fold Change",colnames(cmd.out))]))>2 & abs(as.numeric(cmd.out[,grep("Wilcoxon",colnames(cmd.out))]))<0.05,]
#output.filepath<-file.path(output.dir,paste(pft.names[k],"_pbmc_DEGs_",other.pheno.names[j],"_vs_50-80_pvalue0.05_FC2.0.txt",sep=""))
#write.table(cmd.out,file=output.filepath,append=F,sep="\t",row.names=F,col.names=T,quote=F)

#temp<-paste(cmd.out[,1],"_",cmd.out[,"tss_id"],sep="")
#output.filepath<-file.path(output.dir,paste(pft.names[k],"_pbmc_DEGs_",other.pheno.names[j],"_vs_50-80_pvalue0.05_FC2.0_genelist.txt",sep=""))
#cat(temp,file=output.filepath,append=F,sep="\n")

#cmd.out<-unique(cmd.out[,1])
#cmd.out<-cbind(cmd.out,rep(1,length(cmd.out)))
#output.filepath<-file.path(output.dir,paste(pft.names[k],"_pbmc_DEGs_",other.pheno.names[j],"_vs_50-80_pvalue0.05_FC2.0_genelist_metacore.txt",sep=""))
#write.table(cmd.out,file=output.filepath,append=F,sep="\t",row.names=F,col.names=F,quote=F)
}
}






#module load Apps/R/3.3.2-generic

################################
# 4.11. Identify genes associated with the cell differentials
# matched data. columns of data.matrix are matched with rows in sample.matrix and my.clinic.matrix
# data.matrix	  	---- fpkm.matrix
# sample.matrix   	---- matched.list
# my.clinic.matrix	---- clinic.matrix
#source("/home/fas/kaminski/xy48/Rprogram/GRADS/0_CleanDataLoad.R")

source("/home/antun/Desktop/SUPERvisedA/r_codes/209_local_data_load_AM.R")


cell.filepath<-"/home/antun/Desktop/SUPERvisedA/BAL_Cell_Differentials.xlsx"
library(gdata)
cell.matrix<-read.xls(cell.filepath,sheet=1)
rownames(cell.matrix)<-as.matrix(cell.matrix)[,1]
cell.matrix<-cell.matrix[as.matrix(clinic.matrix)[,1],]

# take an average between the two slides for all the cell differentials
feature.names<-colnames(cell.matrix)[8:ncol(cell.matrix)]
temp.names<-sapply(feature.names,my.element.extract,splitchar="D",index=1)
temp.names<-paste(temp.names,"D",sep="")
feature.list<-split(feature.names,temp.names)
my.cell.matrix.new<-cell.matrix[,1:7]

for(i in 1:length(feature.list)){
temp.matrix<-cell.matrix[,feature.list[[i]]]
temp.matrix[temp.matrix<0]<-NA
temp.matrix<-apply(temp.matrix,1,mean,na.rm=T)
temp.matrix[is.nan(temp.matrix)]<-NA
my.cell.matrix.new<-cbind(my.cell.matrix.new,temp.matrix)
colnames(my.cell.matrix.new)[ncol(my.cell.matrix.new)]<-names(feature.list)[i]
}

cell.matrix<-my.cell.matrix.new

################
# define the function that does the correlation test

my.con.cor<-function(fpkm.matrix.clean,feature.name="AGE",method.name="spearman",tissue.name="BAL",sample.list.all,clinic.matrix,output.dir){

# extract the fpkm matrix for the given tissue
data.matrix<-fpkm.matrix.clean
sample.names<-colnames(data.matrix)
names(sample.names)<-unname(sapply(sample.names,my.element.extract,splitchar="_",index=1))
colnames(data.matrix)<-names(sample.names)

if(tissue.name=="BAL"){
bal.fpkm<-data.matrix[,substr(colnames(data.matrix),3,3)=="B"]
}else{
bal.fpkm<-data.matrix[,substr(colnames(data.matrix),3,3)=="S"]
}

# generate the clinical matrix
if(tissue.name=="BAL"){
temp.list<-sample.list.all
rownames(temp.list)<-sample.list.all[,3]
temp.list<-temp.list[colnames(bal.fpkm),]
my.clinic.matrix<-clinic.matrix[temp.list[,1],]
}else{
temp.list<-sample.list.all
rownames(temp.list)<-sample.list.all[,10]
temp.list<-temp.list[colnames(bal.fpkm),]
my.clinic.matrix<-clinic.matrix[temp.list[,1],]	
}

#bal.fpkm<-data.matrix[,substr(colnames(data.matrix),3,3)=="B"]
#pbmc.fpkm<-data.matrix[,substr(colnames(data.matrix),3,3)=="S"]

clinic.vect<-as.numeric(my.clinic.matrix[,colnames(my.clinic.matrix)==feature.name])

# generate the correlation results
age.pvalue<-apply(bal.fpkm,1,my.cor.test,y=clinic.vect,method.name=method.name)
age.pvalue<-matrix(unlist(age.pvalue),ncol=2,byrow=T)
colnames(age.pvalue)<-c("pvalue","corr_coef")
age.fdr<-p.adjust(age.pvalue[,"pvalue"],method="fdr")
age.pvalue<-cbind(age.pvalue,age.fdr)
colnames(age.pvalue)[ncol(age.pvalue)]<-"fdr"
rownames(age.pvalue)<-rownames(bal.fpkm)

fpkm.info<-apply(bal.fpkm>0,1,sum)
fpkm.info<-cbind(fpkm.info,apply(bal.fpkm>0.1,1,sum))
fpkm.info<-cbind(fpkm.info,apply(bal.fpkm>1,1,sum))
fpkm.info<-cbind(fpkm.info,apply(bal.fpkm,1,median))
fpkm.info<-cbind(fpkm.info,apply(bal.fpkm,1,mean))
colnames(fpkm.info)<-c("fpkm>0_num","fpkm>0.1_num","fpkm>1_num","fpkm_median","fpkm_mean")

age.pvalue<-cbind(age.pvalue,fpkm.info)

if(file.exists(output.dir)==F){
cat("WARNING: output.dir doesn't exists!\n")
return(1)
}

output.filepath<-file.path(output.dir,paste("correlation_",method.name,"_",feature.name,"_allgenes_",tissue.name,".txt",sep=""))
cat("gene_names\t",file=output.filepath,append=F)
write.table(age.pvalue,file=output.filepath,append=T,sep="\t",row.names=T,col.names=T,quote=F)

output.filepath<-file.path(output.dir,paste("correlation_",method.name,"_",feature.name,"_fdr0.05_",tissue.name,".txt",sep=""))
cat("gene_names\t",file=output.filepath,append=F)
write.table(age.pvalue[!is.na(age.pvalue[,3]) & age.pvalue[,3]<0.05,],file=output.filepath,append=T,sep="\t",row.names=T,col.names=T,quote=F)

output.filepath<-file.path(output.dir,paste("correlation_",method.name,"_",feature.name,"_fdr0.05_",tissue.name,"_both_metacore.txt",sep=""))
cat("gene_names\tintensity\n",file=output.filepath,append=F)
sig.genes<-rownames(data.matrix)[!is.na(age.pvalue[,3]) & age.pvalue[,3]<0.05]
sig.genes<-unname(sapply(sig.genes,my.element.extract,splitchar="_",index=1))
sig.genes<-unique(sig.genes)
cmd.out<-cbind(sig.genes,rep(1,length(sig.genes)))
write.table(cmd.out,file=output.filepath,append=T,sep="\t",row.names=F,col.names=F,quote=F)


output.filepath<-file.path(output.dir,paste("correlation_",method.name,"_",feature.name,"_fdr0.05_",tissue.name,"_positive_metacore.txt",sep=""))
cat("gene_names\tintensity\n",file=output.filepath,append=F)
sig.genes<-rownames(data.matrix)[!is.na(age.pvalue[,3]) & age.pvalue[,3]<0.05 & age.pvalue[,2]>0]
sig.genes<-unname(sapply(sig.genes,my.element.extract,splitchar="_",index=1))
sig.genes<-unique(sig.genes)
cmd.out<-cbind(sig.genes,rep(1,length(sig.genes)))
write.table(cmd.out,file=output.filepath,append=T,sep="\t",row.names=F,col.names=F,quote=F)


output.filepath<-file.path(output.dir,paste("correlation_",method.name,"_",feature.name,"_fdr0.05_",tissue.name,"_negative_metacore.txt",sep=""))
cat("gene_names\tintensity\n",file=output.filepath,append=F)
sig.genes<-rownames(data.matrix)[!is.na(age.pvalue[,3]) & age.pvalue[,3]<0.05 & age.pvalue[,2]<0]
sig.genes<-unname(sapply(sig.genes,my.element.extract,splitchar="_",index=1))
sig.genes<-unique(sig.genes)
cmd.out<-cbind(sig.genes,rep(1,length(sig.genes)))
write.table(cmd.out,file=output.filepath,append=T,sep="\t",row.names=F,col.names=F,quote=F)

return(0)
}


# for different cell differentials
#output.dir<-"~/driver_Naftali/GRADS/SARC_results/Results_summary/PFTs_correlated_genes"
output.dir<-"/home/antun/Desktop/SUPERvisedA/Results_summary/CellDifferentials_correlated_genes"
my.con.cor(fpkm.matrix.clean,feature.name="BALCELL",method.name="spearman",tissue.name="BAL",sample.list.all,cell.matrix,output.dir)
#my.con.cor(fpkm.matrix.clean,feature.name="BALCELL",method.name="spearman",tissue.name="PBMC",sample.list.all,cell.matrix,output.dir)
my.con.cor(fpkm.matrix.clean,feature.name="BALCELSP",method.name="spearman",tissue.name="BAL",sample.list.all,cell.matrix,output.dir)
#my.con.cor(fpkm.matrix.clean,feature.name="BALCELSP",method.name="spearman",tissue.name="PBMC",sample.list.all,cell.matrix,output.dir)
my.con.cor(fpkm.matrix.clean,feature.name="AIRBALD",method.name="spearman",tissue.name="BAL",sample.list.all,cell.matrix,output.dir)
#my.con.cor(fpkm.matrix.clean,feature.name="AIRBALD",method.name="spearman",tissue.name="PBMC",sample.list.all,cell.matrix,output.dir)
my.con.cor(fpkm.matrix.clean,feature.name="AIRBWD",method.name="spearman",tissue.name="BAL",sample.list.all,cell.matrix,output.dir)
#my.con.cor(fpkm.matrix.clean,feature.name="AIRBWD",method.name="spearman",tissue.name="PBMC",sample.list.all,cell.matrix,output.dir)
my.con.cor(fpkm.matrix.clean,feature.name="ALVBALD",method.name="spearman",tissue.name="BAL",sample.list.all,cell.matrix,output.dir)
#my.con.cor(fpkm.matrix.clean,feature.name="ALVBALD",method.name="spearman",tissue.name="PBMC",sample.list.all,cell.matrix,output.dir)
my.con.cor(fpkm.matrix.clean,feature.name="ALVBWD",method.name="spearman",tissue.name="BAL",sample.list.all,cell.matrix,output.dir)
#my.con.cor(fpkm.matrix.clean,feature.name="ALVBWD",method.name="spearman",tissue.name="PBMC",sample.list.all,cell.matrix,output.dir)
my.con.cor(fpkm.matrix.clean,feature.name="EOSBALD",method.name="spearman",tissue.name="BAL",sample.list.all,cell.matrix,output.dir)
#my.con.cor(fpkm.matrix.clean,feature.name="EOSBALD",method.name="spearman",tissue.name="PBMC",sample.list.all,cell.matrix,output.dir)
my.con.cor(fpkm.matrix.clean,feature.name="EOSBWD",method.name="spearman",tissue.name="BAL",sample.list.all,cell.matrix,output.dir)
#my.con.cor(fpkm.matrix.clean,feature.name="EOSBWD",method.name="spearman",tissue.name="PBMC",sample.list.all,cell.matrix,output.dir)
my.con.cor(fpkm.matrix.clean,feature.name="LYMBALD",method.name="spearman",tissue.name="BAL",sample.list.all,cell.matrix,output.dir)
#my.con.cor(fpkm.matrix.clean,feature.name="LYMBALD",method.name="spearman",tissue.name="PBMC",sample.list.all,cell.matrix,output.dir)
my.con.cor(fpkm.matrix.clean,feature.name="LYMBWD",method.name="spearman",tissue.name="BAL",sample.list.all,cell.matrix,output.dir)
#my.con.cor(fpkm.matrix.clean,feature.name="LYMBWD",method.name="spearman",tissue.name="PBMC",sample.list.all,cell.matrix,output.dir)
my.con.cor(fpkm.matrix.clean,feature.name="NEUBALD",method.name="spearman",tissue.name="BAL",sample.list.all,cell.matrix,output.dir)
#my.con.cor(fpkm.matrix.clean,feature.name="NEUBALD",method.name="spearman",tissue.name="PBMC",sample.list.all,cell.matrix,output.dir)
my.con.cor(fpkm.matrix.clean,feature.name="NEUBWD",method.name="spearman",tissue.name="BAL",sample.list.all,cell.matrix,output.dir)
#my.con.cor(fpkm.matrix.clean,feature.name="NEUBWD",method.name="spearman",tissue.name="PBMC",sample.list.all,cell.matrix,output.dir)







################################
# 4.12. Identify genes associated with the CT scan variables
# matched data. columns of data.matrix are matched with rows in sample.matrix and my.clinic.matrix
# data.matrix	  	---- fpkm.matrix
# sample.matrix   	---- matched.list
# my.clinic.matrix	---- clinic.matrix
#source("/home/fas/kaminski/xy48/Rprogram/GRADS/0_CleanDataLoad.R")

source("/home/antun/Desktop/SUPERvisedA/r_codes/209_local_data_load_AM.R")

ct.filepath<-"/home/antun/Desktop/SUPERvisedA/Final_sarc_ct_reads.xlsx"
library(gdata)
ct.matrix<-read.xls(ct.filepath,sheet=1)
ct.matrix<-ct.matrix[-149,]
rownames(ct.matrix)<-as.matrix(ct.matrix)[,1]
ct.matrix<-ct.matrix[as.matrix(clinic.matrix)[,1],]
rownames(ct.matrix)<-as.matrix(clinic.matrix)[,1]

# take an average between the two slides for all the cell differentials
feature.names<-c(
"Med_Lymphadenopathy",
"Hilar_Lymphadenopathy",
"Micronodule",
"Bronchial_Wall_Thickening",
"Traction_Bronchiectasis",
"Bronchiectasis_severity",
"Ground_Glass",
"Honeycombing",
"Reticular_Abnormality",
"Pulmonary_Art",
"Tree_in_bud"
)

ct.matrix<-ct.matrix[,feature.names]


################
# change the coding of the CT variables

# Med_Lymphadenopathy: 0=no, 1=left, 2=right, 3=bilateral
# v1: yes vs no
# v2: bilateral vs one side
# v3: numeric 0=no, 1=one side, 2=bilateral
temp.vect<-ct.matrix[,"Med_Lymphadenopathy"]  # yes versus no and billateral versus one side
temp1<-rep(NA,nrow(ct.matrix))
temp2<-temp1
temp3<-temp1

temp1[temp.vect==0]<-0
temp1[temp.vect>0]<-1

temp2[temp.vect==1 | temp.vect==2]<-0
temp2[temp.vect==3]<-1

temp3[temp.vect==0]<-0
temp3[temp.vect==1 | temp.vect==2]<-1
temp3[temp.vect==3]<-2

ct.matrix<-cbind(ct.matrix,temp1,temp2,temp3)
colnames(ct.matrix)[(ncol(ct.matrix)-2):ncol(ct.matrix)]<-c("Med_Lymphadenopathy_yes_no","Med_Lymphadenopathy_oneside_bi","Med_Lymphadenopathy_numeric")


# Hilar_Lymphadenopathy: 0=no, 1=left, 2=right, 3=bilateral
# v1: yes vs no
# v2: bilateral vs one side
# v3: numeric 0=no, 1=one side, 2=bilateral
temp.vect<-ct.matrix[,"Hilar_Lymphadenopathy"]  # yes versus no and billateral versus one side
temp1<-rep(NA,nrow(ct.matrix))
temp2<-temp1
temp3<-temp1

temp1[temp.vect==0]<-0
temp1[temp.vect>0]<-1

temp2[temp.vect==1 | temp.vect==2]<-0
temp2[temp.vect==3]<-1

temp3[temp.vect==0]<-0
temp3[temp.vect==1 | temp.vect==2]<-1
temp3[temp.vect==3]<-2

ct.matrix<-cbind(ct.matrix,temp1,temp2,temp3)
colnames(ct.matrix)[(ncol(ct.matrix)-2):ncol(ct.matrix)]<-c("Hilar_Lymphadenopathy_yes_no","Hilar_Lymphadenopathy_oneside_bi","Hilar_Lymphadenopathy_numeric")



################
# define the function that does the correlation test

my.con.cor<-function(fpkm.matrix.clean,feature.name="AGE",method.name="spearman",tissue.name="BAL",sample.list.all,clinic.matrix,output.dir){

# extract the fpkm matrix for the given tissue
data.matrix<-fpkm.matrix.clean
sample.names<-colnames(data.matrix)
names(sample.names)<-unname(sapply(sample.names,my.element.extract,splitchar="_",index=1))
colnames(data.matrix)<-names(sample.names)

if(tissue.name=="BAL"){
bal.fpkm<-data.matrix[,substr(colnames(data.matrix),3,3)=="B"]
}else{
bal.fpkm<-data.matrix[,substr(colnames(data.matrix),3,3)=="S"]
}

# generate the clinical matrix
if(tissue.name=="BAL"){
temp.list<-sample.list.all
rownames(temp.list)<-sample.list.all[,3]
temp.list<-temp.list[colnames(bal.fpkm),]
my.clinic.matrix<-clinic.matrix[temp.list[,1],]
}else{
temp.list<-sample.list.all
rownames(temp.list)<-sample.list.all[,10]
temp.list<-temp.list[colnames(bal.fpkm),]
my.clinic.matrix<-clinic.matrix[temp.list[,1],]	
}

#bal.fpkm<-data.matrix[,substr(colnames(data.matrix),3,3)=="B"]
#pbmc.fpkm<-data.matrix[,substr(colnames(data.matrix),3,3)=="S"]

clinic.vect<-as.numeric(my.clinic.matrix[,colnames(my.clinic.matrix)==feature.name])

# generate the correlation results
age.pvalue<-apply(bal.fpkm,1,my.cor.test,y=clinic.vect,method.name=method.name)
age.pvalue<-matrix(unlist(age.pvalue),ncol=2,byrow=T)
colnames(age.pvalue)<-c("pvalue","corr_coef")
age.fdr<-p.adjust(age.pvalue[,"pvalue"],method="fdr")
age.pvalue<-cbind(age.pvalue,age.fdr)
colnames(age.pvalue)[ncol(age.pvalue)]<-"fdr"
rownames(age.pvalue)<-rownames(bal.fpkm)

fpkm.info<-apply(bal.fpkm>0,1,sum)
fpkm.info<-cbind(fpkm.info,apply(bal.fpkm>0.1,1,sum))
fpkm.info<-cbind(fpkm.info,apply(bal.fpkm>1,1,sum))
fpkm.info<-cbind(fpkm.info,apply(bal.fpkm,1,median))
fpkm.info<-cbind(fpkm.info,apply(bal.fpkm,1,mean))
colnames(fpkm.info)<-c("fpkm>0_num","fpkm>0.1_num","fpkm>1_num","fpkm_median","fpkm_mean")

age.pvalue<-cbind(age.pvalue,fpkm.info)

if(file.exists(output.dir)==F){
cat("WARNING: output.dir doesn't exists!\n")
return(1)
}

output.filepath<-file.path(output.dir,paste("correlation_",method.name,"_",feature.name,"_allgenes_",tissue.name,".txt",sep=""))
cat("gene_names\t",file=output.filepath,append=F)
write.table(age.pvalue,file=output.filepath,append=T,sep="\t",row.names=T,col.names=T,quote=F)

output.filepath<-file.path(output.dir,paste("correlation_",method.name,"_",feature.name,"_fdr0.05_",tissue.name,".txt",sep=""))
cat("gene_names\t",file=output.filepath,append=F)
write.table(age.pvalue[!is.na(age.pvalue[,3]) & age.pvalue[,3]<0.05,],file=output.filepath,append=T,sep="\t",row.names=T,col.names=T,quote=F)

output.filepath<-file.path(output.dir,paste("correlation_",method.name,"_",feature.name,"_fdr0.05_",tissue.name,"_both_metacore.txt",sep=""))
cat("gene_names\tintensity\n",file=output.filepath,append=F)
sig.genes<-rownames(data.matrix)[!is.na(age.pvalue[,3]) & age.pvalue[,3]<0.05]
sig.genes<-unname(sapply(sig.genes,my.element.extract,splitchar="_",index=1))
sig.genes<-unique(sig.genes)
cmd.out<-cbind(sig.genes,rep(1,length(sig.genes)))
write.table(cmd.out,file=output.filepath,append=T,sep="\t",row.names=F,col.names=F,quote=F)


output.filepath<-file.path(output.dir,paste("correlation_",method.name,"_",feature.name,"_fdr0.05_",tissue.name,"_positive_metacore.txt",sep=""))
cat("gene_names\tintensity\n",file=output.filepath,append=F)
sig.genes<-rownames(data.matrix)[!is.na(age.pvalue[,3]) & age.pvalue[,3]<0.05 & age.pvalue[,2]>0]
sig.genes<-unname(sapply(sig.genes,my.element.extract,splitchar="_",index=1))
sig.genes<-unique(sig.genes)
cmd.out<-cbind(sig.genes,rep(1,length(sig.genes)))
write.table(cmd.out,file=output.filepath,append=T,sep="\t",row.names=F,col.names=F,quote=F)


output.filepath<-file.path(output.dir,paste("correlation_",method.name,"_",feature.name,"_fdr0.05_",tissue.name,"_negative_metacore.txt",sep=""))
cat("gene_names\tintensity\n",file=output.filepath,append=F)
sig.genes<-rownames(data.matrix)[!is.na(age.pvalue[,3]) & age.pvalue[,3]<0.05 & age.pvalue[,2]<0]
sig.genes<-unname(sapply(sig.genes,my.element.extract,splitchar="_",index=1))
sig.genes<-unique(sig.genes)
cmd.out<-cbind(sig.genes,rep(1,length(sig.genes)))
write.table(cmd.out,file=output.filepath,append=T,sep="\t",row.names=F,col.names=F,quote=F)

return(0)
}


# for different cell differentials
#output.dir<-"~/driver_Naftali/GRADS/SARC_results/Results_summary/PFTs_correlated_genes"
output.dir<-"/home/antun/Desktop/SUPERvisedA/Results_summary/CTVariables_correlated_genes"

if(file.exists(output.dir)==F){
dir.create(output.dir)	
}

my.con.cor(fpkm.matrix.clean,feature.name="Med_Lymphadenopathy",method.name="spearman",tissue.name="BAL",sample.list.all,ct.matrix,output.dir)
#my.con.cor(fpkm.matrix.clean,feature.name="Med_Lymphadenopathy",method.name="spearman",tissue.name="PBMC",sample.list.all,ct.matrix,output.dir)
my.con.cor(fpkm.matrix.clean,feature.name="Hilar_Lymphadenopathy",method.name="spearman",tissue.name="BAL",sample.list.all,ct.matrix,output.dir)
#my.con.cor(fpkm.matrix.clean,feature.name="Hilar_Lymphadenopathy",method.name="spearman",tissue.name="PBMC",sample.list.all,ct.matrix,output.dir)
my.con.cor(fpkm.matrix.clean,feature.name="Bronchial_Wall_Thickening",method.name="spearman",tissue.name="BAL",sample.list.all,ct.matrix,output.dir)
#my.con.cor(fpkm.matrix.clean,feature.name="Bronchial_Wall_Thickening",method.name="spearman",tissue.name="PBMC",sample.list.all,ct.matrix,output.dir)
my.con.cor(fpkm.matrix.clean,feature.name="Traction_Bronchiectasis",method.name="spearman",tissue.name="BAL",sample.list.all,ct.matrix,output.dir)
#my.con.cor(fpkm.matrix.clean,feature.name="Traction_Bronchiectasis",method.name="spearman",tissue.name="PBMC",sample.list.all,ct.matrix,output.dir)
my.con.cor(fpkm.matrix.clean,feature.name="Bronchiectasis_severity",method.name="spearman",tissue.name="BAL",sample.list.all,ct.matrix,output.dir)
#my.con.cor(fpkm.matrix.clean,feature.name="Bronchiectasis_severity",method.name="spearman",tissue.name="PBMC",sample.list.all,ct.matrix,output.dir)
my.con.cor(fpkm.matrix.clean,feature.name="Ground_Glass",method.name="spearman",tissue.name="BAL",sample.list.all,ct.matrix,output.dir)
#my.con.cor(fpkm.matrix.clean,feature.name="Ground_Glass",method.name="spearman",tissue.name="PBMC",sample.list.all,ct.matrix,output.dir)
my.con.cor(fpkm.matrix.clean,feature.name="Honeycombing",method.name="spearman",tissue.name="BAL",sample.list.all,ct.matrix,output.dir)
#my.con.cor(fpkm.matrix.clean,feature.name="Honeycombing",method.name="spearman",tissue.name="PBMC",sample.list.all,ct.matrix,output.dir)
my.con.cor(fpkm.matrix.clean,feature.name="Reticular_Abnormality",method.name="spearman",tissue.name="BAL",sample.list.all,ct.matrix,output.dir)
#my.con.cor(fpkm.matrix.clean,feature.name="Reticular_Abnormality",method.name="spearman",tissue.name="PBMC",sample.list.all,ct.matrix,output.dir)
my.con.cor(fpkm.matrix.clean,feature.name="Med_Lymphadenopathy_numeric",method.name="spearman",tissue.name="BAL",sample.list.all,ct.matrix,output.dir)
#my.con.cor(fpkm.matrix.clean,feature.name="Med_Lymphadenopathy_numeric",method.name="spearman",tissue.name="PBMC",sample.list.all,ct.matrix,output.dir)
my.con.cor(fpkm.matrix.clean,feature.name="Hilar_Lymphadenopathy_numeric",method.name="spearman",tissue.name="BAL",sample.list.all,ct.matrix,output.dir)
#my.con.cor(fpkm.matrix.clean,feature.name="Hilar_Lymphadenopathy_numeric",method.name="spearman",tissue.name="PBMC",sample.list.all,ct.matrix,output.dir)






# for the binary CT variables

my.cat.cor<-function(fpkm.matrix.clean,feature.name="GENDER",method.name="wilcox",tissue.name="BAL",samplelist,clinicmatrix,outputdir){

# extract the fpkm matrix for the given tissue
data.matrix<-fpkm.matrix.clean
sample.names<-colnames(data.matrix)
names(sample.names)<-unname(sapply(sample.names,my.element.extract,splitchar="_",index=1))
colnames(data.matrix)<-names(sample.names)
if(tissue.name=="BAL"){
bal.fpkm<-data.matrix[,substr(colnames(data.matrix),3,3)=="B"]
}else{
bal.fpkm<-data.matrix[,substr(colnames(data.matrix),3,3)=="S"]
}
# generate the clinical matrix
if(tissue.name=="BAL"){
#temp.list<-samplelist
#rownames(temp.list)<-samplelist[,3]
#temp.list<-temp.list[colnames(bal.fpkm),]
temp.list<-samplelist[colnames(bal.fpkm),]
temp.match<-match(temp.list[,1],as.matrix(clinicmatrix)[,1])
#my.clinic.matrix<-clinic.matrix[temp.list[,1],]
my.clinic.matrix<-clinicmatrix[temp.match,]
}else{
#temp.list<-samplelist
#rownames(temp.list)<-samplelist[,10]
#temp.list<-temp.list[colnames(bal.fpkm),]
temp.list<-samplelist[colnames(bal.fpkm),]
temp.match<-match(temp.list[,1],as.matrix(clinicmatrix)[,1])
#my.clinic.matrix<-clinic.matrix[temp.list[,1],]
my.clinic.matrix<-clinicmatrix[temp.match,]
}

#line 1 ends
#bal.fpkm<-data.matrix[,substr(colnames(data.matrix),3,3)=="B"]
#pbmc.fpkm<-data.matrix[,substr(colnames(data.matrix),3,3)=="S"]

clinic.vect<-as.numeric(my.clinic.matrix[,colnames(my.clinic.matrix)==feature.name])
if(length(table(clinic.vect))!=2){
cat("ERROR: the phenotype does not have two levels!\n")
return(1)
}


# generate the correlation results
x1<-bal.fpkm[,!is.na(clinic.vect) & clinic.vect==as.numeric(names(table(clinic.vect))[1])]
x2<-bal.fpkm[,!is.na(clinic.vect) & clinic.vect==as.numeric(names(table(clinic.vect))[2])]

cat(sum(!is.na(clinic.vect) & clinic.vect==as.numeric(names(table(clinic.vect))[1])),"samples with phenotype=",as.numeric(names(table(clinic.vect))[1]),"\n")
cat(sum(!is.na(clinic.vect) & clinic.vect==as.numeric(names(table(clinic.vect))[2])),"samples with phenotype=",as.numeric(names(table(clinic.vect))[2]),"\n")

if(sum(!is.na(clinic.vect) & clinic.vect==as.numeric(names(table(clinic.vect))[1]))<=3){
cat("ERROR: there are less than 3 samples with phenotype=",as.numeric(names(table(clinic.vect))[1]),"\n")
return(1)
}

if(sum(!is.na(clinic.vect) & clinic.vect==as.numeric(names(table(clinic.vect))[2]))<=3){
cat("ERROR: there are less than 3 samples with phenotype=",as.numeric(names(table(clinic.vect))[2]),"\n")
return(1)
}

#age.pvalue<-apply(bal.fpkm,1,my.cor.test,y=clinic.vect,method.name=method.name)
age.pvalue<-abs(apply(cbind(x1,x2),1,my.wilcox,n=sum(!is.na(clinic.vect) & clinic.vect==as.numeric(names(table(clinic.vect))[1]))))

age.pvalue.direct<-apply(cbind(x1,x2),1,my.wilcox,n=ncol(x1))

#age.pvalue<-matrix(unlist(age.pvalue),ncol=2,byrow=T)
age.pvalue<-cbind(age.pvalue,rep(NA,length(age.pvalue)))
colnames(age.pvalue)<-c("pvalue","corr_coef")
age.fdr<-p.adjust(age.pvalue[,"pvalue"],method="fdr")
age.pvalue<-cbind(age.pvalue,age.fdr)
colnames(age.pvalue)[ncol(age.pvalue)]<-"fdr"
rownames(age.pvalue)<-rownames(bal.fpkm)

age.pvalue<-cbind(age.pvalue,age.pvalue.direct)
colnames(age.pvalue)[ncol(age.pvalue)]<-"pval_dir"


fpkm.info<-apply(cbind(x1,x2)>0,1,sum)
fpkm.info<-cbind(fpkm.info,apply(cbind(x1,x2)>0.1,1,sum))
fpkm.info<-cbind(fpkm.info,apply(cbind(x1,x2)>1,1,sum))
fpkm.info<-cbind(fpkm.info,apply(cbind(x1,x2),1,median))
fpkm.info<-cbind(fpkm.info,apply(cbind(x1,x2),1,mean))
colnames(fpkm.info)<-c("fpkm>0_num","fpkm>0.1_num","fpkm>1_num","fpkm_median","fpkm_mean")

age.pvalue<-cbind(age.pvalue,fpkm.info)

if(file.exists(outputdir)==F){
cat("WARNING: output.dir doesn't exists!\n")
return(1)
}

output.filepath<-file.path(outputdir,paste("correlation_",method.name,"_",feature.name,"_allgenes_",tissue.name,".txt",sep=""))
cat("gene_names\t",file=output.filepath,append=F)
write.table(age.pvalue,file=output.filepath,append=T,sep="\t",row.names=T,col.names=T,quote=F)

output.filepath<-file.path(outputdir,paste("correlation_",method.name,"_",feature.name,"_fdr0.05_",tissue.name,".txt",sep=""))
cat("gene_names\t",file=output.filepath,append=F)
write.table(age.pvalue[!is.na(age.pvalue[,3]) & age.pvalue[,3]<0.05,],file=output.filepath,append=T,sep="\t",row.names=T,col.names=T,quote=F)

output.filepath<-file.path(outputdir,paste("correlation_",method.name,"_",feature.name,"_fdr0.05_",tissue.name,"_both_metacore.txt",sep=""))
cat("gene_names\tintensity\n",file=output.filepath,append=F)
sig.genes<-rownames(data.matrix)[!is.na(age.pvalue[,3]) & age.pvalue[,3]<0.05]
sig.genes<-unname(sapply(sig.genes,my.element.extract,splitchar="_",index=1))
sig.genes<-unique(sig.genes)
cmd.out<-cbind(sig.genes,rep(1,length(sig.genes)))
write.table(cmd.out,file=output.filepath,append=T,sep="\t",row.names=F,col.names=F,quote=F)


output.filepath<-file.path(outputdir,paste("correlation_",method.name,"_",feature.name,"_fdr0.05_",tissue.name,"_positive_metacore.txt",sep=""))
cat("gene_names\tintensity\n",file=output.filepath,append=F)
sig.genes<-rownames(data.matrix)[!is.na(age.pvalue[,3]) & age.pvalue[,3]<0.05 & age.pvalue[,4]>0]
sig.genes<-unname(sapply(sig.genes,my.element.extract,splitchar="_",index=1))
sig.genes<-unique(sig.genes)
cmd.out<-cbind(sig.genes,rep(1,length(sig.genes)))
write.table(cmd.out,file=output.filepath,append=T,sep="\t",row.names=F,col.names=F,quote=F)


output.filepath<-file.path(outputdir,paste("correlation_",method.name,"_",feature.name,"_fdr0.05_",tissue.name,"_negative_metacore.txt",sep=""))
cat("gene_names\tintensity\n",file=output.filepath,append=F)
sig.genes<-rownames(data.matrix)[!is.na(age.pvalue[,3]) & age.pvalue[,3]<0.05 & age.pvalue[,4]<0]
sig.genes<-unname(sapply(sig.genes,my.element.extract,splitchar="_",index=1))
sig.genes<-unique(sig.genes)
cmd.out<-cbind(sig.genes,rep(1,length(sig.genes)))
write.table(cmd.out,file=output.filepath,append=T,sep="\t",row.names=F,col.names=F,quote=F)

return(0)
}


#output.dir<-"/home/fas/kaminski/am2952/scratch/SARC209clean/Results_summary/CTVariables_correlated_genes"

output.dir<-"/home/antun/Desktop/SUPERvisedA/Results_summary/CTVariables_correlated_genes/Directional_p"


rownames(sample.list.all)<-sample.list.all[,3]
rownames(clinic.matrix)<-as.matrix(clinic.matrix)[,1]
ct.matrix.new<-ct.matrix
ct.matrix.new<-cbind(rownames(ct.matrix),ct.matrix.new)
rownames(ct.matrix.new)<-as.matrix(ct.matrix.new)[,1]
my.cat.cor(fpkm.matrix.clean,feature.name="Hilar_Lymphadenopathy_yes_no",tissue.name="BAL",samplelist=sample.list.all,clinicmatrix=ct.matrix.new,outputdir=output.dir)
#108 samples with phenotype= 0 
#95 samples with phenotype= 1 

my.cat.cor(fpkm.matrix.clean,feature.name="Hilar_Lymphadenopathy_oneside_bi",tissue.name="BAL",samplelist=sample.list.all,clinicmatrix=ct.matrix.new,outputdir=output.dir) # comparing 1 zero to 96 ones
#1 samples with phenotype= 0 
#94 samples with phenotype= 1 
#ERROR: there are less than 3 samples with phenotype= 0 

my.cat.cor(fpkm.matrix.clean,feature.name="Med_Lymphadenopathy_yes_no",tissue.name="BAL",samplelist=sample.list.all,clinicmatrix=ct.matrix.new,outputdir=output.dir) # 77 vs 118
#83 samples with phenotype= 0 
#120 samples with phenotype= 1 

my.cat.cor(fpkm.matrix.clean,feature.name="Med_Lymphadenopathy_oneside_bi",tissue.name="BAL",samplelist=sample.list.all,clinicmatrix=ct.matrix.new,outputdir=output.dir) # comparing 12 vs 106 samples
#12 samples with phenotype= 0 
#108 samples with phenotype= 1 

my.cat.cor(fpkm.matrix.clean,feature.name="Pulmonary_Art",tissue.name="BAL",samplelist=sample.list.all,clinicmatrix=ct.matrix.new,outputdir=output.dir) # 2 vs 193
#200 samples with phenotype= 0 
#3 samples with phenotype= 1 

my.cat.cor(fpkm.matrix.clean,feature.name="Tree_in_bud",tissue.name="BAL",samplelist=sample.list.all,clinicmatrix=ct.matrix.new,outputdir=output.dir) # 0 samples with value of 1
#ERROR: the phenotype does not have two levels!
#[1] 1

my.cat.cor(fpkm.matrix.clean,feature.name="Micronodule",tissue.name="BAL",samplelist=sample.list.all,clinicmatrix=ct.matrix.new,outputdir=output.dir) # 98 vs 97
#102 samples with phenotype= 0 
#101 samples with phenotype= 1 




#rownames(sample.list.all)<-sample.list.all[,10]
#rownames(clinic.matrix)<-as.matrix(clinic.matrix)[,1]
#ct.matrix.new<-ct.matrix
#ct.matrix.new<-cbind(rownames(ct.matrix),ct.matrix.new)
#rownames(ct.matrix.new)<-as.matrix(ct.matrix.new)[,1]
#my.cat.cor(fpkm.matrix.clean,feature.name="Hilar_Lymphadenopathy_yes_no",tissue.name="PBMC",samplelist=sample.list.all,clinicmatrix=ct.matrix.new,outputdir=output.dir) # 102 vs 93
#my.cat.cor(fpkm.matrix.clean,feature.name="Hilar_Lymphadenopathy_oneside_bi",tissue.name="PBMC",samplelist=sample.list.all,clinicmatrix=ct.matrix.new,outputdir=output.dir) # 1 vs 92
#my.cat.cor(fpkm.matrix.clean,feature.name="Med_Lymphadenopathy_yes_no",tissue.name="PBMC",samplelist=sample.list.all,clinicmatrix=ct.matrix.new,outputdir=output.dir) # 76 vs 119
#my.cat.cor(fpkm.matrix.clean,feature.name="Med_Lymphadenopathy_oneside_bi",tissue.name="PBMC",samplelist=sample.list.all,clinicmatrix=ct.matrix.new,outputdir=output.dir) # 12 vs 107
#my.cat.cor(fpkm.matrix.clean,feature.name="Pulmonary_Art",tissue.name="PBMC",samplelist=sample.list.all,clinicmatrix=ct.matrix.new,outputdir=output.dir) # 193 vs. 2
#my.cat.cor(fpkm.matrix.clean,feature.name="Tree_in_bud",tissue.name="PBMC",samplelist=sample.list.all,clinicmatrix=ct.matrix.new,outputdir=output.dir) # 0 samples with value of 1
#my.cat.cor(fpkm.matrix.clean,feature.name="Micronodule",tissue.name="PBMC",samplelist=sample.list.all,clinicmatrix=ct.matrix.new,outputdir=output.dir) # 98 vs 97





#my.con.cor(fpkm.matrix.clean,feature.name="Hilar_Lymphadenopathy_yes_no",method.name="spearman",tissue.name="BAL",sample.list.all,ct.matrix,output.dir)
#my.con.cor(fpkm.matrix.clean,feature.name="Hilar_Lymphadenopathy_yes_no",method.name="spearman",tissue.name="PBMC",sample.list.all,ct.matrix,output.dir)
#my.con.cor(fpkm.matrix.clean,feature.name="Hilar_Lymphadenopathy_oneside_bi",method.name="spearman",tissue.name="BAL",sample.list.all,ct.matrix,output.dir)
#my.con.cor(fpkm.matrix.clean,feature.name="Hilar_Lymphadenopathy_oneside_bi",method.name="spearman",tissue.name="PBMC",sample.list.all,ct.matrix,output.dir)

#my.con.cor(fpkm.matrix.clean,feature.name="Med_Lymphadenopathy_yes_no",method.name="spearman",tissue.name="BAL",sample.list.all,ct.matrix,output.dir)
#my.con.cor(fpkm.matrix.clean,feature.name="Med_Lymphadenopathy_yes_no",method.name="spearman",tissue.name="PBMC",sample.list.all,ct.matrix,output.dir)
#my.con.cor(fpkm.matrix.clean,feature.name="Med_Lymphadenopathy_oneside_bi",method.name="spearman",tissue.name="BAL",sample.list.all,ct.matrix,output.dir)
#my.con.cor(fpkm.matrix.clean,feature.name="Med_Lymphadenopathy_oneside_bi",method.name="spearman",tissue.name="PBMC",sample.list.all,ct.matrix,output.dir)

#my.con.cor(fpkm.matrix.clean,feature.name="Pulmonary_Art",method.name="spearman",tissue.name="BAL",sample.list.all,ct.matrix,output.dir)
#my.con.cor(fpkm.matrix.clean,feature.name="Pulmonary_Art",method.name="spearman",tissue.name="PBMC",sample.list.all,ct.matrix,output.dir)
#my.con.cor(fpkm.matrix.clean,feature.name="Tree_in_bud",method.name="spearman",tissue.name="BAL",sample.list.all,ct.matrix,output.dir)
#my.con.cor(fpkm.matrix.clean,feature.name="Tree_in_bud",method.name="spearman",tissue.name="PBMC",sample.list.all,ct.matrix,output.dir)

#my.con.cor(fpkm.matrix.clean,feature.name="Micronodule",method.name="spearman",tissue.name="BAL",sample.list.all,ct.matrix,output.dir)
#my.con.cor(fpkm.matrix.clean,feature.name="Micronodule",method.name="spearman",tissue.name="PBMC",sample.list.all,ct.matrix,output.dir)



```

### Figure 2a. Overlap of associated genes between different traits
First, we show the number of genes significantly (FDR<0.05) associated with the different clinical traits.
```{r eval=FALSE,echo=FALSE}
data.dir<-file.path(work.dir,"scratch/Antun/SUPERvisedA/Results_summary2/Demographics_correlated_genes")
#data.dir<-"/home/antun/driver_Grace_xy48/scratch/Antun/SUPERvisedA/Results_summary2/Demographics_correlated_genes"
filenames<-list.files(data.dir)

#AGE
temp<-read.table(file.path(data.dir,"correlation_spearman_AGE_fdr0.05_BAL.txt"),sep="\t",header=T,check.names=F,stringsAsFactors=F)
age<-c("AGE",nrow(temp))

#GENDER
temp<-read.table(file.path(data.dir,"correlation_wilcox_GENDER_fdr0.05_BAL.txt"),sep="\t",header=T,check.names=F,stringsAsFactors=F)
gen<-c("GENDER",nrow(temp))

#RACE
temp<-read.table(file.path(data.dir,"correlation_wilcox_RACE_fdr0.05_BAL.txt"),sep="\t",header=T,check.names=F,stringsAsFactors=F)
rac<-c("RACE",nrow(temp))

mytab<-rbind(age,gen,rac)
colnames(mytab)<-c("Features","Number of genes (FDR<0.05)")
write.table(mytab,file="/home/antun/driver_Grace_xy48/scratch/Antun/SUPERvisedA/forDemographicsTable/DemographicsCorrelated.txt",append=F,sep="\t",row.names=F,col.names=T,quote=F)


data.dir<-"/home/antun/driver_Grace_xy48/scratch/Antun/SUPERvisedA/Results_summary2/PFTs_correlated_genes"
#FEV1%
temp<-read.table(file.path(data.dir,"correlation_spearman_FEV1PRED_fdr0.05_BAL.txt"),sep="\t",header=T,check.names=F,stringsAsFactors=F)
fevp<-c("FEV1%",nrow(temp))

#FEV1
temp<-read.table(file.path(data.dir,"correlation_spearman_BDFEV1_fdr0.05_BAL.txt"),sep="\t",header=T,check.names=F,stringsAsFactors=F)
fev<-c("FEV1",nrow(temp))

#FVC%
temp<-read.table(file.path(data.dir,"correlation_spearman_FVCPRED_fdr0.05_BAL.txt"),sep="\t",header=T,check.names=F,stringsAsFactors=F)
fvcp<-c("FVC%",nrow(temp))

#FVC
temp<-read.table(file.path(data.dir,"correlation_spearman_BDFVC_fdr0.05_BAL.txt"),sep="\t",header=T,check.names=F,stringsAsFactors=F)
fvc<-c("FVC",nrow(temp))

#DLCO%
temp<-read.table(file.path(data.dir,"correlation_spearman_PREDDLCO_fdr0.05_BAL.txt"),sep="\t",header=T,check.names=F,stringsAsFactors=F)
dlcop<-c("DLCO%",nrow(temp))

#DLCO
temp<-read.table(file.path(data.dir,"correlation_spearman_BDDLCO_fdr0.05_BAL.txt"),sep="\t",header=T,check.names=F,stringsAsFactors=F)
dlco<-c("DLCO",nrow(temp))

data.dir<-"/home/antun/driver_Grace_xy48/scratch/Antun/SUPERvisedA/Results_summary2/Disease_correlated_genes"
#SCADDING
temp<-read.table(file.path(data.dir,"correlation_spearman_SCADDING_fdr0.05_BAL.txt"),sep="\t",header=T,check.names=F,stringsAsFactors=F)
scad<-c("SCADDING",nrow(temp))

mytab<-rbind(fevp,fev,fvcp,fvc,dlcop,dlco,scad)
colnames(mytab)<-c("Features","Number of genes (FDR<0.05)")
write.table(mytab,file="/home/antun/driver_Grace_xy48/scratch/Antun/SUPERvisedA/forDemographicsTable/PFT_DiseaseCorrelated.txt",append=F,sep="\t",row.names=F,col.names=T,quote=F)

data.dir<-"/home/antun/driver_Grace_xy48/scratch/Antun/SUPERvisedA/Results_summary2/CTVariables_correlated_genes"
#Bronchial Wall Thickening
temp<-read.table(file.path(data.dir,"correlation_spearman_Bronchial_Wall_Thickening_fdr0.05_BAL.txt"),sep="\t",header=T,check.names=F,stringsAsFactors=F)
bwt<-c("Bronchial Wall Thickening",nrow(temp))

#Bronchiectasis Severity
temp<-read.table(file.path(data.dir,"correlation_spearman_Bronchiectasis_severity_fdr0.05_BAL.txt"),sep="\t",header=T,check.names=F,stringsAsFactors=F)
bs<-c("Bronchiectasis Severity",nrow(temp))

#Ground Glass
temp<-read.table(file.path(data.dir,"correlation_spearman_Ground_Glass_fdr0.05_BAL.txt"),sep="\t",header=T,check.names=F,stringsAsFactors=F)
gg<-c("Ground Glass",nrow(temp))

#Honeycombing
temp<-read.table(file.path(data.dir,"correlation_spearman_Honeycombing_fdr0.05_BAL.txt"),sep="\t",header=T,check.names=F,stringsAsFactors=F)
hc<-c("Honeycombing",nrow(temp))

#Mediastinal Lymphadenopathy
temp<-read.table(file.path(data.dir,"correlation_wilcox_Med_Lymphadenopathy_yes_no_fdr0.05_BAL.txt"),sep="\t",header=T,check.names=F,stringsAsFactors=F)
ml<-c("Mediastinal Lymphadenopathy",nrow(temp))

#Reticular Abnormality
temp<-read.table(file.path(data.dir,"correlation_spearman_Reticular_Abnormality_fdr0.05_BAL.txt"),sep="\t",header=T,check.names=F,stringsAsFactors=F)
ra<-c("Reticular Abnormality",nrow(temp))

#Traction Bronchiestasis
temp<-read.table(file.path(data.dir,"correlation_spearman_Traction_Bronchiectasis_fdr0.05_BAL.txt"),sep="\t",header=T,check.names=F,stringsAsFactors=F)
tb<-c("Traction Bronchiestasis",nrow(temp))

#Hilar Lymphadenopathy
temp<-read.table(file.path(data.dir,"correlation_wilcox_Hilar_Lymphadenopathy_yes_no_fdr0.05_BAL.txt"),sep="\t",header=T,check.names=F,stringsAsFactors=F)
hl<-c("Hilar Lymphadenopathy",nrow(temp))


mytab<-rbind(bwt,bs,gg,hc,ml,ra,tb,hl)
colnames(mytab)<-c("Features","Number of genes (FDR<0.05)")
write.table(mytab,file="/home/antun/driver_Grace_xy48/scratch/Antun/SUPERvisedA/forDemographicsTable/CTCorrelated.txt",append=F,sep="\t",row.names=F,col.names=T,quote=F)



data.dir<-"/home/antun/driver_Grace_xy48/scratch/Antun/SUPERvisedA/Results_summary2/CellDifferentials_correlated_genes"
#ALVBALD
temp<-read.table(file.path(data.dir,"correlation_spearman_ALVBALD_fdr0.05_BAL.txt"),sep="\t",header=T,check.names=F,stringsAsFactors=F)
alv<-c("ALVBALD",nrow(temp))

#BALCELL
temp<-read.table(file.path(data.dir,"correlation_spearman_BALCELL_fdr0.05_BAL.txt"),sep="\t",header=T,check.names=F,stringsAsFactors=F)
balc<-c("BALCELL",nrow(temp))

#BALCELSP
temp<-read.table(file.path(data.dir,"correlation_spearman_BALCELSP_fdr0.05_BAL.txt"),sep="\t",header=T,check.names=F,stringsAsFactors=F)
balcsp<-c("BALCELSP",nrow(temp))

#EOSBALD
temp<-read.table(file.path(data.dir,"correlation_spearman_EOSBALD_fdr0.05_BAL.txt"),sep="\t",header=T,check.names=F,stringsAsFactors=F)
eos<-c("EOSBALD",nrow(temp))

#LYMBALD
temp<-read.table(file.path(data.dir,"correlation_spearman_LYMBALD_fdr0.05_BAL.txt"),sep="\t",header=T,check.names=F,stringsAsFactors=F)
lym<-c("LYMBALD",nrow(temp))

#NEUBALD
temp<-read.table(file.path(data.dir,"correlation_spearman_NEUBALD_fdr0.05_BAL.txt"),sep="\t",header=T,check.names=F,stringsAsFactors=F)
neu<-c("NEUBALD",nrow(temp))

mytab<-rbind(alv,balc,balcsp,eos,lym,neu)
colnames(mytab)<-c("Features","Number of genes (FDR<0.05)")
write.table(mytab,file="/home/antun/driver_Grace_xy48/scratch/Antun/SUPERvisedA/forDemographicsTable/CellDiffCorrelated.txt",append=F,sep="\t",row.names=F,col.names=T,quote=F)

```


```{r eval=TRUE, echo=FALSE, cache=TRUE,warning=FALSE,message=FALSE,results='asis'}
data.dir<-file.path(work.dir,"scratch/Antun/SUPERvisedA/Results_summary2")
#data.dir<-"/home/antun/driver_Grace_xy48/scratch/Antun/SUPERvisedA/Results_summary2"
final.list<-list()
# load in the gene lists
var.names<-c("AGE","GENDER","RACE",
             "FVCPRED","BDFVC","FEV1PRED","BDFEV1","PREDDLCO","BDDLCO","FEV1FVCratio",
             "Bronchial_Wall_Thickening","Bronchiectasis_severity","Ground_Glass","Honeycombing","Reticular_Abnormality","Traction_Bronchiectasis","Med_Lymphadenopathy_yes_no","Hilar_Lymphadenopathy_yes_no","SCADDING","BALCELL","ALVBALD","EOSBALD","LYMBALD","NEUBALD")
var.annot<-c("Age","Gender","Race",
             "FVC% PRED","FVC","FEV1% PRED","FEV1","DLCO% PRED","DLCO","FEV1/FVC ratio",
             "Bronchial Wall Thickening","Bronchiectasis Severity","Ground Glass","Honeycombing","Reticular Abnormality","Traction Bronchiestasis","Mediastinal Lymphadenopathy","Hilar Lymphadenopathy","SCADDING","Total Cell Counts","Macrophage %","Eosinophil %","Lymphocyte %","Neutrophil %")
var.cat<-c(rep("Demographics",3),
           rep("PFTs",7),
           rep("CTVariables",8),
           "Disease",
           rep("CellDifferentials",5))
for(i in 1:length(var.names)){
file.names<-list.files(file.path(data.dir,paste(var.cat[i],"_correlated_genes",sep="")))
data.filepath<-file.path(data.dir,paste(var.cat[i],"_correlated_genes",sep=""),file.names[grep(paste(var.names[i],"_allgenes_BAL.txt",sep=""),file.names)])
if(length(data.filepath)>1){
  # For race and gender, there are two results: spearman and wilcox. We use the results by the Wilcoxon Rank Sum test.
  data.filepath<-data.filepath[grep("wilcox",data.filepath)]
}
#data.filepath<-file.path(data.dir,paste(var.cat[i],"_correlated_genes",sep=""),paste("correlation_spearman_",var.names[i],"_allgenes_BAL.txt",sep=""))
temp<-read.table(data.filepath,sep = "\t",header=T,check.names=F,row.names=1)
#  temp<-temp[!is.na(temp[,"fdr"]) & temp[,"fdr"]<0.05,]
final.list[[i]]<-temp
}
names(final.list)<-var.annot

# get all the gene names
name_genes<-character()
for(i in 1:length(final.list)){
temp<-final.list[[i]]

name_genes<-union(name_genes,rownames(temp)[!is.na(temp[,"fdr"]) & temp[,"fdr"]<0.05])
}

name_genes<-unname(sapply(name_genes,my.element.remove,splitchar="_",index=-1))
cat("\n There are in total ",length(name_genes)," significant genes by the supervised analysis on all the variables.\n",sep="")
```

```{r eval=TRUE,fig.width=10,fig.height=10,cache=TRUE,warning=FALSE,message=FALSE,results='asis'}
#http://www.sthda.com/english/wiki/ggplot2-quick-correlation-matrix-heatmap-r-software-and-data-visualization
# calculate the number of overlapping genes
output.dir<-file.path(work.dir,"scratch/GRADS/SARC_results/Results_summary_BAL_hg38/SupervisedAnalysis_Overlap_withFEV1FVCratio")

if(file.exists(output.dir)==F){
  dir.create(output.dir)
}

overlap.n.matrix<-matrix(NA,nrow=length(final.list),ncol=length(final.list))
overlap.p.matrix<-matrix(NA,nrow=length(final.list),ncol=length(final.list))
overlap.n.matrix.pos<-overlap.n.matrix
overlap.n.matrix.neg<-overlap.n.matrix
overlap.n.matrix.directsummary<-overlap.n.matrix

for(i in 1:length(final.list)){
  for(j in 1:length(final.list)){
    if(i==j){
        overlap.n.matrix[i,j]<-sum((!is.na(final.list[[i]][,"fdr"])) & final.list[[i]][,"fdr"]<0.05)  

        overlap.n.matrix.pos[i,j]<-overlap.n.matrix[i,j]
        overlap.n.matrix.neg[i,j]<-0
      
        list.overlap<-rownames(final.list[[i]])[!is.na(final.list[[i]][,"fdr"]) & final.list[[i]][,"fdr"]<0.05]

      if("pval_dir"%in%colnames(final.list[[i]])){
        coef.1<-final.list[[i]][list.overlap,"pval_dir"]
      }else{
        coef.1<-final.list[[i]][list.overlap,"corr_coef"]
      }

      overlap.n.matrix.directsummary[i,j]<-paste(sum(coef.1>0)," +\n",sum(coef.1<0)," -",sep="")
      next
    }
    temp1<-final.list[[i]]
    temp2<-final.list[[j]]
    
    temp.names<-intersect(rownames(temp1),rownames(temp2))

    temp1<-temp1[temp.names,]
    temp2<-temp2[temp.names,]
    
      list1<-rownames(final.list[[i]])[!is.na(final.list[[i]][,"fdr"]) & final.list[[i]][,"fdr"]<0.05]
      list2<-rownames(final.list[[j]])[!is.na(final.list[[j]][,"fdr"]) & final.list[[j]][,"fdr"]<0.05]

    list.overlap<-intersect(list1,list2)
    
    if("pval_dir"%in%colnames(final.list[[i]])){
      coef.1<-final.list[[i]][list.overlap,"pval_dir"]
    }else{
      coef.1<-final.list[[i]][list.overlap,"corr_coef"]
    }

    if("pval_dir"%in%colnames(final.list[[j]])){
      coef.2<-final.list[[j]][list.overlap,"pval_dir"]
    }else{
      coef.2<-final.list[[j]][list.overlap,"corr_coef"]
    }
    
    list.overlap.pos<-list.overlap[coef.1 * coef.2 >0]
    list.overlap.neg<-setdiff(intersect(list1,list2),list.overlap.pos)
    
    if(length(list.overlap.pos)>length(list.overlap.neg)){
      list.1<-list.overlap[coef.1>0 & coef.2>0]
      list.2<-list.overlap[coef.1<0 & coef.2<0]
      overlap.n.matrix.directsummary[i,j]<-paste(length(list.1),"(++)\n",length(list.2),"(--)")
    }
    if(length(list.overlap.pos)==length(list.overlap.neg)){
      list.1<-character()
      list.2<-character()
      overlap.n.matrix.directsummary[i,j]<-paste("0(++)\n0(--)")
    }
    if(length(list.overlap.pos)<length(list.overlap.neg)){
      list.1<-list.overlap[coef.1>0 & coef.2<0]
      list.2<-list.overlap[coef.1<0 & coef.2>0]
      overlap.n.matrix.directsummary[i,j]<-paste(length(list.1),"(+-)\n",length(list.2),"(-+)")
    }
    
    pop.size<-nrow(temp1)
    success.n<-length(list1)
    sample.n<-length(list2)
    x<-length(intersect(list1,list2))
    
    overlap.n.matrix[i,j]<-length(intersect(list1,list2))
    overlap.p.matrix[i,j]<-1-phyper(x-1,success.n,pop.size-success.n,sample.n,lower.tail=TRUE)
    
    overlap.n.matrix.pos[i,j]<-length(list.overlap.pos)
    overlap.n.matrix.neg[i,j]<-length(list.overlap.neg)
    if(i>j){
    # output the list of overlap genes
    if(length(list.overlap)>0){
      output.filepath<-file.path(output.dir,paste(names(final.list)[i],"_VS_",names(final.list)[j],"_all.txt",sep=""))
      cmd.out<-final.list[[i]][list.overlap,]
      colnames(cmd.out)<-paste(colnames(cmd.out)," (",names(final.list)[i],")",sep="")
      cmd.out.1<-final.list[[j]][list.overlap,]
      colnames(cmd.out.1)<-paste(colnames(cmd.out.1)," (",names(final.list)[j],")",sep="")
      cmd.out<-cbind(cmd.out,cmd.out.1)
      cat("tracking_ID\t",file=output.filepath,append=F)
      write.table(cmd.out,file=output.filepath,append=T,sep="\t",row.names=T,col.names=T,quote=F)
    }
    # output the list of overlap genes with the same directions
    if(length(list.overlap.pos)>0){
      output.filepath<-file.path(output.dir,paste(names(final.list)[i],"_VS_",names(final.list)[j],"_samedirection.txt",sep=""))
      cmd.out<-final.list[[i]][list.overlap.pos,]
      colnames(cmd.out)<-paste(colnames(cmd.out)," (",names(final.list)[i],")",sep="")
      cmd.out.1<-final.list[[j]][list.overlap.pos,]
      colnames(cmd.out.1)<-paste(colnames(cmd.out.1)," (",names(final.list)[j],")",sep="")
      cmd.out<-cbind(cmd.out,cmd.out.1)
      cat("tracking_ID\t",file=output.filepath,append=F)
      write.table(cmd.out,file=output.filepath,append=T,sep="\t",row.names=T,col.names=T,quote=F)
    }
    # output the list of overlapped genes with opposite directions
     if(length(list.overlap.neg)>0){
      output.filepath<-file.path(output.dir,paste(names(final.list)[i],"_VS_",names(final.list)[j],"_oppositedirection.txt",sep=""))
      cmd.out<-final.list[[i]][list.overlap.neg,]
      colnames(cmd.out)<-paste(colnames(cmd.out)," (",names(final.list)[i],")",sep="")
      cmd.out.1<-final.list[[j]][list.overlap.neg,]
      colnames(cmd.out.1)<-paste(colnames(cmd.out.1)," (",names(final.list)[j],")",sep="")
      cmd.out<-cbind(cmd.out,cmd.out.1)
      cat("tracking_ID\t",file=output.filepath,append=F)
      write.table(cmd.out,file=output.filepath,append=T,sep="\t",row.names=T,col.names=T,quote=F)
     }
      
     if(length(list.1)>0){
       
       if(length(list.overlap.pos)>length(list.overlap.neg)){
          output.filepath<-file.path(output.dir,paste(names(final.list)[i],"_VS_",names(final.list)[j],"_pospos.txt",sep=""))
       }
       
       if(length(list.overlap.pos)<length(list.overlap.neg)){
          output.filepath<-file.path(output.dir,paste(names(final.list)[i],"_VS_",names(final.list)[j],"_posneg.txt",sep=""))
       }
        cmd.out<-final.list[[i]][list.1,]
        colnames(cmd.out)<-paste(colnames(cmd.out)," (",names(final.list)[i],")",sep="")
        cmd.out.1<-final.list[[j]][list.1,]
        colnames(cmd.out.1)<-paste(colnames(cmd.out.1)," (",names(final.list)[j],")",sep="")
        cmd.out<-cbind(cmd.out,cmd.out.1)
        cat("tracking_ID\t",file=output.filepath,append=F)
        write.table(cmd.out,file=output.filepath,append=T,sep="\t",row.names=T,col.names=T,quote=F)
     }
      
    if(length(list.2)>0){
       if(length(list.overlap.pos)>length(list.overlap.neg)){
          output.filepath<-file.path(output.dir,paste(names(final.list)[i],"_VS_",names(final.list)[j],"_negneg.txt",sep=""))
       }
       
       if(length(list.overlap.pos)<length(list.overlap.neg)){
          output.filepath<-file.path(output.dir,paste(names(final.list)[i],"_VS_",names(final.list)[j],"_negpos.txt",sep=""))
       }
          cmd.out<-final.list[[i]][list.2,]
          colnames(cmd.out)<-paste(colnames(cmd.out)," (",names(final.list)[i],")",sep="")
          cmd.out.1<-final.list[[j]][list.2,]
          colnames(cmd.out.1)<-paste(colnames(cmd.out.1)," (",names(final.list)[j],")",sep="")
          cmd.out<-cbind(cmd.out,cmd.out.1)
          cat("tracking_ID\t",file=output.filepath,append=F)
          write.table(cmd.out,file=output.filepath,append=T,sep="\t",row.names=T,col.names=T,quote=F)
    }
    }
  }
}


rownames(overlap.n.matrix)<-names(final.list)
colnames(overlap.n.matrix)<-names(final.list)

rownames(overlap.p.matrix)<-names(final.list)
colnames(overlap.p.matrix)<-names(final.list)

```


```{r eval=TRUE,cache=FALSE, fig.cap=figure_nums(name="figure_heatmap_geneoverlap_1",caption="Number of significant genes (FDR<0.05) identified by the supervised analysis for the given clinical traits."),echo=FALSE,results='asis'}
```
`r figure_nums(name="figure_heatmap_geneoverlap_1")`
```{r eval=TRUE,fig.width=12,fig.height=12, echo=FALSE,cache=FALSE,results='asis'}
# Get lower triangle of the correlation matrix
get_lower_tri<-function(cormat){
  cormat[upper.tri(cormat)] <- NA
  return(cormat)
}
# Get upper triangle of the correlation matrix
get_upper_tri <- function(cormat){
  cormat[lower.tri(cormat)]<- NA
  return(cormat)
}


# draw the overlap matrix using ggplot2
#c1<-signif(overlap.n.matrix, 2)
c1<-overlap.n.matrix
c2<-overlap.p.matrix
#c2<-signif(overlap.p.matrix, 2)
diag(c2)<-rep(0,length(diag(c2)))
upper_tri <- get_upper_tri(c1)
upper_tri_p<-get_upper_tri(c2)
melted_cormat_n <- melt(upper_tri, na.rm = TRUE)
melted_cormat_p<-melt(upper_tri_p, na.rm=TRUE)

melted_cormat<-melted_cormat_n
melted_cormat[,3]<-melted_cormat_p[,3]
melted_cormat<-cbind(melted_cormat,melted_cormat_n[,3])
colnames(melted_cormat)[4]<-"gnum"
melted_cormat[melted_cormat[,"value"]>=0.05,"value"]<-NA

g<-ggplot(data = melted_cormat, aes(Var2, Var1, fill = value))+
  geom_tile(color = "white")+
  scale_fill_gradient(low = "yellow", high = "white", na.value="lightgray", space = "Lab", name="# of overlapping genes") +
  #scale_fill_gradientn(colours = c("red", "white"),values=scales::rescale(c(0, 0.05, 1)), na.value="white", space = "Lab", name="# of overlapping genes") +
  theme_minimal()+ 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, size = 12, hjust = 1))+
  coord_fixed()
#melted_cormat[,3]<-melted_cormat_n[,3]
g<-g+
  geom_text(aes(Var2, Var1, label = gnum), color = "black", size = 4)
print(g)

output.subdir<-file.path(work.dir,"scratch/GRADS/SARC_results/Results_summary_BAL_hg38/writings")
ggsave(filename = "Figure2a_SupervisedAnalysis_OverlapMatrix_withFEV1FVCratio.eps",path=output.subdir,width=12,height=12,device="eps")
```


Here we generate another figure that shows the breakdown of the genes based on the directionality of the correlation coefficients. Each entry will show the number of overlapping genes and the breakdown of these genes into genes with + and + or - and - correlation with the rows and columns if those two features are positively correlated. The breakdown will be the number of genes with + and - or - and + correlation with the row and column if those two features are negatively correlated.  
```{r fig.width=20,fig.height=20,eval=TRUE}
get_lower_tri<-function(cormat){
  cormat[upper.tri(cormat)] <- NA
  return(cormat)
}
# Get upper triangle of the correlation matrix
get_upper_tri <- function(cormat){
  cormat[lower.tri(cormat)]<- NA
  return(cormat)
}


# draw the overlap matrix using ggplot2
#c1<-signif(overlap.n.matrix, 2)
temp<-overlap.n.matrix.directsummary
temp[!is.na(overlap.p.matrix) & overlap.p.matrix>=0.05]<-""
temp[!is.na(overlap.p.matrix) & overlap.p.matrix<0.05]<-paste("\n",temp[!is.na(overlap.p.matrix) & overlap.p.matrix<0.05],sep="")
temp[is.na(overlap.p.matrix)]<-paste("\n",temp[is.na(overlap.p.matrix)],sep="")

#c1<-matrix(paste(overlap.n.matrix,"\n",overlap.n.matrix.directsummary,sep=""),nrow=nrow(overlap.n.matrix),byrow=F)
c1<-matrix(paste(overlap.n.matrix,temp,sep=""),nrow=nrow(overlap.n.matrix),byrow=F)
rownames(c1)<-rownames(overlap.n.matrix)
colnames(c1)<-rownames(overlap.n.matrix)
c2<-overlap.p.matrix
#c2<-signif(overlap.p.matrix, 2)
diag(c2)<-rep(0,length(diag(c2)))
upper_tri <- get_upper_tri(c1)
upper_tri_p<-get_upper_tri(c2)
melted_cormat_n <- melt(upper_tri, na.rm = TRUE)
melted_cormat_p<-melt(upper_tri_p, na.rm=TRUE)

melted_cormat<-melted_cormat_n
melted_cormat[,3]<-melted_cormat_p[,3]
melted_cormat<-cbind(melted_cormat,melted_cormat_n[,3])
colnames(melted_cormat)[4]<-"gnum"
melted_cormat[melted_cormat[,"value"]>=0.05,"value"]<-NA

g<-ggplot(data = melted_cormat, aes(Var2, Var1, fill = value))+
  geom_tile(color = "white")+
  scale_fill_gradient(low = "yellow", high = "white", na.value="lightgray", space = "Lab", name="# of overlapping genes") +
  #scale_fill_gradientn(colours = c("red", "white"),values=scales::rescale(c(0, 0.05, 1)), na.value="white", space = "Lab", name="# of overlapping genes") +
  theme_minimal()+ 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, size = 18, hjust = 1),axis.text.y = element_text( size = 18))+
  coord_fixed()
#melted_cormat[,3]<-melted_cormat_n[,3]
g<-g+
  geom_text(aes(Var2, Var1, label = gnum), color = "black", size = 4.5,lineheight = .8)
print(g)

#output.subdir<-file.path(output.dir,"scratch/GRADS/SARC_results/Results_summary_BAL_hg38/writings")
ggsave(filename = "Figure2a_SupervisedAnalysis_OverlapMatrix_breakdown.eps",path=output.dir,width=22,height=22,device="eps")
```



### Figure 2b. Sankey plot for supervised analysis
Second, we used the sankey plot to show the functional analysis results of the genes asssociated with given traits.
```{r}

rm(list=setdiff(ls(),c("work.dir","figure_nums","figure_nums_3","table_nums","table_nums_3")))
library(networkD3)
library(rbokeh)
library(gdata)
source("/home/yanxiting/driver_Grace/Rprogram/my_functions.R")

data.dir<-paste(work.dir,"/scratch/Antun/SUPERvisedA/Results_summary2",sep="")
output.dir<-"/home/yanxiting/driver_Grace/scratch/GRADS/SARC_results/Results_summary_BAL_hg38/writings"
top.pathway.num<-5
min.gene.num<-3
log.filepath<-file.path(output.dir,paste("Figure2b_supervised_sankey_fdr0.05_top",top.pathway.num,"_mingenum",min.gene.num,"_withFEV1FVCratio.html.log",sep=""))

cap<-paste("Figure 2b. Sankey plot showing the function analysis results of the clinical traits associated genes identified by the supervised analysis for top.pathway.num=",top.pathway.num,", min.gene.num=",min.gene.num,sep="")
```

```{r figure2b_sankeyplot_supervised,eval=TRUE,cache=FALSE,fig.height=15,fig.width=15,fig.cap=cap}
if(file.exists(log.filepath)){
  file.remove(log.filepath)
}
file.create(log.filepath)

#data.dir<-"/home/antun/driver_Grace_xy48/scratch/Antun/SUPERvisedA/Results_summary2"
MMList<-character()
metacore.pos.list<-character()
metacore.neg.list<-character()
# load in the gene lists
var.names<-c("AGE","GENDER","RACE",
             "FVCPRED","BDFVC","FEV1PRED","BDFEV1","PREDDLCO","BDDLCO","FEV1FVCratio",
             "Bronchial_Wall_Thickening","Bronchiectasis_severity","Ground_Glass","Honeycombing","Reticular_Abnormality","Traction_Bronchiectasis","Med_Lymphadenopathy_yes_no","Hilar_Lymphadenopathy_yes_no","SCADDING","BALCELL","ALVBALD","EOSBALD","LYMBALD","NEUBALD")
var.annot<-c("Age","Gender","Race",
             "FVC% PRED","FVC","FEV1% PRED","FEV1","DLCO% PRED","DLCO","FEV1FVCratio",
             "Bronchial Wall Thickening","Bronchiectasis Severity","Ground Glass","Honeycombing","Reticular Abnormality","Traction Bronchiestasis","Mediastinal Lymphadenopathy","Hilar Lymphadenopathy","SCADDING","Total Cell Counts","Macrophage","Eosinophil","Lymphocyte","Neutrophil")
var.cat<-c(rep("Demographics",3),
           rep("PFTs",7),
           rep("CTVariables",8),
           "Disease",
           rep("CellDifferentials",5))
for(i in 1:length(var.names)){
  
file.names<-list.files(file.path(data.dir,paste(var.cat[i],"_correlated_genes",sep="")))
data.filepath<-file.path(data.dir,paste(var.cat[i],"_correlated_genes",sep=""),file.names[grep(paste(var.names[i],"_allgenes_BAL.txt",sep=""),file.names)])
if(length(data.filepath)>1){
  # For race and gender, there are two results: spearman and wilcox. We use the results by the Wilcoxon Rank Sum test.
  #data.filepath<-data.filepath[2]
  data.filepath<-data.filepath[grep("wilcox",data.filepath)]
}
MMList<-c(MMList,data.filepath)
temp.char<-my.element.extract(data.filepath,splitchar="/",index=-1)
temp.char<-my.element.extract(temp.char,splitchar=".txt",index=1)
temp.char<-my.char.replace(temp.char,orig="_allgenes_",replace="_fdr0.05_")
metacore.pos<-file.path(data.dir,paste(var.cat[i],"_correlated_genes",sep=""),paste(temp.char,"_positive_metacore/Enrichment_analysis.xls",sep=""))
metacore.neg<-file.path(data.dir,paste(var.cat[i],"_correlated_genes",sep=""),paste(temp.char,"_negative_metacore/Enrichment_analysis.xls",sep=""))

metacore.pos.list<-c(metacore.pos.list,metacore.pos)
metacore.neg.list<-c(metacore.neg.list,metacore.neg)
}

names(MMList)<-var.annot


ModColList<-c()
ModValList<-c()
ModToList<-c()
ModTypeList<-c()

for(i in 1:length(MMList)){

#########name convention!!!!!!
#Mmemb<-read.table(file.path(data.dir,MMList[i]),sep="\t",header=T,check.names=F,stringsAsFactors=F)
Mmemb<-read.table(MMList[i],sep="\t",header=T,check.names=F,stringsAsFactors=F)

  if(sum((!is.na(Mmemb[,"fdr"])) & (Mmemb[,"fdr"]<0.05))==0){
    next
  }
  Mmemb<-Mmemb[(!is.na(Mmemb[,"fdr"])) & (Mmemb[,"fdr"]<0.05),]  



first.name<-names(MMList)[i]

test.type<-my.element.extract(MMList[i],splitchar="/",index=-1)
test.type<-my.element.extract(test.type,splitchar="_",index=2)

if(test.type=="spearman"){
Mmemb.pos<-Mmemb[Mmemb[,"corr_coef"]>0,]
pos.length<-sum(Mmemb[,"corr_coef"]>0)

Mmemb.neg<-Mmemb[Mmemb[,"corr_coef"]<0,]
neg.length<-sum(Mmemb[,"corr_coef"]<0)
}else{
Mmemb.pos<-Mmemb[Mmemb[,"pval_dir"]>0,]
pos.length<-sum(Mmemb[,"pval_dir"]>0)

Mmemb.neg<-Mmemb[Mmemb[,"pval_dir"]<0,]
neg.length<-sum(Mmemb[,"pval_dir"]<0)
}

metacore.pos<-metacore.pos.list[i]
metacore.neg<-metacore.neg.list[i]
#metacore.bot<-paste("/home/fas/kaminski/am2952/scratch/sankey_BAL_supervised/allmetacore/",ModCol,"_both_metacore.xls",sep="")


if(pos.length>0){
  ModColList<-c(ModColList,first.name)
  ModToList<-c(ModToList,paste(first.name," positive",sep=""))
  ModValList<-c(ModValList,pos.length)
  ModTypeList<-c(ModTypeList,"L1")
}


if(file.exists(metacore.pos)){
metac<-read.xls(metacore.pos,sheet=1,check.names=F,stringsAsFactors = F)
colnames(metac)<-c("#","Maps","Total","pValue","MinFDR","ID_p-value","ID_FDR","ID_InData","ID_NetworkObjectsfromActiveData")
if(nrow(metac)>=3){

metac<-metac[3:nrow(metac),2:9]
metac$ID_FDR<-as.numeric(metac$ID_FDR)

#take out fdr > 0.05
metac<-metac[metac$ID_FDR < 0.05,]
metac<-metac[order(metac$ID_FDR),]

#only show pathways with min.gene.num genes InData
metac<-metac[as.numeric(metac[,"ID_InData"])>=min.gene.num,]


# merge the pathways that are significant due to exactly the same genes
if(nrow(metac)>0){
temp.index<-split(1:nrow(metac),metac[,"ID_NetworkObjectsfromActiveData"])
my.metac<-numeric()
for(j in 1:length(temp.index)){
  if(length(temp.index[[j]])==1){
    my.metac<-rbind(my.metac,metac[temp.index[[j]],])
  }else{
  temp.matrix<-metac[temp.index[[j]],]  
  temp.result<-temp.matrix[1,]
  cat(first.name," (+):\n",sep="",file=log.filepath,append=T)
  cat(temp.matrix[,1],sep="\n",file=log.filepath,append=T)
  cat("\n",file=log.filepath,append=T)
  #temp.result[1,1]<-paste(temp.matrix[,1],collapse="\r")
  temp.result[1,"ID_FDR"]<-min(temp.matrix[,"ID_FDR"])
  temp.result[1,"ID_p-value"]<-min(temp.matrix[,"ID_p-value"])
  my.metac<-rbind(my.metac,temp.result)
  }
}
metac<-my.metac
metac<-metac[order(as.numeric(metac[,"MinFDR"]),decreasing=F),]
fdr.threshold.pos<-max(as.numeric(metac[1:min(top.pathway.num,nrow(metac)),"MinFDR"]))
metac<-metac[as.numeric(metac[,"MinFDR"])<=fdr.threshold.pos,]
#metac<-metac[1:min(top.pathway.num,nrow(metac)),]

if(nrow(metac)>0){
ModColList<-c(ModColList,rep(paste(first.name," positive",sep=""),nrow(metac)))
ModToList<-c(ModToList,metac[,1])
ModValList<-c(ModValList,metac[,7])
ModTypeList<-c(ModTypeList,rep("L2",nrow(metac)))
}

}
}

}else{
  
  cat("WARNING: ",metacore.pos, " doesn't exist!\n")
}

if(neg.length>0){
  ModColList<-c(ModColList,first.name)
  ModToList<-c(ModToList,paste(first.name," negative",sep=""))
  ModValList<-c(ModValList,neg.length)
  ModTypeList<-c(ModTypeList,"L1")
}

if(file.exists(metacore.neg)){

metac2<-read.xls(metacore.neg,sheet=1,check.names=F,stringsAsFactors = F)
colnames(metac2)<-c("#","Maps","Total","pValue","MinFDR","ID_p-value","ID_FDR","ID_InData","ID_NetworkObjectsfromActiveData")

if(nrow(metac2)>=3){
metac2<-metac2[3:nrow(metac2),2:9]
metac2$ID_FDR<-as.numeric(metac2$ID_FDR)

#take out fdr > 0.05
metac2<-metac2[metac2$ID_FDR < 0.05 ,]
metac2<-metac2[order(metac2$ID_FDR),]
metac2<-metac2[as.numeric(metac2[,"ID_InData"])>=min.gene.num,]

# merge the pathways that are significant due to the same genes
if(nrow(metac2)>0){
temp.index<-split(1:nrow(metac2),metac2[,"ID_NetworkObjectsfromActiveData"])
my.metac<-numeric()
for(j in 1:length(temp.index)){
  if(length(temp.index[[j]])==1){
    my.metac<-rbind(my.metac,metac2[temp.index[[j]],])
  }else{
  temp.matrix<-metac2[temp.index[[j]],]  
  temp.result<-temp.matrix[1,]
  cat(first.name," (-):\n",sep="",file=log.filepath,append=T)
  cat(temp.matrix[,1],sep="\n",file=log.filepath,append=T)
  cat("\n",file=log.filepath,append=T)
  #temp.result[1,1]<-paste(temp.matrix[,1],collapse="\r")
  temp.result[1,"ID_FDR"]<-min(temp.matrix[,"ID_FDR"])
  temp.result[1,"ID_p-value"]<-min(temp.matrix[,"ID_p-value"])
  my.metac<-rbind(my.metac,temp.result)
  }
}
metac2<-my.metac  
metac2<-metac2[order(as.numeric(metac2[,"MinFDR"]),decreasing=F),]
fdr.threshold<-max(as.numeric(metac2[1:min(top.pathway.num,nrow(metac2)),"MinFDR"]))
metac2<-metac2[as.numeric(metac2[,"MinFDR"])<=fdr.threshold,]


if(nrow(metac2)>0){
ModColList<-c(ModColList,rep(paste(first.name," negative",sep=""),times=nrow(metac2)))
ModToList<-c(ModToList,metac2[,1])
ModValList<-c(ModValList,metac2[,7])
ModTypeList<-c(ModTypeList,rep("L2",nrow(metac2)))
}

}
}
}else{
  
  cat("WARNING: ",metacore.neg, " doesn't exist!\n")
}


}

Alledges<-data.frame(ModColList,ModToList,ModValList,ModTypeList,stringsAsFactors=F)
colnames(Alledges)<-c("source","target","value","edgetype")
Alledges<-Alledges[complete.cases(Alledges), ]
write.table(Alledges,file=file.path(output.dir,paste("Figure2b_supervised_sankey_top",top.pathway.num,"_mingenum",min.gene.num,"_fdr0.05_withFEV1FVCratio.txt")),append=F,row.names=F,col.names=T,sep="\t",quote=F)

#alledges<-read.table("/home/yanxiting/driver_Grace/scratch/GRADS/SARC_results/Results_summary_BAL_hg38/writings/Figure2b_supervised_sankey_fdr0.05.txt",sep="\t",header=T,check.names=F,stringsAsFactors=F,quote = "")

# for all nodes in the middle, if there's no edges going out of it, remove all edges going into them
nodes.temp2<-unique(Alledges[Alledges[,4]=="L2",1])
nodes.temp1<-unique(Alledges[Alledges[,4]=="L1",2])
keep.nodes<-intersect(nodes.temp1,nodes.temp2)

alledges<-Alledges[Alledges[,2]%in%keep.nodes | Alledges[,1]%in%keep.nodes ,1:3]

allnodes<-c(alledges$source,alledges$target)
allnodes<-allnodes[!duplicated(allnodes)]

edges2<-alledges

for(i in 1:nrow(alledges)){
  
  edges2[i,1]<-as.integer(match(alledges[i,1],allnodes))-1
  edges2[i,2]<-as.integer(match(alledges[i,2],allnodes))-1
  
}

nodes2<-as.data.frame(allnodes,stringsAsFactors=F)
colnames(nodes2)<-"name"

edges2[,1]<-as.integer(edges2[,1])
edges2[,2]<-as.integer(edges2[,2])
edges2[,3]<-as.numeric(edges2[,3])


edges<-edges2[edges2[,1]==0,]
nodes<-nodes2[unique(as.numeric(as.matrix(edges[,1:2])))+1,]
nodes<-as.data.frame(nodes,stringsAsFactors = F)
colnames(nodes)<-"name"

edges<-rbind(edges,edges2[edges2[,1]==1,])
nodes<-nodes2[unique(as.numeric(as.matrix(edges[,1:2])))+1,]
nodes<-as.data.frame(nodes,stringsAsFactors = F)
colnames(nodes)<-"name"

edges<-rbind(edges,edges2[edges2[,1]==2,])
nodes<-nodes2[unique(as.numeric(as.matrix(edges[,1:2])))+1,]
nodes<-as.data.frame(nodes,stringsAsFactors = F)
colnames(nodes)<-"name"

# specify the nodes domain for the d3.scaleOrdinal
temp.cat<-nodes2[,1]
temp.cat[grep(" positive",temp.cat)]<-"positive"
temp.cat[grep(" negative",temp.cat)]<-"negative"

nodes2<-cbind(nodes2,temp.cat)
colnames(nodes2)[2]<-"ngroup"

# change the node names
nodes2[nodes2[,1]=="ALVBALD",1]<-"Macrophage"
nodes2[nodes2[,1]=="ALVBALD positive",1]<-"Macrophage (+)"
nodes2[nodes2[,1]=="ALVBALD negative",1]<-"Macrophage (-)"

nodes2[nodes2[,1]=="Macrophage positive",1]<-"Macrophage (+)"
nodes2[nodes2[,1]=="Macrophage negative",1]<-"Macrophage (-)"

nodes2[nodes2[,1]=="LYMBALD",1]<-"Lymphocyte"
nodes2[nodes2[,1]=="LYMBALD positive",1]<-"Lymphocyte (+)"
nodes2[nodes2[,1]=="LYMBALD negative",1]<-"Lymphocyte (-)"

nodes2[nodes2[,1]=="Lymphocyte positive",1]<-"Lymphocyte (+)"
nodes2[nodes2[,1]=="Lymphocyte negative",1]<-"Lymphocyte (-)"


nodes2[nodes2[,1]=="EOSBALD",1]<-"Eosinophil"
nodes2[nodes2[,1]=="EOSBALD positive",1]<-"Eosinophil (+)"
nodes2[nodes2[,1]=="EOSBALD negative",1]<-"Eosinophil (-)"

nodes2[nodes2[,1]=="Eosinophil positive",1]<-"Eosinophil (+)"
nodes2[nodes2[,1]=="Eosinophil negative",1]<-"Eosinophil (-)"

nodes2[nodes2[,1]=="Neutrophil positive",1]<-"Neutrophil (+)"
nodes2[nodes2[,1]=="Neutrophil negative",1]<-"Neutrophil (-)"


nodes2[nodes2[,1]=="Total Cell Counts positive",1]<-"Total Cell Counts (+)"
nodes2[nodes2[,1]=="Total Cell Counts negative",1]<-"Total Cell Counts (-)"

#nodes2[nodes2[,1]=="Hilar Lymphadenopathy yes no",1]<-"Hilar Lymphadenopathy"
nodes2[nodes2[,1]=="Hilar Lymphadenopathy positive",1]<-"Hilar Lymphadenopathy (+)"
nodes2[nodes2[,1]=="Hilar Lymphadenopathy negative",1]<-"Hilar Lymphadenopathy (-)"


nodes2[nodes2[,1]=="FEV1PRED",1]<-"FEV1% PRED"
nodes2[nodes2[,1]=="FEV1PRED positive",1]<-"FEV1% PRED (+)"
nodes2[nodes2[,1]=="FEV1PRED negative",1]<-"FEV1% PRED (-)"

nodes2[nodes2[,1]=="FVC positive",1]<-"FVC (+)"
nodes2[nodes2[,1]=="FVC negative",1]<-"FVC (-)"

nodes2[nodes2[,1]=="DLCO positive",1]<-"DLCO (+)"
nodes2[nodes2[,1]=="DLCO negative",1]<-"DLCO (-)"

nodes2[nodes2[,1]=="FEV1 positive",1]<-"FEV1 (+)"
nodes2[nodes2[,1]=="FEV1 negative",1]<-"FEV1 (-)"

nodes2[nodes2[,1]=="FVC% PRED positive",1]<-"FVC% PRED (+)"
nodes2[nodes2[,1]=="FVC% PRED negative",1]<-"FVC% PRED (-)"

nodes2[nodes2[,1]=="DLCO% PRED positive",1]<-"DLCO% PRED (+)"
nodes2[nodes2[,1]=="DLCO% PRED negative",1]<-"DLCO% PRED (-)"

nodes2[nodes2[,1]=="FEV1% PRED positive",1]<-"FEV1% PRED (+)"
nodes2[nodes2[,1]=="FEV1% PRED negative",1]<-"FEV1% PRED (-)"

nodes2[nodes2[,1]=="RACE",1]<-"Race"
nodes2[nodes2[,1]=="RACE positive",1]<-"Race (+)"
nodes2[nodes2[,1]=="RACE negative",1]<-"Race (-)"

nodes2[nodes2[,1]=="Gender positive",1]<-"Gender (+)"
nodes2[nodes2[,1]=="Gender negative",1]<-"Gender (-)"


nodes2[nodes2[,1]=="SCADDING positive",1]<-"SCADDING (+)"
nodes2[nodes2[,1]=="SCADDING negative",1]<-"SCADDING (-)"


nodes2[nodes2[,1]=="Bronchial Wall Thickening positive",1]<-"Bronchial Wall Thickening (+)"
nodes2[nodes2[,1]=="Bronchial Wall Thickening negative",1]<-"Bronchial Wall Thickening (-)"

nodes2[nodes2[,1]=="Reticular Abnormality positive",1]<-"Reticular Abnormality (+)"
nodes2[nodes2[,1]=="Reticular Abnormality negative",1]<-"Reticular Abnormality (-)"

nodes2[nodes2[,1]=="Bronchiectasis Severity positive",1]<-"Bronchiectasis Severity (+)"
nodes2[nodes2[,1]=="Bronchiectasis Severity negative",1]<-"Bronchiectasis Severity (-)"

nodes2[nodes2[,1]=="Ground Glass positive",1]<-"Ground Glass (+)"
nodes2[nodes2[,1]=="Ground Glass negative",1]<-"Ground Glass (-)"

nodes2[nodes2[,1]=="Race positive",1]<-"Race (+)"
nodes2[nodes2[,1]=="Race negative",1]<-"Race (-)"

nodes2[nodes2[,1]=="Traction Bronchiestasis positive",1]<-"Traction Bronchiestasis (+)"
nodes2[nodes2[,1]=="Traction Bronchiestasis negative",1]<-"Traction Bronchiestasis (-)"


nodes2[nodes2[,1]=="Honeycombing positive",1]<-"Honeycombing (+)"
nodes2[nodes2[,1]=="Honeycombing negative",1]<-"Honeycombing (-)"


nodes2[nodes2[,1]=="Honeycombing positive",1]<-"Honeycombing (+)"
nodes2[nodes2[,1]=="Honeycombing negative",1]<-"Honeycombing (-)"

nodes2[nodes2[,1]=="Mediastinal Lymphadenopathy positive",1]<-"Mediastinal Lymphadenopathy (+)"
nodes2[nodes2[,1]=="Mediastinal Lymphadenopathy negative",1]<-"Mediastinal Lymphadenopathy (-)"

# remove nodes that do not have any associated pathways
sn<-sankeyNetwork(Links = edges2, Nodes = nodes2, Source = "source", Target = "target", Value = "value",NodeID="name",fontSize=14,sinksRight = F,nodePadding =12,height=1000,width=800,NodeGroup="ngroup")

sn <- onRender(
  sn,
  '
function(el, x) {
    d3.selectAll(".node text").attr("text-anchor", "begin").attr("x", 30);
  }
  '
)
sn

# save the network into an eps file
saveNetwork(network=sn,file = file.path(output.dir,paste("Figure2b_supervised_sankey_fdr0.05_top",top.pathway.num,"_mingenum",min.gene.num,"_withFEV1FVCratio.html",sep="")),selfcontained=F)
#widget2png(sn, file.path(output.dir,"Figure2b_supervised_sankey_fdr0.05.png"))

```




### Figure 2c. different features of the disease
Third, we show that the bronchial wall thickening, reticular abnormality and the hilar lymphadenopathy are three different features of the disease.

```{r data_loading}
# We first generate a heatmap showing the expression pattern of genes associated with these three features
data.dir<-file.path(work.dir,"scratch/Antun/SUPERvisedA/Results_summary2")
fpkm.filepath<-file.path(work.dir,"scratch/GRADS/SARC_results/Results_summary_BAL_hg38/FPKM_matrix_hg38_Corrected_allGenes_BAL209_withAnnot.txt")
clinic.filepath<-file.path(work.dir,"scratch/GRADS/SARC_results/Data/SARC_combine_clinical_data20170814.txt")
masterkey.filepath<-file.path(work.dir,"scratch/GRADS/SARC_results/Data/GRADS Master Progress Key 6-13-16.xlsx")


############
# 1. load in the fpkm matrix
fpkm.matrix<-read.table(fpkm.filepath,sep="\t",header=T,check.names=F,stringsAsFactors = FALSE)
rownames(fpkm.matrix)<-paste(fpkm.matrix[,"tracking_id"],"_",fpkm.matrix[,"tss_id"],sep="")
fpkm.matrix<-fpkm.matrix[,10:ncol(fpkm.matrix)]

############
# 2. load in the master key file to map the KIT IDs to GRADS IDs
mkey<-read.xls(masterkey.filepath,sheet=3,skip=2,header=T,check.names=F,stringsAsFactors = F)
mkey <- mkey[,1:5]
# remove the GRADS IDs that do not have the FPKMs
temp.names<-colnames(fpkm.matrix)
temp.names<-unname(sapply(temp.names,my.element.extract,splitchar="_",index=1))
matched.list <- mkey[mkey[,4]%in%temp.names,]
matched.list$`Phenotype/Genotype` <- as.factor(matched.list$`Phenotype/Genotype`)
# generate a matrix that match the columns of fpkm.matrix
sample.matrix<-matched.list
rownames(sample.matrix)<-sample.matrix[,4]
sample.matrix<-sample.matrix[temp.names,]


############
# 2. load in the clinical data
clinic.matrix=read.table(clinic.filepath,header=T,sep="\t",check.names=F,stringsAsFactors=F,quote = "")
rownames(clinic.matrix)<-as.matrix(clinic.matrix)[,1] # GRADS ID
clinic.matrix<-clinic.matrix[sample.matrix[,1],]


############
# 3. load in the association test results
final.list<-list()
# load in the gene lists
var.names<-c("AGE","GENDER","RACE",
             "FVCPRED","BDFVC","FEV1PRED","BDFEV1","PREDDLCO","BDDLCO",
             "Bronchial_Wall_Thickening","Bronchiectasis_severity","Ground_Glass","Honeycombing","Reticular_Abnormality","Traction_Bronchiectasis","Med_Lymphadenopathy_yes_no","Hilar_Lymphadenopathy_yes_no","SCADDING","BALCELL","ALVBALD","EOSBALD","LYMBALD","NEUBALD")
var.annot<-c("Age","Gender","Race",
             "FVC% PRED","FVC","FEV1% PRED","FEV1","DLCO% PRED","DLCO",
             "Bronchial Wall Thickening","Bronchiectasis Severity","Ground Glass","Honeycombing","Reticular Abnormality","Traction Bronchiestasis","Mediastinal Lymphadenopathy","Hilar Lymphadenopathy","SCADDING","Total Cell Counts","Macrophage","Eosinophil","Lymphocyte","Neutrophil")
var.cat<-c(rep("Demographics",3),
           rep("PFTs",6),
           rep("CTVariables",8),
           "Disease",
           rep("CellDifferentials",5))
for(i in 1:length(var.names)){
file.names<-list.files(file.path(data.dir,paste(var.cat[i],"_correlated_genes",sep="")))
data.filepath<-file.path(data.dir,paste(var.cat[i],"_correlated_genes",sep=""),file.names[grep(paste(var.names[i],"_allgenes_BAL.txt",sep=""),file.names)])
if(length(data.filepath)>1){
  # For race and gender, there are two results: spearman and wilcox. We use the results by the Wilcoxon Rank Sum test.
  data.filepath<-data.filepath[2]
}
#data.filepath<-file.path(data.dir,paste(var.cat[i],"_correlated_genes",sep=""),paste("correlation_spearman_",var.names[i],"_allgenes_BAL.txt",sep=""))
temp<-read.table(data.filepath,sep = "\t",header=T,check.names=F,row.names=1)
#  temp<-temp[!is.na(temp[,"fdr"]) & temp[,"fdr"]<0.05,]
final.list[[i]]<-temp
}
names(final.list)<-var.annot
```

```{r features_pca}
#####################
# Do a PCA analysis of the three clinical features
# Bronchial Wall Thickening: 0=none, 1=mild, 2=moderate, 3=severe
# Hilar Lymphadenopathy: 0=none, 1=yes
# Reticular Abnormality: 0=no, 1=mild,2=moderate,3=severe
library("factoextra")
library("FactoMineR")
data.matrix<-clinic.matrix[,c("Bronchial_Wall_Thickening","Hilar_Lymphadenopathy","Reticular_Abnormality")]
data.matrix[!is.na(data.matrix[,2]) & data.matrix[,2]>0,2]<-1
temp.sum<-apply(is.na(data.matrix),1,sum)
data.matrix<-data.matrix[temp.sum==0,]

res.pca <- PCA(data.matrix, graph = FALSE)

cat("Here's the percentage of variation explained by each component:\n")
res.pca$eig
fviz_screeplot(res.pca, ncp=3)

cat("Here's the contribution of each feature to the components:\n")
res.pca$var$coord
```

```{r}
##############################
# get the gene lists for the three features
genes.bw<-final.list$"Bronchial Wall Thickening"
temp1<- (!is.na(genes.bw[,"fdr"])) & genes.bw[,"fdr"]<=0.05
genes.bw<-genes.bw[!is.na(genes.bw[,"fdr"]) & genes.bw[,"fdr"]<=0.05,]
genes.bw<-rownames(genes.bw)

genes.hl<-final.list$"Hilar Lymphadenopathy"
temp2<-(!is.na(genes.hl[,"fdr"])) & genes.hl[,"fdr"]<=0.05
genes.hl<-genes.hl[!is.na(genes.hl[,"fdr"]) & genes.hl[,"fdr"]<=0.05,]
genes.hl<-rownames(genes.hl)

genes.ra<-final.list$"Reticular Abnormality"
temp3<-(!is.na(genes.ra[,"fdr"])) & genes.ra[,"fdr"]<=0.05
genes.ra<-genes.ra[!is.na(genes.ra[,"fdr"]) & genes.ra[,"fdr"]<=0.05,]
genes.ra<-rownames(genes.ra)

genes.bw.only<-setdiff(genes.bw,c(genes.hl,genes.ra))
genes.hl.only<-setdiff(genes.hl,c(genes.bw,genes.ra))
genes.ra.only<-setdiff(genes.ra,c(genes.bw,genes.hl))

# generate the Venn Diagram to show the overlaps between these three gene lists
library(limma)
temp<-cbind(temp1,temp2,temp3)
colnames(temp)<-c("BWT","HL","RA")
a<-vennCounts(temp)
vennDiagram(a)

############################################################
# generate the heatmap using the feature specific genes
library("devtools")
library("gplots")
#Load latest version of heatmap.3 function
source_url("https://raw.githubusercontent.com/obigriffith/biostar-tutorials/master/Heatmaps/heatmap.3.R")

bw.col<-rep("white",ncol(clinic.matrix))
bw.col[!is.na(clinic.matrix[,"Bronchial_Wall_Thickening"]) & clinic.matrix[,"Bronchial_Wall_Thickening"]==0]<-colorpanel(n=4,low="purple",mid="grey",high="yellow")[1]
bw.col[!is.na(clinic.matrix[,"Bronchial_Wall_Thickening"]) & clinic.matrix[,"Bronchial_Wall_Thickening"]==1]<-colorpanel(n=4,low="purple",mid="grey",high="yellow")[2]
bw.col[!is.na(clinic.matrix[,"Bronchial_Wall_Thickening"]) & clinic.matrix[,"Bronchial_Wall_Thickening"]==2]<-colorpanel(n=4,low="purple",mid="grey",high="yellow")[3]
bw.col[!is.na(clinic.matrix[,"Bronchial_Wall_Thickening"]) & clinic.matrix[,"Bronchial_Wall_Thickening"]==3]<-colorpanel(n=4,low="purple",mid="grey",high="yellow")[4]

hl.col<-rep("white",ncol(clinic.matrix))
hl.col[!is.na(clinic.matrix[,"Hilar_Lymphadenopathy"]) & clinic.matrix[,"Hilar_Lymphadenopathy"]==0]<-"black"
hl.col[!is.na(clinic.matrix[,"Hilar_Lymphadenopathy"]) & clinic.matrix[,"Hilar_Lymphadenopathy"]!=0]<-"yellow"

ra.col<-rep("white",ncol(clinic.matrix))
ra.col[!is.na(clinic.matrix[,"Reticular_Abnormality"]) & clinic.matrix[,"Reticular_Abnormality"]==0]<-"black"
ra.col[!is.na(clinic.matrix[,"Reticular_Abnormality"]) & clinic.matrix[,"Reticular_Abnormality"]!=0]<-"yellow"

clab=cbind(bw.col,hl.col,ra.col)
colnames(clab)=c("Bronchial Wall Thickening","Hilar Lymphadenopathy","Reticular Abnormality")


fpkm.bw<-fpkm.matrix[genes.bw.only,]
fpkm.hl<-fpkm.matrix[genes.hl.only,]
fpkm.ra<-fpkm.matrix[genes.ra.only,]

# calculate the eigengenes of these three features
eigengene.bw<-svd(fpkm.bw)$v
eigengene.hl<-svd(fpkm.hl)$v
eigengene.ra<-svd(fpkm.ra)$v

# calculate the correlation between the eigen genes
cor.matrix.bw.hl<-matrix(NA,nrow=ncol(eigengene.bw),ncol=ncol(eigengene.hl))
for(i in 1:ncol(eigengene.bw)){
  for(j in 1:ncol(eigengene.hl)){
    cor.matrix.bw.hl[i,j]<-cor.test(eigengene.bw[,i],eigengene.hl[,j],use="complete")$p.value
  }
}

cor.matrix.hl.ra<-matrix(NA,nrow=ncol(eigengene.hl),ncol=ncol(eigengene.ra))
for(i in 1:ncol(eigengene.hl)){
  for(j in 1:ncol(eigengene.ra)){
    cor.matrix.hl.ra[i,j]<-cor.test(eigengene.hl[,i],eigengene.ra[,j],use="complete")$p.value
  }
}


cor.matrix.bw.ra<-matrix(NA,nrow=ncol(eigengene.bw),ncol=ncol(eigengene.ra))
for(i in 1:ncol(eigengene.bw)){
  for(j in 1:ncol(eigengene.ra)){
    cor.matrix.bw.ra[i,j]<-cor.test(eigengene.bw[,i],eigengene.ra[,j],use="complete")$p.value
  }
}

get_lower_tri<-function(cormat){
  cormat[upper.tri(cormat)] <- NA
  return(cormat)
}
# Get upper triangle of the correlation matrix
get_upper_tri <- function(cormat){
  cormat[lower.tri(cormat)]<- NA
  return(cormat)
}


# draw the overlap matrix using ggplot2
#c1<-signif(overlap.n.matrix, 2)
c1<-signif(cor.matrix.bw.hl[1:50,1:50],2)
c2<-cor.matrix.bw.hl[1:50,1:50]
#c2<-signif(overlap.p.matrix, 2)
#diag(c2)<-rep(0,length(diag(c2)))
upper_tri <- get_upper_tri(c1)
upper_tri_p<-get_upper_tri(c2)
melted_cormat_n <- melt(upper_tri, na.rm = TRUE)
melted_cormat_p<-melt(upper_tri_p, na.rm=TRUE)

melted_cormat<-melted_cormat_n
melted_cormat[,3]<-melted_cormat_p[,3]
melted_cormat<-cbind(melted_cormat,melted_cormat_n[,3])
colnames(melted_cormat)[4]<-"gnum"
melted_cormat[melted_cormat[,"value"]>=0.05,"value"]<-NA

g<-ggplot(data = melted_cormat, aes(Var2, Var1, fill = value))+
  geom_tile(color = "white")+
  scale_fill_gradient(low = "yellow", high = "white", na.value="lightgray", space = "Lab", name="correlation test p value") +
  #scale_fill_gradientn(colours = c("red", "white"),values=scales::rescale(c(0, 0.05, 1)), na.value="white", space = "Lab", name="# of overlapping genes") +
  theme_minimal()+ 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, size = 12, hjust = 1))+
  coord_fixed()
#melted_cormat[,3]<-melted_cormat_n[,3]
#g<-g+geom_text(aes(Var2, Var1, label = gnum), color = "black", size = 4)
print(g)

output.subdir<-file.path(work.dir,"scratch/GRADS/SARC_results/Results_summary_BAL_hg38/writings")
ggsave(filename = "Figure2c_cormatrixBWT_HL_top50.pdf",path=output.subdir,width=12,height=12,device="pdf")


# heatmap of the bronchial wall thickening only genes
prob_matrix<-fpkm.bw
rlab=rep("lightgreen",nrow(fpkm.bw))
rlab<-t(as.matrix(rlab))
rownames(rlab)<-"Bronchial Wall Thickening"

mydist=function(c) {dist(c,method="euclidian")}
myclust=function(c) {hclust(c,method="average")}

main_title="Bronchial Wall Thickening Only Genes"
par(cex.main=1)

prob_matrix<-my.normalize(as.matrix(prob_matrix))

output.filepath<-file.path(work.dir,"scratch/GRADS/SARC_results/Results_summary_BAL_hg38/writings/Figure2c_BWT.pdf")
pdf(output.filepath,width=12,height=6)
temp.breaks<-seq(from=quantile(prob_matrix,prob=0.005),to=quantile(prob_matrix,prob=0.995),length=501)
heatmap.3(prob_matrix, hclustfun=myclust, distfun=mydist, na.rm = TRUE, scale="none", dendrogram="column", margins=c(6,12),
Rowv=FALSE, Colv=TRUE, ColSideColors=clab, RowSideColors=rlab, symbreaks=FALSE, key=TRUE, symkey=FALSE,
density.info="none", trace="none", main=main_title, labCol=FALSE, labRow=FALSE, cexRow=1,
ColSideColorsSize=7, RowSideColorsSize=2, KeyValueName="FPKM",col=colorpanel(500,low="purple",mid="grey",high="yellow"),breaks=temp.breaks)
legend("topright",legend=c(paste("BWT ",0:3,sep=""),"yes","no","Bronchial Wall Thickening","Hilar Lymphadenopathy","Reticular Abnormality"),
fill=c(colorpanel(n=4,low="purple",mid="grey",high="yellow"),"black","yellow","lightgreen","lightblue","salmon"), border=FALSE, bty="n", y.intersp = 0.7, cex=0.7)
dev.off()




# heatmap of all three featuers together
prob_matrix<-rbind(fpkm.bw,fpkm.hl,fpkm.ra)


rlab=c(rep("lightgreen",nrow(fpkm.bw)),rep("lightblue",nrow(fpkm.hl)),rep("salmon",nrow(fpkm.ra)))
rlab<-t(as.matrix(rlab))
rownames(rlab)<-"feature"


mydist=function(c) {dist(c,method="euclidian")}
myclust=function(c) {hclust(c,method="average")}

main_title="Drug Response Predictions"
par(cex.main=1)

prob_matrix<-my.normalize(as.matrix(prob_matrix))

output.filepath<-file.path(work.dir,"scratch/GRADS/SARC_results/Results_summary_BAL_hg38/writings/Figure2c_diseasefeatures.pdf")
pdf(output.filepath,width=12,height=6)
temp.breaks<-seq(from=quantile(prob_matrix,prob=0.005),to=quantile(prob_matrix,prob=0.995),length=501)
heatmap.3(prob_matrix, hclustfun=myclust, distfun=mydist, na.rm = TRUE, scale="none", dendrogram="column", margins=c(6,12),
Rowv=FALSE, Colv=TRUE, ColSideColors=clab, RowSideColors=rlab, symbreaks=FALSE, key=TRUE, symkey=FALSE,
density.info="none", trace="none", main=main_title, labCol=FALSE, labRow=FALSE, cexRow=1,
ColSideColorsSize=7, RowSideColorsSize=2, KeyValueName="FPKM",col=colorpanel(500,low="purple",mid="grey",high="yellow"),breaks=temp.breaks)
legend("topright",legend=c(paste("BWT ",0:3,sep=""),"yes","no","Bronchial Wall Thickening","Hilar Lymphadenopathy","Reticular Abnormality"),
fill=c(colorpanel(n=4,low="purple",mid="grey",high="yellow"),"black","yellow","lightgreen","lightblue","salmon"), border=FALSE, bty="n", y.intersp = 0.7, cex=0.7)
dev.off()


#temp.heatmap<-heatmap.2(as.matrix(rbind(fpkm.bw,fpkm.hl,fpkm.ra)),breaks=temp.breaks,scale="none",dendrogram="row",Rowv=T,Colv=F,col=colorpanel(500,low="purple",mid="grey",high="yellow"),cexCol=1,trace="none",density.info="none",key=TRUE,symkey=FALSE,keysize=0.8,na.color="grey",ColSideColors=temp.color,labRow=NULL,labCol="")
```


## Unsupervised Analysis
The WGCNA R package was applied to this data set to identify gene modules. The correlation between the clinical traits and each gene module was also assessed and visualized. We also regress out the effects of cell differentials and redo the WGCNA analysis. The comparison between the WGCNA results of the unadjusted data and the cell differential adjusted data are as follows:
1. [Comparing WGCNA of unadjusted and cell differentials adjusted data](rmarkdown/BAL_WGCNAcompare.nb.html)
2. [Comparing WGCNA of the unadjusted and cell differentials adjusted data but only using the cell differential associated gene modules from the unadjusted data](rmarkdown/BAL_WGCNAcompare_selected.nb.html)


### Figure 3a. Heatmap of module trait correlation
```{r}
# These codes load in the WGCNA results of the unadjusted data to extract some of the information
# We also modify the heatmap by 
# 1. extracting only the chosen modules.
# 2. removing the traits that are meaningless including lymphadenopathy_numeric, lymphadenopathy_oneside/bi and lymphadenopathy
rm(list=setdiff(ls(),c("figure_nums","figure_nums_3","table_nums","table_nums_3","work.dir")))
data.filepath<-"/home/yanxiting/driver_Grace/scratch/GRADS/SARC_results/Results_summary_BAL_hg38/WGCNA_BAL/CV0_compare/wgcna_cutheights.RData"
source("/home/yanxiting/driver_Grace/Rprogram/my_functions.R")
load(data.filepath)
```

```{r figure3a_modulecor_heatmap,fig.width=10,fig.height=5,cache=FALSE,fig.cap="Figure 3a. Heatmap showing the corrleation between the 5 chosen gene modules identified by WGCNA with the clinical traits."}
output.subdir<-"/home/yanxiting/driver_Grace/scratch/GRADS/SARC_results/Results_summary_BAL_hg38/writings"

i<-7
# in the clinical trait data, remove the redundance traits for lymphaodenopathy and add FEV1/FVC ratio, also try to add the disease severity measure by the phenotypic groups within the treated and the untreated group
net<-net.list[[i]]
my.datExpr = datExpr[setdiff(rownames(datExpr),unlist(outlier.list[1:i])), ]

#rename the columns (remove all after _ in sample name)
SarcaleSamples = rownames(my.datExpr)
sampleRows =match(SarcaleSamples,rownames(sample.matrix));
sample.matrix.new=sample.matrix[sampleRows,]
traitRows = match(sample.matrix.new[,1],colnames(datTraits));
my.datTraits = datTraits[,traitRows];

# add the FEV1/FVC ratio
temp.ratio<-my.datTraits["BDFEV1",]/my.datTraits["BDFVC",]
my.datTraits<-rbind(my.datTraits[1:9,],temp.ratio,my.datTraits[10:nrow(my.datTraits),])
rownames(my.datTraits)[10]<-"FEV1/FVC ratio"

# get the MEs
nGenes = ncol(my.datExpr);
nSamples = nrow(my.datExpr);
# Recalculate MEs with color labels
MEs0 = moduleEigengenes(my.datExpr, labels2colors(net$colors))$eigengenes
MEs = orderMEs(MEs0)

geneModuleMembership = as.data.frame(cor(my.datExpr, MEs, use = "p"));
MMPvalue = as.data.frame(corPvalueStudent(as.matrix(geneModuleMembership), nSamples));

geneTraitSignificance = as.data.frame(cor(my.datExpr, t(my.datTraits), use = "p"));
GSPvalue = as.data.frame(corPvalueStudent(as.matrix(geneTraitSignificance), nSamples));

# add the membership results for FEV1/FVC ratio as well but the results will be saved as a different file under the same folder as the previous results without FEV1/FVC ratio
modNames=substring(names(MEs), 3)
output.subdir<-file.path("/home/yanxiting/driver_Grace/scratch/GRADS/SARC_results/Results_summary_BAL_hg38/WGCNA_BAL/CV0_compare",paste("ModuleMembership_GSPvalue_outliers_",i,"_withFEV1FVCratio",sep=""))
if(file.exists(output.subdir)==F){
dir.create(output.subdir)	
}
for(k in 1:length(modNames)){
output.filepath<-file.path(output.subdir,paste("Module_",modNames[k],"_membership_gsig.txt",sep=""))
ps.names<-rownames(my.data.matrix)
gene.names<-unname(sapply(ps.names,my.element.remove,splitchar="_",index=-1))
temp.color.desig<-labels2colors(net$colors)
cmd.out<-cbind(ps.names[temp.color.desig==modNames[k]],gene.names[temp.color.desig==modNames[k]],geneModuleMembership[temp.color.desig==modNames[k],colnames(geneModuleMembership)==paste("ME",modNames[k],sep="")],MMPvalue[temp.color.desig==modNames[k],colnames(MMPvalue)==paste("ME",modNames[k],sep="")],geneTraitSignificance[temp.color.desig==modNames[k],],GSPvalue[temp.color.desig==modNames[k],])
colnames(cmd.out)<-c("ps_names","gene_names","module_membership","module_membership_pvalue",paste("gs_",rownames(my.datTraits),sep=""),paste("gs_",rownames(my.datTraits),"_pvalue",sep=""))
write.table(cmd.out,file=output.filepath,append=F,sep="\t",row.names=F,col.names=T,quote=F)
}

# remove the redundant traits for lymphadenopathy

my.datTraits<-my.datTraits[!rownames(my.datTraits)%in%c("Med_Lymphadenopathy",
                              "Med_Lymphadenopathy_oneside/bi",
                              "Med_Lymphadenopathy_numeric",
                              "Hilar_Lymphadenopathy",
                              "Hilar_Lymphadenopathy_oneside/bi",
                              "Hilar_Lymphadenopathy_numeric"),]

rownames(my.datTraits)[rownames(my.datTraits)=="BDFEV1"]<-"FEV1"
rownames(my.datTraits)[rownames(my.datTraits)=="FEV1PRED"]<-"FEV1% predited"
rownames(my.datTraits)[rownames(my.datTraits)=="BDFVC"]<-"FVC"
rownames(my.datTraits)[rownames(my.datTraits)=="FVCPRED"]<-"FVC% predicted"
rownames(my.datTraits)[rownames(my.datTraits)=="BDDLCO"]<-"DLCO"
rownames(my.datTraits)[rownames(my.datTraits)=="PREDDLCO"]<-"DLCO% predicted"

rownames(my.datTraits)[rownames(my.datTraits)=="Med_Lymphadenopathy_yes/no"]<-"Med_Lymphadenopathy"
rownames(my.datTraits)[rownames(my.datTraits)=="Hilar_Lymphadenopathy_yes/no"]<-"Hilar_Lymphadenopathy"


rownames(my.datTraits)[rownames(my.datTraits)=="ALVBALD"]<-"MACROPHAGE"
rownames(my.datTraits)[rownames(my.datTraits)=="EOSBALD"]<-"EOSINOPHIL"
rownames(my.datTraits)[rownames(my.datTraits)=="LYMBALD"]<-"LYMPHOCYTE"
rownames(my.datTraits)[rownames(my.datTraits)=="NEUBALD"]<-"NEUTROPHIL"
rownames(my.datTraits)[rownames(my.datTraits)=="BALLCELL"]<-"TOTAL CELL COUNTS"
rownames(my.datTraits)[rownames(my.datTraits)=="TREATMENT"]<-"Treatment_Overall"
rownames(my.datTraits)[rownames(my.datTraits)=="TREATMENT_STAGEII-III"]<-"Treatment_Stage II-III"
rownames(my.datTraits)[rownames(my.datTraits)=="TREATMENT_STAGEIV"]<-"Treatment_Stage IV"


nGenes = ncol(my.datExpr)
nSamples = nrow(my.datExpr)
i<-7
power.selected<-power.vect[i]
net<-net.list[[i]]
mergedColors = labels2colors(net$colors)

# generate our own names and the sizes for the modules
my.temp<-table(mergedColors)
my.temp<-sort(my.temp,decreasing=T)
my.names<-paste("Module ",1:length(my.temp),sep="")
modulenames.def<-cbind(names(my.temp),my.names)
modulenames.def<-as.data.frame(modulenames.def)
modulenames.def<-cbind(modulenames.def,as.numeric(my.temp))
rownames(modulenames.def)<-modulenames.def[,1]
colnames(modulenames.def)<-c("wgcna_names","my_names","module_size")

MEList.test<-moduleEigengenes(my.datExpr,colors=mergedColors)
MEs.test<-MEList.test$eigengenes
MEDiss.test<-1-cor(MEs.test)
METree.test=hclust(as.dist(MEDiss.test),method="average")

nGenes = ncol(my.datExpr);
nSamples = nrow(my.datExpr);
# Recalculate MEs with color labels
MEs0 = moduleEigengenes(my.datExpr, labels2colors(net$colors))$eigengenes
MEs = orderMEs(MEs0)

# only show the correlation for chosen gene modules
MEs<-MEs[,colnames(MEs)%in%c("MEturquoise","MEbrown4","MEbrown","MEviolet","MEgrey60")]
my.ME.names<-modulenames.def[sapply(colnames(MEs),my.element.extract,splitchar="ME",index=-1),"my_names"]
my.ME.sizes<-modulenames.def[sapply(colnames(MEs),my.element.extract,splitchar="ME",index=-1),"module_size"]

cp = bicorAndPvalue(MEs, t(my.datTraits))
moduleTraitCor = cp$bicor;
moduleTraitPvalue = cp$p;


textMatrix = paste(signif(moduleTraitCor, 2), "\n(",signif(moduleTraitPvalue, 1), ")", sep = "");
dim(textMatrix) = dim(moduleTraitCor)

# Display the correlation values within a heatmap plot
output.subdir<-"/home/yanxiting/driver_Grace/scratch/GRADS/SARC_results/Results_summary_BAL_hg38/writings"

temp.moduleTraitPvalue<-moduleTraitPvalue
temp.moduleTraitCor<-moduleTraitCor
temp.moduleTraitCor[temp.moduleTraitPvalue>0.05]<-NA

output.filepath<-file.path(output.subdir,paste("Figure3a_heatmap_trait_correlation_modules_outliers_",i,"_modulenames_def.txt",sep=""))
write.table(modulenames.def,file=output.filepath,append=F,sep="\t",col.names=T,row.names=F,quote=F)

output.filepath<-file.path(output.subdir,paste("Figure3a_heatmap_trait_correlation_modules_outliers_",i,"_clean.eps",sep=""))
postscript(output.filepath,width=15,height=4,paper="special")
par(mar = c(8, 8, 3, 1));
labeledHeatmap(Matrix = temp.moduleTraitCor,
xLabels = rownames(my.datTraits),
#yLabels = names(MEs),
yLabels=paste(my.ME.names," (",my.ME.sizes,")",sep=""),
#ySymbols = names(MEs),
ySymbols=paste(my.ME.names," (",my.ME.sizes,")",sep=""),
colorLabels = FALSE,
colors = blueWhiteRed(50),
textMatrix = textMatrix,
#cex.text=0.8,
setStdMargins = FALSE,
cex.text = 0.5,
zlim = c(-1,1),
main = paste("Module-trait relationships"),
naColor="white")
#savePlot(output.filepath,type="jpeg")
dev.off()

par(mar = c(8, 8, 3, 1));
labeledHeatmap(Matrix = temp.moduleTraitCor,
xLabels = rownames(my.datTraits),
#yLabels = names(MEs),
yLabels=paste(my.ME.names," (",my.ME.sizes,")",sep=""),
#ySymbols = names(MEs),
ySymbols=paste(my.ME.names," (",my.ME.sizes,")",sep=""),
colorLabels = FALSE,
colors = blueWhiteRed(50),
textMatrix = textMatrix,
#cex.text=0.8,
setStdMargins = FALSE,
cex.text = 0.5,
zlim = c(-1,1),
main = paste("Module-trait relationships"),
naColor="white")

```

Notes: we have changed the module names but the 5 modules shown in the plot are

```{r results='hold'}
{
cat("Module 1 was turquoise Module.\n")
cat("Module 4 was brown Module.\n")
cat("Module 18 was grey60 Module.\n")
cat("Module 33 was violet Module.\n")
cat("Module 47 was brown4 Module.\n")
}
```



### Figure 3b. Sankey plot of the WGCNA results
To show the functional analysis of the genes in each chosen module, we generate the sankey plot.

```{r}
# sankey plot showing the functional analysis results on the identified gene modules.
source("/home/yanxiting/driver_Grace/Rprogram/my_functions.R")
data.dir<-"/home/yanxiting/driver_Grace/scratch/GRADS/SARC_results/Results_summary_BAL_hg38/WGCNA_BAL/CV0_compare/ModuleMembership_GSPvalue_outliers_7_withFEV1FVCratio"
metacore.dir<-"/home/yanxiting/driver_Grace/scratch/GRADS/SARC_results/Results_summary_BAL_hg38/WGCNA_BAL/CV0_compare/ModuleMembership_GSPvalue_outliers_7"
output.dir<-"/home/yanxiting/driver_Grace/scratch/GRADS/SARC_results/Results_summary_BAL_hg38/writings"
module.choose<-c("turquoise","brown","brown4","violet","grey60")

top.pathway.num<- 5
min.gene.num<-3
log.filepath<-file.path(output.dir,paste("Figure3b_unsupervised_sankey_fdr0.05_top",top.pathway.num,"_mingenum",min.gene.num,".html.log",sep=""))
MMList<-list.files(data.dir)

MMList<-MMList[sapply(MMList,my.element.extract,splitchar="_",index=2)%in%module.choose]
MMList<-MMList[grep("gsig.txt",MMList)]

ModColList<-c()
Alledges<-c()
Allnodes<-c()

# change the column names of the temp.moduleTraitCor to match those in the Module_*_membership_gsig.txt
colnames(temp.moduleTraitCor)[colnames(temp.moduleTraitCor)=="FEV1"]<-"BDFEV1"
colnames(temp.moduleTraitCor)[colnames(temp.moduleTraitCor)=="FEV1% predited"]<-"FEV1PRED"
colnames(temp.moduleTraitCor)[colnames(temp.moduleTraitCor)=="FVC"]<-"BDFVC"
colnames(temp.moduleTraitCor)[colnames(temp.moduleTraitCor)=="FVC% predicted"]<-"FVCPRED"
colnames(temp.moduleTraitCor)[colnames(temp.moduleTraitCor)=="DLCO"]<-"BDDLCO"
colnames(temp.moduleTraitCor)[colnames(temp.moduleTraitCor)=="DLCO% predicted"]<-"PREDDLCO"
colnames(temp.moduleTraitCor)[colnames(temp.moduleTraitCor)=="Med_Lymphadenopathy"]<-"Med_Lymphadenopathy_yes/no"
colnames(temp.moduleTraitCor)[colnames(temp.moduleTraitCor)=="Hilar_Lymphadenopathy"]<-"Hilar_Lymphadenopathy_yes/no"
colnames(temp.moduleTraitCor)[colnames(temp.moduleTraitCor)=="MACROPHAGE"]<-"ALVBALD"
colnames(temp.moduleTraitCor)[colnames(temp.moduleTraitCor)=="EOSINOPHIL"]<-"EOSBALD"
colnames(temp.moduleTraitCor)[colnames(temp.moduleTraitCor)=="LYMPHOCYTE"]<-"LYMBALD"
colnames(temp.moduleTraitCor)[colnames(temp.moduleTraitCor)=="NEUTROPHIL"]<-"NEUBALD"
colnames(temp.moduleTraitCor)[colnames(temp.moduleTraitCor)=="TOTAL CELL COUNTS"]<-"BALLCELL"
colnames(temp.moduleTraitCor)[colnames(temp.moduleTraitCor)=="Treatment_Overall"]<-"TREATMENT"
colnames(temp.moduleTraitCor)[colnames(temp.moduleTraitCor)=="Treatment_Stage II-III"]<-"TREATMENT_STAGEII-III"
colnames(temp.moduleTraitCor)[colnames(temp.moduleTraitCor)=="Treatment_Stage IV"]<-"TREATMENT_STAGEIV"

colnames(temp.moduleTraitPvalue)<-colnames(temp.moduleTraitCor)

node.from<-character()
node.to<-character()
edge.val<-numeric()
edge.type<-character()

for(i in 1:length(MMList)){

  Mmemb<-read.table(file.path(data.dir,MMList[i]),sep="\t",header=T,check.names=F,stringsAsFactors=F)
  trait.show<-colnames(temp.moduleTraitCor)[!is.na(temp.moduleTraitCor[paste("ME",my.element.extract(MMList[i],splitchar="_",index=2),sep=""),])]
  trait.pval<-temp.moduleTraitPvalue[paste("ME",my.element.extract(MMList[i],splitchar="_",index=2),sep=""),trait.show]
  module.size<-nrow(Mmemb)
  count_p<- (-log(trait.pval,base=2))*module.size/sum((-log(trait.pval,base=2)))
  
  #create the edges table
  ModCol<-unname(sapply(MMList[i],my.element.extract,splitchar="Module_",index=2))
  ModCol<-unname(sapply(ModCol,my.element.extract,splitchar="_",index=1))

  node.from<-c(node.from,trait.show)
  node.to<-c(node.to,rep(ModCol, times=length(trait.show)))
  edge.val<-c(edge.val,count_p)
  edge.type<-c(edge.type,rep("L1",length(count_p)))

  # add the edges for the positive and negative genes
  pos.filepath<-file.path(metacore.dir,paste("Module_",ModCol,"_membership_gsig_posgenes.txt",sep=""))
  if(file.exists(pos.filepath)==F){
    cat("WARNING: ",pos.filepath," doesn't exist!\n")
  }else{
    temp<-read.table(pos.filepath,sep="\t",header=T,check.names=F)
    pos.genenum<-nrow(temp)
  }
  
  neg.filepath<-file.path(metacore.dir,paste("Module_",ModCol,"_membership_gsig_neggenes.txt",sep=""))
  if(file.exists(neg.filepath)==F){
    cat("WARNING: ",neg.filepath," doesn't exist!\n",sep="")
  }else{
    temp<-read.table(neg.filepath,sep="\t",header=T,check.names=F)
    neg.genenum<-nrow(temp)
  }
  
  node.from<-c(node.from,ModCol)
  node.to<-c(node.to, paste(ModCol,"(+)",sep=""))
  edge.val<-c(edge.val,pos.genenum)
  edge.type<-c(edge.type,"L2")

  node.from<-c(node.from,ModCol)
  node.to<-c(node.to, paste(ModCol,"(-)",sep=""))
  edge.val<-c(edge.val,neg.genenum)
  edge.type<-c(edge.type,"L2")
  
  # add the edges for the positive metacore results
  metacore.pos.filepath<-file.path(metacore.dir,paste("Module_",ModCol,"_membership_gsig_posgenes_metacore",sep=""),"Enrichment_analysis.xls")
  if(file.exists(metacore.pos.filepath)){
    metac<-read.xls(metacore.pos.filepath,sheet=1,check.names=F,stringsAsFactors = F)
    colnames(metac)<-c("#","Maps","Total","pValue","MinFDR","ID_p-value","ID_FDR","ID_InData","ID_NetworkObjectsfromActiveData")
    
    if(nrow(metac)>=3){

      metac<-metac[3:nrow(metac),2:9]
      metac$ID_FDR<-as.numeric(metac$ID_FDR)

      #take out fdr > 0.05
      metac<-metac[metac$ID_FDR < 0.05,]
      metac<-metac[order(metac$ID_FDR),]

      #only show pathways with min.gene.num genes InData
      metac<-metac[as.numeric(metac[,"ID_InData"])>=min.gene.num,]


      # merge the pathways that are significant due to exactly the same genes
      if(nrow(metac)>0){
        temp.index<-split(1:nrow(metac),metac[,"ID_NetworkObjectsfromActiveData"])
        my.metac<-numeric()
        for(j in 1:length(temp.index)){
          if(length(temp.index[[j]])==1){
            my.metac<-rbind(my.metac,metac[temp.index[[j]],])
          }else{
            temp.matrix<-metac[temp.index[[j]],]  
            temp.result<-temp.matrix[1,]
            cat(ModCol," (+):\n",sep="",file=log.filepath,append=T)
            cat(temp.matrix[,1],sep="\n",file=log.filepath,append=T)
            cat("\n",file=log.filepath,append=T)
            #temp.result[1,1]<-paste(temp.matrix[,1],collapse="\r")
            temp.result[1,"ID_FDR"]<-min(temp.matrix[,"ID_FDR"])
            temp.result[1,"ID_p-value"]<-min(temp.matrix[,"ID_p-value"])
            my.metac<-rbind(my.metac,temp.result)
          }
        }
        metac<-my.metac
        metac<-metac[order(as.numeric(metac[,"MinFDR"]),decreasing=F),]
        fdr.threshold.pos<-max(as.numeric(metac[1:min(top.pathway.num,nrow(metac)),"MinFDR"]))
        metac<-metac[as.numeric(metac[,"MinFDR"])<=fdr.threshold.pos,]
        #metac<-metac[1:min(top.pathway.num,nrow(metac)),]
        
        if(nrow(metac)>0){
          node.from<-c(node.from,rep(paste(ModCol,"(+)",sep=""),nrow(metac)))
          node.to<-c(node.to,metac[,1])
          edge.val<-c(edge.val,metac[,7])
          edge.type<-c(edge.type,rep("L3",nrow(metac)))
        }
      }
    }

    
  }else{
    cat("WARNING: ",metacore.pos.filepath," doesn't exist!",sep="")
  }
  
  
  # add the edges for the negative metacore results
  metacore.neg.filepath<-file.path(metacore.dir,paste("Module_",ModCol,"_membership_gsig_neggenes_metacore",sep=""),"Enrichment_analysis.xls")
  if(file.exists(metacore.neg.filepath)){
    metac2<-read.xls(metacore.neg.filepath,sheet=1,check.names=F,stringsAsFactors = F)
    colnames(metac2)<-c("#","Maps","Total","pValue","MinFDR","ID_p-value","ID_FDR","ID_InData","ID_NetworkObjectsfromActiveData")
    
    if(nrow(metac2)>=3){

      metac2<-metac2[3:nrow(metac2),2:9]
      metac2$ID_FDR<-as.numeric(metac2$ID_FDR)

      #take out fdr > 0.05
      metac2<-metac2[metac2$ID_FDR < 0.05,]
      metac2<-metac2[order(metac2$ID_FDR),]

      #only show pathways with min.gene.num genes InData
      metac2<-metac2[as.numeric(metac2[,"ID_InData"])>=min.gene.num,]


      # merge the pathways that are significant due to exactly the same genes
      if(nrow(metac2)>0){
        temp.index<-split(1:nrow(metac2),metac2[,"ID_NetworkObjectsfromActiveData"])
        my.metac2<-numeric()
        for(j in 1:length(temp.index)){
          if(length(temp.index[[j]])==1){
            my.metac2<-rbind(my.metac2,metac2[temp.index[[j]],])
          }else{
            temp.matrix<-metac2[temp.index[[j]],]  
            temp.result<-temp.matrix[1,]
            cat(ModCol," (-):\n",sep="",file=log.filepath,append=T)
            cat(temp.matrix[,1],sep="\n",file=log.filepath,append=T)
            cat("\n",file=log.filepath,append=T)
            #temp.result[1,1]<-paste(temp.matrix[,1],collapse="\r")
            temp.result[1,"ID_FDR"]<-min(temp.matrix[,"ID_FDR"])
            temp.result[1,"ID_p-value"]<-min(temp.matrix[,"ID_p-value"])
            my.metac2<-rbind(my.metac2,temp.result)
          }
        }
        metac2<-my.metac2
        metac2<-metac2[order(as.numeric(metac2[,"MinFDR"]),decreasing=F),]
        fdr.threshold.neg<-max(as.numeric(metac2[1:min(top.pathway.num,nrow(metac2)),"MinFDR"]))
        metac2<-metac2[as.numeric(metac2[,"MinFDR"])<=fdr.threshold.neg,]
        #metac<-metac[1:min(top.pathway.num,nrow(metac)),]
        
        if(nrow(metac2)>0){
          node.from<-c(node.from,rep(paste(ModCol,"(-)",sep=""),nrow(metac2)))
          node.to<-c(node.to,metac2[,1])
          edge.val<-c(edge.val,metac2[,7])
          edge.type<-c(edge.type,rep("L3",nrow(metac2)))
        }
      }
    }

    
  }else{
    cat("WARNING: ",metacore.neg.filepath," doesn't exist!",sep="")
  }
}  
  
Alledges<-data.frame(node.from,node.to,edge.val,edge.type,stringsAsFactors=F)
colnames(Alledges)<-c("source","target","value","edgetype")
Alledges<-Alledges[complete.cases(Alledges), ]
write.table(Alledges,file=file.path(output.dir,paste("Figure3b_unsupervised_sankey_top",top.pathway.num,"_mingenum",min.gene.num,"_fdr0.05.txt")),append=F,row.names=F,col.names=T,sep="\t",quote=F)

cap<-paste("Figure 3b. Sankey plot showing the association of the chosen modules and the clinical traits and the function analysis results of the chosen modules for top.pathway.num=",top.pathway.num,", min.gene.num=",min.gene.num,sep="")

```

```{r figure3b_sankeyplot_unsupervised,eval=TRUE,cache=FALSE,fig.height=15,fig.width=15,fig.cap=cap}
#nodes.temp2<-unique(Alledges[Alledges[,4]=="L2",1])
#nodes.temp1<-unique(Alledges[Alledges[,4]=="L1",2])
#keep.nodes<-intersect(nodes.temp1,nodes.temp2)

#alledges<-Alledges[Alledges[,2]%in%keep.nodes | Alledges[,1]%in%keep.nodes ,1:3]
alledges<-Alledges[,1:3]

allnodes<-c(alledges$source,alledges$target)
allnodes<-allnodes[!duplicated(allnodes)]

edges2<-alledges

for(i in 1:nrow(alledges)){
  
  edges2[i,1]<-as.integer(match(alledges[i,1],allnodes))-1
  edges2[i,2]<-as.integer(match(alledges[i,2],allnodes))-1
  
}

nodes2<-as.data.frame(allnodes,stringsAsFactors=F)
colnames(nodes2)<-"name"

edges2[,1]<-as.integer(edges2[,1])
edges2[,2]<-as.integer(edges2[,2])
edges2[,3]<-as.numeric(edges2[,3])

temp.cat<-nodes2[,1]
temp.cat[grep("\\(\\+\\)",temp.cat)]<-"positive"
temp.cat[grep("\\(\\-\\)",temp.cat)]<-"negative"

nodes2<-cbind(nodes2,temp.cat)
colnames(nodes2)[2]<-"ngroup"

# change the node names
nodes2[nodes2[,1]=="BDFEV1",1]<-"FEV1"
nodes2[nodes2[,1]=="FEV1PRED",1]<-"FEV1% predicted"
nodes2[nodes2[,1]=="BDFVC",1]<-"FVC"
nodes2[nodes2[,1]=="FVCPRED",1]<-"FVC% predicted"
nodes2[nodes2[,1]=="BDDLCO",1]<-"DLCO"
nodes2[nodes2[,1]=="PREDDLCO",1]<-"DLCO% predicted"
nodes2[nodes2[,1]=="Med_Lymphadenopathy_yes/no",1]<-"Mediastinal Lymphadenopathy"
nodes2[nodes2[,1]=="Hilar_Lymphadenopathy_yes/no",1]<-"Hilar Lymphadenopathy"
nodes2[nodes2[,1]=="ALVBALD",1]<-"MACROPHAGE"
nodes2[nodes2[,1]=="EOSBALD",1]<-"EOSINOPHIL"
nodes2[nodes2[,1]=="LYMBALD",1]<-"LYMPHOCYTE"
nodes2[nodes2[,1]=="NEUBALD",1]<-"NEUTROPHIL"
nodes2[nodes2[,1]=="BALLCELL",1]<-"TOTAL CELL COUNTS"
nodes2[nodes2[,1]=="TREATMENT",1]<-"Treatment (overall)"
nodes2[nodes2[,1]=="TREATMENT_STAGEII-III",1]<-"Treatment (Stage II-III)"
nodes2[nodes2[,1]=="TREATMENT_STAGEIV",1]<-"Treatment (Stage IV)"

# remove nodes that do not have any associated pathways
sn<-sankeyNetwork(Links = edges2, Nodes = nodes2, Source = "source", Target = "target", Value = "value",NodeID="name",fontSize=14,sinksRight = F,nodePadding =12,height=800,width=1200,NodeGroup="ngroup")

sn <- onRender(
  sn,
  '
function(el, x) {
    d3.selectAll(".node text").attr("text-anchor", "begin").attr("x", 30);
  }
  '
)
sn

# save the network into an eps file
saveNetwork(network=sn,file = file.path(output.dir,paste("Figure3b_unsupervised_sankey_fdr0.05_top",top.pathway.num,"_mingenum",min.gene.num,".html",sep="")),selfcontained=F)

```

## Validation

We validated our findings by unsupervised and supervised analysis using this independent cohort of Sarcoidosis patients from Freiburgh, Germany using the following approach.

### Validating the WGCNA results
The WGCNA results were validated in the following way:

1. We conduct a K-means clustering of the patients in GRADS cohort using each chosen gene module identified by WGCNA. The K was chosen based on multiple different internal clustering criteria calculated and data visualization. Then among the patient clusters we found, we chose the two extreme clusters that have the most different expression profiles of the correponding gene module. All available clinical traits are compared between these two clusters.

2. We also cluster the patients in Freiburgh cohort by conducting a K-means clustering on their Freiburgh cohort measured expression profiles of the chosen gene modules identified in the GRADS cohort. 

The clustering was performed using codes in BAL_validation.Rmd.

#### Figure 4. Heatmap of the expression profiles of each chosen module.

```{r}
rm(list=setdiff(ls(),c("figure_nums","figure_nums_3","table_nums","table_nums_3","work.dir")))
data.filepath<-"/home/yanxiting/driver_Grace/scratch/GRADS/SARC_results/Results_summary_BAL_hg38/WGCNA_BAL/CV0_compare/wgcna_cutheights.RData"
source("/home/yanxiting/driver_Grace/Rprogram/my_functions.R")
load(data.filepath)

output.subdir<-"/home/yanxiting/driver_Grace/scratch/GRADS/SARC_results/Results_summary_BAL_hg38/writings"

i<-7
# in the clinical trait data, remove the redundance traits for lymphaodenopathy and add FEV1/FVC ratio, also try to add the disease severity measure by the phenotypic groups within the treated and the untreated group
net<-net.list[[i]]
mergedColors = labels2colors(net$colors)
my.datExpr = datExpr[setdiff(rownames(datExpr),unlist(outlier.list[1:i])), ]

#rename the columns (remove all after _ in sample name)
SarcaleSamples = rownames(my.datExpr)
sampleRows =match(SarcaleSamples,rownames(sample.matrix));
sample.matrix.new=sample.matrix[sampleRows,]
traitRows = match(sample.matrix.new[,1],colnames(datTraits));
my.datTraits = datTraits[,traitRows];

# add the FEV1/FVC ratio
temp.ratio<-my.datTraits["BDFEV1",]/my.datTraits["BDFVC",]
my.datTraits<-rbind(my.datTraits[1:9,],temp.ratio,my.datTraits[10:nrow(my.datTraits),])
rownames(my.datTraits)[10]<-"FEV1/FVC ratio"

```

We plot the K-means clustering results using the heatmap.

```{r figure4,eval=TRUE,fig.width=9.8,fig.height=7,echo=FALSE,cache=FALSE}
array.data.filepath<-"/home/yanxiting/driver_Grace/scratch/Antun/microarray_Jonas/gene_expression_genesymbol_condensed_updated.txt"
array.annot.filepath<-"/home/yanxiting/driver_Grace/scratch/Antun/microarray_Jonas/HuGene-1_0-st-v1.na35.hg19.transcript.csv"
chosen.modules<-c("Turquoise","Brown4","Brown","Violet","Grey60")
array.clustering.dir<-"/home/yanxiting/driver_Grace/scratch/Antun/microarray_Jonas/clustering"
grads.clustering.dir<-"/home/yanxiting/driver_Grace/scratch/Antun/WGCNA_clusthg38/kmeans_selected_updated"
output.dir<-"/home/yanxiting/driver_Grace/scratch/GRADS/SARC_results/Results_summary_BAL_hg38/writings"

fig.list<-list()
heights.vect<-numeric()
# load in the clustering results of the GRADS cohort
for(i in 1:length(chosen.modules)){
  modname<-tolower(chosen.modules[i])
  
  ######################################
  # 1. for the GRADS RNAseq data
  
  # load in the clustering results
  temp.names<-list.files(grads.clustering.dir)
  temp.names<-temp.names[grep("_clustering.txt",temp.names)]
  temp.names<-temp.names[grep(paste(chosen.modules[i],"_",sep=""),temp.names)]
  if(length(temp.names)>1){
    cat("There are >1 clustering files for Module ",chosen.modules[i],"\n",sep="")
    next
  }
  
  clustering.result.filepath<-file.path(grads.clustering.dir,temp.names)
  clustering.result<-read.table(clustering.result.filepath,sep="\t",header=T,check.names = FALSE)
  temp<-clustering.result[,2]
  names(temp)<-clustering.result[,1]
  clustering.result<-temp
  clustering.result<-sort(clustering.result,decreasing=F)
  
  kmeans.k<-my.element.extract(temp.names,splitchar="_",index=2)
  kmeans.k<-as.numeric(my.element.extract(kmeans.k,splitchar="",index=-1))
  
  # load in the gene names in the module
  gene.names<-colnames(my.datExpr)[mergedColors==modname]
  fpkm.mod<-my.normalize(t(my.datExpr[names(clustering.result),gene.names]))
  cat("Module ",modname," quantile (1% and 99%):",quantile(fpkm.mod,prob=c(0.01,0.99)),"\n",sep=" ")

  # within each K means cluster, put similar samples next to each other, do the same thing to genes and obtain the ordered gene names
  
   for(j in 1:kmeans.k){
    temp.matrix<-fpkm.mod[,clustering.result==j]
    temp.matrix<-temp.matrix[,hclust(dist(t(temp.matrix)))$order]
    fpkm.mod[,clustering.result==j]<-temp.matrix
  }
  
  clustering.result<-clustering.result[colnames(fpkm.mod)]

   #p0<-textGrob(panel.names[i],just="top")
   #fig.list[[length(fig.list)+1]]<-p0
  
  # get the order of the genes
  #gene.names.sorted<-rownames(fpkm.mod)[hclust(dist(fpkm.mod))$order]
  
  temp.color<-my.color.gen.cat(paste("cluster_",clustering.result,sep=""))$color.num
  temp.breaks<-seq(from=quantile(fpkm.mod,prob=0.005),to=quantile(fpkm.mod,prob=0.995),length=501)
 
  # generate the heatmap for the K-means clustering results for the GRADS cohort
  temp.heatmap<-heatmap.2(as.matrix(fpkm.mod),breaks=temp.breaks,scale="none",dendrogram="row",Rowv=T,Colv=F,col=colorpanel(500,low="purple",mid="grey",high="yellow"),cexCol=1,trace="none",density.info="none",key=TRUE,symkey=FALSE,keysize=0.8,na.color="grey",ColSideColors=temp.color,labRow=NULL,labCol="")
  gene.names.sorted<-rownames(fpkm.mod)[temp.heatmap$rowInd]
  gene.names.sorted<-gene.names.sorted[length(gene.names.sorted):1]
  
  fig.list[[length(fig.list)+1]]<-as.grob(~heatmap.2(as.matrix(fpkm.mod),breaks=temp.breaks,scale="none",dendrogram="row",Rowv=T,Colv=F,col=colorpanel(500,low="purple",mid="grey",high="yellow"),cexCol=1,trace="none",density.info="none",key=TRUE,symkey=FALSE,keysize=0.8,na.color="grey",ColSideColors=temp.color,labRow="",labCol=""))
  
  ######################################
  # 2. for the array data
  #%up to here: we are regenerating /scratch/Antun/microarray_Jonas/gene_expression_genesymbol_condensed.txt because of the bugs we found in the codes that generated it. The updated codes can be found in BAL_validation.Rmd.
  # load in the gene expression data from the Freiburgh cohort
  array.data<-read.table(array.data.filepath,sep="\t",header=T,check.names=F)
  rownames(array.data)<-as.matrix(array.data)[,1]
  array.data<-array.data[,-1]
#  annot <- read.table(array.annot.filepath, header = TRUE, sep = ",",stringsAsFactors=F)

  # load in the K-means clustering results for them
  clustering.result.filepath<-file.path(array.clustering.dir,temp.names)
  temp<-read.table(clustering.result.filepath,sep="\t",header=T,check.names = FALSE)
  clustering.result.validation<-temp[,2]
  names(clustering.result.validation)<-temp[,1]
  clustering.result.validation<-sort(clustering.result.validation,decreasing=F)
  
  # extract the microarray data for the genes in this module
  temp.names<-unname(sapply(gene.names.sorted,my.element.remove,splitchar="_",index=-1))
  array.mod<-array.data[temp.names,names(clustering.result.validation)]
  array.mod<-my.normalize(array.mod)
  # within each K means cluster, put similar samples next to each other, do the same thing to genes and obtain the ordered gene names
  
  for(j in 1:kmeans.k){
    temp.matrix<-array.mod[,clustering.result.validation==j]
    temp.matrix<-temp.matrix[,hclust(dist(t(temp.matrix)))$order]
    array.mod[,clustering.result.validation==j]<-temp.matrix
  }

  clustering.result.validation<-clustering.result.validation[colnames(array.mod)]
    
  # generate the heatmap for the microarray data
  temp.breaks<-seq(from=quantile(array.mod,prob=0.005,na.rm=T),to=quantile(array.mod,prob=0.995,na.rm=T),length=501)
  temp.color<-my.color.gen.cat(paste("cluster_",clustering.result.validation,sep=""))$color.num

   fig.list[[length(fig.list)+1]]<-as.grob(~heatmap.2(as.matrix(array.mod),breaks=temp.breaks,scale="none",dendrogram="none",Rowv=F,Colv=F,col=colorpanel(500,low="purple",mid="grey",high="yellow"),cexCol=1,trace="none",density.info="none",key=FALSE,symkey=FALSE,keysize=0.8,na.color="grey",ColSideColors=temp.color,labRow="",labCol=""))
   
   heights.vect<-c(heights.vect,length(gene.names.sorted))
   #fig.list[[length(fig.list)+1]]<-arrangeGrob(p1,p2,ncol=2,widths=width.vect)

   # draw the plot for the chosen mnodule
  width.vect<-c(ncol(fpkm.mod),ncol(array.mod))   
  subg<-grid.arrange(grobs=fig.list[(length(fig.list)-1):length(fig.list)],ncol=2,widths=width.vect)
  output.filepath<-file.path(output.dir,paste("Figure4_heatmap_validation_module_",modname,"_",i,".pdf",sep=""))
  ggsave(file=output.filepath,width = 250, height =  180, units = "mm", subg)
   }

#grid.arrange(fig.list[[length(fig.list)-1]],ncol=1)
width.vect<-c(ncol(fpkm.mod),ncol(array.mod))
#g<-grid.arrange(grobs=fig.list,ncol=2,widths=c(0.2*sum(width.vect),width.vect),heights=c(12,2,7,3,9))
g<-grid.arrange(grobs=fig.list,ncol=2,widths=width.vect,heights=c(8,2,7,3,8))

#g<-grid.arrange(grobs=fig.list,ncol=3,heights=heights.vect,widths=c(0.2,width.vect))

#g<-do.call(grid.arrange,c(fig.list,ncol=3,heights=heights.vect,widths=c(0.1,width.vect)))
#g<-grid.arrange(fig.list,ncol=2,heights=heights.vect,widths=c(0.1,2))
output.filepath<-file.path(output.dir,"Figure4_heatmap_validation.pdf")
ggsave(file=output.filepath,width = 250, height =  180, units = "mm", g)

```


```{r debugging,eval=FALSE,fig.width=10,fig.height=10,echo=FALSE,cache=FALSE}
array.data.filepath<-"/home/yanxiting/driver_Grace/scratch/Antun/microarray_Jonas/gene_expression_genesymbol_condensed_updated.txt"
array.annot.filepath<-"/home/yanxiting/driver_Grace/scratch/Antun/microarray_Jonas/HuGene-1_0-st-v1.na35.hg19.transcript.csv"
chosen.modules<-c("Turquoise","Brown4","Brown","Violet","Grey60")
array.clustering.dir<-"/home/yanxiting/driver_Grace/scratch/Antun/microarray_Jonas/clustering"
grads.clustering.dir<-"/home/yanxiting/driver_Grace/scratch/Antun/WGCNA_clusthg38/kmeans_selected_updated"
output.dir<-"/home/yanxiting/driver_Grace/scratch/GRADS/SARC_results/Results_summary_BAL_hg38/writings"

fig.list<-list()
heights.vect<-numeric()
panel.names<-c("a).","b).","c).","d).","e).")
# load in the clustering results of the GRADS cohort
for(i in 1:length(chosen.modules)){
  modname<-tolower(chosen.modules[i])
  
  ######################################
  # 1. for the GRADS RNAseq data
  
  # load in the clustering results
  temp.names<-list.files(grads.clustering.dir)
  temp.names<-temp.names[grep("_clustering.txt",temp.names)]
  temp.names<-temp.names[grep(paste(chosen.modules[i],"_",sep=""),temp.names)]
  if(length(temp.names)>1){
    cat("There are >1 clustering files for Module ",chosen.modules[i],"\n",sep="")
    next
  }
  
  clustering.result.filepath<-file.path(grads.clustering.dir,temp.names)
  clustering.result<-read.table(clustering.result.filepath,sep="\t",header=T,check.names = FALSE)
  temp<-clustering.result[,2]
  names(temp)<-clustering.result[,1]
  clustering.result<-temp
  clustering.result<-sort(clustering.result,decreasing=F)
  
  kmeans.k<-my.element.extract(temp.names,splitchar="_",index=2)
  kmeans.k<-as.numeric(my.element.extract(kmeans.k,splitchar="",index=-1))
  
  # load in the gene names in the module
  gene.names<-colnames(my.datExpr)[mergedColors==modname]
  fpkm.mod<-my.normalize(t(my.datExpr[names(clustering.result),gene.names]))
  cat("Module ",modname," quantile (1% and 99%):",quantile(fpkm.mod,prob=c(0.01,0.99)),"\n",sep=" ")

  # within each K means cluster, put similar samples next to each other, do the same thing to genes and obtain the ordered gene names
  
   for(j in 1:kmeans.k){
    temp.matrix<-fpkm.mod[,clustering.result==j]
    temp.matrix<-temp.matrix[,hclust(dist(t(temp.matrix)))$order]
    fpkm.mod[,clustering.result==j]<-temp.matrix
  }
  
  clustering.result<-clustering.result[colnames(fpkm.mod)]

   p0<-textGrob(panel.names[i],just="top")
   fig.list[[length(fig.list)+1]]<-p0
  
  # get the order of the genes
  #gene.names.sorted<-rownames(fpkm.mod)[hclust(dist(fpkm.mod))$order]
  
  temp.color<-my.color.gen.cat(paste("cluster_",clustering.result,sep=""))$color.num
  temp.breaks<-seq(from=quantile(fpkm.mod,prob=0.005),to=quantile(fpkm.mod,prob=0.995),length=501)
 
  
  # generate the heatmap for the K-means clustering results for the GRADS cohort
  temp.heatmap<-heatmap.2(as.matrix(fpkm.mod),breaks=temp.breaks,scale="none",dendrogram="row",Rowv=T,Colv=F,col=colorpanel(500,low="purple",mid="grey",high="yellow"),cexCol=1,trace="none",density.info="none",key=TRUE,symkey=FALSE,keysize=0.8,na.color="grey",ColSideColors=temp.color,labRow=NULL,labCol="")
  gene.names.sorted<-rownames(fpkm.mod)[temp.heatmap$rowInd]
  gene.names.sorted<-gene.names.sorted[length(gene.names.sorted):1]
  
  fig.list[[length(fig.list)+1]]<-as.grob(~heatmap.2(as.matrix(fpkm.mod),breaks=temp.breaks,scale="none",dendrogram="row",Rowv=T,Colv=F,col=colorpanel(500,low="purple",mid="grey",high="yellow"),cexCol=1,trace="none",density.info="none",key=TRUE,symkey=FALSE,keysize=0.8,na.color="grey",ColSideColors=temp.color,labRow="",labCol=""))
  
  ######################################
  # 2. for the array data
  #%up to here: we are regenerating /scratch/Antun/microarray_Jonas/gene_expression_genesymbol_condensed.txt because of the bugs we found in the codes that generated it. The updated codes can be found in BAL_validation.Rmd.
  # load in the gene expression data from the Freiburgh cohort
  array.data<-read.table(array.data.filepath,sep="\t",header=T,check.names=F)
  rownames(array.data)<-as.matrix(array.data)[,1]
  array.data<-array.data[,-1]
#  annot <- read.table(array.annot.filepath, header = TRUE, sep = ",",stringsAsFactors=F)

  # load in the K-means clustering results for them
  clustering.result.filepath<-file.path(array.clustering.dir,temp.names)
  temp<-read.table(clustering.result.filepath,sep="\t",header=T,check.names = FALSE)
  clustering.result.validation<-temp[,2]
  names(clustering.result.validation)<-temp[,1]
  clustering.result.validation<-sort(clustering.result.validation,decreasing=F)
  
  # extract the microarray data for the genes in this module
  temp.names<-unname(sapply(gene.names.sorted,my.element.remove,splitchar="_",index=-1))
  array.mod<-array.data[temp.names,names(clustering.result.validation)]
  array.mod<-my.normalize(array.mod)
  # within each K means cluster, put similar samples next to each other, do the same thing to genes and obtain the ordered gene names
  
  for(j in 1:kmeans.k){
    temp.matrix<-array.mod[,clustering.result.validation==j]
    temp.matrix<-temp.matrix[,hclust(dist(t(temp.matrix)))$order]
    array.mod[,clustering.result.validation==j]<-temp.matrix
  }

  clustering.result.validation<-clustering.result.validation[colnames(array.mod)]
    
  # generate the heatmap for the microarray data
  temp.breaks<-seq(from=quantile(array.mod,prob=0.005,na.rm=T),to=quantile(array.mod,prob=0.995,na.rm=T),length=501)
  temp.color<-my.color.gen.cat(paste("cluster_",clustering.result.validation,sep=""))$color.num

   fig.list[[length(fig.list)+1]]<-as.grob(~heatmap.2(as.matrix(array.mod),breaks=temp.breaks,scale="none",dendrogram="none",Rowv=F,Colv=F,col=colorpanel(500,low="purple",mid="grey",high="yellow"),cexCol=1,trace="none",density.info="none",key=FALSE,symkey=FALSE,keysize=0.8,na.color="grey",ColSideColors=temp.color,labRow="",labCol=""))
   
   heights.vect<-c(heights.vect,length(gene.names.sorted))
   #fig.list[[length(fig.list)+1]]<-arrangeGrob(p1,p2,ncol=2,widths=width.vect)
}

#grid.arrange(fig.list[[length(fig.list)-1]],ncol=1)
width.vect<-c(ncol(fpkm.mod),ncol(array.mod))
g<-grid.arrange(grobs=fig.list,ncol=3,widths=c(0.2*sum(width.vect),width.vect),heights=c(12,2,7,3,9))

#g<-grid.arrange(grobs=fig.list,ncol=3,heights=heights.vect,widths=c(0.2,width.vect))

#g<-do.call(grid.arrange,c(fig.list,ncol=3,heights=heights.vect,widths=c(0.1,width.vect)))
#g<-grid.arrange(fig.list,ncol=2,heights=heights.vect,widths=c(0.1,2))
output.filepath<-file.path(output.dir,"Figure4_heatmap_validation.pdf")
ggsave(file=output.filepath,width = 200, height =  1000, units = "mm", g)

```




### Figure 5. Overlap of the supervised analysis between the GRADS cohort and the Freiburgh cohort.
```{r pft_overlap_genes}
freiburgh.dir<-file.path(work.dir,"scratch/Antun/microarray_Jonas/supervised_analysis/PFTs_correlated_genes")
grads.dir<-file.path(work.dir,"scratch/GRADS/SARC_results/Results_summary_BAL_hg38/GRADS_BAL_SupervisedAnalysisHG38/Results_summary2/PFTs_correlated_genes")
output.subdir<-file.path(work.dir,"scratch/Antun/microarray_Jonas/overlap_2cohorts")

if(file.exists(output.subdir)==F){
  dir.create(output.subdir)
}

##############################
# for FVC% predicted
freiburgh.filepath<-file.path(freiburgh.dir,"correlation_spearman_BDFVCPRED_allgenes_Microarray.txt")
grads.filepath<-file.path(grads.dir,"correlation_spearman_FVCPRED_allgenes_BAL.txt")

temp1<-read.table(freiburgh.filepath,sep="\t",header=T,check.names=F)
temp2<-read.table(grads.filepath,sep="\t",header=T,check.names=F)

temp1.sig<-as.matrix(temp1)[temp1[,"pvalue"]<0.05,1]
temp1.nonsig<-as.matrix(temp1)[temp1[,"pvalue"]>=0.05,1]
temp2.sig<-as.matrix(temp2)[!is.na(temp2[,"pvalue"]) & temp2[,"pvalue"]<0.05,1]
temp2.sig<-unname(sapply(temp2.sig,my.element.remove,splitchar="_",index=-1))
temp2.nonsig<-as.matrix(temp2)[!is.na(temp2[,"pvalue"]) & temp2[,"pvalue"]>=0.05,1]
temp2.nonsig<-unname(sapply(temp2.nonsig,my.element.remove,splitchar="_",index=-1))
# output the list of overlapped genes
common.sig<-intersect(temp1.sig,temp2.sig)
cmd1.out<-temp1[temp1[,1]%in%common.sig,]
cmd2.out<-temp2[unname(sapply(as.matrix(temp2)[,1],my.element.remove,splitchar="_",index=-1))%in%common.sig,]
cmd1.out<-cmd1.out[,1:4]
cmd2.out<-cmd2.out[,1:4]
rownames(cmd1.out)<-as.matrix(cmd1.out)[,1]
rownames(cmd2.out)<-unname(sapply(as.matrix(cmd2.out)[,1],my.element.remove,splitchar="_",index=-1))
cmd1.out<-cmd1.out[common.sig,]
cmd2.out<-cmd2.out[common.sig,]

colnames(cmd1.out)[2:4]<-paste("Freiburgh_",colnames(cmd1.out)[2:4],sep="")
colnames(cmd2.out)[2:4]<-paste("GRADS_",colnames(cmd2.out)[2:4],sep="")

cmd.out<-cbind(cmd1.out,cmd2.out[,2:4])
output.filepath<-file.path(output.subdir,"common_sig_genes_pvalue0.05_FVCPRED_all.txt")
#write.table(cmd.out,file=output.filepath,append=F,sep="\t",row.names=F,col.names=T,quote=F)

cmd.out<-cmd.out[cmd.out[,"Freiburgh_corr_coef"]*cmd.out[,"GRADS_corr_coef"]>0,]
output.filepath<-file.path(output.subdir,"common_sig_genes_pvalue0.05_FVCPRED_consistent.txt")
#write.table(cmd.out,file=output.filepath,append=F,sep="\t",row.names=F,col.names=T,quote=F)


###################################
# for FEV1% predicted
freiburgh.filepath<-file.path(freiburgh.dir,"correlation_spearman_BDFEV1PRED_allgenes_Microarray.txt")
grads.filepath<-file.path(grads.dir,"correlation_spearman_FEV1PRED_allgenes_BAL.txt")

temp1<-read.table(freiburgh.filepath,sep="\t",header=T,check.names=F)
temp2<-read.table(grads.filepath,sep="\t",header=T,check.names=F)

temp1.sig<-as.matrix(temp1)[temp1[,"pvalue"]<0.05,1]
temp2.sig<-as.matrix(temp2)[(!is.na(temp2[,"pvalue"])) & (temp2[,"pvalue"]<0.05),1]
temp2.sig.tss<-temp2.sig
temp2.sig<-unname(sapply(temp2.sig,my.element.remove,splitchar="_",index=-1))
# output the list of overlapped genes
common.sig<-intersect(temp1.sig,temp2.sig)
cmd1.out<-temp1[temp1[,1]%in%common.sig,]
cmd2.out<-temp2[as.matrix(temp2)[,1]%in%temp2.sig.tss[temp2.sig%in%common.sig],]
cmd1.out<-cmd1.out[,1:4]
cmd2.out<-cmd2.out[,1:4]
rownames(cmd1.out)<-as.matrix(cmd1.out)[,1]
rownames(cmd2.out)<-unname(sapply(as.matrix(cmd2.out)[,1],my.element.remove,splitchar="_",index=-1))
cmd1.out<-cmd1.out[common.sig,]
cmd2.out<-cmd2.out[common.sig,]

colnames(cmd1.out)[2:4]<-paste("Freiburgh_",colnames(cmd1.out)[2:4],sep="")
colnames(cmd2.out)[2:4]<-paste("GRADS_",colnames(cmd2.out)[2:4],sep="")

cmd.out<-cbind(cmd1.out,cmd2.out[,2:4])
output.filepath<-file.path(output.subdir,"common_sig_genes_pvalue0.05_FEV1PRED_all.txt")
write.table(cmd.out,file=output.filepath,append=F,sep="\t",row.names=F,col.names=T,quote=F)

cmd.out<-cmd.out[cmd.out[,"Freiburgh_corr_coef"]*cmd.out[,"GRADS_corr_coef"]>0,]
output.filepath<-file.path(output.subdir,"common_sig_genes_pvalue0.05_FEV1PRED_consistent.txt")
write.table(cmd.out,file=output.filepath,append=F,sep="\t",row.names=F,col.names=T,quote=F)


```

```{r figure5_mosaic_plot_unique_genes_fvcpred}

#################################################
# calculate the contingency table to calculate the significance of overlap

freiburgh.dir<-file.path(work.dir,"scratch/Antun/microarray_Jonas/supervised_analysis/PFTs_correlated_genes")
grads.dir<-file.path(work.dir,"scratch/GRADS/SARC_results/Results_summary_BAL_hg38/GRADS_BAL_SupervisedAnalysisHG38/Results_summary2/PFTs_correlated_genes")
output.subdir<-file.path(work.dir,"scratch/Antun/microarray_Jonas/overlap_2cohorts")

if(file.exists(output.subdir)==F){
  dir.create(output.subdir)
}

freiburgh.filepath<-file.path(freiburgh.dir,"correlation_spearman_BDFVCPRED_allgenes_Microarray.txt")
grads.filepath<-file.path(grads.dir,"correlation_spearman_FVCPRED_allgenes_BAL.txt")

temp1<-read.table(freiburgh.filepath,sep="\t",header=T,check.names=F)
temp2<-read.table(grads.filepath,sep="\t",header=T,check.names=F)

temp2<-temp2[!is.na(temp2[,"corr_coef"]),]
temp1<-temp1[!is.na(temp1[,"corr_coef"]),]

temp2.list<-split(temp2[,"gene_names"],unname(sapply(as.matrix(temp2)[,"gene_names"],my.element.remove,splitchar="_",index=-1)))
temp2.length<-unlist(lapply(temp2.list,length))
temp2<-temp2[unname(sapply(as.matrix(temp2)[,"gene_names"],my.element.remove,splitchar="_",index=-1))%in%names(temp2.list)[temp2.length==1],]

rownames(temp1)<-as.matrix(temp1)[,1]
rownames(temp2)<-unname(sapply(as.matrix(temp2)[,"gene_names"],my.element.remove,splitchar="_",index=-1))

common.genes<-intersect(rownames(temp1),rownames(temp2))
temp1<-temp1[common.genes,]
temp2<-temp2[common.genes,]

sig.overlap<-rownames(temp1)[temp1[,"pvalue"]<0.05 & temp2[,"pvalue"]<0.05 & temp1[,"corr_coef"]*temp2[,"corr_coef"]>0]
nonsig.overlap<-rownames(temp1)[temp1[,"pvalue"]>=0.05 & temp2[,"pvalue"]>=0.05 & temp1[,"corr_coef"]*temp2[,"corr_coef"]>0]
sig.temp1.only<-rownames(temp1)[temp1[,"pvalue"]<0.05]
sig.temp1.only<-setdiff(sig.temp1.only,sig.overlap)
sig.temp2.only<-rownames(temp2)[temp2[,"pvalue"]<0.05]
sig.temp2.only<-setdiff(sig.temp2.only,sig.overlap)

con.table.fvcpred<-matrix(c(length(sig.overlap),length(sig.temp1.only),length(sig.temp2.only),length(nonsig.overlap)),nrow=2,byrow=T)

colnames(con.table.fvcpred)<-c("sig","non-sig")
rownames(con.table.fvcpred)<-c("sig","non-sig")

#output.filepath<-file.path(output.dir,"figure5_FVCPRED.eps")
#postscript(output.filepath,height=3.5,width=3.5)
mosaic(con.table.fvcpred,shade=TRUE,legend=TRUE)

cat("There are in total ",nrow(temp1)," genes in the analysis.\n",sep="")
write.table(con.table.fvcpred,sep="\t",row.names=T,col.names=T,quote=F)

cmd1.out<-temp1[sig.overlap,1:4]
cmd2.out<-temp2[sig.overlap,1:4]

colnames(cmd1.out)[2:4]<-paste("Freiburgh_",colnames(cmd1.out)[2:4],sep="")
colnames(cmd2.out)[2:4]<-paste("GRADS_",colnames(cmd2.out)[2:4],sep="")

cmd.out<-cbind(cmd1.out,cmd2.out[,2:4])
output.filepath<-file.path(output.subdir,"common_sig_genes_pvalue0.05_FEV1PRED_all.txt")
write.table(cmd.out,file=output.filepath,append=F,sep="\t",row.names=F,col.names=T,quote=F)

```


```{r figure5_mosaic_plot_unique_genes_fev1pred}

#################################################
# calculate the contingency table to calculate the significance of overlap

freiburgh.dir<-file.path(work.dir,"scratch/Antun/microarray_Jonas/supervised_analysis/PFTs_correlated_genes")
grads.dir<-file.path(work.dir,"scratch/GRADS/SARC_results/Results_summary_BAL_hg38/GRADS_BAL_SupervisedAnalysisHG38/Results_summary2/PFTs_correlated_genes")
output.subdir<-file.path(work.dir,"scratch/Antun/microarray_Jonas/overlap_2cohorts")

if(file.exists(output.subdir)==F){
  dir.create(output.subdir)
}


freiburgh.filepath<-file.path(freiburgh.dir,"correlation_spearman_BDFEV1PRED_allgenes_Microarray.txt")
grads.filepath<-file.path(grads.dir,"correlation_spearman_FEV1PRED_allgenes_BAL.txt")

temp1<-read.table(freiburgh.filepath,sep="\t",header=T,check.names=F)
temp2<-read.table(grads.filepath,sep="\t",header=T,check.names=F)

temp2<-temp2[!is.na(temp2[,"corr_coef"]),]
temp1<-temp1[!is.na(temp1[,"corr_coef"]),]

temp2.list<-split(temp2[,"gene_names"],unname(sapply(as.matrix(temp2)[,"gene_names"],my.element.remove,splitchar="_",index=-1)))
temp2.length<-unlist(lapply(temp2.list,length))
temp2<-temp2[unname(sapply(as.matrix(temp2)[,"gene_names"],my.element.remove,splitchar="_",index=-1))%in%names(temp2.list)[temp2.length==1],]

rownames(temp1)<-as.matrix(temp1)[,1]
rownames(temp2)<-unname(sapply(as.matrix(temp2)[,"gene_names"],my.element.remove,splitchar="_",index=-1))

common.genes<-intersect(rownames(temp1),rownames(temp2))
temp1<-temp1[common.genes,]
temp2<-temp2[common.genes,]

sig.overlap<-rownames(temp1)[temp1[,"pvalue"]<0.05 & temp2[,"pvalue"]<0.05 & temp1[,"corr_coef"]*temp2[,"corr_coef"]>0]
nonsig.overlap<-rownames(temp1)[temp1[,"pvalue"]>=0.05 & temp2[,"pvalue"]>=0.05 & temp1[,"corr_coef"]*temp2[,"corr_coef"]>0]
sig.temp1.only<-rownames(temp1)[temp1[,"pvalue"]<0.05]
sig.temp1.only<-setdiff(sig.temp1.only,sig.overlap)
sig.temp2.only<-rownames(temp2)[temp2[,"pvalue"]<0.05]
sig.temp2.only<-setdiff(sig.temp2.only,sig.overlap)

con.table.fev1pred<-matrix(c(length(sig.overlap),length(sig.temp1.only),length(sig.temp2.only),length(nonsig.overlap)),nrow=2,byrow=T)

colnames(con.table.fev1pred)<-c("sig","non-sig")
rownames(con.table.fev1pred)<-c("sig","non-sig")

#output.filepath<-file.path(output.dir,"figure5_FVCPRED.eps")
#postscript(output.filepath,height=3.5,width=3.5)
mosaic(con.table.fev1pred,shade=TRUE,legend=TRUE)
cat("There are in total ",nrow(temp1)," genes in the analysis.\n",sep="")
write.table(con.table.fev1pred,sep="\t",row.names=T,col.names=T,quote=F)
```

```{r figure5_mosaic_plot}
# These codes were run on the laptop
library(vcd)
output.dir<-"/Users/yanxiting/MyVolumes/GRACE/scratch/GRADS/SARC_results/Results_summary_BAL_hg38/writings"
con.table.fvcpred<-matrix(c(44,2631,628,17140),nrow=2,byrow=T)
con.table.fvcpred<-matrix(c(44,2587,585,17141),nrow=2,byrow=T)

colnames(con.table.fvcpred)<-c("sig","non-sig")
rownames(con.table.fvcpred)<-c("sig","non-sig")

output.filepath<-file.path(output.dir,"figure5_FVCPRED.eps")
postscript(output.filepath,height=3.5,width=3.5)
mosaic(con.table.fvcpred,shade=TRUE,legend=TRUE)
dev.off()


con.table.fev1pred<-matrix(c(48,3264,354,16785),nrow=2,byrow=T)
colnames(con.table.fev1pred)<-c("sig","non-sig")
rownames(con.table.fev1pred)<-c("sig","non-sig")

output.filepath<-file.path(output.dir,"figure5_FEV1PRED.eps")
postscript(output.filepath,height=3.5,width=3.5)
mosaic(con.table.fev1pred,shade=TRUE,legend=TRUE)
dev.off()



con.table.bdfvc<-matrix(c(115,2087,931,17460),nrow=2,byrow=T)
colnames(con.table.bdfvc)<-c("sig","non-sig")
rownames(con.table.bdfvc)<-c("sig","non-sig")

output.filepath<-file.path(output.dir,"figure5_BDFVC.eps")
postscript(output.filepath,height=3.5,width=3.5)
mosaic(con.table.bdfvc,shade=TRUE,legend=TRUE)
dev.off()


con.table.bdfev1<-matrix(c(126,2907,658,16930),nrow=2,byrow=T)
colnames(con.table.bdfev1)<-c("sig","non-sig")
rownames(con.table.bdfev1)<-c("sig","non-sig")

output.filepath<-file.path(output.dir,"figure5_BDFEV1.eps")
postscript(output.filepath,height=3.5,width=3.5)
mosaic(con.table.bdfev1,shade=TRUE,legend=TRUE)
dev.off()

```


### Data Sharing with Pittsburgh
```{r fastq_extraction}
fastq.dir<-file.path(work.dir,"scratch_kaminski/public/Backup/GRADS_BQ/fastq_JH/GRADS_fastq/fastq")
fpkm.filepath<-file.path(work.dir,"scratch/GRADS/SARC_results/Results_summary_BAL_hg38/FPKM_matrix_hg38_Corrected_allGenes_BAL209.txt")
fpkm.matrix<-read.table(fpkm.filepath,sep="\t",header=T,check.names=F,row.names=1)
sample.names<-colnames(fpkm.matrix)

filenames<-list.files(fastq.dir)
filenames<-filenames[substr(filenames,1,1)%in%c("0","1")]
temp1<-unname(sapply(filenames,my.element.extract,splitchar="_",index=1))
temp2<-unname(sapply(filenames,my.element.extract,splitchar="_",index=3))
filenames.id<-paste(temp1,"_",temp2,sep="")

filenames<-filenames[filenames.id%in%sample.names]
output.dir<-file.path(work.dir,"scratch60/GRADS_SARC_share")
for(i in 1:length(filenames)){
  file.copy(from=file.path(fastq.dir,filenames[i]),to=file.path(output.dir,filenames[i]))
}

```





## pca plot labeled for age, gender, race, phenotype group

## supervised analysis

### nominal p value based
#### scatterplot: x axis is the % predicted, y is the correlation coefficient, dot size is the p value, show gene names for some of them. color represent whether the genes overlap betwen the different traits.
analysis to show the overlap

#### 3 heatmaps to show the selected genes for the clinical traits with clinical traits showing on the top (phenotype, scadding and PFT categorical)
#### heatmaps for genes only pass FDR<0.05
#### consider violin plot for different stages

### FDR based

adjust the data for cell differentials and covariates (age, gender and race)
